// @algorandfoundation/algokit-utils/packages/indexer_client/src/apis/api-service.js
const require_model_runtime = require("../core/model-runtime.js");
const require_block = require("../models/block.js");
const require_box = require("../models/box.js");
const require_health_check = require("../models/health-check.js");
const require_account_response = require("../models/account-response.js");
const require_asset_holdings_response = require("../models/asset-holdings-response.js");
const require_accounts_response = require("../models/accounts-response.js");
const require_asset_balances_response = require("../models/asset-balances-response.js");
const require_application_response = require("../models/application-response.js");
const require_applications_response = require("../models/applications-response.js");
const require_application_logs_response = require("../models/application-logs-response.js");
const require_application_local_states_response = require("../models/application-local-states-response.js");
const require_asset_response = require("../models/asset-response.js");
const require_boxes_response = require("../models/boxes-response.js");
const require_assets_response = require("../models/assets-response.js");
const require_block_headers_response = require("../models/block-headers-response.js");
const require_transaction_response = require("../models/transaction-response.js");
const require_transactions_response = require("../models/transactions-response.js");
//#region packages/indexer_client/src/apis/api-service.ts
var IndexerApi = class {
	constructor(httpRequest) {
		this.httpRequest = httpRequest;
	}
	mimeTypeFor(format) {
		return format === "json" ? "application/json" : format === "msgpack" ? "application/msgpack" : "text/plain";
	}
	async healthCheck() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/health",
			path: {},
			query: {},
			headers,
			body: void 0
		}), require_health_check.HealthCheckMeta);
	}
	/**
	* Lookup an account's asset holdings, optionally for a specific ID.
	*/
	async lookupAccountAppLocalStates(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/apps-local-state",
			path: { "account-id": account },
			query: {
				"application-id": params?.applicationId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), require_application_local_states_response.ApplicationLocalStatesResponseMeta);
	}
	/**
	* Lookup an account's asset holdings, optionally for a specific ID.
	*/
	async lookupAccountAssets(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/assets",
			path: { "account-id": account },
			query: {
				"asset-id": params?.assetId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), require_asset_holdings_response.AssetHoldingsResponseMeta);
	}
	/**
	* Lookup account information.
	*/
	async lookupAccountById(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}",
			path: { "account-id": account },
			query: {
				round: params?.round,
				"include-all": params?.includeAll,
				exclude: params?.exclude
			},
			headers,
			body: void 0
		}), require_account_response.AccountResponseMeta);
	}
	/**
	* Lookup an account's created application parameters, optionally for a specific ID.
	*/
	async lookupAccountCreatedApplications(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/created-applications",
			path: { "account-id": account },
			query: {
				"application-id": params?.applicationId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), require_applications_response.ApplicationsResponseMeta);
	}
	/**
	* Lookup an account's created asset parameters, optionally for a specific ID.
	*/
	async lookupAccountCreatedAssets(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/created-assets",
			path: { "account-id": account },
			query: {
				"asset-id": params?.assetId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), require_assets_response.AssetsResponseMeta);
	}
	/**
	* Lookup account transactions. Transactions are returned newest to oldest.
	*/
	async lookupAccountTransactions(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/transactions",
			path: { "account-id": account },
			query: {
				limit: params?.limit,
				next: params?.next,
				"note-prefix": params?.notePrefix,
				"tx-type": params?.txType,
				"sig-type": params?.sigType,
				txid: params?.txId,
				round: params?.round,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"asset-id": params?.assetId,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan,
				"rekey-to": params?.rekeyTo
			},
			headers,
			body: void 0
		}), require_transactions_response.TransactionsResponseMeta);
	}
	/**
	* Given an application ID and box name, returns base64 encoded box name and value. Box names must be in the goal app call arg form 'encoding:value'. For ints, use the form 'int:1234'. For raw bytes, encode base 64 and use 'b64' prefix as in 'b64:A=='. For printable strings, use the form 'str:hello'. For addresses, use the form 'addr:XYZ...'.
	*/
	async _lookupApplicationBoxByIdAndName(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/box",
			path: { "application-id": applicationId },
			query: { name: params?.name },
			headers,
			body: void 0
		}), require_box.BoxMeta);
	}
	/**
	* Lookup application.
	*/
	async lookupApplicationById(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}",
			path: { "application-id": applicationId },
			query: { "include-all": params?.includeAll },
			headers,
			body: void 0
		}), require_application_response.ApplicationResponseMeta);
	}
	/**
	* Lookup application logs.
	*/
	async lookupApplicationLogsById(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/logs",
			path: { "application-id": applicationId },
			query: {
				limit: params?.limit,
				next: params?.next,
				txid: params?.txId,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"sender-address": params?.senderAddress
			},
			headers,
			body: void 0
		}), require_application_logs_response.ApplicationLogsResponseMeta);
	}
	/**
	* Lookup the list of accounts who hold this asset
	*/
	async lookupAssetBalances(assetId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}/balances",
			path: { "asset-id": assetId },
			query: {
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan
			},
			headers,
			body: void 0
		}), require_asset_balances_response.AssetBalancesResponseMeta);
	}
	/**
	* Lookup asset information.
	*/
	async lookupAssetById(assetId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}",
			path: { "asset-id": assetId },
			query: { "include-all": params?.includeAll },
			headers,
			body: void 0
		}), require_asset_response.AssetResponseMeta);
	}
	/**
	* Lookup transactions for an asset. Transactions are returned oldest to newest.
	*/
	async lookupAssetTransactions(assetId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}/transactions",
			path: { "asset-id": assetId },
			query: {
				limit: params?.limit,
				next: params?.next,
				"note-prefix": params?.notePrefix,
				"tx-type": params?.txType,
				"sig-type": params?.sigType,
				txid: params?.txId,
				round: params?.round,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan,
				address: params?.address,
				"address-role": params?.addressRole,
				"exclude-close-to": params?.excludeCloseTo,
				"rekey-to": params?.rekeyTo
			},
			headers,
			body: void 0
		}), require_transactions_response.TransactionsResponseMeta);
	}
	/**
	* Lookup block.
	*/
	async lookupBlock(roundNumber, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round-number}",
			path: { "round-number": roundNumber },
			query: { "header-only": params?.headerOnly },
			headers,
			body: void 0
		}), require_block.BlockMeta);
	}
	/**
	* Lookup a single transaction.
	*/
	async lookupTransactionById(txId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions/{txid}",
			path: { txid: txId },
			query: {},
			headers,
			body: void 0
		}), require_transaction_response.TransactionResponseMeta);
	}
	/**
	* Search for accounts.
	*/
	async searchForAccounts(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts",
			path: {},
			query: {
				"asset-id": params?.assetId,
				limit: params?.limit,
				next: params?.next,
				"currency-greater-than": params?.currencyGreaterThan,
				"include-all": params?.includeAll,
				exclude: params?.exclude,
				"currency-less-than": params?.currencyLessThan,
				"auth-addr": params?.authAddr,
				round: params?.round,
				"application-id": params?.applicationId,
				"online-only": params?.onlineOnly
			},
			headers,
			body: void 0
		}), require_accounts_response.AccountsResponseMeta);
	}
	/**
	* Given an application ID, returns the box names of that application sorted lexicographically.
	*/
	async searchForApplicationBoxes(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/boxes",
			path: { "application-id": applicationId },
			query: {
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), require_boxes_response.BoxesResponseMeta);
	}
	/**
	* Search for applications
	*/
	async searchForApplications(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications",
			path: {},
			query: {
				"application-id": params?.applicationId,
				creator: params?.creator,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), require_applications_response.ApplicationsResponseMeta);
	}
	/**
	* Search for assets.
	*/
	async searchForAssets(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets",
			path: {},
			query: {
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next,
				creator: params?.creator,
				name: params?.name,
				unit: params?.unit,
				"asset-id": params?.assetId
			},
			headers,
			body: void 0
		}), require_assets_response.AssetsResponseMeta);
	}
	/**
	* Search for block headers. Block headers are returned in ascending round order. Transactions are not included in the output.
	*/
	async searchForBlockHeaders(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/block-headers",
			path: {},
			query: {
				limit: params?.limit,
				next: params?.next,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				proposers: params?.proposers,
				expired: params?.expired,
				absent: params?.absent
			},
			headers,
			body: void 0
		}), require_block_headers_response.BlockHeadersResponseMeta);
	}
	/**
	* Search for transactions. Transactions are returned oldest to newest unless the address parameter is used, in which case results are returned newest to oldest.
	*/
	async searchForTransactions(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions",
			path: {},
			query: {
				limit: params?.limit,
				next: params?.next,
				"note-prefix": params?.notePrefix,
				"tx-type": params?.txType,
				"sig-type": params?.sigType,
				"group-id": params?.groupId,
				txid: params?.txId,
				round: params?.round,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"asset-id": params?.assetId,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan,
				address: params?.address,
				"address-role": params?.addressRole,
				"exclude-close-to": params?.excludeCloseTo,
				"rekey-to": params?.rekeyTo,
				"application-id": params?.applicationId
			},
			headers,
			body: void 0
		}), require_transactions_response.TransactionsResponseMeta);
	}
	/**
	* Given an application ID and box name, it returns the round, box name, and value.
	*/
	async lookupApplicationBoxByIdAndName(applicationId, boxName) {
		const name = `b64:${Buffer.from(boxName).toString("base64")}`;
		return this._lookupApplicationBoxByIdAndName(applicationId, { name });
	}
};
//#endregion
exports.IndexerApi = IndexerApi;
