// @algorandfoundation/algokit-utils/packages/indexer_client/src/apis/api-service.mjs
import { decodeJson } from "../core/model-runtime.mjs";
import { BlockMeta } from "../models/block.mjs";
import { BoxMeta } from "../models/box.mjs";
import { HealthCheckMeta } from "../models/health-check.mjs";
import { AccountResponseMeta } from "../models/account-response.mjs";
import { AssetHoldingsResponseMeta } from "../models/asset-holdings-response.mjs";
import { AccountsResponseMeta } from "../models/accounts-response.mjs";
import { AssetBalancesResponseMeta } from "../models/asset-balances-response.mjs";
import { ApplicationResponseMeta } from "../models/application-response.mjs";
import { ApplicationsResponseMeta } from "../models/applications-response.mjs";
import { ApplicationLogsResponseMeta } from "../models/application-logs-response.mjs";
import { ApplicationLocalStatesResponseMeta } from "../models/application-local-states-response.mjs";
import { AssetResponseMeta } from "../models/asset-response.mjs";
import { BoxesResponseMeta } from "../models/boxes-response.mjs";
import { AssetsResponseMeta } from "../models/assets-response.mjs";
import { BlockHeadersResponseMeta } from "../models/block-headers-response.mjs";
import { TransactionResponseMeta } from "../models/transaction-response.mjs";
import { TransactionsResponseMeta } from "../models/transactions-response.mjs";
//#region packages/indexer_client/src/apis/api-service.ts
var IndexerApi = class {
	constructor(httpRequest) {
		this.httpRequest = httpRequest;
	}
	mimeTypeFor(format) {
		return format === "json" ? "application/json" : format === "msgpack" ? "application/msgpack" : "text/plain";
	}
	async healthCheck() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/health",
			path: {},
			query: {},
			headers,
			body: void 0
		}), HealthCheckMeta);
	}
	/**
	* Lookup an account's asset holdings, optionally for a specific ID.
	*/
	async lookupAccountAppLocalStates(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/apps-local-state",
			path: { "account-id": account },
			query: {
				"application-id": params?.applicationId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), ApplicationLocalStatesResponseMeta);
	}
	/**
	* Lookup an account's asset holdings, optionally for a specific ID.
	*/
	async lookupAccountAssets(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/assets",
			path: { "account-id": account },
			query: {
				"asset-id": params?.assetId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), AssetHoldingsResponseMeta);
	}
	/**
	* Lookup account information.
	*/
	async lookupAccountById(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}",
			path: { "account-id": account },
			query: {
				round: params?.round,
				"include-all": params?.includeAll,
				exclude: params?.exclude
			},
			headers,
			body: void 0
		}), AccountResponseMeta);
	}
	/**
	* Lookup an account's created application parameters, optionally for a specific ID.
	*/
	async lookupAccountCreatedApplications(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/created-applications",
			path: { "account-id": account },
			query: {
				"application-id": params?.applicationId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), ApplicationsResponseMeta);
	}
	/**
	* Lookup an account's created asset parameters, optionally for a specific ID.
	*/
	async lookupAccountCreatedAssets(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/created-assets",
			path: { "account-id": account },
			query: {
				"asset-id": params?.assetId,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), AssetsResponseMeta);
	}
	/**
	* Lookup account transactions. Transactions are returned newest to oldest.
	*/
	async lookupAccountTransactions(account, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{account-id}/transactions",
			path: { "account-id": account },
			query: {
				limit: params?.limit,
				next: params?.next,
				"note-prefix": params?.notePrefix,
				"tx-type": params?.txType,
				"sig-type": params?.sigType,
				txid: params?.txId,
				round: params?.round,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"asset-id": params?.assetId,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan,
				"rekey-to": params?.rekeyTo
			},
			headers,
			body: void 0
		}), TransactionsResponseMeta);
	}
	/**
	* Given an application ID and box name, returns base64 encoded box name and value. Box names must be in the goal app call arg form 'encoding:value'. For ints, use the form 'int:1234'. For raw bytes, encode base 64 and use 'b64' prefix as in 'b64:A=='. For printable strings, use the form 'str:hello'. For addresses, use the form 'addr:XYZ...'.
	*/
	async _lookupApplicationBoxByIdAndName(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/box",
			path: { "application-id": applicationId },
			query: { name: params?.name },
			headers,
			body: void 0
		}), BoxMeta);
	}
	/**
	* Lookup application.
	*/
	async lookupApplicationById(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}",
			path: { "application-id": applicationId },
			query: { "include-all": params?.includeAll },
			headers,
			body: void 0
		}), ApplicationResponseMeta);
	}
	/**
	* Lookup application logs.
	*/
	async lookupApplicationLogsById(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/logs",
			path: { "application-id": applicationId },
			query: {
				limit: params?.limit,
				next: params?.next,
				txid: params?.txId,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"sender-address": params?.senderAddress
			},
			headers,
			body: void 0
		}), ApplicationLogsResponseMeta);
	}
	/**
	* Lookup the list of accounts who hold this asset
	*/
	async lookupAssetBalances(assetId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}/balances",
			path: { "asset-id": assetId },
			query: {
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan
			},
			headers,
			body: void 0
		}), AssetBalancesResponseMeta);
	}
	/**
	* Lookup asset information.
	*/
	async lookupAssetById(assetId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}",
			path: { "asset-id": assetId },
			query: { "include-all": params?.includeAll },
			headers,
			body: void 0
		}), AssetResponseMeta);
	}
	/**
	* Lookup transactions for an asset. Transactions are returned oldest to newest.
	*/
	async lookupAssetTransactions(assetId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}/transactions",
			path: { "asset-id": assetId },
			query: {
				limit: params?.limit,
				next: params?.next,
				"note-prefix": params?.notePrefix,
				"tx-type": params?.txType,
				"sig-type": params?.sigType,
				txid: params?.txId,
				round: params?.round,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan,
				address: params?.address,
				"address-role": params?.addressRole,
				"exclude-close-to": params?.excludeCloseTo,
				"rekey-to": params?.rekeyTo
			},
			headers,
			body: void 0
		}), TransactionsResponseMeta);
	}
	/**
	* Lookup block.
	*/
	async lookupBlock(roundNumber, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round-number}",
			path: { "round-number": roundNumber },
			query: { "header-only": params?.headerOnly },
			headers,
			body: void 0
		}), BlockMeta);
	}
	/**
	* Lookup a single transaction.
	*/
	async lookupTransactionById(txId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions/{txid}",
			path: { txid: txId },
			query: {},
			headers,
			body: void 0
		}), TransactionResponseMeta);
	}
	/**
	* Search for accounts.
	*/
	async searchForAccounts(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts",
			path: {},
			query: {
				"asset-id": params?.assetId,
				limit: params?.limit,
				next: params?.next,
				"currency-greater-than": params?.currencyGreaterThan,
				"include-all": params?.includeAll,
				exclude: params?.exclude,
				"currency-less-than": params?.currencyLessThan,
				"auth-addr": params?.authAddr,
				round: params?.round,
				"application-id": params?.applicationId,
				"online-only": params?.onlineOnly
			},
			headers,
			body: void 0
		}), AccountsResponseMeta);
	}
	/**
	* Given an application ID, returns the box names of that application sorted lexicographically.
	*/
	async searchForApplicationBoxes(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/boxes",
			path: { "application-id": applicationId },
			query: {
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), BoxesResponseMeta);
	}
	/**
	* Search for applications
	*/
	async searchForApplications(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications",
			path: {},
			query: {
				"application-id": params?.applicationId,
				creator: params?.creator,
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next
			},
			headers,
			body: void 0
		}), ApplicationsResponseMeta);
	}
	/**
	* Search for assets.
	*/
	async searchForAssets(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets",
			path: {},
			query: {
				"include-all": params?.includeAll,
				limit: params?.limit,
				next: params?.next,
				creator: params?.creator,
				name: params?.name,
				unit: params?.unit,
				"asset-id": params?.assetId
			},
			headers,
			body: void 0
		}), AssetsResponseMeta);
	}
	/**
	* Search for block headers. Block headers are returned in ascending round order. Transactions are not included in the output.
	*/
	async searchForBlockHeaders(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/block-headers",
			path: {},
			query: {
				limit: params?.limit,
				next: params?.next,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				proposers: params?.proposers,
				expired: params?.expired,
				absent: params?.absent
			},
			headers,
			body: void 0
		}), BlockHeadersResponseMeta);
	}
	/**
	* Search for transactions. Transactions are returned oldest to newest unless the address parameter is used, in which case results are returned newest to oldest.
	*/
	async searchForTransactions(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions",
			path: {},
			query: {
				limit: params?.limit,
				next: params?.next,
				"note-prefix": params?.notePrefix,
				"tx-type": params?.txType,
				"sig-type": params?.sigType,
				"group-id": params?.groupId,
				txid: params?.txId,
				round: params?.round,
				"min-round": params?.minRound,
				"max-round": params?.maxRound,
				"asset-id": params?.assetId,
				"before-time": params?.beforeTime,
				"after-time": params?.afterTime,
				"currency-greater-than": params?.currencyGreaterThan,
				"currency-less-than": params?.currencyLessThan,
				address: params?.address,
				"address-role": params?.addressRole,
				"exclude-close-to": params?.excludeCloseTo,
				"rekey-to": params?.rekeyTo,
				"application-id": params?.applicationId
			},
			headers,
			body: void 0
		}), TransactionsResponseMeta);
	}
	/**
	* Given an application ID and box name, it returns the round, box name, and value.
	*/
	async lookupApplicationBoxByIdAndName(applicationId, boxName) {
		const name = `b64:${Buffer.from(boxName).toString("base64")}`;
		return this._lookupApplicationBoxByIdAndName(applicationId, { name });
	}
};
//#endregion
export { IndexerApi };
