// @algorandfoundation/algokit-utils/packages/common/src/codecs/primitives/fixed-bytes.mjs
import { Codec } from "../codec.mjs";
import { Buffer } from "buffer";

//#region packages/common/src/codecs/primitives/fixed-bytes.ts
var FixedBytesCodec = class extends Codec {
	constructor(length) {
		super();
		this.length = length;
	}
	defaultValue() {
		return new Uint8Array(this.length);
	}
	toEncoded(value, format) {
		if (format === "json") return Buffer.from(value).toString("base64");
		return value;
	}
	fromEncoded(value, _format) {
		if (value instanceof Uint8Array) return value;
		if (typeof value === "string") return new Uint8Array(Buffer.from(value, "base64"));
		throw new Error(`Cannot decode fixed ${this.length} bytes from ${typeof value}`);
	}
	isDefaultValue(value) {
		if (value.byteLength !== this.length) return false;
		return value.every((byte) => byte === 0);
	}
};
const fixedBytes32Codec = new FixedBytesCodec(32);
const fixedBytes64Codec = new FixedBytesCodec(64);
const fixedBytes1793Codec = new FixedBytesCodec(1793);

//#endregion
export { fixedBytes1793Codec, fixedBytes32Codec, fixedBytes64Codec };
