// @algorandfoundation/algokit-utils/packages/common/src/codecs/primitives/unknown.mjs
import { ADDRESS_LENGTH, PUBLIC_KEY_BYTE_LENGTH, SIGNATURE_BYTE_LENGTH } from "../../constants.mjs";
import { Codec } from "../codec.mjs";
import { bigIntCodec } from "./bigint.mjs";
import { booleanCodec } from "./boolean.mjs";
import { numberCodec } from "./number.mjs";
import { stringCodec } from "./string.mjs";
import { ArrayCodec } from "../composite/array.mjs";
import { RecordCodec } from "../composite/record.mjs";

//#region packages/common/src/codecs/primitives/unknown.ts
/**
* Unknown codec - passthrough for unknown/any types
* Converts Maps with Uint8Array keys to objects with string keys recursively
*/
var UnknownCodec = class extends Codec {
	textDecoder = new TextDecoder("utf-8", { fatal: true });
	recordCodec;
	arrayCodec;
	constructor() {
		super();
		this.recordCodec = new RecordCodec(this);
		this.arrayCodec = new ArrayCodec(this);
	}
	defaultValue() {}
	toEncoded(_value, _format) {
		throw new Error("UnknownCodec does not support encoding");
	}
	fromEncoded(value, format) {
		return this.processValue(value, format);
	}
	maybeDecodeAsUtf8(bytes) {
		try {
			if ((bytes.length === ADDRESS_LENGTH || bytes.length === PUBLIC_KEY_BYTE_LENGTH || bytes.length === SIGNATURE_BYTE_LENGTH) && bytes.every((byte) => byte === 0)) return bytes;
			return this.textDecoder.decode(bytes);
		} catch {
			return bytes;
		}
	}
	/**
	* Recursively process unknown values, using various codecs where appropriate
	*/
	processValue(value, format) {
		if (value === null || value === void 0) return value;
		if (value instanceof Uint8Array) return this.maybeDecodeAsUtf8(value);
		if (typeof value === "bigint") return bigIntCodec.decode(value, format);
		if (typeof value === "number") return numberCodec.decode(value, format);
		if (typeof value === "string") return stringCodec.decode(value, format);
		if (typeof value === "boolean") return booleanCodec.decode(value, format);
		if (Array.isArray(value)) return this.arrayCodec.decode(value, format);
		if (value instanceof Map || typeof value === "object") return this.recordCodec.decode(value, format);
		return value;
	}
};
const unknownCodec = new UnknownCodec();

//#endregion
export { unknownCodec };
