// @x402/avm/dist/esm/chunk-QEY4CIDB.mjs
import {
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH
} from "./chunk-MWOLCZUZ.mjs";

// src/utils.ts
import {
  decodeTransaction as decodeUnsignedTxn,
  decodeSignedTransaction as decodeSignedTxn
} from "@algorandfoundation/algokit-utils/transact";
import { isValidAddress } from "@algorandfoundation/algokit-utils/common";
import { convertToTokenAmount, numberToDecimalString } from "@x402/core/utils";
import { Address, encodeAddress, decodeAddress } from "@algorandfoundation/algokit-utils/common";
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function decodeSignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return decodeSignedTxn(bytes);
}
function decodeUnsignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return decodeUnsignedTxn(bytes);
}
function isValidAlgorandAddress(address) {
  return isValidAddress(address);
}
function getSenderFromTransaction(txnBytes, isSigned = true) {
  if (isSigned) {
    const signedTxn = decodeSignedTxn(txnBytes);
    return signedTxn.txn.sender.toString();
  }
  const txn = decodeUnsignedTxn(txnBytes);
  return txn.sender.toString();
}
function convertFromTokenAmount(atomicAmount, decimals) {
  const amount = BigInt(atomicAmount);
  const divisor = BigInt(10 ** decimals);
  const intPart = amount / divisor;
  const decPart = amount % divisor;
  if (decPart === BigInt(0)) {
    return intPart.toString();
  }
  const decStr = decPart.toString().padStart(decimals, "0");
  const trimmedDec = decStr.replace(/0+$/, "");
  return `${intPart}.${trimmedDec}`;
}
function getNetworkFromCaip2(caip2) {
  if (!caip2.startsWith("algorand:")) {
    return null;
  }
  const genesisHash = caip2.slice("algorand:".length);
  if (genesisHash === ALGORAND_MAINNET_GENESIS_HASH) {
    return "mainnet";
  }
  if (genesisHash === ALGORAND_TESTNET_GENESIS_HASH) {
    return "testnet";
  }
  return null;
}
function isAlgorandNetwork(network) {
  return network.startsWith("algorand:");
}
function isTestnetNetwork(network) {
  return network === ALGORAND_TESTNET_CAIP2;
}
function getGenesisHashFromTransaction(txn) {
  if (!txn.genesisHash) {
    throw new Error("Transaction does not have a genesis hash");
  }
  return Buffer.from(txn.genesisHash).toString("base64");
}
function validateGroupId(txns) {
  if (txns.length <= 1) {
    return true;
  }
  let expectedGroupId = null;
  for (const txnBytes of txns) {
    const txn = decodeUnsignedTxn(txnBytes);
    const groupId = txn.group ? Buffer.from(txn.group).toString("base64") : null;
    if (expectedGroupId === null) {
      expectedGroupId = groupId;
    } else if (groupId !== expectedGroupId) {
      return false;
    }
  }
  return true;
}
function getTransactionId(signedTxnBytes) {
  const signedTxn = decodeSignedTxn(signedTxnBytes);
  return signedTxn.txn.txId();
}
function hasSignature(signedTxnBytes) {
  const signedTxn = decodeSignedTxn(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

export {
  encodeTransaction,
  decodeTransaction,
  decodeSignedTransaction,
  decodeUnsignedTransaction,
  isValidAlgorandAddress,
  getSenderFromTransaction,
  convertFromTokenAmount,
  getNetworkFromCaip2,
  isAlgorandNetwork,
  isTestnetNetwork,
  getGenesisHashFromTransaction,
  validateGroupId,
  getTransactionId,
  hasSignature,
  convertToTokenAmount
};
