// @x402/avm/dist/esm/chunk-OMKK3QVS.mjs
import {
  encodeTransaction,
  isTestnetNetwork
} from "./chunk-QEY4CIDB.mjs";
import {
  USDC_CONFIG
} from "./chunk-MWOLCZUZ.mjs";

// src/exact/client/scheme.ts
import { AlgorandClient } from "@algorandfoundation/algokit-utils/algorand-client";
import {
  Transaction,
  encodeTransactionRaw,
  groupTransactions,
  makeEmptyTransactionSigner
} from "@algorandfoundation/algokit-utils/transact";
import { microAlgo } from "@algorandfoundation/algokit-utils/amount";
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   *
   * Constructs an atomic transaction group with:
   * - Optional fee payer transaction (if feePayer specified in requirements.extra)
   * - ASA transfer transaction to payTo address
   *
   * Uses TransactionComposer for automatic suggested params, group ID assignment,
   * and fee pooling. For sponsored (gasless) transactions, exact fees are calculated
   * from actual encoded transaction sizes using the protocol fee formula:
   *   fee = max(fee_per_byte × txn_size_in_bytes, min_fee)
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const { amount, asset, payTo, network, extra } = paymentRequirements;
    const algorandClient = this.getAlgorandClient(network);
    const assetId = this.getAssetId(asset, network);
    const feePayer = extra?.feePayer;
    let paymentIndex = 0;
    const emptySigner = makeEmptyTransactionSigner();
    const composer = algorandClient.newGroup();
    if (feePayer) {
      composer.addPayment({
        sender: feePayer,
        receiver: feePayer,
        amount: microAlgo(0),
        note: `x402-fee-payer-${Date.now()}`,
        signer: emptySigner
      });
      paymentIndex = 1;
    }
    composer.addAssetTransfer({
      sender: this.signer.address,
      receiver: payTo,
      assetId: BigInt(assetId),
      amount: BigInt(amount),
      staticFee: feePayer ? microAlgo(0) : void 0,
      // 0 fee when fee payer covers
      note: `x402-payment-v${x402Version}-${Date.now()}`,
      signer: emptySigner
    });
    const built = await composer.build();
    let transactions = built.transactions.map((tws) => tws.txn);
    if (feePayer) {
      const sp = await algorandClient.getSuggestedParams();
      const feePerByte = Number(sp.fee);
      const minFee = Number(sp.minFee);
      let totalGroupFee = BigInt(0);
      for (const txn of transactions) {
        const txnSize = encodeTransactionRaw(txn).length;
        const txnFee = feePerByte > 0 ? Math.max(feePerByte * txnSize, minFee) : minFee;
        totalGroupFee += BigInt(txnFee);
      }
      const ungrouped = transactions.map((txn, i) => {
        const fee = i === 0 ? totalGroupFee : BigInt(0);
        return new Transaction({ ...txn, fee, group: void 0 });
      });
      transactions = groupTransactions(ungrouped);
    }
    const encodedTxns = transactions.map((txn) => encodeTransactionRaw(txn));
    const clientIndexes = transactions.map((txn, i) => txn.sender.toString() === this.signer.address ? i : -1).filter((i) => i !== -1);
    const signedTxns = await this.signer.signTransactions(encodedTxns, clientIndexes);
    const paymentGroup = encodedTxns.map((txnBytes, i) => {
      const signedTxn = signedTxns[i];
      if (signedTxn) {
        return encodeTransaction(signedTxn);
      }
      return encodeTransaction(txnBytes);
    });
    const payload = {
      paymentGroup,
      paymentIndex
    };
    return {
      x402Version,
      payload
    };
  }
  /**
   * Creates or retrieves an AlgorandClient for the given network.
   *
   * @param network - Network identifier (CAIP-2 format)
   * @returns AlgorandClient instance
   */
  getAlgorandClient(network) {
    if (this.config?.algorandClient) {
      return this.config.algorandClient;
    }
    if (this.config?.algodUrl) {
      return AlgorandClient.fromConfig({
        algodConfig: {
          server: this.config.algodUrl,
          token: this.config.algodToken ?? ""
        }
      });
    }
    return isTestnetNetwork(network) ? AlgorandClient.testNet() : AlgorandClient.mainNet();
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - Asset identifier from requirements
   * @param network - Network identifier
   * @returns Asset ID as string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

export {
  ExactAvmScheme
};
