// @x402/avm/dist/esm/index.mjs
import {
  ExactAvmScheme
} from "./chunk-OMKK3QVS.mjs";
import {
  isExactAvmPayload
} from "./chunk-AY5WYLS5.mjs";
import {
  convertFromTokenAmount,
  convertToTokenAmount,
  decodeSignedTransaction,
  decodeTransaction,
  decodeUnsignedTransaction,
  encodeTransaction,
  getGenesisHashFromTransaction,
  getNetworkFromCaip2,
  getSenderFromTransaction,
  getTransactionId,
  hasSignature,
  isAlgorandNetwork,
  isTestnetNetwork,
  isValidAlgorandAddress,
  validateGroupId
} from "./chunk-QEY4CIDB.mjs";
import {
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS,
  MAX_REASONABLE_FEE_PER_TXN,
  USDC_CONFIG,
  USDC_DECIMALS,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  maxReasonableGroupFee
} from "./chunk-MWOLCZUZ.mjs";

// src/signer.ts
import { AlgorandClient } from "@algorandfoundation/algokit-utils/algorand-client";
import { ed25519Generator } from "@algorandfoundation/algokit-utils/crypto";
import {
  decodeTransaction as decodeTransaction2,
  generateAddressWithSigners
} from "@algorandfoundation/algokit-utils/transact";
import { waitForConfirmation } from "@algorandfoundation/algokit-utils/transaction";
var ALGOKIT_SIGNER = Symbol("algokit-signer");
function isAvmSignerWallet(wallet) {
  return typeof wallet === "object" && wallet !== null && "address" in wallet && typeof wallet.address === "string" && "signTransactions" in wallet && typeof wallet.signTransactions === "function";
}
function decodePrivateKey(privateKeyBase64) {
  const secretKey = Buffer.from(privateKeyBase64, "base64");
  if (secretKey.length !== 64) {
    throw new Error(
      "AVM private key must be a Base64-encoded 64-byte key (32-byte seed + 32-byte public key)"
    );
  }
  const seed = secretKey.subarray(0, 32);
  return ed25519Generator(seed);
}
function toClientAvmSigner(privateKeyBase64) {
  const { ed25519Pubkey, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  const algokitSigners = generateAddressWithSigners({ ed25519Pubkey, rawEd25519Signer });
  const address = algokitSigners.addr.toString();
  const signer = {
    address,
    signTransactions: async (txns, indexesToSign) => {
      return Promise.all(
        txns.map(async (txn, i) => {
          if (indexesToSign && !indexesToSign.includes(i)) return null;
          const decoded = decodeTransaction2(txn);
          const signedBytes = await algokitSigners.signer([decoded], [0]);
          return signedBytes[0];
        })
      );
    }
  };
  Object.defineProperty(signer, ALGOKIT_SIGNER, {
    value: algokitSigners,
    enumerable: false,
    writable: false
  });
  return signer;
}
function getAlgokitSigner(signer) {
  const internal = signer[ALGOKIT_SIGNER];
  if (internal && "addr" in internal && "signer" in internal) {
    return { addr: internal.addr, signer: internal.signer };
  }
  return null;
}
function isTestnet(network) {
  return network === ALGORAND_TESTNET_CAIP2;
}
function toFacilitatorAvmSigner(privateKeyBase64, config) {
  const { ed25519Pubkey, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  const algokitSigners = generateAddressWithSigners({ ed25519Pubkey, rawEd25519Signer });
  const address = algokitSigners.addr.toString();
  const getAlgorandClientForNetwork = (network) => {
    if (isTestnet(network)) {
      if (config?.testnetUrl) {
        return AlgorandClient.fromConfig({
          algodConfig: { server: config.testnetUrl, token: config.algodToken ?? "" }
        });
      }
      return AlgorandClient.testNet();
    }
    if (config?.mainnetUrl) {
      return AlgorandClient.fromConfig({
        algodConfig: { server: config.mainnetUrl, token: config.algodToken ?? "" }
      });
    }
    return AlgorandClient.mainNet();
  };
  const clientCache = /* @__PURE__ */ new Map();
  const getClient = (network) => {
    const key = isTestnet(network) ? "testnet" : "mainnet";
    let client = clientCache.get(key);
    if (!client) {
      client = getAlgorandClientForNetwork(network);
      clientCache.set(key, client);
    }
    return client;
  };
  return {
    getAddresses: () => [address],
    signTransaction: async (txn, _) => {
      const decoded = decodeTransaction2(txn);
      const signedBytes = await algokitSigners.signer([decoded], [0]);
      return signedBytes[0];
    },
    getAlgodClient: (network) => getClient(network).client.algod,
    simulateTransactions: async (txns, network) => {
      const algod = getClient(network).client.algod;
      return await algod.simulateRawTransactions(txns);
    },
    sendTransactions: async (signedTxns, network) => {
      const algod = getClient(network).client.algod;
      const response = await algod.sendRawTransaction(signedTxns);
      return response.txId;
    },
    waitForConfirmation: async (txId, network, waitRounds = 5) => {
      const algod = getClient(network).client.algod;
      return await waitForConfirmation(txId, waitRounds, algod);
    }
  };
}

// src/index.ts
import { ALGORAND_ADDRESS_LENGTH } from "@algorandfoundation/algokit-utils/common";
import { ALGORAND_MIN_TX_FEE } from "@algorandfoundation/algokit-utils/amount";
export {
  ALGOKIT_SIGNER,
  ALGORAND_ADDRESS_LENGTH,
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_MIN_TX_FEE,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS,
  ExactAvmScheme,
  MAX_REASONABLE_FEE_PER_TXN,
  USDC_CONFIG,
  USDC_DECIMALS,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  convertFromTokenAmount,
  convertToTokenAmount,
  decodeSignedTransaction,
  decodeTransaction,
  decodeUnsignedTransaction,
  encodeTransaction,
  getAlgokitSigner,
  getGenesisHashFromTransaction,
  getNetworkFromCaip2,
  getSenderFromTransaction,
  getTransactionId,
  hasSignature,
  isAlgorandNetwork,
  isAvmSignerWallet,
  isExactAvmPayload,
  isTestnetNetwork,
  isValidAlgorandAddress,
  maxReasonableGroupFee,
  toClientAvmSigner,
  toFacilitatorAvmSigner,
  validateGroupId
};
