// @algorandfoundation/algokit-utils/packages/common/src/codecs/composite/array.js
const require_codec = require('../codec.js');
const require_address = require('../primitives/address.js');
const require_bigint = require('../primitives/bigint.js');
const require_boolean = require('../primitives/boolean.js');
const require_bytes = require('../primitives/bytes.js');
const require_number = require('../primitives/number.js');
const require_string = require('../primitives/string.js');

//#region packages/common/src/codecs/composite/array.ts
/**
* Array codec - encodes each element using the item codec
*/
var ArrayCodec = class extends require_codec.Codec {
	constructor(itemCodec) {
		super();
		this.itemCodec = itemCodec;
	}
	defaultValue() {
		return [];
	}
	toEncoded(value, format) {
		return value.map((item) => this.itemCodec.encode(item, format));
	}
	fromEncoded(value, format) {
		return value.map((item) => this.itemCodec.decode(item, format));
	}
	isDefaultValue(value) {
		return value.length === 0;
	}
};
const bytesArrayCodec = new ArrayCodec(require_bytes.bytesCodec);
const addressArrayCodec = new ArrayCodec(require_address.addressCodec);
const bigIntArrayCodec = new ArrayCodec(require_bigint.bigIntCodec);
const numberArrayCodec = new ArrayCodec(require_number.numberCodec);
const booleanArrayCodec = new ArrayCodec(require_boolean.booleanCodec);
const stringArrayCodec = new ArrayCodec(require_string.stringCodec);

//#endregion
exports.ArrayCodec = ArrayCodec;
exports.addressArrayCodec = addressArrayCodec;
exports.bigIntArrayCodec = bigIntArrayCodec;
exports.booleanArrayCodec = booleanArrayCodec;
exports.bytesArrayCodec = bytesArrayCodec;
exports.numberArrayCodec = numberArrayCodec;
exports.stringArrayCodec = stringArrayCodec;
