// @algorandfoundation/algokit-utils/packages/common/src/codecs/composite/array.mjs
import { Codec } from "../codec.mjs";
import { addressCodec } from "../primitives/address.mjs";
import { bigIntCodec } from "../primitives/bigint.mjs";
import { booleanCodec } from "../primitives/boolean.mjs";
import { bytesCodec } from "../primitives/bytes.mjs";
import { numberCodec } from "../primitives/number.mjs";
import { stringCodec } from "../primitives/string.mjs";

//#region packages/common/src/codecs/composite/array.ts
/**
* Array codec - encodes each element using the item codec
*/
var ArrayCodec = class extends Codec {
	constructor(itemCodec) {
		super();
		this.itemCodec = itemCodec;
	}
	defaultValue() {
		return [];
	}
	toEncoded(value, format) {
		return value.map((item) => this.itemCodec.encode(item, format));
	}
	fromEncoded(value, format) {
		return value.map((item) => this.itemCodec.decode(item, format));
	}
	isDefaultValue(value) {
		return value.length === 0;
	}
};
const bytesArrayCodec = new ArrayCodec(bytesCodec);
const addressArrayCodec = new ArrayCodec(addressCodec);
const bigIntArrayCodec = new ArrayCodec(bigIntCodec);
const numberArrayCodec = new ArrayCodec(numberCodec);
const booleanArrayCodec = new ArrayCodec(booleanCodec);
const stringArrayCodec = new ArrayCodec(stringCodec);

//#endregion
export { ArrayCodec, addressArrayCodec, bigIntArrayCodec, booleanArrayCodec, bytesArrayCodec, numberArrayCodec, stringArrayCodec };
