// @algorandfoundation/algokit-utils/packages/common/src/codecs/composite/map.js
const require_codec = require('../codec.js');

//#region packages/common/src/codecs/composite/map.ts
/**
* Map codec - handles Maps with any key type (including Uint8Array, bigint, number)
* Depending on the encoding format, the map is encoded differently:
* - json: Supports string and bigint keys. The map is represented as an object when encoding.
*   Bigint keys are converted to/from strings (e.g., 1n becomes "1").
*   An exception is thrown upon encountering an unsupported key type.
* - msgpack: Preserves key types and is represented as a Map when encoding.
*/
var MapCodec = class extends require_codec.Codec {
	keyType;
	constructor(keyCodec, valueCodec) {
		super();
		this.keyCodec = keyCodec;
		this.valueCodec = valueCodec;
		const defaultKeyValue = this.keyCodec.defaultValue();
		this.keyType = defaultKeyValue instanceof Uint8Array ? "Uint8Array" : typeof defaultKeyValue;
	}
	defaultValue() {
		return /* @__PURE__ */ new Map();
	}
	ensureKeyIsSupported(format) {
		if (format === "msgpack") return true;
		if (this.keyType !== "string" && this.keyType !== "bigint") throw new Error(`Map key of type '${this.keyType}' is not supported in ${format} format`);
	}
	toEncoded(value, format) {
		this.ensureKeyIsSupported(format);
		const entries = [];
		for (const [k, v] of value.entries()) {
			const encodedKey = this.keyCodec.encode(k, format);
			const encodedValue = this.valueCodec.encode(v, format);
			entries.push([encodedKey, encodedValue]);
		}
		if (format === "json") {
			const obj = {};
			for (const [k, v] of entries) {
				const keyStr = typeof k === "string" ? k : String(k);
				obj[keyStr] = v;
			}
			return obj;
		}
		return new Map(entries);
	}
	fromEncoded(value, format) {
		this.ensureKeyIsSupported(format);
		const result = /* @__PURE__ */ new Map();
		let entries;
		if (value instanceof Map) entries = Array.from(value.entries());
		else entries = Object.entries(value);
		for (const [encodedKey, encodedValue] of entries) {
			let keyToDecode = encodedKey;
			if (format === "json" && this.keyType === "bigint" && typeof encodedKey === "string") keyToDecode = BigInt(encodedKey);
			const key = this.keyCodec.decode(keyToDecode, format);
			const val = this.valueCodec.decode(encodedValue, format);
			result.set(key, val);
		}
		return result;
	}
	isDefaultValue(value) {
		return value.size === 0;
	}
};

//#endregion
exports.MapCodec = MapCodec;
