// @algorandfoundation/algokit-utils/packages/common/src/codecs/composite/record.js
const require_runtime = require('../../../../../_virtual/_rolldown/runtime.js');
const require_codec = require('../codec.js');
let buffer = require("buffer");

//#region packages/common/src/codecs/composite/record.ts
/**
* Record codec - for string-keyed objects with homogeneous values
*/
var RecordCodec = class extends require_codec.Codec {
	constructor(valueCodec) {
		super();
		this.valueCodec = valueCodec;
	}
	defaultValue() {
		return {};
	}
	toEncoded(value, format) {
		const result = {};
		for (const [key, val] of Object.entries(value)) result[key] = this.valueCodec.encode(val, format);
		return result;
	}
	fromEncoded(value, format) {
		const result = {};
		if (value instanceof Map) for (const [_key, val] of value.entries()) {
			const keyType = typeof _key;
			if (keyType === "number" || keyType === "bigint") throw new Error(`RecordCodec received a non-string key of type ${keyType}`);
			const key = _key instanceof Uint8Array ? buffer.Buffer.from(_key).toString("utf-8") : String(_key);
			result[key] = this.valueCodec.decode(val, format);
		}
		else for (const [key, val] of Object.entries(value)) result[key] = this.valueCodec.decode(val, format);
		return result;
	}
	isDefaultValue(value) {
		return Object.keys(value).length === 0;
	}
};

//#endregion
exports.RecordCodec = RecordCodec;
