// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/account.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { AccountParticipationMeta } from "./account-participation.mjs";
import { ApplicationStateSchemaMeta } from "./application-state-schema.mjs";
import { ApplicationMeta } from "./application.mjs";
import { ApplicationLocalStateMeta } from "./application-local-state.mjs";
import { AssetMeta } from "./asset.mjs";
import { AssetHoldingMeta } from "./asset-holding.mjs";
//#region packages/indexer_client/src/models/account.ts
const AccountMeta = {
	name: "Account",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: false,
			codec: stringCodec
		},
		{
			name: "amount",
			wireKey: "amount",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "minBalance",
			wireKey: "min-balance",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "amountWithoutPendingRewards",
			wireKey: "amount-without-pending-rewards",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "appsLocalState",
			wireKey: "apps-local-state",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ApplicationLocalStateMeta))
		},
		{
			name: "appsTotalSchema",
			wireKey: "apps-total-schema",
			optional: true,
			codec: new ObjectModelCodec(ApplicationStateSchemaMeta)
		},
		{
			name: "appsTotalExtraPages",
			wireKey: "apps-total-extra-pages",
			optional: true,
			codec: numberCodec
		},
		{
			name: "assets",
			wireKey: "assets",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(AssetHoldingMeta))
		},
		{
			name: "createdApps",
			wireKey: "created-apps",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ApplicationMeta))
		},
		{
			name: "createdAssets",
			wireKey: "created-assets",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(AssetMeta))
		},
		{
			name: "participation",
			wireKey: "participation",
			optional: true,
			codec: new ObjectModelCodec(AccountParticipationMeta)
		},
		{
			name: "incentiveEligible",
			wireKey: "incentive-eligible",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "pendingRewards",
			wireKey: "pending-rewards",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "rewardBase",
			wireKey: "reward-base",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "rewards",
			wireKey: "rewards",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "status",
			wireKey: "status",
			optional: false,
			codec: stringCodec
		},
		{
			name: "sigType",
			wireKey: "sig-type",
			optional: true,
			codec: stringCodec
		},
		{
			name: "totalAppsOptedIn",
			wireKey: "total-apps-opted-in",
			optional: false,
			codec: numberCodec
		},
		{
			name: "totalAssetsOptedIn",
			wireKey: "total-assets-opted-in",
			optional: false,
			codec: numberCodec
		},
		{
			name: "totalBoxBytes",
			wireKey: "total-box-bytes",
			optional: false,
			codec: numberCodec
		},
		{
			name: "totalBoxes",
			wireKey: "total-boxes",
			optional: false,
			codec: numberCodec
		},
		{
			name: "totalCreatedApps",
			wireKey: "total-created-apps",
			optional: false,
			codec: numberCodec
		},
		{
			name: "totalCreatedAssets",
			wireKey: "total-created-assets",
			optional: false,
			codec: numberCodec
		},
		{
			name: "authAddr",
			wireKey: "auth-addr",
			optional: true,
			codec: addressCodec
		},
		{
			name: "lastProposed",
			wireKey: "last-proposed",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "lastHeartbeat",
			wireKey: "last-heartbeat",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "createdAtRound",
			wireKey: "created-at-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "closedAtRound",
			wireKey: "closed-at-round",
			optional: true,
			codec: bigIntCodec
		}
	]
};
//#endregion
export { AccountMeta };
