// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/accounts-response.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_account = require("./account.js");
//#region packages/indexer_client/src/models/accounts-response.ts
const AccountsResponseMeta = {
	name: "AccountsResponse",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "accounts",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_account.AccountMeta))
		},
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};
//#endregion
exports.AccountsResponseMeta = AccountsResponseMeta;
