// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/account.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_account_participation = require("./account-participation.js");
const require_application_state_schema = require("./application-state-schema.js");
const require_application = require("./application.js");
const require_application_local_state = require("./application-local-state.js");
const require_asset = require("./asset.js");
const require_asset_holding = require("./asset-holding.js");
//#region packages/indexer_client/src/models/account.ts
const AccountMeta = {
	name: "Account",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "amount",
			wireKey: "amount",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "minBalance",
			wireKey: "min-balance",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "amountWithoutPendingRewards",
			wireKey: "amount-without-pending-rewards",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "appsLocalState",
			wireKey: "apps-local-state",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_application_local_state.ApplicationLocalStateMeta))
		},
		{
			name: "appsTotalSchema",
			wireKey: "apps-total-schema",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_state_schema.ApplicationStateSchemaMeta)
		},
		{
			name: "appsTotalExtraPages",
			wireKey: "apps-total-extra-pages",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "assets",
			wireKey: "assets",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_asset_holding.AssetHoldingMeta))
		},
		{
			name: "createdApps",
			wireKey: "created-apps",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_application.ApplicationMeta))
		},
		{
			name: "createdAssets",
			wireKey: "created-assets",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_asset.AssetMeta))
		},
		{
			name: "participation",
			wireKey: "participation",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_account_participation.AccountParticipationMeta)
		},
		{
			name: "incentiveEligible",
			wireKey: "incentive-eligible",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "pendingRewards",
			wireKey: "pending-rewards",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardBase",
			wireKey: "reward-base",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewards",
			wireKey: "rewards",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "status",
			wireKey: "status",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "sigType",
			wireKey: "sig-type",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "totalAppsOptedIn",
			wireKey: "total-apps-opted-in",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalAssetsOptedIn",
			wireKey: "total-assets-opted-in",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalBoxBytes",
			wireKey: "total-box-bytes",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalBoxes",
			wireKey: "total-boxes",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalCreatedApps",
			wireKey: "total-created-apps",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalCreatedAssets",
			wireKey: "total-created-assets",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "authAddr",
			wireKey: "auth-addr",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "lastProposed",
			wireKey: "last-proposed",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastHeartbeat",
			wireKey: "last-heartbeat",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "createdAtRound",
			wireKey: "created-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closedAtRound",
			wireKey: "closed-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};
//#endregion
exports.AccountMeta = AccountMeta;
