// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { ApplicationParamsMeta } from "./application-params.mjs";
//#region packages/indexer_client/src/models/application.ts
const ApplicationMeta = {
	name: "Application",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "createdAtRound",
			wireKey: "created-at-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "deletedAtRound",
			wireKey: "deleted-at-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "params",
			wireKey: "params",
			optional: false,
			codec: new ObjectModelCodec(ApplicationParamsMeta)
		}
	]
};
//#endregion
export { ApplicationMeta };
