// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application-logs-response.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_application_log_data = require("./application-log-data.js");
//#region packages/indexer_client/src/models/application-logs-response.ts
const ApplicationLogsResponseMeta = {
	name: "ApplicationLogsResponse",
	kind: "object",
	fields: [
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "logData",
			wireKey: "log-data",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_application_log_data.ApplicationLogDataMeta))
		}
	]
};
//#endregion
exports.ApplicationLogsResponseMeta = ApplicationLogsResponseMeta;
