// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_application_params = require("./application-params.js");
//#region packages/indexer_client/src/models/application.ts
const ApplicationMeta = {
	name: "Application",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "createdAtRound",
			wireKey: "created-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "deletedAtRound",
			wireKey: "deleted-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "params",
			wireKey: "params",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_application_params.ApplicationParamsMeta)
		}
	]
};
//#endregion
exports.ApplicationMeta = ApplicationMeta;
