// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application-local-state.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_array_model = require("../../../common/src/codecs/models/array-model.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_application_state_schema = require("./application-state-schema.js");
const require_teal_key_value_store = require("./teal-key-value-store.js");
//#region packages/indexer_client/src/models/application-local-state.ts
const ApplicationLocalStateMeta = {
	name: "ApplicationLocalState",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "optedInAtRound",
			wireKey: "opted-in-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closedOutAtRound",
			wireKey: "closed-out-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "schema",
			wireKey: "schema",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_application_state_schema.ApplicationStateSchemaMeta)
		},
		{
			name: "keyValue",
			wireKey: "key-value",
			optional: true,
			codec: new require_array_model.ArrayModelCodec(require_teal_key_value_store.TealKeyValueStoreMeta)
		}
	]
};
//#endregion
exports.ApplicationLocalStateMeta = ApplicationLocalStateMeta;
