// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/asset.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_asset_params = require("./asset-params.js");
//#region packages/indexer_client/src/models/asset.ts
const AssetMeta = {
	name: "Asset",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "index",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "createdAtRound",
			wireKey: "created-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "destroyedAtRound",
			wireKey: "destroyed-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "params",
			wireKey: "params",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_asset_params.AssetParamsMeta)
		}
	]
};
//#endregion
exports.AssetMeta = AssetMeta;
