// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/asset-balances-response.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_mini_asset_holding = require("./mini-asset-holding.js");
//#region packages/indexer_client/src/models/asset-balances-response.ts
const AssetBalancesResponseMeta = {
	name: "AssetBalancesResponse",
	kind: "object",
	fields: [
		{
			name: "balances",
			wireKey: "balances",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_mini_asset_holding.MiniAssetHoldingMeta))
		},
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};
//#endregion
exports.AssetBalancesResponseMeta = AssetBalancesResponseMeta;
