// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/asset-params.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
//#region packages/indexer_client/src/models/asset-params.ts
const AssetParamsMeta = {
	name: "AssetParams",
	kind: "object",
	fields: [
		{
			name: "clawback",
			wireKey: "clawback",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "creator",
			wireKey: "creator",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "decimals",
			wireKey: "decimals",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "defaultFrozen",
			wireKey: "default-frozen",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "freeze",
			wireKey: "freeze",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "manager",
			wireKey: "manager",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "metadataHash",
			wireKey: "metadata-hash",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "name",
			wireKey: "name",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "nameB64",
			wireKey: "name-b64",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "reserve",
			wireKey: "reserve",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "total",
			wireKey: "total",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "unitName",
			wireKey: "unit-name",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "unitNameB64",
			wireKey: "unit-name-b64",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "url",
			wireKey: "url",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "urlB64",
			wireKey: "url-b64",
			optional: true,
			codec: require_bytes.bytesCodec
		}
	]
};
//#endregion
exports.AssetParamsMeta = AssetParamsMeta;
