// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/block.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_block_rewards = require("./block-rewards.js");
const require_block_upgrade_state = require("./block-upgrade-state.js");
const require_block_upgrade_vote = require("./block-upgrade-vote.js");
const require_participation_updates = require("./participation-updates.js");
const require_state_proof_tracking = require("./state-proof-tracking.js");
const require_transaction = require("./transaction.js");
//#region packages/indexer_client/src/models/block.ts
const BlockMeta = {
	name: "Block",
	kind: "object",
	fields: [
		{
			name: "proposer",
			wireKey: "proposer",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "feesCollected",
			wireKey: "fees-collected",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "bonus",
			wireKey: "bonus",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "proposerPayout",
			wireKey: "proposer-payout",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "genesisHash",
			wireKey: "genesis-hash",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "genesis-id",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "previousBlockHash",
			wireKey: "previous-block-hash",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "previousBlockHash512",
			wireKey: "previous-block-hash-512",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		},
		{
			name: "rewards",
			wireKey: "rewards",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_block_rewards.BlockRewardsMeta)
		},
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "seed",
			wireKey: "seed",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "stateProofTracking",
			wireKey: "state-proof-tracking",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_state_proof_tracking.StateProofTrackingMeta))
		},
		{
			name: "timestamp",
			wireKey: "timestamp",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "transactions",
			wireKey: "transactions",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_transaction.TransactionMeta))
		},
		{
			name: "transactionsRoot",
			wireKey: "transactions-root",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "transactionsRootSha256",
			wireKey: "transactions-root-sha256",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "transactionsRootSha512",
			wireKey: "transactions-root-sha512",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		},
		{
			name: "txnCounter",
			wireKey: "txn-counter",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "upgradeState",
			wireKey: "upgrade-state",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_block_upgrade_state.BlockUpgradeStateMeta)
		},
		{
			name: "upgradeVote",
			wireKey: "upgrade-vote",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_block_upgrade_vote.BlockUpgradeVoteMeta)
		},
		{
			name: "participationUpdates",
			wireKey: "participation-updates",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_participation_updates.ParticipationUpdatesMeta)
		}
	]
};
//#endregion
exports.BlockMeta = BlockMeta;
