// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/boxes-response.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_box_descriptor = require("./box-descriptor.js");
//#region packages/indexer_client/src/models/boxes-response.ts
const BoxesResponseMeta = {
	name: "BoxesResponse",
	kind: "object",
	fields: [
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "boxes",
			wireKey: "boxes",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_box_descriptor.BoxDescriptorMeta))
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};
//#endregion
exports.BoxesResponseMeta = BoxesResponseMeta;
