// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/health-check.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_record = require("../../../common/src/codecs/composite/record.js");
const require_unknown = require("../../../common/src/codecs/primitives/unknown.js");
//#region packages/indexer_client/src/models/health-check.ts
const HealthCheckMeta = {
	name: "HealthCheck",
	kind: "object",
	fields: [
		{
			name: "version",
			wireKey: "version",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "data",
			wireKey: "data",
			optional: true,
			codec: new require_record.RecordCodec(require_unknown.unknownCodec)
		},
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "isMigrating",
			wireKey: "is-migrating",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "dbAvailable",
			wireKey: "db-available",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "message",
			wireKey: "message",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "errors",
			wireKey: "errors",
			optional: true,
			codec: require_array.stringArrayCodec
		}
	]
};
//#endregion
exports.HealthCheckMeta = HealthCheckMeta;
