// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/health-check.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { stringArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { RecordCodec } from "../../../common/src/codecs/composite/record.mjs";
import { unknownCodec } from "../../../common/src/codecs/primitives/unknown.mjs";
//#region packages/indexer_client/src/models/health-check.ts
const HealthCheckMeta = {
	name: "HealthCheck",
	kind: "object",
	fields: [
		{
			name: "version",
			wireKey: "version",
			optional: false,
			codec: stringCodec
		},
		{
			name: "data",
			wireKey: "data",
			optional: true,
			codec: new RecordCodec(unknownCodec)
		},
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "isMigrating",
			wireKey: "is-migrating",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "dbAvailable",
			wireKey: "db-available",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "message",
			wireKey: "message",
			optional: false,
			codec: stringCodec
		},
		{
			name: "errors",
			wireKey: "errors",
			optional: true,
			codec: stringArrayCodec
		}
	]
};
//#endregion
export { HealthCheckMeta };
