// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/mini-asset-holding.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
//#region packages/indexer_client/src/models/mini-asset-holding.ts
const MiniAssetHoldingMeta = {
	name: "MiniAssetHolding",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: false,
			codec: stringCodec
		},
		{
			name: "amount",
			wireKey: "amount",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "isFrozen",
			wireKey: "is-frozen",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "optedInAtRound",
			wireKey: "opted-in-at-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "optedOutAtRound",
			wireKey: "opted-out-at-round",
			optional: true,
			codec: bigIntCodec
		}
	]
};
//#endregion
export { MiniAssetHoldingMeta };
