// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/state-proof-fields.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_merkle_array_proof = require("./merkle-array-proof.js");
const require_state_proof_reveal = require("./state-proof-reveal.js");
//#region packages/indexer_client/src/models/state-proof-fields.ts
const StateProofFieldsMeta = {
	name: "StateProofFields",
	kind: "object",
	fields: [
		{
			name: "sigCommit",
			wireKey: "sig-commit",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "signedWeight",
			wireKey: "signed-weight",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "sigProofs",
			wireKey: "sig-proofs",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_merkle_array_proof.MerkleArrayProofMeta)
		},
		{
			name: "partProofs",
			wireKey: "part-proofs",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_merkle_array_proof.MerkleArrayProofMeta)
		},
		{
			name: "saltVersion",
			wireKey: "salt-version",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "reveals",
			wireKey: "reveals",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_state_proof_reveal.StateProofRevealMeta))
		},
		{
			name: "positionsToReveal",
			wireKey: "positions-to-reveal",
			optional: true,
			codec: require_array.bigIntArrayCodec
		}
	]
};
//#endregion
exports.StateProofFieldsMeta = StateProofFieldsMeta;
