// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-signature-logicsig.js
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_transaction_signature_multisig = require("./transaction-signature-multisig.js");
//#region packages/indexer_client/src/models/transaction-signature-logicsig.ts
const TransactionSignatureLogicsigMeta = {
	name: "TransactionSignatureLogicsig",
	kind: "object",
	fields: [
		{
			name: "args",
			wireKey: "args",
			optional: true,
			codec: require_array.bytesArrayCodec
		},
		{
			name: "logic",
			wireKey: "logic",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "multisigSignature",
			wireKey: "multisig-signature",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_signature_multisig.TransactionSignatureMultisigMeta)
		},
		{
			name: "logicMultisigSignature",
			wireKey: "logic-multisig-signature",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_signature_multisig.TransactionSignatureMultisigMeta)
		},
		{
			name: "signature",
			wireKey: "signature",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		}
	]
};
//#endregion
exports.TransactionSignatureLogicsigMeta = TransactionSignatureLogicsigMeta;
