// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-heartbeat.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_hb_proof_fields = require("./hb-proof-fields.js");
//#region packages/indexer_client/src/models/transaction-heartbeat.ts
const TransactionHeartbeatMeta = {
	name: "TransactionHeartbeat",
	kind: "object",
	fields: [
		{
			name: "hbAddress",
			wireKey: "hb-address",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "hbProof",
			wireKey: "hb-proof",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_hb_proof_fields.HbProofFieldsMeta)
		},
		{
			name: "hbSeed",
			wireKey: "hb-seed",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "hbVoteId",
			wireKey: "hb-vote-id",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "hbKeyDilution",
			wireKey: "hb-key-dilution",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};
//#endregion
exports.TransactionHeartbeatMeta = TransactionHeartbeatMeta;
