// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_array_model = require("../../../common/src/codecs/models/array-model.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_state_delta = require("./state-delta.js");
const require_account_state_delta = require("./account-state-delta.js");
const require_transaction_application = require("./transaction-application.js");
const require_transaction_asset_config = require("./transaction-asset-config.js");
const require_transaction_asset_freeze = require("./transaction-asset-freeze.js");
const require_transaction_asset_transfer = require("./transaction-asset-transfer.js");
const require_transaction_heartbeat = require("./transaction-heartbeat.js");
const require_transaction_keyreg = require("./transaction-keyreg.js");
const require_transaction_payment = require("./transaction-payment.js");
const require_transaction_signature = require("./transaction-signature.js");
const require_transaction_state_proof = require("./transaction-state-proof.js");
//#region packages/indexer_client/src/models/transaction.ts
const TransactionMeta = {
	name: "Transaction",
	kind: "object",
	fields: [
		{
			name: "applicationTransaction",
			wireKey: "application-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_application.TransactionApplicationMeta)
		},
		{
			name: "assetConfigTransaction",
			wireKey: "asset-config-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_asset_config.TransactionAssetConfigMeta)
		},
		{
			name: "assetFreezeTransaction",
			wireKey: "asset-freeze-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_asset_freeze.TransactionAssetFreezeMeta)
		},
		{
			name: "assetTransferTransaction",
			wireKey: "asset-transfer-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_asset_transfer.TransactionAssetTransferMeta)
		},
		{
			name: "stateProofTransaction",
			wireKey: "state-proof-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_state_proof.TransactionStateProofMeta)
		},
		{
			name: "heartbeatTransaction",
			wireKey: "heartbeat-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_heartbeat.TransactionHeartbeatMeta)
		},
		{
			name: "authAddr",
			wireKey: "auth-addr",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "closeRewards",
			wireKey: "close-rewards",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closingAmount",
			wireKey: "closing-amount",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "confirmedRound",
			wireKey: "confirmed-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "createdAppId",
			wireKey: "created-application-index",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "createdAssetId",
			wireKey: "created-asset-index",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "fee",
			wireKey: "fee",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "firstValid",
			wireKey: "first-valid",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "genesisHash",
			wireKey: "genesis-hash",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "genesis-id",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "group",
			wireKey: "group",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "id",
			wireKey: "id",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "intraRoundOffset",
			wireKey: "intra-round-offset",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "keyregTransaction",
			wireKey: "keyreg-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_keyreg.TransactionKeyregMeta)
		},
		{
			name: "lastValid",
			wireKey: "last-valid",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lease",
			wireKey: "lease",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "note",
			wireKey: "note",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "paymentTransaction",
			wireKey: "payment-transaction",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_payment.TransactionPaymentMeta)
		},
		{
			name: "receiverRewards",
			wireKey: "receiver-rewards",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rekeyTo",
			wireKey: "rekey-to",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "roundTime",
			wireKey: "round-time",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "sender",
			wireKey: "sender",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "senderRewards",
			wireKey: "sender-rewards",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "signature",
			wireKey: "signature",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_signature.TransactionSignatureMeta)
		},
		{
			name: "txType",
			wireKey: "tx-type",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "localStateDelta",
			wireKey: "local-state-delta",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_account_state_delta.AccountStateDeltaMeta))
		},
		{
			name: "globalStateDelta",
			wireKey: "global-state-delta",
			optional: true,
			codec: new require_array_model.ArrayModelCodec(require_state_delta.StateDeltaMeta)
		},
		{
			name: "logs",
			wireKey: "logs",
			optional: true,
			codec: require_array.bytesArrayCodec
		},
		{
			name: "innerTxns",
			wireKey: "inner-txns",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(() => TransactionMeta))
		}
	]
};
//#endregion
exports.TransactionMeta = TransactionMeta;
