// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-signature-multisig.js
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_transaction_signature_multisig_subsignature = require("./transaction-signature-multisig-subsignature.js");
//#region packages/indexer_client/src/models/transaction-signature-multisig.ts
const TransactionSignatureMultisigMeta = {
	name: "TransactionSignatureMultisig",
	kind: "object",
	fields: [
		{
			name: "subsignature",
			wireKey: "subsignature",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_transaction_signature_multisig_subsignature.TransactionSignatureMultisigSubsignatureMeta))
		},
		{
			name: "threshold",
			wireKey: "threshold",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "version",
			wireKey: "version",
			optional: true,
			codec: require_number.numberCodec
		}
	]
};
//#endregion
exports.TransactionSignatureMultisigMeta = TransactionSignatureMultisigMeta;
