// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-asset-transfer.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
//#region packages/indexer_client/src/models/transaction-asset-transfer.ts
const TransactionAssetTransferMeta = {
	name: "TransactionAssetTransfer",
	kind: "object",
	fields: [
		{
			name: "amount",
			wireKey: "amount",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "assetId",
			wireKey: "asset-id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closeAmount",
			wireKey: "close-amount",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closeTo",
			wireKey: "close-to",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "receiver",
			wireKey: "receiver",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "sender",
			wireKey: "sender",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};
//#endregion
exports.TransactionAssetTransferMeta = TransactionAssetTransferMeta;
