// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-heartbeat.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes32Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { HbProofFieldsMeta } from "./hb-proof-fields.mjs";
//#region packages/indexer_client/src/models/transaction-heartbeat.ts
const TransactionHeartbeatMeta = {
	name: "TransactionHeartbeat",
	kind: "object",
	fields: [
		{
			name: "hbAddress",
			wireKey: "hb-address",
			optional: false,
			codec: stringCodec
		},
		{
			name: "hbProof",
			wireKey: "hb-proof",
			optional: false,
			codec: new ObjectModelCodec(HbProofFieldsMeta)
		},
		{
			name: "hbSeed",
			wireKey: "hb-seed",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "hbVoteId",
			wireKey: "hb-vote-id",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "hbKeyDilution",
			wireKey: "hb-key-dilution",
			optional: false,
			codec: bigIntCodec
		}
	]
};
//#endregion
export { TransactionHeartbeatMeta };
