// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transactions-response.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_transaction = require("./transaction.js");
//#region packages/indexer_client/src/models/transactions-response.ts
const TransactionsResponseMeta = {
	name: "TransactionsResponse",
	kind: "object",
	fields: [
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "transactions",
			wireKey: "transactions",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_transaction.TransactionMeta))
		}
	]
};
//#endregion
exports.TransactionsResponseMeta = TransactionsResponseMeta;
