// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-keyreg.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { fixedBytes32Codec, fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
//#region packages/indexer_client/src/models/transaction-keyreg.ts
const TransactionKeyregMeta = {
	name: "TransactionKeyreg",
	kind: "object",
	fields: [
		{
			name: "nonParticipation",
			wireKey: "non-participation",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "selectionParticipationKey",
			wireKey: "selection-participation-key",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "voteFirstValid",
			wireKey: "vote-first-valid",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "vote-key-dilution",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteLastValid",
			wireKey: "vote-last-valid",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteParticipationKey",
			wireKey: "vote-participation-key",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "stateProofKey",
			wireKey: "state-proof-key",
			optional: true,
			codec: fixedBytes64Codec
		}
	]
};
//#endregion
export { TransactionKeyregMeta };
