// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-signature-logicsig.mjs
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { TransactionSignatureMultisigMeta } from "./transaction-signature-multisig.mjs";
//#region packages/indexer_client/src/models/transaction-signature-logicsig.ts
const TransactionSignatureLogicsigMeta = {
	name: "TransactionSignatureLogicsig",
	kind: "object",
	fields: [
		{
			name: "args",
			wireKey: "args",
			optional: true,
			codec: bytesArrayCodec
		},
		{
			name: "logic",
			wireKey: "logic",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "multisigSignature",
			wireKey: "multisig-signature",
			optional: true,
			codec: new ObjectModelCodec(TransactionSignatureMultisigMeta)
		},
		{
			name: "logicMultisigSignature",
			wireKey: "logic-multisig-signature",
			optional: true,
			codec: new ObjectModelCodec(TransactionSignatureMultisigMeta)
		},
		{
			name: "signature",
			wireKey: "signature",
			optional: true,
			codec: fixedBytes64Codec
		}
	]
};
//#endregion
export { TransactionSignatureLogicsigMeta };
