// @algorandfoundation/algokit-utils/packages/kmd_client/src/models/sign-program-multisig-request.js
const require_address = require('../../../common/src/codecs/primitives/address.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_multisig_sig = require('./multisig-sig.js');

//#region packages/kmd_client/src/models/sign-program-multisig-request.ts
const SignProgramMultisigRequestMeta = {
	name: "SignProgramMultisigRequest",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "program",
			wireKey: "data",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "partialMultisig",
			wireKey: "partial_multisig",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_multisig_sig.MultisigSigMeta)
		},
		{
			name: "publicKey",
			wireKey: "public_key",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "useLegacyMsig",
			wireKey: "use_legacy_msig",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "walletHandleToken",
			wireKey: "wallet_handle_token",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "walletPassword",
			wireKey: "wallet_password",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};

//#endregion
exports.SignProgramMultisigRequestMeta = SignProgramMultisigRequestMeta;
