// @x402/avm/dist/esm/exact/server/index.mjs
import {
  USDC_CONFIG,
  USDC_DECIMALS
} from "../../chunk-MWOLCZUZ.mjs";

// src/exact/server/scheme.ts
import { convertToTokenAmount, numberToDecimalString } from "@x402/core/utils";
var ExactAvmScheme = class {
  constructor() {
    this.scheme = "exact";
    this.moneyParsers = [];
  }
  /**
   * Register a custom money parser in the parser chain.
   * Multiple parsers can be registered - they will be tried in registration order.
   * Each parser receives a decimal amount (e.g., 1.50 for $1.50).
   * If a parser returns null, the next parser in the chain will be tried.
   * The default parser is always the final fallback.
   *
   * @param parser - Custom function to convert amount to AssetAmount (or null to skip)
   * @returns The server instance for chaining
   *
   * @example
   * ```typescript
   * avmServer.registerMoneyParser(async (amount, network) => {
   *   // Custom conversion logic for non-USDC assets
   *   if (amount > 100) {
   *     return { amount: (amount * 1e6).toString(), asset: "12345678" };
   *   }
   *   return null; // Use next parser
   * });
   * ```
   */
  registerMoneyParser(parser) {
    this.moneyParsers.push(parser);
    return this;
  }
  /**
   * Parses a price into an asset amount.
   * If price is already an AssetAmount, returns it directly.
   * If price is Money (string | number), parses to decimal and tries custom parsers.
   * Falls back to default conversion if all custom parsers return null.
   *
   * @param price - The price to parse
   * @param network - The network to use
   * @returns Promise that resolves to the parsed asset amount
   */
  async parsePrice(price, network) {
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!price.asset) {
        throw new Error(`Asset ID must be specified for AssetAmount on network ${network}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {}
      };
    }
    const amount = this.parseMoneyToDecimal(price);
    for (const parser of this.moneyParsers) {
      const result = await parser(amount, network);
      if (result !== null) {
        return result;
      }
    }
    return this.defaultMoneyConversion(amount, network);
  }
  /**
   * Build payment requirements for this scheme/network combination
   *
   * @param paymentRequirements - The base payment requirements
   * @param supportedKind - The supported kind from facilitator (contains extra data like feePayer)
   * @param supportedKind.x402Version - The x402 version
   * @param supportedKind.scheme - The logical payment scheme
   * @param supportedKind.network - The network identifier in CAIP-2 format
   * @param supportedKind.extra - Optional extra metadata (e.g., feePayer address)
   * @param extensionKeys - Extension keys supported by the facilitator
   * @returns Payment requirements ready to be sent to clients
   */
  enhancePaymentRequirements(paymentRequirements, supportedKind, extensionKeys) {
    void extensionKeys;
    const usdcConfig = USDC_CONFIG[supportedKind.network];
    const decimals = usdcConfig?.decimals ?? USDC_DECIMALS;
    const enhanced = {
      ...paymentRequirements,
      extra: {
        ...paymentRequirements.extra,
        decimals
      }
    };
    if (supportedKind.extra?.feePayer) {
      enhanced.extra = {
        ...enhanced.extra,
        feePayer: supportedKind.extra.feePayer
      };
    }
    return Promise.resolve(enhanced);
  }
  /**
   * Parse Money (string | number) to a decimal number.
   * Handles formats like "$1.50", "1.50", 1.50, etc.
   *
   * @param money - The money value to parse
   * @returns Decimal number
   */
  parseMoneyToDecimal(money) {
    if (typeof money === "number") {
      return money;
    }
    const cleanMoney = money.replace(/^\$/, "").trim();
    const amount = parseFloat(cleanMoney);
    if (isNaN(amount)) {
      throw new Error(`Invalid money format: ${money}`);
    }
    return amount;
  }
  /**
   * Default money conversion implementation.
   * Converts decimal amount to the default stablecoin (USDC) on the specified network.
   *
   * @param amount - The decimal amount (e.g., 1.50)
   * @param network - The network to use
   * @returns The parsed asset amount in USDC
   */
  defaultMoneyConversion(amount, network) {
    const assetInfo = this.getDefaultAsset(network);
    const tokenAmount = convertToTokenAmount(numberToDecimalString(amount), assetInfo.decimals);
    return {
      amount: tokenAmount,
      asset: assetInfo.asaId
    };
  }
  /**
   * Get the default asset info for a network (USDC)
   *
   * @param network - The network to get asset info for
   * @returns The asset information including ASA ID, name, and decimals
   */
  getDefaultAsset(network) {
    const assetInfo = USDC_CONFIG[network];
    if (!assetInfo) {
      throw new Error(`No default asset configured for network ${network}`);
    }
    return assetInfo;
  }
};
export {
  ExactAvmScheme
};
