// @x402/avm/dist/cjs/exact/facilitator/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/facilitator/index.ts
var facilitator_exports = {};
__export(facilitator_exports, {
  Errors: () => errors_exports,
  ExactAvmScheme: () => ExactAvmScheme
});
module.exports = __toCommonJS(facilitator_exports);

// src/exact/facilitator/scheme.ts
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_crypto = require("@algorandfoundation/algokit-utils/crypto");

// src/types.ts
function isExactAvmPayload(payload) {
  return typeof payload === "object" && payload !== null && "paymentGroup" in payload && "paymentIndex" in payload && Array.isArray(payload.paymentGroup) && typeof payload.paymentIndex === "number" && payload.paymentGroup.every((item) => typeof item === "string");
}

// src/exact/facilitator/scheme.ts
var import_common3 = require("@algorandfoundation/algokit-utils/common");

// src/utils.ts
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var MAX_REASONABLE_FEE_PER_TXN = 5e3;
function maxReasonableGroupFee(groupSize) {
  return MAX_REASONABLE_FEE_PER_TXN * groupSize;
}

// src/utils.ts
var import_utils = require("@x402/core/utils");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function hasSignature(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

// src/exact/facilitator/errors.ts
var errors_exports = {};
__export(errors_exports, {
  ErrAmountMismatch: () => ErrAmountMismatch,
  ErrAssetMismatch: () => ErrAssetMismatch,
  ErrConfirmationFailed: () => ErrConfirmationFailed,
  ErrFacilitatorTransferring: () => ErrFacilitatorTransferring,
  ErrFeeTooHigh: () => ErrFeeTooHigh,
  ErrGroupSizeExceeded: () => ErrGroupSizeExceeded,
  ErrInvalidFeePayer: () => ErrInvalidFeePayer,
  ErrInvalidGroupId: () => ErrInvalidGroupId,
  ErrInvalidPayload: () => ErrInvalidPayload,
  ErrInvalidPaymentIndex: () => ErrInvalidPaymentIndex,
  ErrInvalidScheme: () => ErrInvalidScheme,
  ErrInvalidSignature: () => ErrInvalidSignature,
  ErrInvalidTransaction: () => ErrInvalidTransaction,
  ErrInvalidVersion: () => ErrInvalidVersion,
  ErrNetworkMismatch: () => ErrNetworkMismatch,
  ErrNotAssetTransfer: () => ErrNotAssetTransfer,
  ErrPaymentNotSigned: () => ErrPaymentNotSigned,
  ErrReceiverMismatch: () => ErrReceiverMismatch,
  ErrSettleFailed: () => ErrSettleFailed,
  ErrSimulationFailed: () => ErrSimulationFailed,
  ErrUnsignedNonFacilitator: () => ErrUnsignedNonFacilitator
});
var ErrInvalidScheme = "invalid_exact_avm_scheme";
var ErrNetworkMismatch = "invalid_exact_avm_network_mismatch";
var ErrInvalidVersion = "invalid_exact_avm_invalid_version";
var ErrInvalidPayload = "invalid_exact_avm_payload";
var ErrGroupSizeExceeded = "invalid_exact_avm_group_size_exceeded";
var ErrInvalidPaymentIndex = "invalid_exact_avm_payment_index";
var ErrInvalidTransaction = "invalid_exact_avm_invalid_transaction";
var ErrInvalidGroupId = "invalid_exact_avm_invalid_group_id";
var ErrNotAssetTransfer = "invalid_exact_avm_not_asset_transfer";
var ErrAmountMismatch = "invalid_exact_avm_amount_mismatch";
var ErrReceiverMismatch = "invalid_exact_avm_receiver_mismatch";
var ErrAssetMismatch = "invalid_exact_avm_asset_mismatch";
var ErrInvalidFeePayer = "invalid_exact_avm_invalid_fee_payer";
var ErrFeeTooHigh = "invalid_exact_avm_fee_too_high";
var ErrPaymentNotSigned = "invalid_exact_avm_payment_not_signed";
var ErrInvalidSignature = "invalid_exact_avm_invalid_signature";
var ErrSimulationFailed = "invalid_exact_avm_simulation_failed";
var ErrFacilitatorTransferring = "invalid_exact_avm_facilitator_transferring";
var ErrUnsignedNonFacilitator = "invalid_exact_avm_unsigned_non_facilitator";
var ErrSettleFailed = "invalid_exact_avm_settlement_failed";
var ErrConfirmationFailed = "invalid_exact_avm_confirmation_failed";

// src/exact/facilitator/scheme.ts
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme facilitator instance.
   *
   * @param signer - The AVM signer for facilitator operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "algorand:*";
  }
  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   * For AVM, returns the feePayer address for gasless transactions.
   *
   * @param _ - The network identifier (unused, feePayer is network-agnostic)
   * @returns Extra data with feePayer address
   */
  getExtra(_) {
    const addresses = this.signer.getAddresses();
    if (addresses.length === 0) {
      return void 0;
    }
    const randomIndex = Math.floor(Math.random() * addresses.length);
    return { feePayer: addresses[randomIndex] };
  }
  /**
   * Get signer addresses used by this facilitator.
   * Returns all addresses this facilitator can use for signing fee payer transactions.
   *
   * @param _ - The network identifier (unused, addresses are network-agnostic)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a payment payload.
   *
   * Verification steps:
   * 1. Validate x402Version, scheme, and network
   * 2. Validate payload format and structure
   * 3. Check group size does not exceed maximum (16)
   * 4. Decode and validate transaction group
   * 5. Verify payment transaction (amount, receiver, asset)
   * 6. Prepare signed group (verify fee payer safety + sign)
   * 7. Simulate transaction group
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements) {
    try {
      if (payload.x402Version !== 2) {
        return {
          isValid: false,
          invalidReason: ErrInvalidVersion,
          invalidMessage: `Expected x402Version 2, got ${payload.x402Version}`
        };
      }
      if (payload.accepted.scheme !== "exact" || requirements.scheme !== "exact") {
        return {
          isValid: false,
          invalidReason: ErrInvalidScheme,
          invalidMessage: `Expected scheme "exact", got payload="${payload.accepted.scheme}" requirements="${requirements.scheme}"`
        };
      }
      if (payload.accepted.network !== requirements.network) {
        return {
          isValid: false,
          invalidReason: ErrNetworkMismatch,
          invalidMessage: `Network mismatch: payload="${payload.accepted.network}" requirements="${requirements.network}"`
        };
      }
      const rawPayload = payload.payload;
      if (!isExactAvmPayload(rawPayload)) {
        return {
          isValid: false,
          invalidReason: ErrInvalidPayload,
          invalidMessage: "Payload does not match ExactAvmPayloadV2 format"
        };
      }
      const { paymentGroup, paymentIndex } = rawPayload;
      if (paymentGroup.length > import_common3.MAX_TRANSACTION_GROUP_SIZE) {
        return {
          isValid: false,
          invalidReason: ErrGroupSizeExceeded,
          invalidMessage: `Transaction group has ${paymentGroup.length} transactions, maximum is ${import_common3.MAX_TRANSACTION_GROUP_SIZE}`
        };
      }
      if (paymentIndex < 0 || paymentIndex >= paymentGroup.length) {
        return {
          isValid: false,
          invalidReason: ErrInvalidPaymentIndex,
          invalidMessage: `Payment index ${paymentIndex} out of bounds for group of ${paymentGroup.length}`
        };
      }
      const facilitatorAddresses = this.signer.getAddresses();
      const decoded = this.decodeTransactionGroup(paymentGroup, facilitatorAddresses);
      if ("error" in decoded) return decoded.error;
      const paymentTxn = decoded.txns[paymentIndex].txn;
      const payer = paymentTxn.sender.toString();
      if (facilitatorAddresses.includes(payer)) {
        return {
          isValid: false,
          invalidReason: ErrFacilitatorTransferring,
          invalidMessage: "Facilitator signer cannot be the payment sender"
        };
      }
      const paymentCheck = await this.verifyPaymentTransaction(
        decoded.txns[paymentIndex],
        requirements,
        paymentGroup[paymentIndex]
      );
      if (!paymentCheck.isValid) return paymentCheck;
      const prepared = await this.prepareSignedGroup(decoded.txns, paymentGroup);
      if ("error" in prepared) return prepared.error;
      const simResult = await this.simulateTransactionGroup(
        prepared.signedTxns,
        requirements.network
      );
      if (!simResult.isValid) return simResult;
      return { isValid: true, payer };
    } catch (error) {
      return {
        isValid: false,
        invalidReason: ErrInvalidPayload,
        invalidMessage: `Unexpected error: ${error instanceof Error ? error.message : "Unknown"}`
      };
    }
  }
  /**
   * Settles a payment by submitting the transaction group.
   *
   * Settlement steps:
   * 1. Verify the payment first
   * 2. Decode and sign fee payer transactions (reuses shared decode/sign logic)
   * 3. Submit transaction group
   * 4. Wait for on-chain confirmation
   * 5. Return transaction ID
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements) {
    const verification = await this.verify(payload, requirements);
    if (!verification.isValid) {
      return {
        success: false,
        errorReason: verification.invalidReason,
        errorMessage: verification.invalidMessage,
        transaction: "",
        network: requirements.network,
        payer: verification.payer
      };
    }
    const avmPayload = payload.payload;
    const { paymentGroup, paymentIndex } = avmPayload;
    const facilitatorAddresses = this.signer.getAddresses();
    const decoded = this.decodeTransactionGroup(paymentGroup, facilitatorAddresses);
    if ("error" in decoded) {
      return {
        success: false,
        errorReason: ErrSettleFailed,
        errorMessage: decoded.error.invalidMessage ?? decoded.error.invalidReason,
        transaction: "",
        network: requirements.network,
        payer: verification.payer
      };
    }
    const prepared = await this.prepareSignedGroup(decoded.txns, paymentGroup);
    if ("error" in prepared) {
      return {
        success: false,
        errorReason: ErrSettleFailed,
        errorMessage: prepared.error.invalidMessage ?? prepared.error.invalidReason,
        transaction: "",
        network: requirements.network,
        payer: verification.payer
      };
    }
    const paymentTxnBytes = prepared.signedTxns[paymentIndex];
    const paymentStxn = (0, import_transact2.decodeSignedTransaction)(paymentTxnBytes);
    const paymentTxId = paymentStxn.txn.txId();
    try {
      await this.signer.sendTransactions(prepared.signedTxns, requirements.network);
    } catch (error) {
      return {
        success: false,
        errorReason: ErrSettleFailed,
        errorMessage: `Failed to submit transaction: ${error instanceof Error ? error.message : "Unknown error"}`,
        transaction: paymentTxId,
        network: requirements.network,
        payer: verification.payer
      };
    }
    try {
      await this.signer.waitForConfirmation(paymentTxId, requirements.network, 10);
    } catch (error) {
      return {
        success: false,
        errorReason: ErrConfirmationFailed,
        errorMessage: `Transaction submitted but confirmation failed: ${error instanceof Error ? error.message : "Unknown error"}`,
        transaction: paymentTxId,
        network: requirements.network,
        payer: verification.payer
      };
    }
    return {
      success: true,
      transaction: paymentTxId,
      network: requirements.network,
      payer: verification.payer
    };
  }
  /**
   * Decodes all transactions in the group and validates structure.
   *
   * - Signed transactions are decoded as-is
   * - Unsigned transactions are only accepted from facilitator addresses (fee payers)
   * - Verifies group ID consistency across all transactions
   *
   * @param paymentGroup - Array of base64-encoded transaction strings
   * @param facilitatorAddresses - Addresses controlled by this facilitator
   * @returns Decoded transactions or an error response
   */
  decodeTransactionGroup(paymentGroup, facilitatorAddresses) {
    const txns = [];
    for (let i = 0; i < paymentGroup.length; i++) {
      try {
        const bytes = decodeTransaction(paymentGroup[i]);
        try {
          const stxn = (0, import_transact2.decodeSignedTransaction)(bytes);
          if (!stxn.txn.type || stxn.txn.type === "unknown") {
            throw new Error("Invalid signed transaction: missing type");
          }
          txns.push(stxn);
        } catch {
          const unsignedTxn = (0, import_transact2.decodeTransaction)(bytes);
          const sender = unsignedTxn.sender.toString();
          if (!facilitatorAddresses.includes(sender)) {
            return {
              error: {
                isValid: false,
                invalidReason: ErrUnsignedNonFacilitator,
                invalidMessage: `Unsigned transaction at index ${i} from ${sender} is not a facilitator address`
              }
            };
          }
          const encodedForSimulate = (0, import_transact2.encodeSignedTransaction)({ txn: unsignedTxn });
          txns.push((0, import_transact2.decodeSignedTransaction)(encodedForSimulate));
        }
      } catch {
        return {
          error: {
            isValid: false,
            invalidReason: ErrInvalidTransaction,
            invalidMessage: `Failed to decode transaction at index ${i}`
          }
        };
      }
    }
    if (txns.length > 1) {
      const firstGroup = txns[0].txn.group;
      const firstGroupId = firstGroup ? Buffer.from(firstGroup).toString("base64") : null;
      for (let i = 1; i < txns.length; i++) {
        const group = txns[i].txn.group;
        const groupId = group ? Buffer.from(group).toString("base64") : null;
        if (groupId !== firstGroupId) {
          return {
            error: {
              isValid: false,
              invalidReason: ErrInvalidGroupId,
              invalidMessage: "Transactions have inconsistent group IDs"
            }
          };
        }
      }
    }
    return { txns };
  }
  /**
   * Verifies fee payer transactions and signs them, returning the assembled group
   * ready for simulation or submission.
   *
   * @param decodedTxns - Decoded signed transaction objects
   * @param paymentGroup - Original base64-encoded transaction strings
   * @returns Signed transaction bytes or an error response
   */
  async prepareSignedGroup(decodedTxns, paymentGroup) {
    const facilitatorAddresses = this.signer.getAddresses();
    const signedTxns = [];
    for (let i = 0; i < decodedTxns.length; i++) {
      const txn = decodedTxns[i].txn;
      const sender = txn.sender.toString();
      if (facilitatorAddresses.includes(sender)) {
        const feeCheck = this.verifyFeePayerTransaction(txn, decodedTxns.length);
        if (!feeCheck.isValid) return { error: feeCheck };
        try {
          const signedTxn = await this.signer.signTransaction((0, import_transact2.encodeTransactionRaw)(txn), sender);
          signedTxns.push(signedTxn);
        } catch (error) {
          return {
            error: {
              isValid: false,
              invalidReason: ErrInvalidFeePayer,
              invalidMessage: `Failed to sign fee payer transaction: ${error instanceof Error ? error.message : "Unknown error"}`
            }
          };
        }
      } else {
        signedTxns.push(decodeTransaction(paymentGroup[i]));
      }
    }
    return { signedTxns };
  }
  /**
   * Simulates the transaction group and returns the verification result.
   *
   * @param signedTxns - Signed transaction bytes to simulate
   * @param network - Target network for simulation
   * @returns Verification result from simulation
   */
  async simulateTransactionGroup(signedTxns, network) {
    try {
      const simResult = await this.signer.simulateTransactions(signedTxns, network);
      if (simResult.txnGroups?.[0]?.failureMessage) {
        return {
          isValid: false,
          invalidReason: ErrSimulationFailed,
          invalidMessage: simResult.txnGroups[0].failureMessage
        };
      }
      return { isValid: true };
    } catch (error) {
      return {
        isValid: false,
        invalidReason: ErrSimulationFailed,
        invalidMessage: error instanceof Error ? error.message : "Unknown error"
      };
    }
  }
  /**
   * Verifies the payment transaction matches requirements
   *
   * @param stxn - The signed payment transaction
   * @param requirements - Payment requirements to verify against
   * @param encodedTxn - Base64-encoded transaction for signature check
   * @returns Verification result
   */
  async verifyPaymentTransaction(stxn, requirements, encodedTxn) {
    const txn = stxn.txn;
    if (txn.type !== "axfer") {
      return {
        isValid: false,
        invalidReason: ErrNotAssetTransfer,
        invalidMessage: `Expected asset transfer, got "${txn.type}"`
      };
    }
    const assetTransfer = txn.assetTransfer;
    if (!assetTransfer) {
      return {
        isValid: false,
        invalidReason: ErrNotAssetTransfer,
        invalidMessage: "Missing assetTransfer data"
      };
    }
    const amount = assetTransfer.amount ?? BigInt(0);
    if (amount !== BigInt(requirements.amount)) {
      return {
        isValid: false,
        invalidReason: ErrAmountMismatch,
        invalidMessage: `Expected ${requirements.amount}, got ${amount.toString()}`
      };
    }
    const receiver = assetTransfer.receiver ? assetTransfer.receiver.toString() : "";
    if (receiver !== requirements.payTo) {
      return {
        isValid: false,
        invalidReason: ErrReceiverMismatch,
        invalidMessage: `Expected ${requirements.payTo}, got ${receiver}`
      };
    }
    const assetId = assetTransfer.assetId?.toString() ?? "";
    if (assetId !== requirements.asset) {
      return {
        isValid: false,
        invalidReason: ErrAssetMismatch,
        invalidMessage: `Expected asset ${requirements.asset}, got ${assetId}`
      };
    }
    const txnBytes = decodeTransaction(encodedTxn);
    if (!hasSignature(txnBytes)) {
      return {
        isValid: false,
        invalidReason: ErrPaymentNotSigned,
        invalidMessage: "Payment transaction is not signed"
      };
    }
    if (stxn.sig) {
      const signedMsg = import_transact2.bytesForSigning.transaction(txn);
      const isValidSig = await (0, import_crypto.ed25519Verifier)(stxn.sig, signedMsg, txn.sender.publicKey);
      if (!isValidSig) {
        return {
          isValid: false,
          invalidReason: ErrInvalidSignature,
          invalidMessage: "Payment transaction signature does not match sender"
        };
      }
    }
    return { isValid: true };
  }
  /**
   * Verifies a fee payer transaction is safe to sign
   *
   * @param txn - The fee payer transaction to validate
   * @param groupSize - Number of transactions in the atomic group (for fee cap calculation)
   * @returns Verification result
   */
  verifyFeePayerTransaction(txn, groupSize) {
    if (txn.type !== "pay") {
      return {
        isValid: false,
        invalidReason: ErrInvalidFeePayer,
        invalidMessage: `Expected payment transaction, got ${txn.type}`
      };
    }
    const paymentFields = txn.payment;
    const payAmount = paymentFields?.amount ?? BigInt(0);
    if (payAmount > BigInt(0)) {
      return {
        isValid: false,
        invalidReason: ErrInvalidFeePayer,
        invalidMessage: "Fee payer amount must be 0"
      };
    }
    if (paymentFields?.receiver) {
      const receiverAddr = paymentFields.receiver.toString();
      const senderAddr = txn.sender.toString();
      if (receiverAddr !== senderAddr) {
        return {
          isValid: false,
          invalidReason: ErrInvalidFeePayer,
          invalidMessage: "Fee payer receiver must be same as sender (self-payment)"
        };
      }
    }
    if (paymentFields?.closeRemainderTo) {
      return {
        isValid: false,
        invalidReason: ErrInvalidFeePayer,
        invalidMessage: "closeRemainderTo not allowed on fee payer"
      };
    }
    if (txn.rekeyTo) {
      return {
        isValid: false,
        invalidReason: ErrInvalidFeePayer,
        invalidMessage: "rekeyTo not allowed on fee payer"
      };
    }
    const fee = Number(txn.fee ?? 0);
    const maxFee = maxReasonableGroupFee(groupSize);
    if (fee > maxFee) {
      return {
        isValid: false,
        invalidReason: ErrFeeTooHigh,
        invalidMessage: `Fee ${fee} exceeds maximum ${maxFee} (${groupSize} txns \xD7 5000 \xB5Algo)`
      };
    }
    return { isValid: true };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Errors,
  ExactAvmScheme
});
