// @x402/avm/dist/cjs/exact/client/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/client/index.ts
var client_exports = {};
__export(client_exports, {
  ExactAvmScheme: () => ExactAvmScheme
});
module.exports = __toCommonJS(client_exports);

// src/exact/client/scheme.ts
var import_algorand_client = require("@algorandfoundation/algokit-utils/algorand-client");
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_amount = require("@algorandfoundation/algokit-utils/amount");

// src/utils.ts
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};

// src/utils.ts
var import_utils = require("@x402/core/utils");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}
function isTestnetNetwork(network) {
  return network === ALGORAND_TESTNET_CAIP2;
}

// src/exact/client/scheme.ts
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   *
   * Constructs an atomic transaction group with:
   * - Optional fee payer transaction (if feePayer specified in requirements.extra)
   * - ASA transfer transaction to payTo address
   *
   * Uses TransactionComposer for automatic suggested params, group ID assignment,
   * and fee pooling. For sponsored (gasless) transactions, exact fees are calculated
   * from actual encoded transaction sizes using the protocol fee formula:
   *   fee = max(fee_per_byte × txn_size_in_bytes, min_fee)
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const { amount, asset, payTo, network, extra } = paymentRequirements;
    const algorandClient = this.getAlgorandClient(network);
    const assetId = this.getAssetId(asset, network);
    const feePayer = extra?.feePayer;
    let paymentIndex = 0;
    const emptySigner = (0, import_transact2.makeEmptyTransactionSigner)();
    const composer = algorandClient.newGroup();
    if (feePayer) {
      composer.addPayment({
        sender: feePayer,
        receiver: feePayer,
        amount: (0, import_amount.microAlgo)(0),
        note: `x402-fee-payer-${Date.now()}`,
        signer: emptySigner
      });
      paymentIndex = 1;
    }
    composer.addAssetTransfer({
      sender: this.signer.address,
      receiver: payTo,
      assetId: BigInt(assetId),
      amount: BigInt(amount),
      staticFee: feePayer ? (0, import_amount.microAlgo)(0) : void 0,
      // 0 fee when fee payer covers
      note: `x402-payment-v${x402Version}-${Date.now()}`,
      signer: emptySigner
    });
    const built = await composer.build();
    let transactions = built.transactions.map((tws) => tws.txn);
    if (feePayer) {
      const sp = await algorandClient.getSuggestedParams();
      const feePerByte = Number(sp.fee);
      const minFee = Number(sp.minFee);
      let totalGroupFee = BigInt(0);
      for (const txn of transactions) {
        const txnSize = (0, import_transact2.encodeTransactionRaw)(txn).length;
        const txnFee = feePerByte > 0 ? Math.max(feePerByte * txnSize, minFee) : minFee;
        totalGroupFee += BigInt(txnFee);
      }
      const ungrouped = transactions.map((txn, i) => {
        const fee = i === 0 ? totalGroupFee : BigInt(0);
        return new import_transact2.Transaction({ ...txn, fee, group: void 0 });
      });
      transactions = (0, import_transact2.groupTransactions)(ungrouped);
    }
    const encodedTxns = transactions.map((txn) => (0, import_transact2.encodeTransactionRaw)(txn));
    const clientIndexes = transactions.map((txn, i) => txn.sender.toString() === this.signer.address ? i : -1).filter((i) => i !== -1);
    const signedTxns = await this.signer.signTransactions(encodedTxns, clientIndexes);
    const paymentGroup = encodedTxns.map((txnBytes, i) => {
      const signedTxn = signedTxns[i];
      if (signedTxn) {
        return encodeTransaction(signedTxn);
      }
      return encodeTransaction(txnBytes);
    });
    const payload = {
      paymentGroup,
      paymentIndex
    };
    return {
      x402Version,
      payload
    };
  }
  /**
   * Creates or retrieves an AlgorandClient for the given network.
   *
   * @param network - Network identifier (CAIP-2 format)
   * @returns AlgorandClient instance
   */
  getAlgorandClient(network) {
    if (this.config?.algorandClient) {
      return this.config.algorandClient;
    }
    if (this.config?.algodUrl) {
      return import_algorand_client.AlgorandClient.fromConfig({
        algodConfig: {
          server: this.config.algodUrl,
          token: this.config.algodToken ?? ""
        }
      });
    }
    return isTestnetNetwork(network) ? import_algorand_client.AlgorandClient.testNet() : import_algorand_client.AlgorandClient.mainNet();
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - Asset identifier from requirements
   * @param network - Network identifier
   * @returns Asset ID as string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactAvmScheme
});
