// @algorandfoundation/algokit-utils/network-manager.js
const require_amount = require('./amount.js');

//#region src/network-manager.ts
/**
* Manager for LocalNet-specific network operations.
* These methods only work on LocalNet and will throw an error if called on other networks.
*/
var LocalNetManager = class {
	_algod;
	_networkManager;
	_algorand;
	/** @param algod The algod client to use for LocalNet operations */
	constructor(algod, networkManager, algorand) {
		this._algod = algod;
		this._networkManager = networkManager;
		this._algorand = algorand;
	}
	async _ensureLocalNet() {
		if (!await this._algorand.client.isLocalNet()) throw new Error("LocalNet methods are only available when connected to a LocalNet network");
	}
	/**
	* Generates a single block by sending a zero-amount payment transaction from the dispenser to itself.
	* This triggers block creation in LocalNet dev mode.
	*/
	async _generateBlock() {
		const algorand = this._algorand;
		const dispenser = await algorand.account.localNetDispenser();
		await algorand.send.payment({
			sender: dispenser,
			receiver: dispenser,
			amount: require_amount.AlgoAmount.MicroAlgo(0),
			note: "algokit-utils:block-generation"
		});
	}
	/**
	* Advances the blockchain by generating blocks until the target round is reached.
	* @param targetRound The round number to advance to
	* @throws Error if not connected to LocalNet
	* @example
	* ```typescript
	* // Advance LocalNet to round 100
	* await algorand.network.localNet.blockWarp(100n)
	* ```
	*/
	async blockWarp(targetRound) {
		await this._ensureLocalNet();
		while (await this._networkManager.getLastRound() < targetRound) await this._generateBlock();
	}
	/**
	* Advances the blockchain timestamp to a target Unix timestamp.
	* @param targetTimestamp The target Unix timestamp in seconds
	* @throws Error if not connected to LocalNet
	* @example
	* ```typescript
	* // Advance LocalNet time by 1 hour
	* const oneHourFromNow = BigInt(Math.floor(Date.now() / 1000)) + 3600n
	* await algorand.network.localNet.timeWarp(oneHourFromNow)
	* ```
	*/
	async timeWarp(targetTimestamp) {
		await this._ensureLocalNet();
		const currentTimestamp = await this._networkManager.getLatestTimestamp();
		const offsetSeconds = Number(targetTimestamp - currentTimestamp);
		await this._algod.setBlockTimeStampOffset(offsetSeconds);
		await this._generateBlock();
		await this._algod.setBlockTimeStampOffset(0);
	}
};
/**
* Manager for network-related operations.
* Provides utilities for querying blockchain state and waiting for specific conditions.
*/
var NetworkManager = class {
	_algod;
	_algorand;
	_localNet;
	/**
	* Create a new NetworkManager instance.
	* @param algod The algod client to use for network operations
	* @param algorand The AlgorandClient instance (for LocalNet operations)
	*/
	constructor(algod, algorand) {
		this._algod = algod;
		this._algorand = algorand;
	}
	/**
	* Get the last committed round number.
	* @returns The last round number
	* @example
	* ```typescript
	* const lastRound = await algorand.network.getLastRound()
	* console.log(`Current round: ${lastRound}`)
	* ```
	*/
	async getLastRound() {
		return (await this._algod.status()).lastRound;
	}
	/**
	* Get the Unix timestamp of the latest block.
	*
	* Note: This method makes two sequential API calls (status then block fetch).
	* The round may advance between these calls, so the returned timestamp may not
	* be from the actual latest timestamp.
	*
	* @returns The UNIX timestamp of the last round
	* @example
	* ```typescript
	* const timestamp = await algorand.network.getLatestTimestamp()
	* console.log(`Latest block time: ${new Date(Number(timestamp) * 1000)}`)
	* ```
	*/
	async getLatestTimestamp() {
		const status = await this._algod.status();
		return (await this._algod.block(status.lastRound)).block.header.timestamp;
	}
	/**
	* Wait until a specific round is reached.
	*
	* @param targetRound The round number to wait for
	* @throws Error if timeout is reached before the target round
	* @example
	* ```typescript
	* // Wait for round 1000
	* await algorand.network.waitUntilRound(1000n)
	*
	* // Wait with a 30 second timeout
	* await algorand.network.waitUntilRound(1000n, 30000)
	* ```
	*/
	async waitUntilRound(targetRound) {
		let status = await this._algod.status();
		while (status.lastRound < targetRound) status = await this._algod.statusAfterBlock(status.lastRound);
	}
	/**
	* Wait until a specific Unix timestamp is reached on the blockchain.
	*
	* @param targetTimestamp The target Unix timestamp in seconds
	* @param options Optional parameters for waiting
	* @example
	* ```typescript
	* // Wait until a specific time
	* const futureTime = BigInt(Math.floor(Date.now() / 1000)) + 60n // 1 minute from now
	* await algorand.network.waitUntilTimestamp(futureTime)
	*
	* // Wait with custom timing parameters
	* await algorand.network.waitUntilTimestamp(futureTime, { blockTimeSeconds: 3.0, pollingIntervalMs: 500 })
	* ```
	*/
	async waitUntilTimestamp(targetTimestamp, options) {
		const blockTimeSeconds = options?.blockTimeSeconds ?? 2.7;
		const pollingIntervalMs = options?.pollingIntervalMs ?? 1e3;
		const pollingTimeoutMs = options?.pollingTimeoutMs ?? 5e3;
		const currentRound = (await this._algod.status()).lastRound;
		const currentBlockTimestamp = (await this._algod.block(currentRound)).block.header.timestamp;
		if (currentBlockTimestamp >= targetTimestamp) return;
		const secondsToWait = Number(targetTimestamp - currentBlockTimestamp);
		const roundsToWait = Math.floor(secondsToWait / blockTimeSeconds);
		if (roundsToWait > 1) await this.waitUntilRound(currentRound + BigInt(roundsToWait));
		const pollingStartTime = Date.now();
		while (true) {
			const latestTimeStamp = await this.getLatestTimestamp();
			if (latestTimeStamp > targetTimestamp) return;
			if (Date.now() - pollingStartTime >= pollingTimeoutMs) throw new Error(`Timeout waiting for timestamp ${targetTimestamp} (current: ${latestTimeStamp})`);
			await new Promise((resolve) => setTimeout(resolve, pollingIntervalMs));
		}
	}
	/**
	* Get LocalNet-specific network utilities.
	* These methods only work when connected to a LocalNet network and will throw
	* an error if called on other networks.
	*
	* @returns The LocalNetManager instance
	* @example
	* ```typescript
	* // Block warp on LocalNet
	* await algorand.network.localNet.blockWarp(100n)
	*
	* // Time warp on LocalNet
	* await algorand.network.localNet.timeWarp(BigInt(Date.now() / 1000) + 3600n)
	* ```
	*/
	get localNet() {
		if (!this._localNet) this._localNet = new LocalNetManager(this._algod, this, this._algorand);
		return this._localNet;
	}
};

//#endregion
exports.LocalNetManager = LocalNetManager;
exports.NetworkManager = NetworkManager;
