// @algorandfoundation/algokit-utils/packages/crypto/src/index.js
const require_runtime = require("../../../_virtual/_rolldown/runtime.js");
require("./ed25519.js");
require("./hash.js");
require("./hd.js");
let _noble_hashes_sha2_js = require("@noble/hashes/sha2.js");
let _noble_ed25519 = require("@noble/ed25519");
_noble_ed25519 = require_runtime.__toESM(_noble_ed25519);
let _noble_curves_ed25519_js = require("@noble/curves/ed25519.js");
let _noble_curves_utils_js = require("@noble/curves/utils.js");
let _noble_curves_abstract_modular_js = require("@noble/curves/abstract/modular.js");
//#region packages/crypto/src/index.ts
const ED25519_SEED_LENGTH = 32;
const ED25519_EXTENDED_PRIVATE_KEY_LENGTH = 96;
const assertEd25519SecretLength = (secret, secretType) => {
	let expectedLength;
	switch (secretType) {
		case "ed25519 seed":
			expectedLength = ED25519_SEED_LENGTH;
			break;
		case "HD extended key":
			expectedLength = ED25519_EXTENDED_PRIVATE_KEY_LENGTH;
			break;
		default: throw new Error(`Unknown secret type: ${secretType}`);
	}
	if (secret.length !== expectedLength) throw new Error(`Expected unwrapped ${secretType} to be ${expectedLength} bytes, got ${secret.length}.`);
};
const throwWrapUnwrapErrors = (operationError, wrapError, operationName) => {
	throw new AggregateError([operationError, wrapError], `${operationName} failed and failed to re-wrap Ed25519 secret. Check both errors for details.`);
};
function rawSign(extendedSecretKey, data) {
	const scalar = (0, _noble_curves_utils_js.bytesToNumberLE)(extendedSecretKey.subarray(0, 32));
	const kR = extendedSecretKey.subarray(32, 64);
	const publicKey = rawPubkey(extendedSecretKey);
	const r = (0, _noble_curves_abstract_modular_js.mod)((0, _noble_curves_utils_js.bytesToNumberLE)((0, _noble_hashes_sha2_js.sha512)(new Uint8Array([...kR, ...data]))), _noble_curves_ed25519_js.ed25519.Point.Fn.ORDER);
	const R = _noble_curves_ed25519_js.ed25519.Point.BASE.multiply(r);
	const S = (0, _noble_curves_abstract_modular_js.mod)(r + (0, _noble_curves_abstract_modular_js.mod)((0, _noble_curves_utils_js.bytesToNumberLE)((0, _noble_hashes_sha2_js.sha512)(new Uint8Array([
		...R.toBytes(),
		...publicKey,
		...data
	]))), _noble_curves_ed25519_js.ed25519.Point.Fn.ORDER) * scalar, _noble_curves_ed25519_js.ed25519.Point.Fn.ORDER);
	return new Uint8Array([...R.toBytes(), ...(0, _noble_curves_utils_js.numberToBytesLE)(S, 32)]);
}
function rawPubkey(extendedSecretKey) {
	const scalar = (0, _noble_curves_utils_js.bytesToNumberLE)(extendedSecretKey.subarray(0, 32));
	if ((scalar & 1n << 255n) !== 0n) throw new Error("Invalid HD-expanded Ed25519 secret scalar: most-significant bit (bit 255) of the 32-byte scalar must be 0 for rawSign/rawPubkey inputs.");
	const reducedScalar = (0, _noble_curves_abstract_modular_js.mod)(scalar, _noble_curves_ed25519_js.ed25519.Point.Fn.ORDER);
	return _noble_curves_ed25519_js.ed25519.Point.BASE.multiply(reducedScalar).toBytes();
}
/**
* Creates an Ed25519 signing key from a wrapped secret using the @noble/ed25519 implementation.
*
* NOTE: This function will zero out the unwrapped secret after the wrap function is called.
*
* @param wrapUnwrap - The wrapped secret provider that unwraps and re-wraps the Ed25519 secret.
* @returns A promise that resolves to an Ed25519 signing key containing the public key and raw signer.
*/
const nobleEd25519SigningKeyFromWrappedSecret = async (wrapUnwrap) => {
	let wrapFunction;
	if ("wrapEd25519Seed" in wrapUnwrap) wrapFunction = wrapUnwrap.wrapEd25519Seed;
	else if ("wrapHdExtendedPrivateKey" in wrapUnwrap) wrapFunction = wrapUnwrap.wrapHdExtendedPrivateKey;
	else throw new Error("Invalid WrappedEd25519Secret: missing wrap function");
	const signer = async (bytesToSign) => {
		let secret = void 0;
		let signature = void 0;
		let signingError;
		let wrapError;
		try {
			if ("unwrapEd25519Seed" in wrapUnwrap) {
				secret = await wrapUnwrap.unwrapEd25519Seed();
				assertEd25519SecretLength(secret, "ed25519 seed");
				signature = await _noble_ed25519.signAsync(bytesToSign, secret);
			} else if ("unwrapHdExtendedPrivateKey" in wrapUnwrap) {
				secret = await wrapUnwrap.unwrapHdExtendedPrivateKey();
				assertEd25519SecretLength(secret, "HD extended key");
				signature = rawSign(secret.subarray(0, 64), bytesToSign);
			} else throw new Error("Invalid WrappedEd25519Secret: missing unwrap function");
		} catch (error) {
			signingError = error;
		} finally {
			try {
				await wrapFunction();
			} catch (error) {
				wrapError = error;
			} finally {
				secret?.fill(0);
			}
		}
		if (signingError !== void 0 && wrapError !== void 0) throwWrapUnwrapErrors(signingError, wrapError, "Signing");
		if (signingError !== void 0) throw signingError;
		if (wrapError !== void 0) throw wrapError;
		if (signature === void 0) throw new Error("Signing failed unexpectedly without an error.");
		return signature;
	};
	let pubkey = void 0;
	let pubkeyError;
	let wrapError;
	let secret = void 0;
	try {
		if ("unwrapEd25519Seed" in wrapUnwrap) {
			secret = await wrapUnwrap.unwrapEd25519Seed();
			assertEd25519SecretLength(secret, "ed25519 seed");
			pubkey = await _noble_ed25519.getPublicKeyAsync(secret);
		} else if ("unwrapHdExtendedPrivateKey" in wrapUnwrap) {
			secret = await wrapUnwrap.unwrapHdExtendedPrivateKey();
			assertEd25519SecretLength(secret, "HD extended key");
			pubkey = rawPubkey(secret.subarray(0, 64));
		} else throw new Error("Invalid WrappedEd25519Secret: missing unwrap function");
	} catch (error) {
		pubkeyError = error;
	} finally {
		try {
			await wrapFunction();
		} catch (error) {
			wrapError = error;
		} finally {
			secret?.fill(0);
		}
	}
	if (pubkeyError !== void 0 && wrapError !== void 0) throwWrapUnwrapErrors(pubkeyError, wrapError, "Deriving Ed25519 public key");
	if (pubkeyError !== void 0) throw pubkeyError;
	if (wrapError !== void 0) throw wrapError;
	if (pubkey === void 0) throw new Error("Deriving Ed25519 public key failed unexpectedly without an error.");
	return {
		ed25519Pubkey: pubkey,
		rawEd25519Signer: signer
	};
};
/**
* Creates an ed25519 signing key from a wrapped secret using the default ed25519 implementation (currently @noble/ed25519).
* The implementation may change in the future. To explicitly use the @noble/ed25519 implementation, use `nobleEd25519SigningKeyFromWrappedSecret`.
*
* NOTE: This function will zero out the unwrapped secret after the wrap function is called.
*
* @param wrapUnwrap - The wrapped secret provider that unwraps and re-wraps the ed25519 secret.
* @returns A promise that resolves to an ed25519 signing key with public key and raw signer.
*/
const ed25519SigningKeyFromWrappedSecret = nobleEd25519SigningKeyFromWrappedSecret;
//#endregion
exports.ed25519SigningKeyFromWrappedSecret = ed25519SigningKeyFromWrappedSecret;
exports.nobleEd25519SigningKeyFromWrappedSecret = nobleEd25519SigningKeyFromWrappedSecret;
