// @algorandfoundation/algokit-utils/packages/common/src/address.js
const require_runtime = require('../../../_virtual/_rolldown/runtime.js');
const require_array = require('./array.js');
const require_constants = require('./constants.js');
let hi_base32 = require("hi-base32");
hi_base32 = require_runtime.__toESM(hi_base32);
let _noble_hashes_sha2_js = require("@noble/hashes/sha2.js");

//#region packages/common/src/address.ts
const ALGORAND_ADDRESS_BYTE_LENGTH = 36;
const ALGORAND_CHECKSUM_BYTE_LENGTH = 4;
const ALGORAND_ADDRESS_LENGTH = 58;
const ALGORAND_ZERO_ADDRESS_STRING = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ";
const MALFORMED_ADDRESS_ERROR_MSG = "address seems to be malformed";
const CHECKSUM_ADDRESS_ERROR_MSG = "wrong checksum for address";
function checksumFromPublicKey(publicKey) {
	return Uint8Array.from((0, _noble_hashes_sha2_js.sha512_256)(publicKey).slice(require_constants.HASH_BYTES_LENGTH - require_constants.CHECKSUM_BYTE_LENGTH, require_constants.HASH_BYTES_LENGTH));
}
function bytesToHex(bytes) {
	return Buffer.from(bytes).toString("hex");
}
function encodeUint64(num) {
	if (!(typeof num === "bigint" || Number.isInteger(num)) || num < 0 || num > BigInt("0xffffffffffffffff")) throw new Error("Input is not a 64-bit unsigned integer");
	const encoding = new Uint8Array(8);
	new DataView(encoding.buffer).setBigUint64(0, BigInt(num));
	return encoding;
}
/** Symbol used for instanceof checks across packages (CJS/ESM) */
const ADDR_SYMBOL = Symbol.for("algokit_common:Address");
/**
* Represents an Algorand address
*/
var Address = class Address {
	/**
	* The binary form of the address. For standard accounts, this is the public key.
	*/
	publicKey;
	/** @internal */
	[ADDR_SYMBOL];
	/**
	* Create a new Address object from its binary form.
	* @param publicKey - The binary form of the address. Must be 32 bytes.
	*/
	constructor(publicKey) {
		this[ADDR_SYMBOL] = true;
		if (publicKey.length !== ALGORAND_ADDRESS_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH) throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: 0x${bytesToHex(publicKey)}, length ${publicKey.length}`);
		this.publicKey = publicKey;
	}
	/**
	* Check if the address is equal to another address.
	*/
	equals(other) {
		return other instanceof Address && require_array.arrayEqual(this.publicKey, other.publicKey);
	}
	/**
	* Compute the 4 byte checksum of the address.
	*/
	checksum() {
		return checksumFromPublicKey(this.publicKey);
	}
	/**
	* Encode the address into a string form.
	*/
	toString() {
		return hi_base32.default.encode(require_array.concatArrays(this.publicKey, this.checksum())).slice(0, ALGORAND_ADDRESS_LENGTH);
	}
	/**
	* Decode an address from a string.
	* @param address - The address to decode. Must be 58 characters long.
	* @returns An Address object corresponding to the input string.
	*/
	static fromString(address) {
		if (typeof address !== "string") throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: expected string, got ${typeof address}, ${address}`);
		if (address.length !== ALGORAND_ADDRESS_LENGTH) throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: expected length ${ALGORAND_ADDRESS_LENGTH}, got ${address.length}: ${address}`);
		const decoded = hi_base32.default.decode.asBytes(address);
		if (decoded.length !== ALGORAND_ADDRESS_BYTE_LENGTH) throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: expected byte length ${ALGORAND_ADDRESS_BYTE_LENGTH}, got ${decoded.length}`);
		const pk = new Uint8Array(decoded.slice(0, ALGORAND_ADDRESS_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH));
		const cs = new Uint8Array(decoded.slice(ALGORAND_ADDRESS_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH, ALGORAND_ADDRESS_BYTE_LENGTH));
		const checksum = checksumFromPublicKey(pk);
		if (!require_array.arrayEqual(checksum, cs)) throw new Error(`${CHECKSUM_ADDRESS_ERROR_MSG}: ${address} (${cs}, ${checksum})`);
		return new Address(pk);
	}
	/**
	* Get the zero address.
	*/
	static zeroAddress() {
		return new Address(new Uint8Array(ALGORAND_ADDRESS_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH));
	}
};
Object.defineProperty(Address, Symbol.hasInstance, { value: function(obj) {
	if (obj instanceof Object && Object.getPrototypeOf(obj) === Address.prototype) return true;
	return Boolean(obj && typeof obj === "object" && ADDR_SYMBOL in obj && obj[ADDR_SYMBOL]);
} });
function getAddress(addr) {
	if (typeof addr == "string") return Address.fromString(addr);
	else if ("addr" in addr) return addr.addr;
	else return addr;
}
function getOptionalAddress(addr) {
	if (addr === void 0) return;
	return getAddress(addr);
}
/**
* isValidAddress checks if a string is a valid Algorand address.
* @param address - an Algorand address with checksum.
* @returns true if valid, false otherwise
*/
function isValidAddress(address) {
	try {
		Address.fromString(address);
	} catch {
		return false;
	}
	return true;
}
const APP_ID_PREFIX = new TextEncoder().encode("appID");
/**
* Get the escrow address of an application.
* @param appID - The ID of the application.
* @returns The address corresponding to that application's escrow account.
*/
function getApplicationAddress(appID) {
	return new Address((0, _noble_hashes_sha2_js.sha512_256)(require_array.concatArrays(APP_ID_PREFIX, encodeUint64(appID))));
}
/**
* decodeAddress takes an Algorand address in string form and decodes it into a Uint8Array.
* @param address - an Algorand address with checksum.
* @returns the decoded form of the address's public key and checksum
*/
function decodeAddress(address) {
	return Address.fromString(address);
}
/**
* encodeAddress takes an Algorand address as a Uint8Array and encodes it into a string with checksum.
* @param address - a raw Algorand address
* @returns the address and checksum encoded as a string.
*/
function encodeAddress(address) {
	return new Address(address).toString();
}

//#endregion
exports.ALGORAND_ADDRESS_BYTE_LENGTH = ALGORAND_ADDRESS_BYTE_LENGTH;
exports.ALGORAND_ADDRESS_LENGTH = ALGORAND_ADDRESS_LENGTH;
exports.ALGORAND_CHECKSUM_BYTE_LENGTH = ALGORAND_CHECKSUM_BYTE_LENGTH;
exports.ALGORAND_ZERO_ADDRESS_STRING = ALGORAND_ZERO_ADDRESS_STRING;
exports.Address = Address;
exports.CHECKSUM_ADDRESS_ERROR_MSG = CHECKSUM_ADDRESS_ERROR_MSG;
exports.MALFORMED_ADDRESS_ERROR_MSG = MALFORMED_ADDRESS_ERROR_MSG;
exports.checksumFromPublicKey = checksumFromPublicKey;
exports.decodeAddress = decodeAddress;
exports.encodeAddress = encodeAddress;
exports.encodeUint64 = encodeUint64;
exports.getAddress = getAddress;
exports.getApplicationAddress = getApplicationAddress;
exports.getOptionalAddress = getOptionalAddress;
exports.isValidAddress = isValidAddress;
