// @algorandfoundation/algokit-utils/packages/common/src/constants.js

//#region packages/common/src/constants.ts
const TRANSACTION_DOMAIN_SEPARATOR = "TX";
const TRANSACTION_GROUP_DOMAIN_SEPARATOR = "TG";
const MULTISIG_DOMAIN_SEPARATOR = "MultisigAddr";
const SIGNATURE_ENCODING_INCR = 75;
const HASH_BYTES_LENGTH = 32;
const PUBLIC_KEY_BYTE_LENGTH = 32;
const MAX_TRANSACTION_GROUP_SIZE = 16;
const CHECKSUM_BYTE_LENGTH = 4;
const ADDRESS_LENGTH = 58;
const TRANSACTION_ID_LENGTH = 52;
const LENGTH_ENCODE_BYTE_SIZE = 2;
const BOOL_TRUE_BYTE = 128;
const BOOL_FALSE_BYTE = 0;
const SIGNATURE_BYTE_LENGTH = 64;
const EMPTY_SIGNATURE = new Uint8Array(SIGNATURE_BYTE_LENGTH);
const MAX_EXTRA_PROGRAM_PAGES = 3;
const PROGRAM_PAGE_SIZE = 2048;
const MAX_APP_ARGS = 16;
const MAX_ARGS_SIZE = 2048;
const MAX_OVERALL_REFERENCES = 8;
const MAX_ACCOUNT_REFERENCES = 8;
const MAX_APP_REFERENCES = 8;
const MAX_ASSET_REFERENCES = 8;
const MAX_BOX_REFERENCES = 8;
const MAX_GLOBAL_STATE_KEYS = 64;
const MAX_LOCAL_STATE_KEYS = 16;
const MAX_ASSET_NAME_LENGTH = 32;
const MAX_ASSET_UNIT_NAME_LENGTH = 8;
const MAX_ASSET_URL_LENGTH = 96;
const MAX_ASSET_DECIMALS = 19;

//#endregion
exports.ADDRESS_LENGTH = ADDRESS_LENGTH;
exports.BOOL_FALSE_BYTE = BOOL_FALSE_BYTE;
exports.BOOL_TRUE_BYTE = BOOL_TRUE_BYTE;
exports.CHECKSUM_BYTE_LENGTH = CHECKSUM_BYTE_LENGTH;
exports.EMPTY_SIGNATURE = EMPTY_SIGNATURE;
exports.HASH_BYTES_LENGTH = HASH_BYTES_LENGTH;
exports.LENGTH_ENCODE_BYTE_SIZE = LENGTH_ENCODE_BYTE_SIZE;
exports.MAX_ACCOUNT_REFERENCES = MAX_ACCOUNT_REFERENCES;
exports.MAX_APP_ARGS = MAX_APP_ARGS;
exports.MAX_APP_REFERENCES = MAX_APP_REFERENCES;
exports.MAX_ARGS_SIZE = MAX_ARGS_SIZE;
exports.MAX_ASSET_DECIMALS = MAX_ASSET_DECIMALS;
exports.MAX_ASSET_NAME_LENGTH = MAX_ASSET_NAME_LENGTH;
exports.MAX_ASSET_REFERENCES = MAX_ASSET_REFERENCES;
exports.MAX_ASSET_UNIT_NAME_LENGTH = MAX_ASSET_UNIT_NAME_LENGTH;
exports.MAX_ASSET_URL_LENGTH = MAX_ASSET_URL_LENGTH;
exports.MAX_BOX_REFERENCES = MAX_BOX_REFERENCES;
exports.MAX_EXTRA_PROGRAM_PAGES = MAX_EXTRA_PROGRAM_PAGES;
exports.MAX_GLOBAL_STATE_KEYS = MAX_GLOBAL_STATE_KEYS;
exports.MAX_LOCAL_STATE_KEYS = MAX_LOCAL_STATE_KEYS;
exports.MAX_OVERALL_REFERENCES = MAX_OVERALL_REFERENCES;
exports.MAX_TRANSACTION_GROUP_SIZE = MAX_TRANSACTION_GROUP_SIZE;
exports.MULTISIG_DOMAIN_SEPARATOR = MULTISIG_DOMAIN_SEPARATOR;
exports.PROGRAM_PAGE_SIZE = PROGRAM_PAGE_SIZE;
exports.PUBLIC_KEY_BYTE_LENGTH = PUBLIC_KEY_BYTE_LENGTH;
exports.SIGNATURE_BYTE_LENGTH = SIGNATURE_BYTE_LENGTH;
exports.SIGNATURE_ENCODING_INCR = SIGNATURE_ENCODING_INCR;
exports.TRANSACTION_DOMAIN_SEPARATOR = TRANSACTION_DOMAIN_SEPARATOR;
exports.TRANSACTION_GROUP_DOMAIN_SEPARATOR = TRANSACTION_GROUP_DOMAIN_SEPARATOR;
exports.TRANSACTION_ID_LENGTH = TRANSACTION_ID_LENGTH;
