// @algorandfoundation/algokit-utils/packages/common/src/sourcemap.mjs
import * as vlq from "vlq";

//#region packages/common/src/sourcemap.ts
/**
* Contains a mapping from TEAL program PC to source file location.
*/
var ProgramSourceMap = class {
	version;
	/**
	* A list of original sources used by the "mappings" entry.
	*/
	sources;
	/**
	* A list of symbol names used by the "mappings" entry.
	*/
	names;
	/**
	* A string with the encoded mapping data.
	*/
	mappings;
	pcToLocation;
	sourceAndLineToPc;
	constructor({ version, sources, names, mappings }) {
		this.version = version;
		this.sources = sources;
		this.names = names;
		this.mappings = mappings;
		if (this.version !== 3) throw new Error(`Only version 3 is supported, got ${this.version}`);
		if (this.mappings === void 0) throw new Error("mapping undefined, cannot build source map without `mapping`");
		const pcList = this.mappings.split(";").map(vlq.decode);
		this.pcToLocation = /* @__PURE__ */ new Map();
		this.sourceAndLineToPc = /* @__PURE__ */ new Map();
		const lastLocation = {
			line: 0,
			column: 0,
			sourceIndex: 0,
			nameIndex: 0
		};
		for (const [pc, data] of pcList.entries()) {
			const dataArray = data;
			if (dataArray.length < 4) continue;
			const nameDelta = dataArray.length > 4 ? dataArray[4] : void 0;
			const [, sourceDelta, lineDelta, columnDelta] = dataArray;
			lastLocation.sourceIndex += sourceDelta;
			lastLocation.line += lineDelta;
			lastLocation.column += columnDelta;
			if (typeof nameDelta !== "undefined") lastLocation.nameIndex += nameDelta;
			const sourceAndLineKey = `${lastLocation.sourceIndex}:${lastLocation.line}`;
			let pcsForSourceAndLine = this.sourceAndLineToPc.get(sourceAndLineKey);
			if (pcsForSourceAndLine === void 0) {
				pcsForSourceAndLine = [];
				this.sourceAndLineToPc.set(sourceAndLineKey, pcsForSourceAndLine);
			}
			const pcInLine = {
				pc,
				column: lastLocation.column
			};
			const pcLocation = {
				line: lastLocation.line,
				column: lastLocation.column,
				sourceIndex: lastLocation.sourceIndex
			};
			if (typeof nameDelta !== "undefined") {
				pcInLine.nameIndex = lastLocation.nameIndex;
				pcLocation.nameIndex = lastLocation.nameIndex;
			}
			pcsForSourceAndLine.push(pcInLine);
			this.pcToLocation.set(pc, pcLocation);
		}
	}
	getPcs() {
		return Array.from(this.pcToLocation.keys());
	}
	getLocationForPc(pc) {
		return this.pcToLocation.get(pc);
	}
	getPcsOnSourceLine(sourceIndex, line) {
		const pcs = this.sourceAndLineToPc.get(`${sourceIndex}:${line}`);
		if (pcs === void 0) return [];
		return pcs;
	}
};

//#endregion
export { ProgramSourceMap };
