// @algorandfoundation/algokit-utils/packages/algod_client/src/models/account-application-response.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_application_params = require('./application-params.js');
const require_application_local_state = require('./application-local-state.js');

//#region packages/algod_client/src/models/account-application-response.ts
const AccountApplicationResponseMeta = {
	name: "AccountApplicationResponse",
	kind: "object",
	fields: [
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "appLocalState",
			wireKey: "app-local-state",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_local_state.ApplicationLocalStateMeta)
		},
		{
			name: "createdApp",
			wireKey: "created-app",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_params.ApplicationParamsMeta)
		}
	]
};

//#endregion
exports.AccountApplicationResponseMeta = AccountApplicationResponseMeta;
