// @algorandfoundation/algokit-utils/packages/algod_client/src/models/application-initial-states.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_application_kv_storage = require('./application-kv-storage.js');

//#region packages/algod_client/src/models/application-initial-states.ts
const ApplicationInitialStatesMeta = {
	name: "ApplicationInitialStates",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "appLocals",
			wireKey: "app-locals",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_application_kv_storage.ApplicationKvStorageMeta))
		},
		{
			name: "appGlobals",
			wireKey: "app-globals",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_kv_storage.ApplicationKvStorageMeta)
		},
		{
			name: "appBoxes",
			wireKey: "app-boxes",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_kv_storage.ApplicationKvStorageMeta)
		}
	]
};

//#endregion
exports.ApplicationInitialStatesMeta = ApplicationInitialStatesMeta;
