// @algorandfoundation/algokit-utils/packages/algod_client/src/models/application-state-operation.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { AvmValueMeta } from "./avm-value.mjs";

//#region packages/algod_client/src/models/application-state-operation.ts
const ApplicationStateOperationMeta = {
	name: "ApplicationStateOperation",
	kind: "object",
	fields: [
		{
			name: "operation",
			wireKey: "operation",
			optional: false,
			codec: stringCodec
		},
		{
			name: "appStateType",
			wireKey: "app-state-type",
			optional: false,
			codec: stringCodec
		},
		{
			name: "key",
			wireKey: "key",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "newValue",
			wireKey: "new-value",
			optional: true,
			codec: new ObjectModelCodec(AvmValueMeta)
		},
		{
			name: "account",
			wireKey: "account",
			optional: true,
			codec: addressCodec
		}
	]
};

//#endregion
export { ApplicationStateOperationMeta };
