// @algorandfoundation/algokit-utils/packages/algod_client/src/models/application-initial-states.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { ApplicationKvStorageMeta } from "./application-kv-storage.mjs";

//#region packages/algod_client/src/models/application-initial-states.ts
const ApplicationInitialStatesMeta = {
	name: "ApplicationInitialStates",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "appLocals",
			wireKey: "app-locals",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ApplicationKvStorageMeta))
		},
		{
			name: "appGlobals",
			wireKey: "app-globals",
			optional: true,
			codec: new ObjectModelCodec(ApplicationKvStorageMeta)
		},
		{
			name: "appBoxes",
			wireKey: "app-boxes",
			optional: true,
			codec: new ObjectModelCodec(ApplicationKvStorageMeta)
		}
	]
};

//#endregion
export { ApplicationInitialStatesMeta };
