// @algorandfoundation/algokit-utils/packages/algod_client/src/models/block.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes32Codec, fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec, addressArrayCodec, bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { MapCodec } from "../../../common/src/codecs/composite/map.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { SignedTransactionMeta } from "../../../transact/src/transactions/signed-transaction-meta.mjs";

//#region packages/algod_client/src/models/block.ts
const BlockEvalDeltaMeta = {
	name: "BlockEvalDelta",
	kind: "object",
	fields: [
		{
			name: "action",
			wireKey: "at",
			optional: false,
			codec: numberCodec
		},
		{
			name: "bytes",
			wireKey: "bs",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "uint",
			wireKey: "ui",
			optional: true,
			codec: bigIntCodec
		}
	]
};
const BlockAppEvalDeltaMeta = {
	name: "BlockAppEvalDelta",
	kind: "object",
	fields: [
		{
			name: "globalDelta",
			wireKey: "gd",
			optional: true,
			codec: new MapCodec(bytesCodec, new ObjectModelCodec(BlockEvalDeltaMeta))
		},
		{
			name: "localDeltas",
			wireKey: "ld",
			optional: true,
			codec: new MapCodec(numberCodec, new MapCodec(bytesCodec, new ObjectModelCodec(BlockEvalDeltaMeta)))
		},
		{
			name: "innerTxns",
			wireKey: "itx",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(() => SignedTxnWithADMeta))
		},
		{
			name: "sharedAccounts",
			wireKey: "sa",
			optional: true,
			codec: addressArrayCodec
		},
		{
			name: "logs",
			wireKey: "lg",
			optional: true,
			codec: bytesArrayCodec
		}
	]
};
const BlockStateProofTrackingDataMeta = {
	name: "BlockStateProofTrackingData",
	kind: "object",
	fields: [
		{
			name: "stateProofVotersCommitment",
			wireKey: "v",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "stateProofOnlineTotalWeight",
			wireKey: "t",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "stateProofNextRound",
			wireKey: "n",
			optional: true,
			codec: bigIntCodec
		}
	]
};
const ApplyDataMeta = {
	name: "SignedTxnInBlock",
	kind: "object",
	fields: [
		{
			name: "closingAmount",
			wireKey: "ca",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "assetClosingAmount",
			wireKey: "aca",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "senderRewards",
			wireKey: "rs",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "receiverRewards",
			wireKey: "rr",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "closeRewards",
			wireKey: "rc",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "evalDelta",
			wireKey: "dt",
			optional: true,
			codec: new ObjectModelCodec(BlockAppEvalDeltaMeta)
		},
		{
			name: "configAsset",
			wireKey: "caid",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "applicationId",
			wireKey: "apid",
			optional: true,
			codec: bigIntCodec
		}
	]
};
const SignedTxnWithADMeta = {
	name: "SignedTxnWithAD",
	kind: "object",
	fields: [{
		name: "signedTxn",
		flattened: true,
		optional: false,
		codec: new ObjectModelCodec(SignedTransactionMeta)
	}, {
		name: "applyData",
		flattened: true,
		optional: true,
		codec: new ObjectModelCodec(ApplyDataMeta)
	}]
};
const SignedTxnInBlockMeta = {
	name: "SignedTxnInBlock",
	kind: "object",
	fields: [
		{
			name: "signedTxn",
			flattened: true,
			optional: false,
			codec: new ObjectModelCodec(SignedTxnWithADMeta)
		},
		{
			name: "hasGenesisId",
			wireKey: "hgi",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "hasGenesisHash",
			wireKey: "hgh",
			optional: true,
			codec: booleanCodec
		}
	]
};
const ParticipationUpdatesMeta = {
	name: "ParticipationUpdates",
	kind: "object",
	fields: [{
		name: "expiredParticipationAccounts",
		wireKey: "partupdrmv",
		optional: false,
		codec: addressArrayCodec
	}, {
		name: "absentParticipationAccounts",
		wireKey: "partupdabs",
		optional: false,
		codec: addressArrayCodec
	}]
};
const TxnCommitmentsMeta = {
	name: "TxnCommitments",
	kind: "object",
	fields: [
		{
			name: "nativeSha512_256Commitment",
			wireKey: "txn",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "sha256Commitment",
			wireKey: "txn256",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "sha512Commitment",
			wireKey: "txn512",
			optional: true,
			codec: fixedBytes64Codec
		}
	]
};
const RewardStateMeta = {
	name: "RewardState",
	kind: "object",
	fields: [
		{
			name: "feeSink",
			wireKey: "fees",
			optional: false,
			codec: addressCodec
		},
		{
			name: "rewardsPool",
			wireKey: "rwd",
			optional: false,
			codec: addressCodec
		},
		{
			name: "rewardsLevel",
			wireKey: "earn",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "rewardsRate",
			wireKey: "rate",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "rewardsResidue",
			wireKey: "frac",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "rewardsRecalculationRound",
			wireKey: "rwcalr",
			optional: false,
			codec: bigIntCodec
		}
	]
};
const UpgradeStateMeta = {
	name: "UpgradeState",
	kind: "object",
	fields: [
		{
			name: "currentProtocol",
			wireKey: "proto",
			optional: false,
			codec: stringCodec
		},
		{
			name: "nextProtocol",
			wireKey: "nextproto",
			optional: true,
			codec: stringCodec
		},
		{
			name: "nextProtocolApprovals",
			wireKey: "nextyes",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "nextProtocolVoteBefore",
			wireKey: "nextbefore",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "nextProtocolSwitchOn",
			wireKey: "nextswitch",
			optional: true,
			codec: bigIntCodec
		}
	]
};
const UpgradeVoteMeta = {
	name: "UpgradeVote",
	kind: "object",
	fields: [
		{
			name: "upgradePropose",
			wireKey: "upgradeprop",
			optional: true,
			codec: stringCodec
		},
		{
			name: "upgradeDelay",
			wireKey: "upgradedelay",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "upgradeApprove",
			wireKey: "upgradeyes",
			optional: true,
			codec: booleanCodec
		}
	]
};
const BlockMeta = {
	name: "Block",
	kind: "object",
	fields: [{
		name: "header",
		flattened: true,
		optional: false,
		codec: new ObjectModelCodec({
			name: "BlockHeader",
			kind: "object",
			fields: [
				{
					name: "round",
					wireKey: "rnd",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "previousBlockHash",
					wireKey: "prev",
					optional: false,
					codec: fixedBytes32Codec
				},
				{
					name: "previousBlockHash512",
					wireKey: "prev512",
					optional: true,
					codec: fixedBytes64Codec
				},
				{
					name: "seed",
					wireKey: "seed",
					optional: false,
					codec: bytesCodec
				},
				{
					name: "txnCommitments",
					flattened: true,
					optional: false,
					codec: new ObjectModelCodec(TxnCommitmentsMeta)
				},
				{
					name: "timestamp",
					wireKey: "ts",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "genesisId",
					wireKey: "gen",
					optional: false,
					codec: stringCodec
				},
				{
					name: "genesisHash",
					wireKey: "gh",
					optional: false,
					codec: fixedBytes32Codec
				},
				{
					name: "proposer",
					wireKey: "prp",
					optional: true,
					codec: addressCodec
				},
				{
					name: "feesCollected",
					wireKey: "fc",
					optional: true,
					codec: bigIntCodec
				},
				{
					name: "bonus",
					wireKey: "bi",
					optional: true,
					codec: bigIntCodec
				},
				{
					name: "proposerPayout",
					wireKey: "pp",
					optional: true,
					codec: bigIntCodec
				},
				{
					name: "rewardState",
					flattened: true,
					optional: false,
					codec: new ObjectModelCodec(RewardStateMeta)
				},
				{
					name: "upgradeState",
					flattened: true,
					optional: false,
					codec: new ObjectModelCodec(UpgradeStateMeta)
				},
				{
					name: "upgradeVote",
					flattened: true,
					optional: true,
					codec: new ObjectModelCodec(UpgradeVoteMeta)
				},
				{
					name: "txnCounter",
					wireKey: "tc",
					optional: true,
					codec: bigIntCodec
				},
				{
					name: "stateProofTracking",
					wireKey: "spt",
					optional: true,
					codec: new MapCodec(numberCodec, new ObjectModelCodec(BlockStateProofTrackingDataMeta))
				},
				{
					name: "participationUpdates",
					flattened: true,
					optional: false,
					codec: new ObjectModelCodec(ParticipationUpdatesMeta)
				}
			]
		})
	}, {
		name: "payset",
		wireKey: "txns",
		optional: false,
		codec: new ArrayCodec(new ObjectModelCodec(SignedTxnInBlockMeta))
	}]
};
/**
* Custom codec for Block that populates genesis information on transactions after decoding.
*
* When blocks are returned from algod, transactions may not include the genesisId
* and genesisHash fields even though they are required for correct transaction ID calculation.
* The block contains `hasGenesisId` and `hasGenesisHash` flags that indicate whether these
* fields should be populated from the block header.
*
* This codec automatically populates these fields after decoding to ensure transaction IDs
* can be calculated correctly.
*/
var BlockCodec = class extends ObjectModelCodec {
	constructor() {
		super(BlockMeta);
	}
	fromEncoded(value, format) {
		const block = super.fromEncoded(value, format);
		const genesisId = block.header.genesisId;
		const genesisHash = block.header.genesisHash;
		for (const txnInBlock of block.payset ?? []) {
			const txn = txnInBlock.signedTxn.signedTxn.txn;
			if (txnInBlock.hasGenesisId && txn.genesisId === void 0) txn.genesisId = genesisId;
			if (txnInBlock.hasGenesisHash !== false && txn.genesisHash === void 0) txn.genesisHash = genesisHash;
		}
		return block;
	}
};
const blockCodec = new BlockCodec();

//#endregion
export { blockCodec };
