// @algorandfoundation/algokit-utils/packages/algod_client/src/models/build-version.js
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');

//#region packages/algod_client/src/models/build-version.ts
const BuildVersionMeta = {
	name: "BuildVersion",
	kind: "object",
	fields: [
		{
			name: "branch",
			wireKey: "branch",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "buildNumber",
			wireKey: "build_number",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "channel",
			wireKey: "channel",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "commitHash",
			wireKey: "commit_hash",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "major",
			wireKey: "major",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "minor",
			wireKey: "minor",
			optional: false,
			codec: require_number.numberCodec
		}
	]
};

//#endregion
exports.BuildVersionMeta = BuildVersionMeta;
