// @algorandfoundation/algokit-utils/packages/algod_client/src/models/genesis-allocation.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";

//#region packages/algod_client/src/models/genesis-allocation.ts
const GenesisAllocationStateMeta = {
	name: "GenesisAllocationStateMeta",
	kind: "object",
	fields: [
		{
			name: "algo",
			wireKey: "algo",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "onl",
			wireKey: "onl",
			optional: false,
			codec: numberCodec
		},
		{
			name: "sel",
			wireKey: "sel",
			optional: true,
			codec: stringCodec
		},
		{
			name: "stprf",
			wireKey: "stprf",
			optional: true,
			codec: stringCodec
		},
		{
			name: "vote",
			wireKey: "vote",
			optional: true,
			codec: stringCodec
		},
		{
			name: "voteKd",
			wireKey: "voteKD",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteFst",
			wireKey: "voteFst",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteLst",
			wireKey: "voteLst",
			optional: true,
			codec: bigIntCodec
		}
	]
};
const GenesisAllocationMeta = {
	name: "GenesisAllocation",
	kind: "object",
	fields: [
		{
			name: "addr",
			wireKey: "addr",
			optional: false,
			codec: stringCodec
		},
		{
			name: "comment",
			wireKey: "comment",
			optional: false,
			codec: stringCodec
		},
		{
			name: "state",
			wireKey: "state",
			optional: false,
			codec: new ObjectModelCodec(GenesisAllocationStateMeta)
		}
	]
};

//#endregion
export { GenesisAllocationMeta };
