// @algorandfoundation/algokit-utils/packages/algod_client/src/models/genesis.mjs
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { GenesisAllocationMeta } from "./genesis-allocation.mjs";

//#region packages/algod_client/src/models/genesis.ts
const GenesisMeta = {
	name: "Genesis",
	kind: "object",
	fields: [
		{
			name: "alloc",
			wireKey: "alloc",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec(GenesisAllocationMeta))
		},
		{
			name: "comment",
			wireKey: "comment",
			optional: true,
			codec: stringCodec
		},
		{
			name: "devmode",
			wireKey: "devmode",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "fees",
			wireKey: "fees",
			optional: false,
			codec: stringCodec
		},
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: stringCodec
		},
		{
			name: "network",
			wireKey: "network",
			optional: false,
			codec: stringCodec
		},
		{
			name: "proto",
			wireKey: "proto",
			optional: false,
			codec: stringCodec
		},
		{
			name: "rwd",
			wireKey: "rwd",
			optional: false,
			codec: stringCodec
		},
		{
			name: "timestamp",
			wireKey: "timestamp",
			optional: true,
			codec: numberCodec
		}
	]
};

//#endregion
export { GenesisMeta };
