// @algorandfoundation/algokit-utils/packages/algod_client/src/models/ledger-state-delta.js
const require_address = require('../../../common/src/codecs/primitives/address.js');
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_map = require('../../../common/src/codecs/composite/map.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_block = require('./block.js');

//#region packages/algod_client/src/models/ledger-state-delta.ts
const LedgerTealValueMeta = {
	name: "LedgerTealValue",
	kind: "object",
	fields: [
		{
			name: "type",
			wireKey: "tt",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "bytes",
			wireKey: "tb",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "uint",
			wireKey: "ui",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};
const LedgerStateSchemaMeta = {
	name: "LedgerStateSchema",
	kind: "object",
	fields: [{
		name: "numUints",
		wireKey: "nui",
		optional: true,
		codec: require_bigint.bigIntCodec
	}, {
		name: "numByteSlices",
		wireKey: "nbs",
		optional: true,
		codec: require_bigint.bigIntCodec
	}]
};
const LedgerAppParamsMeta = {
	name: "LedgerAppParams",
	kind: "object",
	fields: [
		{
			name: "approvalProgram",
			wireKey: "approv",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "clearp",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "localStateSchema",
			wireKey: "lsch",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerStateSchemaMeta)
		},
		{
			name: "globalStateSchema",
			wireKey: "gsch",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerStateSchemaMeta)
		},
		{
			name: "extraProgramPages",
			wireKey: "epp",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "version",
			wireKey: "v",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "sizeSponsor",
			wireKey: "ss",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "globalState",
			wireKey: "gs",
			optional: true,
			codec: new require_map.MapCodec(require_bytes.bytesCodec, new require_object_model.ObjectModelCodec(LedgerTealValueMeta))
		}
	]
};
const LedgerAppLocalStateMeta = {
	name: "LedgerAppLocalState",
	kind: "object",
	fields: [{
		name: "schema",
		wireKey: "hsch",
		optional: false,
		codec: new require_object_model.ObjectModelCodec(LedgerStateSchemaMeta)
	}, {
		name: "keyValue",
		wireKey: "tkv",
		optional: true,
		codec: new require_map.MapCodec(require_bytes.bytesCodec, new require_object_model.ObjectModelCodec(LedgerTealValueMeta))
	}]
};
const LedgerAppLocalStateDeltaMeta = {
	name: "LedgerAppLocalStateDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: require_boolean.booleanCodec
	}, {
		name: "localState",
		wireKey: "LocalState",
		optional: true,
		codec: new require_object_model.ObjectModelCodec(LedgerAppLocalStateMeta)
	}]
};
const LedgerAppParamsDeltaMeta = {
	name: "LedgerAppParamsDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: require_boolean.booleanCodec
	}, {
		name: "params",
		wireKey: "Params",
		optional: true,
		codec: new require_object_model.ObjectModelCodec(LedgerAppParamsMeta)
	}]
};
const LedgerAppResourceRecordMeta = {
	name: "LedgerAppResourceRecord",
	kind: "object",
	fields: [
		{
			name: "appId",
			wireKey: "Aidx",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "address",
			wireKey: "Addr",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "params",
			wireKey: "Params",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAppParamsDeltaMeta)
		},
		{
			name: "state",
			wireKey: "State",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAppLocalStateDeltaMeta)
		}
	]
};
const LedgerAssetHoldingMeta = {
	name: "LedgerAssetHolding",
	kind: "object",
	fields: [{
		name: "amount",
		wireKey: "a",
		optional: false,
		codec: require_bigint.bigIntCodec
	}, {
		name: "frozen",
		wireKey: "f",
		optional: false,
		codec: require_boolean.booleanCodec
	}]
};
const LedgerAssetHoldingDeltaMeta = {
	name: "LedgerAssetHoldingDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: require_boolean.booleanCodec
	}, {
		name: "holding",
		wireKey: "Holding",
		optional: true,
		codec: new require_object_model.ObjectModelCodec(LedgerAssetHoldingMeta)
	}]
};
const LedgerAssetParamsMeta = {
	name: "LedgerAssetParams",
	kind: "object",
	fields: [
		{
			name: "total",
			wireKey: "t",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "decimals",
			wireKey: "dc",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "defaultFrozen",
			wireKey: "df",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "unitName",
			wireKey: "un",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "assetName",
			wireKey: "an",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "url",
			wireKey: "au",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "metadataHash",
			wireKey: "am",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "manager",
			wireKey: "m",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "reserve",
			wireKey: "r",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "freeze",
			wireKey: "f",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "clawback",
			wireKey: "c",
			optional: true,
			codec: require_address.addressCodec
		}
	]
};
const LedgerAssetParamsDeltaMeta = {
	name: "LedgerAssetParamsDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: require_boolean.booleanCodec
	}, {
		name: "params",
		wireKey: "Params",
		optional: true,
		codec: new require_object_model.ObjectModelCodec(LedgerAssetParamsMeta)
	}]
};
const LedgerAssetResourceRecordMeta = {
	name: "LedgerAssetResourceRecord",
	kind: "object",
	fields: [
		{
			name: "assetId",
			wireKey: "Aidx",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "address",
			wireKey: "Addr",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "params",
			wireKey: "Params",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAssetParamsDeltaMeta)
		},
		{
			name: "holding",
			wireKey: "Holding",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAssetHoldingDeltaMeta)
		}
	]
};
const LedgerVotingDataMeta = {
	name: "LedgerVotingData",
	kind: "object",
	fields: [
		{
			name: "voteId",
			wireKey: "VoteID",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "selectionId",
			wireKey: "SelectionID",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "stateProofId",
			wireKey: "StateProofID",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "voteFirstValid",
			wireKey: "VoteFirstValid",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteLastValid",
			wireKey: "VoteLastValid",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "VoteKeyDilution",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};
const LedgerAccountBaseDataMeta = {
	name: "LedgerAccountBaseData",
	kind: "object",
	fields: [
		{
			name: "status",
			wireKey: "Status",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "microAlgos",
			wireKey: "MicroAlgos",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsBase",
			wireKey: "RewardsBase",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardedMicroAlgos",
			wireKey: "RewardedMicroAlgos",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "authAddress",
			wireKey: "AuthAddr",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "incentiveEligible",
			wireKey: "IncentiveEligible",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "totalAppSchema",
			wireKey: "TotalAppSchema",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerStateSchemaMeta)
		},
		{
			name: "totalExtraAppPages",
			wireKey: "TotalExtraAppPages",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalAppParams",
			wireKey: "TotalAppParams",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalAppLocalStates",
			wireKey: "TotalAppLocalStates",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalAssetParams",
			wireKey: "TotalAssetParams",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalAssets",
			wireKey: "TotalAssets",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "totalBoxes",
			wireKey: "TotalBoxes",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "totalBoxBytes",
			wireKey: "TotalBoxBytes",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastProposed",
			wireKey: "LastProposed",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastHeartbeat",
			wireKey: "LastHeartbeat",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};
const LedgerAccountDataMeta = {
	name: "LedgerAccountData",
	kind: "object",
	fields: [{
		name: "accountBaseData",
		flattened: true,
		optional: false,
		codec: new require_object_model.ObjectModelCodec(LedgerAccountBaseDataMeta)
	}, {
		name: "votingData",
		flattened: true,
		optional: false,
		codec: new require_object_model.ObjectModelCodec(LedgerVotingDataMeta)
	}]
};
const LedgerBalanceRecordMeta = {
	name: "LedgerBalanceRecord",
	kind: "object",
	fields: [{
		name: "address",
		wireKey: "Addr",
		optional: false,
		codec: require_address.addressCodec
	}, {
		name: "accountData",
		flattened: true,
		optional: false,
		codec: new require_object_model.ObjectModelCodec(LedgerAccountDataMeta)
	}]
};
const LedgerAccountDeltasMeta = {
	name: "LedgerAccountDeltas",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "Accts",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(LedgerBalanceRecordMeta))
		},
		{
			name: "appResources",
			wireKey: "AppResources",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(LedgerAppResourceRecordMeta))
		},
		{
			name: "assetResources",
			wireKey: "AssetResources",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(LedgerAssetResourceRecordMeta))
		}
	]
};
const LedgerKvValueDeltaMeta = {
	name: "LedgerKvValueDelta",
	kind: "object",
	fields: [{
		name: "data",
		wireKey: "Data",
		optional: true,
		codec: require_bytes.bytesCodec
	}, {
		name: "oldData",
		wireKey: "OldData",
		optional: true,
		codec: require_bytes.bytesCodec
	}]
};
const LedgerIncludedTransactionsMeta = {
	name: "LedgerIncludedTransactions",
	kind: "object",
	fields: [{
		name: "lastValid",
		wireKey: "LastValid",
		optional: false,
		codec: require_bigint.bigIntCodec
	}, {
		name: "intra",
		wireKey: "Intra",
		optional: false,
		codec: require_number.numberCodec
	}]
};
const LedgerModifiedCreatableMeta = {
	name: "LedgerModifiedCreatable",
	kind: "object",
	fields: [
		{
			name: "creatableType",
			wireKey: "Ctype",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "created",
			wireKey: "Created",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "creator",
			wireKey: "Creator",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "ndeltas",
			wireKey: "Ndeltas",
			optional: false,
			codec: require_number.numberCodec
		}
	]
};
const LedgerAlgoCountMeta = {
	name: "LedgerAlgoCount",
	kind: "object",
	fields: [{
		name: "money",
		wireKey: "mon",
		optional: false,
		codec: require_bigint.bigIntCodec
	}, {
		name: "rewardUnits",
		wireKey: "rwd",
		optional: false,
		codec: require_bigint.bigIntCodec
	}]
};
const LedgerAccountTotalsMeta = {
	name: "LedgerAccountTotals",
	kind: "object",
	fields: [
		{
			name: "online",
			wireKey: "online",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAlgoCountMeta)
		},
		{
			name: "offline",
			wireKey: "offline",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAlgoCountMeta)
		},
		{
			name: "notParticipating",
			wireKey: "notpart",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAlgoCountMeta)
		},
		{
			name: "rewardsLevel",
			wireKey: "rwdlvl",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};
const LedgerStateDeltaMeta = {
	name: "LedgerStateDelta",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "Accts",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAccountDeltasMeta)
		},
		{
			name: "block",
			wireKey: "Hdr",
			optional: false,
			codec: require_block.blockCodec
		},
		{
			name: "stateProofNext",
			wireKey: "StateProofNext",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "prevTimestamp",
			wireKey: "PrevTimestamp",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "totals",
			wireKey: "Totals",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(LedgerAccountTotalsMeta)
		},
		{
			name: "kvMods",
			wireKey: "KvMods",
			optional: true,
			codec: new require_map.MapCodec(require_bytes.bytesCodec, new require_object_model.ObjectModelCodec(LedgerKvValueDeltaMeta))
		},
		{
			name: "txIds",
			wireKey: "Txids",
			optional: true,
			codec: new require_map.MapCodec(require_bytes.bytesCodec, new require_object_model.ObjectModelCodec(LedgerIncludedTransactionsMeta))
		},
		{
			name: "creatables",
			wireKey: "Creatables",
			optional: true,
			codec: new require_map.MapCodec(require_number.numberCodec, new require_object_model.ObjectModelCodec(LedgerModifiedCreatableMeta))
		}
	]
};

//#endregion
exports.LedgerStateDeltaMeta = LedgerStateDeltaMeta;
