// @algorandfoundation/algokit-utils/packages/algod_client/src/models/node-status-response.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');

//#region packages/algod_client/src/models/node-status-response.ts
const NodeStatusResponseMeta = {
	name: "NodeStatusResponse",
	kind: "object",
	fields: [
		{
			name: "catchupTime",
			wireKey: "catchup-time",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastRound",
			wireKey: "last-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastVersion",
			wireKey: "last-version",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "nextVersion",
			wireKey: "next-version",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "nextVersionRound",
			wireKey: "next-version-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextVersionSupported",
			wireKey: "next-version-supported",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "stoppedAtUnsupportedRound",
			wireKey: "stopped-at-unsupported-round",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "timeSinceLastRound",
			wireKey: "time-since-last-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastCatchpoint",
			wireKey: "last-catchpoint",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "catchpoint",
			wireKey: "catchpoint",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "catchpointTotalAccounts",
			wireKey: "catchpoint-total-accounts",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointProcessedAccounts",
			wireKey: "catchpoint-processed-accounts",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointVerifiedAccounts",
			wireKey: "catchpoint-verified-accounts",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointTotalKvs",
			wireKey: "catchpoint-total-kvs",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointProcessedKvs",
			wireKey: "catchpoint-processed-kvs",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointVerifiedKvs",
			wireKey: "catchpoint-verified-kvs",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointTotalBlocks",
			wireKey: "catchpoint-total-blocks",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "catchpointAcquiredBlocks",
			wireKey: "catchpoint-acquired-blocks",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "upgradeDelay",
			wireKey: "upgrade-delay",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "upgradeNodeVote",
			wireKey: "upgrade-node-vote",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "upgradeVotesRequired",
			wireKey: "upgrade-votes-required",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "upgradeVotes",
			wireKey: "upgrade-votes",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "upgradeYesVotes",
			wireKey: "upgrade-yes-votes",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "upgradeNoVotes",
			wireKey: "upgrade-no-votes",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "upgradeNextProtocolVoteBefore",
			wireKey: "upgrade-next-protocol-vote-before",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "upgradeVoteRounds",
			wireKey: "upgrade-vote-rounds",
			optional: true,
			codec: require_number.numberCodec
		}
	]
};

//#endregion
exports.NodeStatusResponseMeta = NodeStatusResponseMeta;
