// @algorandfoundation/algokit-utils/packages/algod_client/src/models/pending-transaction-response.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_array_model = require('../../../common/src/codecs/models/array-model.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_signed_transaction_meta = require('../../../transact/src/transactions/signed-transaction-meta.js');
const require_state_delta = require('./state-delta.js');
const require_account_state_delta = require('./account-state-delta.js');

//#region packages/algod_client/src/models/pending-transaction-response.ts
const PendingTransactionResponseMeta = {
	name: "PendingTransactionResponse",
	kind: "object",
	fields: [
		{
			name: "assetId",
			wireKey: "asset-index",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "appId",
			wireKey: "application-index",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closeRewards",
			wireKey: "close-rewards",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closingAmount",
			wireKey: "closing-amount",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "assetClosingAmount",
			wireKey: "asset-closing-amount",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "confirmedRound",
			wireKey: "confirmed-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "poolError",
			wireKey: "pool-error",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "receiverRewards",
			wireKey: "receiver-rewards",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "senderRewards",
			wireKey: "sender-rewards",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "localStateDelta",
			wireKey: "local-state-delta",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_account_state_delta.AccountStateDeltaMeta))
		},
		{
			name: "globalStateDelta",
			wireKey: "global-state-delta",
			optional: true,
			codec: new require_array_model.ArrayModelCodec(require_state_delta.StateDeltaMeta)
		},
		{
			name: "logs",
			wireKey: "logs",
			optional: true,
			codec: require_array.bytesArrayCodec
		},
		{
			name: "innerTxns",
			wireKey: "inner-txns",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(() => PendingTransactionResponseMeta))
		},
		{
			name: "txn",
			wireKey: "txn",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_signed_transaction_meta.SignedTransactionMeta)
		}
	]
};

//#endregion
exports.PendingTransactionResponseMeta = PendingTransactionResponseMeta;
