// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-transaction-group-result.mjs
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec, numberArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { SimulateUnnamedResourcesAccessedMeta } from "./simulate-unnamed-resources-accessed.mjs";
import { SimulateTransactionResultMeta } from "./simulate-transaction-result.mjs";

//#region packages/algod_client/src/models/simulate-transaction-group-result.ts
const SimulateTransactionGroupResultMeta = {
	name: "SimulateTransactionGroupResult",
	kind: "object",
	fields: [
		{
			name: "txnResults",
			wireKey: "txn-results",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec(SimulateTransactionResultMeta))
		},
		{
			name: "failureMessage",
			wireKey: "failure-message",
			optional: true,
			codec: stringCodec
		},
		{
			name: "failedAt",
			wireKey: "failed-at",
			optional: true,
			codec: numberArrayCodec
		},
		{
			name: "appBudgetAdded",
			wireKey: "app-budget-added",
			optional: true,
			codec: numberCodec
		},
		{
			name: "appBudgetConsumed",
			wireKey: "app-budget-consumed",
			optional: true,
			codec: numberCodec
		},
		{
			name: "unnamedResourcesAccessed",
			wireKey: "unnamed-resources-accessed",
			optional: true,
			codec: new ObjectModelCodec(SimulateUnnamedResourcesAccessedMeta)
		}
	]
};

//#endregion
export { SimulateTransactionGroupResultMeta };
