// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-unnamed-resources-accessed.mjs
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec, addressArrayCodec, bigIntArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { BoxReferenceMeta, HoldingReferenceMeta, LocalsReferenceMeta } from "../../../transact/src/transactions/reference-types-meta.mjs";

//#region packages/algod_client/src/models/simulate-unnamed-resources-accessed.ts
const SimulateUnnamedResourcesAccessedMeta = {
	name: "SimulateUnnamedResourcesAccessed",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "accounts",
			optional: true,
			codec: addressArrayCodec
		},
		{
			name: "assets",
			wireKey: "assets",
			optional: true,
			codec: bigIntArrayCodec
		},
		{
			name: "apps",
			wireKey: "apps",
			optional: true,
			codec: bigIntArrayCodec
		},
		{
			name: "boxes",
			wireKey: "boxes",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(BoxReferenceMeta))
		},
		{
			name: "extraBoxRefs",
			wireKey: "extra-box-refs",
			optional: true,
			codec: numberCodec
		},
		{
			name: "assetHoldings",
			wireKey: "asset-holdings",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(HoldingReferenceMeta))
		},
		{
			name: "appLocals",
			wireKey: "app-locals",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(LocalsReferenceMeta))
		}
	]
};

//#endregion
export { SimulateUnnamedResourcesAccessedMeta };
