// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulation-opcode-trace-unit.js
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_avm_value = require('./avm-value.js');
const require_application_state_operation = require('./application-state-operation.js');
const require_scratch_change = require('./scratch-change.js');

//#region packages/algod_client/src/models/simulation-opcode-trace-unit.ts
const SimulationOpcodeTraceUnitMeta = {
	name: "SimulationOpcodeTraceUnit",
	kind: "object",
	fields: [
		{
			name: "pc",
			wireKey: "pc",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "scratchChanges",
			wireKey: "scratch-changes",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_scratch_change.ScratchChangeMeta))
		},
		{
			name: "stateChanges",
			wireKey: "state-changes",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_application_state_operation.ApplicationStateOperationMeta))
		},
		{
			name: "spawnedInners",
			wireKey: "spawned-inners",
			optional: true,
			codec: require_array.numberArrayCodec
		},
		{
			name: "stackPopCount",
			wireKey: "stack-pop-count",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "stackAdditions",
			wireKey: "stack-additions",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_avm_value.AvmValueMeta))
		}
	]
};

//#endregion
exports.SimulationOpcodeTraceUnitMeta = SimulationOpcodeTraceUnitMeta;
