// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-request.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_simulate_trace_config = require('./simulate-trace-config.js');
const require_simulate_request_transaction_group = require('./simulate-request-transaction-group.js');

//#region packages/algod_client/src/models/simulate-request.ts
const SimulateRequestMeta = {
	name: "SimulateRequest",
	kind: "object",
	fields: [
		{
			name: "txnGroups",
			wireKey: "txn-groups",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_simulate_request_transaction_group.SimulateRequestTransactionGroupMeta))
		},
		{
			name: "round",
			wireKey: "round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "allowEmptySignatures",
			wireKey: "allow-empty-signatures",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "allowMoreLogging",
			wireKey: "allow-more-logging",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "allowUnnamedResources",
			wireKey: "allow-unnamed-resources",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "extraOpcodeBudget",
			wireKey: "extra-opcode-budget",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "execTraceConfig",
			wireKey: "exec-trace-config",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulate_trace_config.SimulateTraceConfigMeta)
		},
		{
			name: "fixSigners",
			wireKey: "fix-signers",
			optional: true,
			codec: require_boolean.booleanCodec
		}
	]
};

//#endregion
exports.SimulateRequestMeta = SimulateRequestMeta;
