// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulation-transaction-exec-trace.mjs
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { SimulationOpcodeTraceUnitMeta } from "./simulation-opcode-trace-unit.mjs";

//#region packages/algod_client/src/models/simulation-transaction-exec-trace.ts
const SimulationTransactionExecTraceMeta = {
	name: "SimulationTransactionExecTrace",
	kind: "object",
	fields: [
		{
			name: "approvalProgramTrace",
			wireKey: "approval-program-trace",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(SimulationOpcodeTraceUnitMeta))
		},
		{
			name: "approvalProgramHash",
			wireKey: "approval-program-hash",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "clearStateProgramTrace",
			wireKey: "clear-state-program-trace",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(SimulationOpcodeTraceUnitMeta))
		},
		{
			name: "clearStateProgramHash",
			wireKey: "clear-state-program-hash",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "clearStateRollback",
			wireKey: "clear-state-rollback",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "clearStateRollbackError",
			wireKey: "clear-state-rollback-error",
			optional: true,
			codec: stringCodec
		},
		{
			name: "logicSigTrace",
			wireKey: "logic-sig-trace",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(SimulationOpcodeTraceUnitMeta))
		},
		{
			name: "logicSigHash",
			wireKey: "logic-sig-hash",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "innerTrace",
			wireKey: "inner-trace",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(() => SimulationTransactionExecTraceMeta))
		}
	]
};

//#endregion
export { SimulationTransactionExecTraceMeta };
