// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-response.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_simulate_initial_states = require('./simulate-initial-states.js');
const require_simulate_trace_config = require('./simulate-trace-config.js');
const require_simulate_transaction_group_result = require('./simulate-transaction-group-result.js');
const require_simulation_eval_overrides = require('./simulation-eval-overrides.js');
const require_model_runtime = require('../core/model-runtime.js');

//#region packages/algod_client/src/models/simulate-response.ts
const SimulateResponseMeta = {
	name: "SimulateResponse",
	kind: "object",
	fields: [
		{
			name: "version",
			wireKey: "version",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "lastRound",
			wireKey: "last-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "txnGroups",
			wireKey: "txn-groups",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_simulate_transaction_group_result.SimulateTransactionGroupResultMeta))
		},
		{
			name: "evalOverrides",
			wireKey: "eval-overrides",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulation_eval_overrides.SimulationEvalOverridesMeta)
		},
		{
			name: "execTraceConfig",
			wireKey: "exec-trace-config",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulate_trace_config.SimulateTraceConfigMeta)
		},
		{
			name: "initialStates",
			wireKey: "initial-states",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulate_initial_states.SimulateInitialStatesMeta)
		}
	]
};
function encodeSimulateResponseToJson(simulateResponse) {
	return require_model_runtime.encodeJson(simulateResponse, SimulateResponseMeta, 2);
}
function decodeSimulateResponseFromJson(json) {
	return require_model_runtime.decodeJson(json, SimulateResponseMeta);
}

//#endregion
exports.SimulateResponseMeta = SimulateResponseMeta;
exports.decodeSimulateResponseFromJson = decodeSimulateResponseFromJson;
exports.encodeSimulateResponseToJson = encodeSimulateResponseToJson;
