// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-response.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { SimulateInitialStatesMeta } from "./simulate-initial-states.mjs";
import { SimulateTraceConfigMeta } from "./simulate-trace-config.mjs";
import { SimulateTransactionGroupResultMeta } from "./simulate-transaction-group-result.mjs";
import { SimulationEvalOverridesMeta } from "./simulation-eval-overrides.mjs";
import { decodeJson, encodeJson } from "../core/model-runtime.mjs";

//#region packages/algod_client/src/models/simulate-response.ts
const SimulateResponseMeta = {
	name: "SimulateResponse",
	kind: "object",
	fields: [
		{
			name: "version",
			wireKey: "version",
			optional: false,
			codec: numberCodec
		},
		{
			name: "lastRound",
			wireKey: "last-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "txnGroups",
			wireKey: "txn-groups",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec(SimulateTransactionGroupResultMeta))
		},
		{
			name: "evalOverrides",
			wireKey: "eval-overrides",
			optional: true,
			codec: new ObjectModelCodec(SimulationEvalOverridesMeta)
		},
		{
			name: "execTraceConfig",
			wireKey: "exec-trace-config",
			optional: true,
			codec: new ObjectModelCodec(SimulateTraceConfigMeta)
		},
		{
			name: "initialStates",
			wireKey: "initial-states",
			optional: true,
			codec: new ObjectModelCodec(SimulateInitialStatesMeta)
		}
	]
};
function encodeSimulateResponseToJson(simulateResponse) {
	return encodeJson(simulateResponse, SimulateResponseMeta, 2);
}
function decodeSimulateResponseFromJson(json) {
	return decodeJson(json, SimulateResponseMeta);
}

//#endregion
export { SimulateResponseMeta, decodeSimulateResponseFromJson, encodeSimulateResponseToJson };
