// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulation-opcode-trace-unit.mjs
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec, numberArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { AvmValueMeta } from "./avm-value.mjs";
import { ApplicationStateOperationMeta } from "./application-state-operation.mjs";
import { ScratchChangeMeta } from "./scratch-change.mjs";

//#region packages/algod_client/src/models/simulation-opcode-trace-unit.ts
const SimulationOpcodeTraceUnitMeta = {
	name: "SimulationOpcodeTraceUnit",
	kind: "object",
	fields: [
		{
			name: "pc",
			wireKey: "pc",
			optional: false,
			codec: numberCodec
		},
		{
			name: "scratchChanges",
			wireKey: "scratch-changes",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ScratchChangeMeta))
		},
		{
			name: "stateChanges",
			wireKey: "state-changes",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ApplicationStateOperationMeta))
		},
		{
			name: "spawnedInners",
			wireKey: "spawned-inners",
			optional: true,
			codec: numberArrayCodec
		},
		{
			name: "stackPopCount",
			wireKey: "stack-pop-count",
			optional: true,
			codec: numberCodec
		},
		{
			name: "stackAdditions",
			wireKey: "stack-additions",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(AvmValueMeta))
		}
	]
};

//#endregion
export { SimulationOpcodeTraceUnitMeta };
