// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-request.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { SimulateTraceConfigMeta } from "./simulate-trace-config.mjs";
import { SimulateRequestTransactionGroupMeta } from "./simulate-request-transaction-group.mjs";

//#region packages/algod_client/src/models/simulate-request.ts
const SimulateRequestMeta = {
	name: "SimulateRequest",
	kind: "object",
	fields: [
		{
			name: "txnGroups",
			wireKey: "txn-groups",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec(SimulateRequestTransactionGroupMeta))
		},
		{
			name: "round",
			wireKey: "round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "allowEmptySignatures",
			wireKey: "allow-empty-signatures",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "allowMoreLogging",
			wireKey: "allow-more-logging",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "allowUnnamedResources",
			wireKey: "allow-unnamed-resources",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "extraOpcodeBudget",
			wireKey: "extra-opcode-budget",
			optional: true,
			codec: numberCodec
		},
		{
			name: "execTraceConfig",
			wireKey: "exec-trace-config",
			optional: true,
			codec: new ObjectModelCodec(SimulateTraceConfigMeta)
		},
		{
			name: "fixSigners",
			wireKey: "fix-signers",
			optional: true,
			codec: booleanCodec
		}
	]
};

//#endregion
export { SimulateRequestMeta };
