// @algorandfoundation/algokit-utils/packages/algod_client/src/models/transaction-parameters-response.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { fixedBytes32Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";

//#region packages/algod_client/src/models/transaction-parameters-response.ts
const TransactionParametersResponseMeta = {
	name: "TransactionParametersResponse",
	kind: "object",
	fields: [
		{
			name: "consensusVersion",
			wireKey: "consensus-version",
			optional: false,
			codec: stringCodec
		},
		{
			name: "fee",
			wireKey: "fee",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "genesisHash",
			wireKey: "genesis-hash",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "genesis-id",
			optional: false,
			codec: stringCodec
		},
		{
			name: "lastRound",
			wireKey: "last-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "minFee",
			wireKey: "min-fee",
			optional: false,
			codec: bigIntCodec
		}
	]
};

//#endregion
export { TransactionParametersResponseMeta };
