// @algorandfoundation/algokit-utils/packages/algod_client/src/models/transaction-parameters-response.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_fixed_bytes = require('../../../common/src/codecs/primitives/fixed-bytes.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');

//#region packages/algod_client/src/models/transaction-parameters-response.ts
const TransactionParametersResponseMeta = {
	name: "TransactionParametersResponse",
	kind: "object",
	fields: [
		{
			name: "consensusVersion",
			wireKey: "consensus-version",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "fee",
			wireKey: "fee",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "genesisHash",
			wireKey: "genesis-hash",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "genesis-id",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "lastRound",
			wireKey: "last-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "minFee",
			wireKey: "min-fee",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};

//#endregion
exports.TransactionParametersResponseMeta = TransactionParametersResponseMeta;
