// @algorandfoundation/algokit-utils/packages/algod_client/src/models/version.js
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_build_version = require('./build-version.js');

//#region packages/algod_client/src/models/version.ts
const VersionMeta = {
	name: "Version",
	kind: "object",
	fields: [
		{
			name: "build",
			wireKey: "build",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_build_version.BuildVersionMeta)
		},
		{
			name: "genesisHashB64",
			wireKey: "genesis_hash_b64",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "genesisId",
			wireKey: "genesis_id",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "versions",
			wireKey: "versions",
			optional: false,
			codec: require_array.stringArrayCodec
		}
	]
};

//#endregion
exports.VersionMeta = VersionMeta;
