// @algorandfoundation/algokit-utils/transactions/asset-transfer.js
const require_address = require('../packages/common/src/address.js');
const require_transaction_type = require('../packages/transact/src/transactions/transaction-type.js');
const require_transaction = require('../packages/transact/src/transactions/transaction.js');
const require_common = require('./common.js');

//#region src/transactions/asset-transfer.ts
const buildAssetTransfer = (params, suggestedParams, defaultValidityWindow) => {
	return new require_transaction.Transaction({
		...require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: require_transaction_type.TransactionType.AssetTransfer,
		assetTransfer: {
			assetId: params.assetId,
			amount: params.amount,
			receiver: require_address.getAddress(params.receiver),
			assetSender: require_address.getOptionalAddress(params.clawbackTarget),
			closeRemainderTo: require_address.getOptionalAddress(params.closeAssetTo)
		}
	});
};
const buildAssetOptIn = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	return new require_transaction.Transaction({
		...commonData,
		type: require_transaction_type.TransactionType.AssetTransfer,
		assetTransfer: {
			assetId: params.assetId,
			amount: 0n,
			receiver: commonData.sender
		}
	});
};
const buildAssetOptOut = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	return new require_transaction.Transaction({
		...commonData,
		type: require_transaction_type.TransactionType.AssetTransfer,
		assetTransfer: {
			assetId: params.assetId,
			amount: 0n,
			receiver: commonData.sender,
			closeRemainderTo: require_address.getOptionalAddress(params.creator)
		}
	});
};

//#endregion
exports.buildAssetOptIn = buildAssetOptIn;
exports.buildAssetOptOut = buildAssetOptOut;
exports.buildAssetTransfer = buildAssetTransfer;
