// @algorandfoundation/algokit-utils/transactions/asset-config.mjs
import { getAddress, getOptionalAddress } from "../packages/common/src/address.mjs";
import { TransactionType } from "../packages/transact/src/transactions/transaction-type.mjs";
import { Transaction } from "../packages/transact/src/transactions/transaction.mjs";
import { buildTransactionCommonData, ensureString } from "./common.mjs";

//#region src/transactions/asset-config.ts
const buildAssetCreate = (params, suggestedParams, defaultValidityWindow) => {
	return new Transaction({
		...buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: TransactionType.AssetConfig,
		assetConfig: {
			assetId: 0n,
			total: params.total,
			decimals: params.decimals,
			defaultFrozen: params.defaultFrozen,
			assetName: params.assetName,
			unitName: params.unitName,
			url: params.url,
			metadataHash: ensureString(params.metadataHash),
			manager: getOptionalAddress(params.manager),
			reserve: getOptionalAddress(params.reserve),
			freeze: getOptionalAddress(params.freeze),
			clawback: getOptionalAddress(params.clawback)
		}
	});
};
const buildAssetConfig = (params, suggestedParams, defaultValidityWindow) => {
	return new Transaction({
		...buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: TransactionType.AssetConfig,
		assetConfig: {
			assetId: params.assetId,
			manager: getOptionalAddress(params.manager),
			reserve: getOptionalAddress(params.reserve),
			freeze: getOptionalAddress(params.freeze),
			clawback: getOptionalAddress(params.clawback)
		}
	});
};
const buildAssetFreeze = (params, suggestedParams, defaultValidityWindow) => {
	return new Transaction({
		...buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: TransactionType.AssetFreeze,
		assetFreeze: {
			assetId: params.assetId,
			freezeTarget: getAddress(params.freezeTarget),
			frozen: params.frozen
		}
	});
};
const buildAssetDestroy = (params, suggestedParams, defaultValidityWindow) => {
	return new Transaction({
		...buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: TransactionType.AssetConfig,
		assetConfig: { assetId: params.assetId }
	});
};

//#endregion
export { buildAssetConfig, buildAssetCreate, buildAssetDestroy, buildAssetFreeze };
