// @algorandfoundation/algokit-utils/transactions/common.js
const require_address = require('../packages/common/src/address.js');
const require_transaction = require('../transaction/transaction.js');

//#region src/transactions/common.ts
const ensureString = (data) => {
	if (data === void 0) return void 0;
	const encoder = new TextEncoder();
	return typeof data === "string" ? encoder.encode(data) : data;
};
const buildTransactionCommonData = (commonParams, suggestedParams, defaultValidityWindow) => {
	const firstValid = commonParams.firstValidRound ?? suggestedParams.firstValid;
	const lease = commonParams.lease === void 0 ? void 0 : require_transaction.encodeLease(commonParams.lease);
	const note = ensureString(commonParams.note);
	return {
		sender: require_address.getAddress(commonParams.sender),
		rekeyTo: require_address.getOptionalAddress(commonParams.rekeyTo),
		note,
		lease,
		fee: commonParams.staticFee?.microAlgos,
		genesisId: suggestedParams.genesisId,
		genesisHash: suggestedParams.genesisHash,
		firstValid,
		lastValid: commonParams.lastValidRound ?? (commonParams.validityWindow !== void 0 ? firstValid + BigInt(commonParams.validityWindow) : firstValid + defaultValidityWindow),
		group: void 0
	};
};

//#endregion
exports.buildTransactionCommonData = buildTransactionCommonData;
exports.ensureString = ensureString;
