// @algorandfoundation/algokit-utils/transactions/common.mjs
import { getAddress, getOptionalAddress } from "../packages/common/src/address.mjs";
import { encodeLease } from "../transaction/transaction.mjs";

//#region src/transactions/common.ts
const ensureString = (data) => {
	if (data === void 0) return void 0;
	const encoder = new TextEncoder();
	return typeof data === "string" ? encoder.encode(data) : data;
};
const buildTransactionCommonData = (commonParams, suggestedParams, defaultValidityWindow) => {
	const firstValid = commonParams.firstValidRound ?? suggestedParams.firstValid;
	const lease = commonParams.lease === void 0 ? void 0 : encodeLease(commonParams.lease);
	const note = ensureString(commonParams.note);
	return {
		sender: getAddress(commonParams.sender),
		rekeyTo: getOptionalAddress(commonParams.rekeyTo),
		note,
		lease,
		fee: commonParams.staticFee?.microAlgos,
		genesisId: suggestedParams.genesisId,
		genesisHash: suggestedParams.genesisHash,
		firstValid,
		lastValid: commonParams.lastValidRound ?? (commonParams.validityWindow !== void 0 ? firstValid + BigInt(commonParams.validityWindow) : firstValid + defaultValidityWindow),
		group: void 0
	};
};

//#endregion
export { buildTransactionCommonData, ensureString };
