// @algorandfoundation/algokit-utils/transactions/key-registration.mjs
import { TransactionType } from "../packages/transact/src/transactions/transaction-type.mjs";
import { Transaction } from "../packages/transact/src/transactions/transaction.mjs";
import { buildTransactionCommonData } from "./common.mjs";

//#region src/transactions/key-registration.ts
const buildKeyReg = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	if ("voteKey" in params) return new Transaction({
		...commonData,
		type: TransactionType.KeyRegistration,
		keyRegistration: {
			voteKey: params.voteKey,
			selectionKey: params.selectionKey,
			voteFirst: params.voteFirst,
			voteLast: params.voteLast,
			voteKeyDilution: params.voteKeyDilution,
			nonParticipation: false,
			stateProofKey: params.stateProofKey
		}
	});
	else return new Transaction({
		...commonData,
		type: TransactionType.KeyRegistration,
		keyRegistration: { nonParticipation: params.preventAccountFromEverParticipatingAgain }
	});
};

//#endregion
export { buildKeyReg };
