// @algorandfoundation/algokit-utils/packages/common/src/codecs/models/array-model.js
const require_codec = require('../codec.js');

//#region packages/common/src/codecs/models/array-model.ts
var ArrayModelCodec = class extends require_codec.Codec {
	resolvedMetadata = void 0;
	constructor(metadata) {
		super();
		this.metadata = metadata;
	}
	getMetadata() {
		if (!this.resolvedMetadata) this.resolvedMetadata = typeof this.metadata === "function" ? this.metadata() : this.metadata;
		return this.resolvedMetadata;
	}
	defaultValue() {
		return this.getMetadata().codec.defaultValue();
	}
	isDefaultValue(value) {
		return this.getMetadata().codec.isDefaultValue(value);
	}
	encode(value, format) {
		return this.getMetadata().codec.encode(value, format);
	}
	encodeOptional(value, format) {
		return this.getMetadata().codec.encodeOptional(value, format);
	}
	decode(value, format) {
		return this.getMetadata().codec.decode(value, format);
	}
	decodeOptional(value, format) {
		return this.getMetadata().codec.decodeOptional(value, format);
	}
};

//#endregion
exports.ArrayModelCodec = ArrayModelCodec;
