// @algorandfoundation/algokit-utils/packages/common/src/codecs/models/object-model.js
const require_codec = require('../codec.js');
const require_wire = require('../wire.js');

//#region packages/common/src/codecs/models/object-model.ts
function isEmptyObject(value) {
	if (value === null || value === void 0) return true;
	if (typeof value !== "object" || Array.isArray(value) || value instanceof Uint8Array) return false;
	if (value instanceof Map) return value.size === 0;
	const keys = Object.keys(value);
	return keys.length === 0 || keys.every((key) => value[key] === void 0);
}
var ObjectModelCodec = class extends require_codec.Codec {
	resolvedMetadata = void 0;
	resolvedDefaultValue = void 0;
	constructor(metadata) {
		super();
		this.metadata = metadata;
	}
	decode(value, format) {
		if (value === void 0 || value === null) return this.defaultValue();
		return this.fromEncoded(value, format);
	}
	getMetadata() {
		if (!this.resolvedMetadata) this.resolvedMetadata = typeof this.metadata === "function" ? this.metadata() : this.metadata;
		return this.resolvedMetadata;
	}
	defaultValue() {
		if (this.resolvedDefaultValue === void 0) {
			const metadata = this.getMetadata();
			const result = {};
			for (const field of metadata.fields) if (!field.optional) result[field.name] = field.codec.defaultValue();
			this.resolvedDefaultValue = result;
		}
		return this.resolvedDefaultValue;
	}
	isDefaultValue(value) {
		const metadata = this.getMetadata();
		for (const field of metadata.fields) {
			const fieldValue = value[field.name];
			if (field.optional && (fieldValue === void 0 || fieldValue === null)) continue;
			if (!field.codec.isDefaultValue(fieldValue)) return false;
		}
		return true;
	}
	toEncoded(value, format) {
		const metadata = this.getMetadata();
		const result = {};
		for (const field of metadata.fields) {
			const fieldValue = value[field.name];
			if (field.flattened) {
				const encoded = this.encodeFlattenedField(field, fieldValue, format);
				if (encoded !== void 0) Object.assign(result, encoded);
				continue;
			}
			const wireKey = field.wireKey ?? field.name;
			const encoded = field.codec.encodeOptional(fieldValue, format);
			if (encoded !== void 0 && !isEmptyObject(encoded)) result[wireKey] = encoded;
		}
		return result;
	}
	fromEncoded(value, format) {
		const metadata = this.getMetadata();
		const normalizedWireObject = require_wire.normalizeWireObject(value);
		const result = {};
		for (const field of metadata.fields) if (field.flattened) {
			const decoded = this.decodeFieldValue(field, normalizedWireObject, format);
			if (!isEmptyObject(decoded) || !field.optional) result[field.name] = decoded;
		} else {
			const wireValue = normalizedWireObject[field.wireKey || field.name];
			const decoded = this.decodeFieldValue(field, wireValue, format);
			if (!isEmptyObject(decoded) || !field.optional) result[field.name] = decoded;
		}
		return result;
	}
	encodeFlattenedField(field, fieldValue, format) {
		const encoded = field.codec.encodeOptional(fieldValue, format);
		if (encoded !== void 0 && typeof encoded === "object" && !Array.isArray(encoded)) return require_wire.normalizeWireObject(encoded);
		return {};
	}
	decodeFieldValue(field, wireValue, format) {
		return field.optional ? field.codec.decodeOptional(wireValue, format) : field.codec.decode(wireValue, format);
	}
};

//#endregion
exports.ObjectModelCodec = ObjectModelCodec;
