// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/account-participation.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { fixedBytes32Codec, fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";

//#region packages/indexer_client/src/models/account-participation.ts
const AccountParticipationMeta = {
	name: "AccountParticipation",
	kind: "object",
	fields: [
		{
			name: "selectionParticipationKey",
			wireKey: "selection-participation-key",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "voteFirstValid",
			wireKey: "vote-first-valid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "vote-key-dilution",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "voteLastValid",
			wireKey: "vote-last-valid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "voteParticipationKey",
			wireKey: "vote-participation-key",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "stateProofKey",
			wireKey: "state-proof-key",
			optional: true,
			codec: fixedBytes64Codec
		}
	]
};

//#endregion
export { AccountParticipationMeta };
