// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application-local-state.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { ArrayModelCodec } from "../../../common/src/codecs/models/array-model.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { ApplicationStateSchemaMeta } from "./application-state-schema.mjs";
import { TealKeyValueStoreMeta } from "./teal-key-value-store.mjs";

//#region packages/indexer_client/src/models/application-local-state.ts
const ApplicationLocalStateMeta = {
	name: "ApplicationLocalState",
	kind: "object",
	fields: [
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "optedInAtRound",
			wireKey: "opted-in-at-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "closedOutAtRound",
			wireKey: "closed-out-at-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "schema",
			wireKey: "schema",
			optional: false,
			codec: new ObjectModelCodec(ApplicationStateSchemaMeta)
		},
		{
			name: "keyValue",
			wireKey: "key-value",
			optional: true,
			codec: new ArrayModelCodec(TealKeyValueStoreMeta)
		}
	]
};

//#endregion
export { ApplicationLocalStateMeta };
