// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application-local-states-response.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { ApplicationLocalStateMeta } from "./application-local-state.mjs";

//#region packages/indexer_client/src/models/application-local-states-response.ts
const ApplicationLocalStatesResponseMeta = {
	name: "ApplicationLocalStatesResponse",
	kind: "object",
	fields: [
		{
			name: "appsLocalStates",
			wireKey: "apps-local-states",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec(ApplicationLocalStateMeta))
		},
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: stringCodec
		}
	]
};

//#endregion
export { ApplicationLocalStatesResponseMeta };
