// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application-local-states-response.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_application_local_state = require('./application-local-state.js');

//#region packages/indexer_client/src/models/application-local-states-response.ts
const ApplicationLocalStatesResponseMeta = {
	name: "ApplicationLocalStatesResponse",
	kind: "object",
	fields: [
		{
			name: "appsLocalStates",
			wireKey: "apps-local-states",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_application_local_state.ApplicationLocalStateMeta))
		},
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};

//#endregion
exports.ApplicationLocalStatesResponseMeta = ApplicationLocalStatesResponseMeta;
