// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/application-logs-response.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { ApplicationLogDataMeta } from "./application-log-data.mjs";

//#region packages/indexer_client/src/models/application-logs-response.ts
const ApplicationLogsResponseMeta = {
	name: "ApplicationLogsResponse",
	kind: "object",
	fields: [
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "currentRound",
			wireKey: "current-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "nextToken",
			wireKey: "next-token",
			optional: true,
			codec: stringCodec
		},
		{
			name: "logData",
			wireKey: "log-data",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ApplicationLogDataMeta))
		}
	]
};

//#endregion
export { ApplicationLogsResponseMeta };
