// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/asset-holding.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');

//#region packages/indexer_client/src/models/asset-holding.ts
const AssetHoldingMeta = {
	name: "AssetHolding",
	kind: "object",
	fields: [
		{
			name: "amount",
			wireKey: "amount",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "assetId",
			wireKey: "asset-id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "isFrozen",
			wireKey: "is-frozen",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "optedInAtRound",
			wireKey: "opted-in-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "optedOutAtRound",
			wireKey: "opted-out-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};

//#endregion
exports.AssetHoldingMeta = AssetHoldingMeta;
