// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/block-rewards.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');

//#region packages/indexer_client/src/models/block-rewards.ts
const BlockRewardsMeta = {
	name: "BlockRewards",
	kind: "object",
	fields: [
		{
			name: "feeSink",
			wireKey: "fee-sink",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "rewardsCalculationRound",
			wireKey: "rewards-calculation-round",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsLevel",
			wireKey: "rewards-level",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsPool",
			wireKey: "rewards-pool",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "rewardsRate",
			wireKey: "rewards-rate",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsResidue",
			wireKey: "rewards-residue",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};

//#endregion
exports.BlockRewardsMeta = BlockRewardsMeta;
