// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/block-upgrade-state.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";

//#region packages/indexer_client/src/models/block-upgrade-state.ts
const BlockUpgradeStateMeta = {
	name: "BlockUpgradeState",
	kind: "object",
	fields: [
		{
			name: "currentProtocol",
			wireKey: "current-protocol",
			optional: false,
			codec: stringCodec
		},
		{
			name: "nextProtocol",
			wireKey: "next-protocol",
			optional: true,
			codec: stringCodec
		},
		{
			name: "nextProtocolApprovals",
			wireKey: "next-protocol-approvals",
			optional: true,
			codec: numberCodec
		},
		{
			name: "nextProtocolSwitchOn",
			wireKey: "next-protocol-switch-on",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "nextProtocolVoteBefore",
			wireKey: "next-protocol-vote-before",
			optional: true,
			codec: bigIntCodec
		}
	]
};

//#endregion
export { BlockUpgradeStateMeta };
