// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/block-upgrade-state.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');

//#region packages/indexer_client/src/models/block-upgrade-state.ts
const BlockUpgradeStateMeta = {
	name: "BlockUpgradeState",
	kind: "object",
	fields: [
		{
			name: "currentProtocol",
			wireKey: "current-protocol",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "nextProtocol",
			wireKey: "next-protocol",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "nextProtocolApprovals",
			wireKey: "next-protocol-approvals",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "nextProtocolSwitchOn",
			wireKey: "next-protocol-switch-on",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextProtocolVoteBefore",
			wireKey: "next-protocol-vote-before",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};

//#endregion
exports.BlockUpgradeStateMeta = BlockUpgradeStateMeta;
