// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/block.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { fixedBytes32Codec, fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { BlockRewardsMeta } from "./block-rewards.mjs";
import { BlockUpgradeStateMeta } from "./block-upgrade-state.mjs";
import { BlockUpgradeVoteMeta } from "./block-upgrade-vote.mjs";
import { ParticipationUpdatesMeta } from "./participation-updates.mjs";
import { StateProofTrackingMeta } from "./state-proof-tracking.mjs";
import { TransactionMeta } from "./transaction.mjs";

//#region packages/indexer_client/src/models/block.ts
const BlockMeta = {
	name: "Block",
	kind: "object",
	fields: [
		{
			name: "proposer",
			wireKey: "proposer",
			optional: true,
			codec: addressCodec
		},
		{
			name: "feesCollected",
			wireKey: "fees-collected",
			optional: true,
			codec: numberCodec
		},
		{
			name: "bonus",
			wireKey: "bonus",
			optional: true,
			codec: numberCodec
		},
		{
			name: "proposerPayout",
			wireKey: "proposer-payout",
			optional: true,
			codec: numberCodec
		},
		{
			name: "genesisHash",
			wireKey: "genesis-hash",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "genesis-id",
			optional: false,
			codec: stringCodec
		},
		{
			name: "previousBlockHash",
			wireKey: "previous-block-hash",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "previousBlockHash512",
			wireKey: "previous-block-hash-512",
			optional: true,
			codec: fixedBytes64Codec
		},
		{
			name: "rewards",
			wireKey: "rewards",
			optional: false,
			codec: new ObjectModelCodec(BlockRewardsMeta)
		},
		{
			name: "round",
			wireKey: "round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "seed",
			wireKey: "seed",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "stateProofTracking",
			wireKey: "state-proof-tracking",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(StateProofTrackingMeta))
		},
		{
			name: "timestamp",
			wireKey: "timestamp",
			optional: false,
			codec: numberCodec
		},
		{
			name: "transactions",
			wireKey: "transactions",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec(TransactionMeta))
		},
		{
			name: "transactionsRoot",
			wireKey: "transactions-root",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "transactionsRootSha256",
			wireKey: "transactions-root-sha256",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "transactionsRootSha512",
			wireKey: "transactions-root-sha512",
			optional: true,
			codec: fixedBytes64Codec
		},
		{
			name: "txnCounter",
			wireKey: "txn-counter",
			optional: true,
			codec: numberCodec
		},
		{
			name: "upgradeState",
			wireKey: "upgrade-state",
			optional: false,
			codec: new ObjectModelCodec(BlockUpgradeStateMeta)
		},
		{
			name: "upgradeVote",
			wireKey: "upgrade-vote",
			optional: true,
			codec: new ObjectModelCodec(BlockUpgradeVoteMeta)
		},
		{
			name: "participationUpdates",
			wireKey: "participation-updates",
			optional: false,
			codec: new ObjectModelCodec(ParticipationUpdatesMeta)
		}
	]
};

//#endregion
export { BlockMeta };
