// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/indexer-state-proof-message.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');

//#region packages/indexer_client/src/models/indexer-state-proof-message.ts
const IndexerStateProofMessageMeta = {
	name: "IndexerStateProofMessage",
	kind: "object",
	fields: [
		{
			name: "blockHeadersCommitment",
			wireKey: "block-headers-commitment",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "votersCommitment",
			wireKey: "voters-commitment",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "lnProvenWeight",
			wireKey: "ln-proven-weight",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "firstAttestedRound",
			wireKey: "first-attested-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "latestAttestedRound",
			wireKey: "latest-attested-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};

//#endregion
exports.IndexerStateProofMessageMeta = IndexerStateProofMessageMeta;
