// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/mini-asset-holding.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');
const require_string = require('../../../common/src/codecs/primitives/string.js');

//#region packages/indexer_client/src/models/mini-asset-holding.ts
const MiniAssetHoldingMeta = {
	name: "MiniAssetHolding",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "amount",
			wireKey: "amount",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "isFrozen",
			wireKey: "is-frozen",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "deleted",
			wireKey: "deleted",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "optedInAtRound",
			wireKey: "opted-in-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "optedOutAtRound",
			wireKey: "opted-out-at-round",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};

//#endregion
exports.MiniAssetHoldingMeta = MiniAssetHoldingMeta;
