// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/resource-ref.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { BoxReferenceMeta } from "./box-reference.mjs";
import { HoldingRefMeta } from "./holding-ref.mjs";
import { LocalsRefMeta } from "./locals-ref.mjs";

//#region packages/indexer_client/src/models/resource-ref.ts
const ResourceRefMeta = {
	name: "ResourceRef",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: true,
			codec: addressCodec
		},
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "assetId",
			wireKey: "asset-id",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "box",
			wireKey: "box",
			optional: true,
			codec: new ObjectModelCodec(BoxReferenceMeta)
		},
		{
			name: "holding",
			wireKey: "holding",
			optional: true,
			codec: new ObjectModelCodec(HoldingRefMeta)
		},
		{
			name: "local",
			wireKey: "local",
			optional: true,
			codec: new ObjectModelCodec(LocalsRefMeta)
		}
	]
};

//#endregion
export { ResourceRefMeta };
