// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/resource-ref.js
const require_address = require('../../../common/src/codecs/primitives/address.js');
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_box_reference = require('./box-reference.js');
const require_holding_ref = require('./holding-ref.js');
const require_locals_ref = require('./locals-ref.js');

//#region packages/indexer_client/src/models/resource-ref.ts
const ResourceRefMeta = {
	name: "ResourceRef",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: true,
			codec: require_address.addressCodec
		},
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "assetId",
			wireKey: "asset-id",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "box",
			wireKey: "box",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_box_reference.BoxReferenceMeta)
		},
		{
			name: "holding",
			wireKey: "holding",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_holding_ref.HoldingRefMeta)
		},
		{
			name: "local",
			wireKey: "local",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_locals_ref.LocalsRefMeta)
		}
	]
};

//#endregion
exports.ResourceRefMeta = ResourceRefMeta;
