// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/state-proof-reveal.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_state_proof_participant = require('./state-proof-participant.js');
const require_state_proof_sig_slot = require('./state-proof-sig-slot.js');

//#region packages/indexer_client/src/models/state-proof-reveal.ts
const StateProofRevealMeta = {
	name: "StateProofReveal",
	kind: "object",
	fields: [
		{
			name: "position",
			wireKey: "position",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "sigSlot",
			wireKey: "sig-slot",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_state_proof_sig_slot.StateProofSigSlotMeta)
		},
		{
			name: "participant",
			wireKey: "participant",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_state_proof_participant.StateProofParticipantMeta)
		}
	]
};

//#endregion
exports.StateProofRevealMeta = StateProofRevealMeta;
