// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/state-proof-fields.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec, bigIntArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { MerkleArrayProofMeta } from "./merkle-array-proof.mjs";
import { StateProofRevealMeta } from "./state-proof-reveal.mjs";

//#region packages/indexer_client/src/models/state-proof-fields.ts
const StateProofFieldsMeta = {
	name: "StateProofFields",
	kind: "object",
	fields: [
		{
			name: "sigCommit",
			wireKey: "sig-commit",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "signedWeight",
			wireKey: "signed-weight",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "sigProofs",
			wireKey: "sig-proofs",
			optional: true,
			codec: new ObjectModelCodec(MerkleArrayProofMeta)
		},
		{
			name: "partProofs",
			wireKey: "part-proofs",
			optional: true,
			codec: new ObjectModelCodec(MerkleArrayProofMeta)
		},
		{
			name: "saltVersion",
			wireKey: "salt-version",
			optional: true,
			codec: numberCodec
		},
		{
			name: "reveals",
			wireKey: "reveals",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(StateProofRevealMeta))
		},
		{
			name: "positionsToReveal",
			wireKey: "positions-to-reveal",
			optional: true,
			codec: bigIntArrayCodec
		}
	]
};

//#endregion
export { StateProofFieldsMeta };
