// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/state-proof-signature.js
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_merkle_array_proof = require('./merkle-array-proof.js');

//#region packages/indexer_client/src/models/state-proof-signature.ts
const StateProofSignatureMeta = {
	name: "StateProofSignature",
	kind: "object",
	fields: [
		{
			name: "falconSignature",
			wireKey: "falcon-signature",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "merkleArrayIndex",
			wireKey: "merkle-array-index",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "proof",
			wireKey: "proof",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_merkle_array_proof.MerkleArrayProofMeta)
		},
		{
			name: "verifyingKey",
			wireKey: "verifying-key",
			optional: true,
			codec: require_bytes.bytesCodec
		}
	]
};

//#endregion
exports.StateProofSignatureMeta = StateProofSignatureMeta;
