// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-state-proof.js
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_indexer_state_proof_message = require('./indexer-state-proof-message.js');
const require_state_proof_fields = require('./state-proof-fields.js');

//#region packages/indexer_client/src/models/transaction-state-proof.ts
const TransactionStateProofMeta = {
	name: "TransactionStateProof",
	kind: "object",
	fields: [
		{
			name: "stateProofType",
			wireKey: "state-proof-type",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "stateProof",
			wireKey: "state-proof",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_state_proof_fields.StateProofFieldsMeta)
		},
		{
			name: "message",
			wireKey: "message",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_indexer_state_proof_message.IndexerStateProofMessageMeta)
		}
	]
};

//#endregion
exports.TransactionStateProofMeta = TransactionStateProofMeta;
