// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-application.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec, addressArrayCodec, bigIntArrayCodec, bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { PrimitiveModelCodec } from "../../../common/src/codecs/models/primitive-model.mjs";
import { BoxReferenceMeta } from "./box-reference.mjs";
import { OnCompletionMeta } from "./on-completion.mjs";
import { ResourceRefMeta } from "./resource-ref.mjs";
import { StateSchemaMeta } from "./state-schema.mjs";

//#region packages/indexer_client/src/models/transaction-application.ts
const TransactionApplicationMeta = {
	name: "TransactionApplication",
	kind: "object",
	fields: [
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "onCompletion",
			wireKey: "on-completion",
			optional: false,
			codec: new PrimitiveModelCodec(OnCompletionMeta)
		},
		{
			name: "applicationArgs",
			wireKey: "application-args",
			optional: true,
			codec: bytesArrayCodec
		},
		{
			name: "access",
			wireKey: "access",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(ResourceRefMeta))
		},
		{
			name: "accounts",
			wireKey: "accounts",
			optional: true,
			codec: addressArrayCodec
		},
		{
			name: "boxReferences",
			wireKey: "box-references",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(BoxReferenceMeta))
		},
		{
			name: "foreignApps",
			wireKey: "foreign-apps",
			optional: true,
			codec: bigIntArrayCodec
		},
		{
			name: "foreignAssets",
			wireKey: "foreign-assets",
			optional: true,
			codec: bigIntArrayCodec
		},
		{
			name: "localStateSchema",
			wireKey: "local-state-schema",
			optional: true,
			codec: new ObjectModelCodec(StateSchemaMeta)
		},
		{
			name: "globalStateSchema",
			wireKey: "global-state-schema",
			optional: true,
			codec: new ObjectModelCodec(StateSchemaMeta)
		},
		{
			name: "approvalProgram",
			wireKey: "approval-program",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "clear-state-program",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "extraProgramPages",
			wireKey: "extra-program-pages",
			optional: true,
			codec: numberCodec
		},
		{
			name: "rejectVersion",
			wireKey: "reject-version",
			optional: true,
			codec: numberCodec
		}
	]
};

//#endregion
export { TransactionApplicationMeta };
