// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes32Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec, bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ArrayModelCodec } from "../../../common/src/codecs/models/array-model.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { StateDeltaMeta } from "./state-delta.mjs";
import { AccountStateDeltaMeta } from "./account-state-delta.mjs";
import { TransactionApplicationMeta } from "./transaction-application.mjs";
import { TransactionAssetConfigMeta } from "./transaction-asset-config.mjs";
import { TransactionAssetFreezeMeta } from "./transaction-asset-freeze.mjs";
import { TransactionAssetTransferMeta } from "./transaction-asset-transfer.mjs";
import { TransactionHeartbeatMeta } from "./transaction-heartbeat.mjs";
import { TransactionKeyregMeta } from "./transaction-keyreg.mjs";
import { TransactionPaymentMeta } from "./transaction-payment.mjs";
import { TransactionSignatureMeta } from "./transaction-signature.mjs";
import { TransactionStateProofMeta } from "./transaction-state-proof.mjs";

//#region packages/indexer_client/src/models/transaction.ts
const TransactionMeta = {
	name: "Transaction",
	kind: "object",
	fields: [
		{
			name: "applicationTransaction",
			wireKey: "application-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionApplicationMeta)
		},
		{
			name: "assetConfigTransaction",
			wireKey: "asset-config-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionAssetConfigMeta)
		},
		{
			name: "assetFreezeTransaction",
			wireKey: "asset-freeze-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionAssetFreezeMeta)
		},
		{
			name: "assetTransferTransaction",
			wireKey: "asset-transfer-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionAssetTransferMeta)
		},
		{
			name: "stateProofTransaction",
			wireKey: "state-proof-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionStateProofMeta)
		},
		{
			name: "heartbeatTransaction",
			wireKey: "heartbeat-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionHeartbeatMeta)
		},
		{
			name: "authAddr",
			wireKey: "auth-addr",
			optional: true,
			codec: addressCodec
		},
		{
			name: "closeRewards",
			wireKey: "close-rewards",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "closingAmount",
			wireKey: "closing-amount",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "confirmedRound",
			wireKey: "confirmed-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "createdAppId",
			wireKey: "created-application-index",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "createdAssetId",
			wireKey: "created-asset-index",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "fee",
			wireKey: "fee",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "firstValid",
			wireKey: "first-valid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "genesisHash",
			wireKey: "genesis-hash",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "genesis-id",
			optional: true,
			codec: stringCodec
		},
		{
			name: "group",
			wireKey: "group",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "id",
			wireKey: "id",
			optional: true,
			codec: stringCodec
		},
		{
			name: "intraRoundOffset",
			wireKey: "intra-round-offset",
			optional: true,
			codec: numberCodec
		},
		{
			name: "keyregTransaction",
			wireKey: "keyreg-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionKeyregMeta)
		},
		{
			name: "lastValid",
			wireKey: "last-valid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "lease",
			wireKey: "lease",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "note",
			wireKey: "note",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "paymentTransaction",
			wireKey: "payment-transaction",
			optional: true,
			codec: new ObjectModelCodec(TransactionPaymentMeta)
		},
		{
			name: "receiverRewards",
			wireKey: "receiver-rewards",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "rekeyTo",
			wireKey: "rekey-to",
			optional: true,
			codec: addressCodec
		},
		{
			name: "roundTime",
			wireKey: "round-time",
			optional: true,
			codec: numberCodec
		},
		{
			name: "sender",
			wireKey: "sender",
			optional: false,
			codec: stringCodec
		},
		{
			name: "senderRewards",
			wireKey: "sender-rewards",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "signature",
			wireKey: "signature",
			optional: true,
			codec: new ObjectModelCodec(TransactionSignatureMeta)
		},
		{
			name: "txType",
			wireKey: "tx-type",
			optional: false,
			codec: stringCodec
		},
		{
			name: "localStateDelta",
			wireKey: "local-state-delta",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(AccountStateDeltaMeta))
		},
		{
			name: "globalStateDelta",
			wireKey: "global-state-delta",
			optional: true,
			codec: new ArrayModelCodec(StateDeltaMeta)
		},
		{
			name: "logs",
			wireKey: "logs",
			optional: true,
			codec: bytesArrayCodec
		},
		{
			name: "innerTxns",
			wireKey: "inner-txns",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(() => TransactionMeta))
		}
	]
};

//#endregion
export { TransactionMeta };
