// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-keyreg.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_boolean = require('../../../common/src/codecs/primitives/boolean.js');
const require_fixed_bytes = require('../../../common/src/codecs/primitives/fixed-bytes.js');

//#region packages/indexer_client/src/models/transaction-keyreg.ts
const TransactionKeyregMeta = {
	name: "TransactionKeyreg",
	kind: "object",
	fields: [
		{
			name: "nonParticipation",
			wireKey: "non-participation",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "selectionParticipationKey",
			wireKey: "selection-participation-key",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "voteFirstValid",
			wireKey: "vote-first-valid",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "vote-key-dilution",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteLastValid",
			wireKey: "vote-last-valid",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteParticipationKey",
			wireKey: "vote-participation-key",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "stateProofKey",
			wireKey: "state-proof-key",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		}
	]
};

//#endregion
exports.TransactionKeyregMeta = TransactionKeyregMeta;
