// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-application.js
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_primitive_model = require('../../../common/src/codecs/models/primitive-model.js');
const require_box_reference = require('./box-reference.js');
const require_on_completion = require('./on-completion.js');
const require_resource_ref = require('./resource-ref.js');
const require_state_schema = require('./state-schema.js');

//#region packages/indexer_client/src/models/transaction-application.ts
const TransactionApplicationMeta = {
	name: "TransactionApplication",
	kind: "object",
	fields: [
		{
			name: "applicationId",
			wireKey: "application-id",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "onCompletion",
			wireKey: "on-completion",
			optional: false,
			codec: new require_primitive_model.PrimitiveModelCodec(require_on_completion.OnCompletionMeta)
		},
		{
			name: "applicationArgs",
			wireKey: "application-args",
			optional: true,
			codec: require_array.bytesArrayCodec
		},
		{
			name: "access",
			wireKey: "access",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_resource_ref.ResourceRefMeta))
		},
		{
			name: "accounts",
			wireKey: "accounts",
			optional: true,
			codec: require_array.addressArrayCodec
		},
		{
			name: "boxReferences",
			wireKey: "box-references",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_box_reference.BoxReferenceMeta))
		},
		{
			name: "foreignApps",
			wireKey: "foreign-apps",
			optional: true,
			codec: require_array.bigIntArrayCodec
		},
		{
			name: "foreignAssets",
			wireKey: "foreign-assets",
			optional: true,
			codec: require_array.bigIntArrayCodec
		},
		{
			name: "localStateSchema",
			wireKey: "local-state-schema",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_state_schema.StateSchemaMeta)
		},
		{
			name: "globalStateSchema",
			wireKey: "global-state-schema",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_state_schema.StateSchemaMeta)
		},
		{
			name: "approvalProgram",
			wireKey: "approval-program",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "clear-state-program",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "extraProgramPages",
			wireKey: "extra-program-pages",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "rejectVersion",
			wireKey: "reject-version",
			optional: true,
			codec: require_number.numberCodec
		}
	]
};

//#endregion
exports.TransactionApplicationMeta = TransactionApplicationMeta;
