// @algorandfoundation/algokit-utils/packages/indexer_client/src/models/transaction-signature.js
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_transaction_signature_multisig = require('./transaction-signature-multisig.js');
const require_transaction_signature_logicsig = require('./transaction-signature-logicsig.js');

//#region packages/indexer_client/src/models/transaction-signature.ts
const TransactionSignatureMeta = {
	name: "TransactionSignature",
	kind: "object",
	fields: [
		{
			name: "logicsig",
			wireKey: "logicsig",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_signature_logicsig.TransactionSignatureLogicsigMeta)
		},
		{
			name: "multisig",
			wireKey: "multisig",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_transaction_signature_multisig.TransactionSignatureMultisigMeta)
		},
		{
			name: "sig",
			wireKey: "sig",
			optional: true,
			codec: require_bytes.bytesCodec
		}
	]
};

//#endregion
exports.TransactionSignatureMeta = TransactionSignatureMeta;
