// @algorandfoundation/algokit-utils/transaction/perform-transaction-composer-simulate.mjs
//#region src/transaction/perform-transaction-composer-simulate.ts
/**
* @deprecated Use `composer.simulate` with
*  - `allowEmptySignatures` flag set to true
*  - `resultOnFailure` flag set to true
*
* Performs a simulation of the transactions loaded into the given TransactionComposer.
* Uses empty transaction signers for all transactions.
*
* @param composer The TransactionComposer with transaction(s) loaded.
* @returns The simulation result, which includes various details about how the transactions would be processed.
*/
async function performTransactionComposerSimulate(composer, options) {
	const simulateOptions = { ...options ?? {
		skipSignatures: true,
		allowEmptySignatures: true,
		fixSigners: true,
		allowMoreLogging: true,
		execTraceConfig: {
			enable: true,
			scratchChange: true,
			stackChange: true,
			stateChange: true
		},
		resultOnFailure: true
	} };
	return (await composer.simulate(simulateOptions)).simulateResponse;
}

//#endregion
export { performTransactionComposerSimulate };
