// @algorandfoundation/algokit-utils/packages/transact/src/logicsig.mjs
import { concatArrays } from "../../common/src/array.mjs";
import { Address } from "../../common/src/address.mjs";
import { decodeMsgpack } from "../../common/src/msgpack.mjs";
import { hash } from "../../crypto/src/hash.mjs";
import { logicSigSignatureCodec } from "./transactions/signed-transaction-meta.mjs";
import { encodeSignedTransaction, validateSignedTransaction } from "./transactions/signed-transaction.mjs";
import { MultisigAccount } from "./multisig.mjs";

//#region packages/transact/src/logicsig.ts
const PROGRAM_TAG = new TextEncoder().encode("Program");
const MSIG_PROGRAM_TAG = new TextEncoder().encode("MsigProgram");
const SIGN_PROGRAM_DATA_PREFIX = new TextEncoder().encode("ProgData");
var LogicSig = class LogicSig {
	logic;
	args;
	_addr;
	constructor(program, programArgs) {
		this.logic = program;
		this.args = programArgs ?? [];
		this._addr = new Address(hash(concatArrays(PROGRAM_TAG, this.logic)));
	}
	static fromSignature(signature) {
		return new LogicSig(signature.logic, signature.args || []);
	}
	static fromBytes(encodedLsig) {
		const decoded = decodeMsgpack(encodedLsig);
		const lsigSignature = logicSigSignatureCodec.decode(decoded, "msgpack");
		return LogicSig.fromSignature(lsigSignature);
	}
	address() {
		return this._addr;
	}
	get addr() {
		return this._addr;
	}
	bytesToSignForDelegation(msig) {
		if (msig) return concatArrays(MSIG_PROGRAM_TAG, msig.addr.publicKey, this.logic);
		else return concatArrays(PROGRAM_TAG, this.logic);
	}
	signProgramData(data, signer) {
		return signer(data, this);
	}
	programDataToSign(data) {
		return concatArrays(SIGN_PROGRAM_DATA_PREFIX, this.address().publicKey, data);
	}
	account() {
		return new LogicSigAccount(this.logic, this.args);
	}
	delegatedAccount(delegator) {
		return new LogicSigAccount(this.logic, this.args, delegator);
	}
};
var LogicSigAccount = class LogicSigAccount extends LogicSig {
	sig;
	msig;
	lmsig;
	static fromSignature(signature, delegator) {
		if (signature.lmsig || signature.msig) {
			const msigAddr = MultisigAccount.fromSignature(signature.lmsig || signature.msig).addr;
			if (delegator && !msigAddr.equals(delegator)) throw new Error("Provided delegator address does not match multisig address");
			const lsig = new LogicSigAccount(signature.logic, signature.args || [], msigAddr);
			lsig.lmsig = signature.lmsig;
			lsig.msig = signature.msig;
			return lsig;
		}
		const lsigAccount = new LogicSigAccount(signature.logic, signature.args || [], delegator);
		if (signature.sig && delegator === void 0) throw new Error("Delegated address must be provided when logic sig has a signature");
		if (signature.sig) {
			lsigAccount.sig = signature.sig;
			return lsigAccount;
		}
		return lsigAccount;
	}
	static fromBytes(encodedLsig, delegator) {
		const decoded = decodeMsgpack(encodedLsig);
		const lsigSignature = logicSigSignatureCodec.decode(decoded, "msgpack");
		return LogicSigAccount.fromSignature(lsigSignature, delegator);
	}
	constructor(program, programArgs, delegator) {
		super(program, programArgs ?? void 0);
		this._addr = delegator ?? this._addr;
	}
	get signer() {
		return async (txns, indexes) => {
			const signedTxns = [];
			for (const index of indexes) {
				const stxn = {
					txn: txns[index],
					lsig: {
						logic: this.logic,
						args: this.args,
						msig: this.msig,
						lmsig: this.lmsig,
						sig: this.sig
					}
				};
				if (!stxn.txn.sender.equals(this.addr)) stxn.authAddress = this.addr;
				validateSignedTransaction(stxn);
				signedTxns.push(encodeSignedTransaction(stxn));
			}
			return signedTxns;
		};
	}
	async signForDelegation(delegator) {
		const result = await delegator.lsigSigner(this);
		if (!result.addr.equals(this._addr)) throw new Error(`Delegator address from signer does not match expected delegator address. Expected: ${this._addr.toString()}, got: ${result.addr.toString()}`);
		if ("sig" in result && result.sig) this.sig = result.sig;
		else if ("lmsig" in result && result.lmsig) this.lmsig = result.lmsig;
		else throw new Error("Delegated lsig signer must return either a sig or lmsig");
	}
};

//#endregion
export { LogicSig, LogicSigAccount };
