// @algorandfoundation/algokit-utils/packages/transact/src/multisig.mjs
import { arrayEqual } from "../../common/src/array.mjs";
import { PUBLIC_KEY_BYTE_LENGTH, SIGNATURE_BYTE_LENGTH } from "../../common/src/constants.mjs";
import { ALGORAND_ADDRESS_BYTE_LENGTH, ALGORAND_CHECKSUM_BYTE_LENGTH, Address, getAddress } from "../../common/src/address.mjs";
import { hash } from "../../crypto/src/hash.mjs";
import { decodeSignedTransaction, encodeSignedTransaction, validateSignedTransaction } from "./transactions/signed-transaction.mjs";

//#region packages/transact/src/multisig.ts
const toPublicKeys = (addrs) => addrs.map((addr) => getAddress(addr).publicKey);
/**
* Applies a subsignature for a participant to a multisignature signature, replacing any existing signature.
*
* This method applies the signature to ALL instances of the given public key (to support weighted multisig).
* Since ed25519 signatures are deterministic, there's only one valid signature for a given message and public key.
*/
function applyMultisigSubsignature(multisigSignature, participant, signature) {
	let found = false;
	const newSubsignatures = multisigSignature.subsigs.map((subsig) => {
		if (arrayEqual(subsig.publicKey, participant)) {
			found = true;
			return {
				...subsig,
				sig: signature
			};
		}
		return subsig;
	});
	if (!found) throw new Error("Public key not found in multisig signature");
	return {
		...multisigSignature,
		subsigs: newSubsignatures
	};
}
const MULTISIG_PREIMG2ADDR_PREFIX = new Uint8Array([
	77,
	117,
	108,
	116,
	105,
	115,
	105,
	103,
	65,
	100,
	100,
	114
]);
const INVALID_MSIG_VERSION_ERROR_MSG = "invalid multisig version";
const INVALID_MSIG_THRESHOLD_ERROR_MSG = "bad multisig threshold";
const INVALID_MSIG_PK_ERROR_MSG = "bad multisig public key - wrong length";
const UNEXPECTED_PK_LEN_ERROR_MSG = "nacl public key length is not 32 bytes";
const MULTISIG_MERGE_LESSTHANTWO_ERROR_MSG = "Not enough multisig transactions to merge. Need at least two";
const MULTISIG_MERGE_MISMATCH_ERROR_MSG = "Cannot merge txs. txIDs differ";
const MULTISIG_MERGE_MISMATCH_AUTH_ADDR_MSG = "Cannot merge txs. Auth addrs differ";
const MULTISIG_MERGE_WRONG_PREIMAGE_ERROR_MSG = "Cannot merge txs. Multisig preimages differ";
const MULTISIG_MERGE_SIG_MISMATCH_ERROR_MSG = "Cannot merge txs. subsigs are mismatched.";
const MULTISIG_SIGNATURE_LENGTH_ERROR_MSG = "Cannot add multisig signature. Signature is not of the correct length.";
const MULTISIG_KEY_NOT_EXIST_ERROR_MSG = "Key does not exist";
/**
* creates a raw, multisig transaction blob without any signatures.
* @param txn - the actual transaction.
* @param version - multisig version
* @param threshold - multisig threshold
* @param pks - ordered list of public keys in this multisig
* @returns encoded multisig blob
*/
function createMultisigTransaction(txn, { version, threshold, addrs }) {
	const pks = toPublicKeys(addrs);
	const msig = {
		version,
		threshold,
		subsigs: pks.map((pk) => ({
			publicKey: pk,
			sig: void 0
		}))
	};
	const msigAddr = addressFromMultisigPreImg({
		version,
		threshold,
		publicKeys: pks
	});
	let authAddress;
	if (!msigAddr.equals(txn.sender)) authAddress = msigAddr;
	return {
		txn,
		msig,
		authAddress
	};
}
/**
* creates a multisig transaction blob with an included signature.
* @param txn - the actual transaction to sign.
* @param rawSig - a Uint8Array raw signature of that transaction
* @param myPk - a public key that corresponds with rawSig
* @param version - multisig version
* @param threshold - multisig threshold
* @param pks - ordered list of public keys in this multisig
* @returns encoded multisig blob
*/
function createMultisigTransactionWithSignature(txn, { rawSig, myPk }, { version, threshold, publicKeys }) {
	const signedTxn = createMultisigTransaction(txn, {
		version,
		threshold,
		addrs: publicKeys.map((pk) => new Address(pk))
	});
	let keyExist = false;
	const updatedSubsigs = signedTxn.msig.subsigs.map((subsig) => {
		if (arrayEqual(subsig.publicKey, myPk)) {
			keyExist = true;
			return {
				...subsig,
				sig: rawSig
			};
		}
		return subsig;
	});
	if (!keyExist) throw new Error(MULTISIG_KEY_NOT_EXIST_ERROR_MSG);
	return {
		...signedTxn,
		msig: {
			...signedTxn.msig,
			subsigs: updatedSubsigs
		}
	};
}
/**
* takes a list of multisig transaction blobs, and merges them.
* @param multisigTxnBlobs - a list of blobs representing encoded multisig txns
* @returns typed array msg-pack encoded multisig txn
*/
function mergeMultisigTransactions(multisigTxnBlobs) {
	if (multisigTxnBlobs.length < 2) throw new Error(MULTISIG_MERGE_LESSTHANTWO_ERROR_MSG);
	const refSigTx = multisigTxnBlobs[0];
	if (!refSigTx.msig) throw new Error("Invalid multisig transaction, multisig structure missing at index 0");
	const refTxID = refSigTx.txn.txId();
	const refAuthAddr = refSigTx.authAddress;
	const refMsigAddr = addressFromMultisigPreImg({
		version: refSigTx.msig.version,
		threshold: refSigTx.msig.threshold,
		publicKeys: refSigTx.msig.subsigs.map((subsig) => subsig.publicKey)
	});
	const newSubsigs = refSigTx.msig.subsigs.map((sig) => ({ ...sig }));
	for (let i = 1; i < multisigTxnBlobs.length; i++) {
		const unisig = multisigTxnBlobs[i];
		if (!unisig.msig) throw new Error(`Invalid multisig transaction, multisig structure missing at index ${i}`);
		if (unisig.txn.txId() !== refTxID) throw new Error(MULTISIG_MERGE_MISMATCH_ERROR_MSG);
		if (refAuthAddr !== unisig.authAddress) throw new Error(MULTISIG_MERGE_MISMATCH_AUTH_ADDR_MSG);
		if (unisig.msig.subsigs.length !== refSigTx.msig.subsigs.length) throw new Error(MULTISIG_MERGE_WRONG_PREIMAGE_ERROR_MSG);
		const msgigAddr = addressFromMultisigPreImg({
			version: unisig.msig.version,
			threshold: unisig.msig.threshold,
			publicKeys: unisig.msig.subsigs.map((subsig) => subsig.publicKey)
		});
		if (refMsigAddr.toString() !== msgigAddr.toString()) throw new Error(MULTISIG_MERGE_WRONG_PREIMAGE_ERROR_MSG);
		unisig.msig.subsigs.forEach((uniSubsig, index) => {
			if (!uniSubsig.sig) return;
			const current = newSubsigs[index];
			if (current.sig && !arrayEqual(uniSubsig.sig, current.sig)) throw new Error(MULTISIG_MERGE_SIG_MISMATCH_ERROR_MSG);
			current.sig = uniSubsig.sig;
		});
	}
	return {
		txn: refSigTx.txn,
		msig: {
			version: refSigTx.msig.version,
			threshold: refSigTx.msig.threshold,
			subsigs: newSubsigs
		},
		authAddress: refAuthAddr
	};
}
/**
* Partially signs this transaction with an external raw multisig signature and returns
* a partially-signed multisig transaction, encoded with msgpack as a typed array.
* @param transaction - The transaction to sign
* @param metadata - multisig metadata
* @param signerAddr - address of the signer
* @param signature - raw multisig signature
* @returns an encoded, partially signed multisig transaction.
*/
function partialSignWithMultisigSignature(transaction, metadata, signerAddr, signature) {
	if (signature.length != SIGNATURE_BYTE_LENGTH) throw new Error(MULTISIG_SIGNATURE_LENGTH_ERROR_MSG);
	return createMultisigTransactionWithSignature(transaction, {
		rawSig: signature,
		myPk: (typeof signerAddr === "string" ? Address.fromString(signerAddr) : signerAddr).publicKey
	}, metadata);
}
/**
* Takes a multisig transaction blob, and appends a given raw signature to it.
* This makes it possible to compile a multisig signature using only raw signatures from external methods.
* @param multisigTxnBlob - an encoded multisig txn. Supports non-payment txn types.
* @param version - multisig version
* @param threshold - multisig threshold
* @param addrs - a list of Algorand addresses representing possible signers for this multisig. Order is important.
* @param signerAddr - address of the signer
* @param signature - raw multisig signature
* @returns object containing txID, and blob representing encoded multisig txn
*/
function appendSignRawMultisigSignature(multisigTxn, { version, threshold, addrs }, signerAddr, signature) {
	const publicKeys = toPublicKeys(addrs);
	return mergeMultisigTransactions([multisigTxn, partialSignWithMultisigSignature(multisigTxn.txn, {
		version,
		threshold,
		publicKeys
	}, signerAddr, signature)]);
}
/**
* Takes multisig parameters and returns a 32 byte typed array public key,
* representing an address that identifies the "exact group, version, and public keys" that are required for signing.
* Hash("MultisigAddr" || version uint8 || threshold uint8 || PK1 || PK2 || ...)
* Encoding this output yields a human readable address.
* @param version - multisig version
* @param threshold - multisig threshold
* @param pks - array of typed array public keys
*/
function addressFromMultisigPreImg({ version, threshold, publicKeys }) {
	if (version > 255 || version < 0) throw new Error(`${INVALID_MSIG_VERSION_ERROR_MSG}: ${version}`);
	if (threshold === 0 || publicKeys.length === 0 || threshold > publicKeys.length || threshold > 255) throw new Error(INVALID_MSIG_THRESHOLD_ERROR_MSG);
	const pkLen = ALGORAND_ADDRESS_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH;
	if (pkLen !== PUBLIC_KEY_BYTE_LENGTH) throw new Error(UNEXPECTED_PK_LEN_ERROR_MSG);
	const merged = new Uint8Array(MULTISIG_PREIMG2ADDR_PREFIX.length + 2 + pkLen * publicKeys.length);
	merged.set(MULTISIG_PREIMG2ADDR_PREFIX, 0);
	merged.set([version], MULTISIG_PREIMG2ADDR_PREFIX.length);
	merged.set([threshold], MULTISIG_PREIMG2ADDR_PREFIX.length + 1);
	for (let i = 0; i < publicKeys.length; i++) {
		if (publicKeys[i].length !== pkLen) throw new Error(INVALID_MSIG_PK_ERROR_MSG);
		merged.set(publicKeys[i], MULTISIG_PREIMG2ADDR_PREFIX.length + 2 + i * pkLen);
	}
	return new Address(Uint8Array.from(hash(merged)));
}
/**
* Takes multisig parameters and returns a human readable Algorand address.
* This is equivalent to fromMultisigPreImg, but interfaces with encoded addresses.
* @param version - multisig version
* @param threshold - multisig threshold
* @param addrs - array of encoded addresses
*/
function addressFromMultisigPreImgAddrs({ version, threshold, addrs }) {
	return addressFromMultisigPreImg({
		version,
		threshold,
		publicKeys: toPublicKeys(addrs)
	});
}
/**
* Takes multisig metadata (preimage) and returns the corresponding human readable Algorand address.
* @param version - multisig version
* @param threshold - multisig threshold
* @param addrs - list of Algorand addresses
*/
const multisigAddress = addressFromMultisigPreImgAddrs;
/** Account wrapper that supports partial or full multisig signing. */
var MultisigAccount = class MultisigAccount {
	_params;
	_subSigners;
	_addr;
	_signer;
	_lsigSigner;
	/** The parameters for the multisig account */
	get params() {
		return this._params;
	}
	/** The list of accounts that are present to sign transactions or lsigs */
	get subSigners() {
		return this._subSigners;
	}
	/** The address of the multisig account */
	get addr() {
		return this._addr;
	}
	/** The transaction signer for the multisig account */
	get signer() {
		return this._signer;
	}
	get lsigSigner() {
		return this._lsigSigner;
	}
	static fromSignature(signature) {
		return new MultisigAccount({
			version: signature.version,
			threshold: signature.threshold,
			addrs: signature.subsigs.map((subsig) => new Address(subsig.publicKey))
		}, []);
	}
	constructor(multisigParams, subSigners) {
		this._params = multisigParams;
		this._subSigners = subSigners;
		this._addr = multisigAddress(multisigParams);
		this._signer = async (txns, indexesToSign) => {
			const txnsToSign = txns.filter((_, index) => indexesToSign.includes(index));
			const signedMsigTxns = [];
			for (const txn of txnsToSign) {
				let signedMsigTxn = createMultisigTransaction(txn, this._params);
				for (const subSigner of this.subSigners) {
					const stxn = (await subSigner.signer([txn], [0]))[0];
					const sig = decodeSignedTransaction(stxn).sig;
					if (!sig) throw new Error(`Signer for address ${subSigner.addr.toString()} did not produce a valid signature when signing ${txn.txId()} for multisig account ${this._addr.toString()}`);
					signedMsigTxn = appendSignRawMultisigSignature(signedMsigTxn, this._params, subSigner.addr, sig);
				}
				signedMsigTxns.push(signedMsigTxn);
			}
			return signedMsigTxns.map((stxn) => {
				validateSignedTransaction(stxn);
				return encodeSignedTransaction(stxn);
			});
		};
		this._lsigSigner = async (lsig, _) => {
			let lmsig = lsig.lmsig ?? this.createMultisigSignature();
			for (const addrWithSigner of this.subSigners) {
				const { lsigSigner, addr } = addrWithSigner;
				const result = await lsigSigner(lsig, this);
				if (!("sig" in result) || !result.sig) throw new Error(`Signer for address ${addr.toString()} did not produce a valid signature when signing logic sig for multisig account ${this._addr.toString()}`);
				lmsig = this.applySignature(lmsig, addr.publicKey, result.sig);
			}
			return {
				addr: this.addr,
				lmsig
			};
		};
	}
	createMultisigTransaction(txn) {
		return createMultisigTransaction(txn, this._params);
	}
	createMultisigSignature() {
		const subsignatures = toPublicKeys(this._params.addrs).map((pk) => ({
			publicKey: pk,
			sig: void 0
		}));
		return {
			version: this._params.version,
			threshold: this._params.threshold,
			subsigs: subsignatures
		};
	}
	applySignatureToTxn(txn, pubkey, signature) {
		if (!txn.msig) txn.msig = this.createMultisigTransaction(txn.txn).msig;
		txn.msig = applyMultisigSubsignature(txn.msig, pubkey, signature);
	}
	applySignature(msigSignature, pubkey, signature) {
		if (msigSignature.version !== this._params.version || msigSignature.threshold !== this._params.threshold) {
			const thisParams = {
				version: this._params.version,
				threshold: this._params.threshold,
				participants: this._params.addrs.map((addr) => addr.toString())
			};
			const givenParams = {
				version: msigSignature.version,
				threshold: msigSignature.threshold,
				participants: msigSignature.subsigs.map((subsig) => new Address(subsig.publicKey).toString())
			};
			throw new Error(`Multisig signature parameters do not match expected multisig parameters. Multisig params: ${JSON.stringify(thisParams)}, signature: ${JSON.stringify(givenParams)}`);
		}
		return applyMultisigSubsignature(msigSignature, pubkey, signature);
	}
};

//#endregion
export { MultisigAccount };
