// @algorandfoundation/algokit-utils/packages/kmd_client/src/apis/api-service.mjs
import { encodeTransactionRaw } from "../../../transact/src/transactions/transaction.mjs";
import { decodeJson, encodeJson } from "../core/model-runtime.mjs";
import { ListWalletsResponseMeta } from "../models/list-wallets-response.mjs";
import { ExportKeyResponseMeta } from "../models/export-key-response.mjs";
import { ImportKeyResponseMeta } from "../models/import-key-response.mjs";
import { ListKeysResponseMeta } from "../models/list-keys-response.mjs";
import { GenerateKeyResponseMeta } from "../models/generate-key-response.mjs";
import { ExportMasterKeyResponseMeta } from "../models/export-master-key-response.mjs";
import { ExportMultisigResponseMeta } from "../models/export-multisig-response.mjs";
import { ImportMultisigResponseMeta } from "../models/import-multisig-response.mjs";
import { ListMultisigResponseMeta } from "../models/list-multisig-response.mjs";
import { SignProgramMultisigResponseMeta } from "../models/sign-program-multisig-response.mjs";
import { SignMultisigResponseMeta } from "../models/sign-multisig-response.mjs";
import { SignProgramResponseMeta } from "../models/sign-program-response.mjs";
import { SignTransactionResponseMeta } from "../models/sign-transaction-response.mjs";
import { WalletInfoResponseMeta } from "../models/wallet-info-response.mjs";
import { InitWalletHandleTokenResponseMeta } from "../models/init-wallet-handle-token-response.mjs";
import { RenameWalletResponseMeta } from "../models/rename-wallet-response.mjs";
import { RenewWalletHandleTokenResponseMeta } from "../models/renew-wallet-handle-token-response.mjs";
import { CreateWalletResponseMeta } from "../models/create-wallet-response.mjs";
import { CreateWalletRequestMeta } from "../models/create-wallet-request.mjs";
import { DeleteKeyRequestMeta } from "../models/delete-key-request.mjs";
import { DeleteMultisigRequestMeta } from "../models/delete-multisig-request.mjs";
import { ExportKeyRequestMeta } from "../models/export-key-request.mjs";
import { ExportMasterKeyRequestMeta } from "../models/export-master-key-request.mjs";
import { ExportMultisigRequestMeta } from "../models/export-multisig-request.mjs";
import { GenerateKeyRequestMeta } from "../models/generate-key-request.mjs";
import { ImportKeyRequestMeta } from "../models/import-key-request.mjs";
import { ImportMultisigRequestMeta } from "../models/import-multisig-request.mjs";
import { InitWalletHandleTokenRequestMeta } from "../models/init-wallet-handle-token-request.mjs";
import { ListKeysRequestMeta } from "../models/list-keys-request.mjs";
import { ListMultisigRequestMeta } from "../models/list-multisig-request.mjs";
import { ReleaseWalletHandleTokenRequestMeta } from "../models/release-wallet-handle-token-request.mjs";
import { RenameWalletRequestMeta } from "../models/rename-wallet-request.mjs";
import { RenewWalletHandleTokenRequestMeta } from "../models/renew-wallet-handle-token-request.mjs";
import { SignMultisigTxnRequestMeta } from "../models/sign-multisig-txn-request.mjs";
import { SignProgramMultisigRequestMeta } from "../models/sign-program-multisig-request.mjs";
import { SignProgramRequestMeta } from "../models/sign-program-request.mjs";
import { SignTxnRequestMeta } from "../models/sign-txn-request.mjs";
import { VersionsResponseMeta } from "../models/versions-response.mjs";
import { WalletInfoRequestMeta } from "../models/wallet-info-request.mjs";

//#region packages/kmd_client/src/apis/api-service.ts
var KmdApi = class {
	constructor(httpRequest) {
		this.httpRequest = httpRequest;
	}
	mimeTypeFor(format) {
		return format === "json" ? "application/json" : format === "msgpack" ? "application/msgpack" : "text/plain";
	}
	/**
	* Create a new wallet (collection of keys) with the given parameters.
	*/
	async _createWallet(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = CreateWalletRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), CreateWalletResponseMeta);
	}
	/**
	* Deletes the key with the passed public key from the wallet.
	*/
	async deleteKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = DeleteKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		await this.httpRequest.request({
			method: "DELETE",
			url: "/v1/key",
			path: {},
			query: {},
			headers,
			body: serializedBody
		});
	}
	/**
	* Deletes multisig preimage information for the passed address from the wallet.
	*/
	async deleteMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = DeleteMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		await this.httpRequest.request({
			method: "DELETE",
			url: "/v1/multisig",
			path: {},
			query: {},
			headers,
			body: serializedBody
		});
	}
	/**
	* Export the secret key associated with the passed public key.
	*/
	async exportKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ExportKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key/export",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ExportKeyResponseMeta);
	}
	/**
	* Export the master derivation key from the wallet. This key is a master "backup" key for the underlying wallet. With it, you can regenerate all of the wallets that have been generated with this wallet's `POST /v1/key` endpoint. This key will not allow you to recover keys imported from other wallets, however.
	*/
	async exportMasterKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ExportMasterKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/master-key/export",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ExportMasterKeyResponseMeta);
	}
	/**
	* Given a multisig address whose preimage this wallet stores, returns the information used to generate the address, including public keys, threshold, and multisig version.
	*/
	async exportMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ExportMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/export",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ExportMultisigResponseMeta);
	}
	/**
	* Generates the next key in the deterministic key sequence (as determined by the master derivation key) and adds it to the wallet, returning the public key.
	*/
	async generateKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = GenerateKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), GenerateKeyResponseMeta);
	}
	/**
	* Import an externally generated key into the wallet. Note that if you wish to back up the imported key, you must do so by backing up the entire wallet database, because imported keys were not derived from the wallet's master derivation key.
	*/
	async importKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ImportKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key/import",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ImportKeyResponseMeta);
	}
	/**
	* Generates a multisig account from the passed public keys array and multisig metadata, and stores all of this in the wallet.
	*/
	async importMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ImportMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/import",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ImportMultisigResponseMeta);
	}
	/**
	* Unlock the wallet and return a wallet handle token that can be used for subsequent operations. These tokens expire periodically and must be renewed. You can `POST` the token to `/v1/wallet/info` to see how much time remains until expiration, and renew it with `/v1/wallet/renew`. When you're done, you can invalidate the token with `/v1/wallet/release`.
	*/
	async initWalletHandle(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = InitWalletHandleTokenRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/init",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), InitWalletHandleTokenResponseMeta);
	}
	/**
	* Lists all of the public keys in this wallet. All of them have a stored private key.
	*/
	async listKeysInWallet(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ListKeysRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key/list",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ListKeysResponseMeta);
	}
	/**
	* Lists all of the multisig accounts whose preimages this wallet stores
	*/
	async listMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ListMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/list",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), ListMultisigResponseMeta);
	}
	/**
	* Lists all of the wallets that kmd is aware of.
	*/
	async listWallets() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v1/wallets",
			path: {},
			query: {},
			headers,
			body: void 0
		}), ListWalletsResponseMeta);
	}
	/**
	* Invalidate the passed wallet handle token, making it invalid for use in subsequent requests.
	*/
	async releaseWalletHandleToken(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = ReleaseWalletHandleTokenRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/release",
			path: {},
			query: {},
			headers,
			body: serializedBody
		});
	}
	/**
	* Rename the underlying wallet to something else
	*/
	async renameWallet(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = RenameWalletRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/rename",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), RenameWalletResponseMeta);
	}
	/**
	* Renew a wallet handle token, increasing its expiration duration to its initial value
	*/
	async renewWalletHandleToken(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = RenewWalletHandleTokenRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/renew",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), RenewWalletHandleTokenResponseMeta);
	}
	/**
	* Start a multisig signature, or add a signature to a partially completed multisig signature object.
	*/
	async signMultisigProgram(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = SignProgramMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/signprogram",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), SignProgramMultisigResponseMeta);
	}
	/**
	* Start a multisig signature, or add a signature to a partially completed multisig signature object.
	*/
	async _signMultisigTransaction(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = SignMultisigTxnRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/sign",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), SignMultisigResponseMeta);
	}
	/**
	* Signs the passed program with a key from the wallet, determined by the account named in the request.
	*/
	async signProgram(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = SignProgramRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/program/sign",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), SignProgramResponseMeta);
	}
	/**
	* Signs the passed transaction with a key from the wallet, determined by the sender encoded in the transaction.
	*/
	async _signTransaction(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = SignTxnRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/transaction/sign",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), SignTransactionResponseMeta);
	}
	async version() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/versions",
			path: {},
			query: {},
			headers,
			body: void 0
		}), VersionsResponseMeta);
	}
	/**
	* Returns information about the wallet associated with the passed wallet handle token. Additionally returns expiration information about the token itself.
	*/
	async walletInfo(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = WalletInfoRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeJson(body, bodyMeta) : void 0;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/info",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), WalletInfoResponseMeta);
	}
	/**
	* Create a new wallet (collection of keys) with the given parameters.
	*/
	async createWallet(body) {
		const requestBody = {
			...body,
			walletDriverName: body.walletDriverName ?? "sqlite"
		};
		return await this._createWallet(requestBody);
	}
	/**
	* Enables the signing of a transaction using the provided wallet and multisig info.
	* The public key is used to identify which multisig account key to use for signing.
	* When a signer is provided it is used to resolve the private key and sign the transaction, enabling rekeyed account signing.
	* @returns A multisig signature or partial signature, which can be used to form a signed transaction.
	*/
	async signMultisigTransaction(body) {
		const requestBody = {
			...body,
			transaction: encodeTransactionRaw(body.transaction)
		};
		return this._signMultisigTransaction(requestBody);
	}
	/**
	* Enables the signing of a transaction using the provided wallet info.
	* When a public key is provided it is used to resolve the private key and sign the transaction, enabling rekeyed account signing.
	* @returns An encoded, signed transaction.
	*/
	async signTransaction(body) {
		const requestBody = {
			...body,
			transaction: encodeTransactionRaw(body.transaction)
		};
		return this._signTransaction(requestBody);
	}
};

//#endregion
export { KmdApi };
