// @algorandfoundation/algokit-utils/types/app.mjs
import { ABI_RETURN_PREFIX as ABI_RETURN_PREFIX$1, APP_DEPLOY_NOTE_DAPP as APP_DEPLOY_NOTE_DAPP$1, APP_PAGE_MAX_SIZE as APP_PAGE_MAX_SIZE$1, DELETABLE_TEMPLATE_NAME as DELETABLE_TEMPLATE_NAME$1, OnSchemaBreak as OnSchemaBreak$1, OnUpdate as OnUpdate$1, UPDATABLE_TEMPLATE_NAME as UPDATABLE_TEMPLATE_NAME$1 } from "../app.mjs";

//#region src/types/app.ts
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const UPDATABLE_TEMPLATE_NAME = UPDATABLE_TEMPLATE_NAME$1;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const DELETABLE_TEMPLATE_NAME = DELETABLE_TEMPLATE_NAME$1;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const APP_DEPLOY_NOTE_DAPP = APP_DEPLOY_NOTE_DAPP$1;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const APP_PAGE_MAX_SIZE = APP_PAGE_MAX_SIZE$1;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const ABI_RETURN_PREFIX = ABI_RETURN_PREFIX$1;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const OnUpdate = OnUpdate$1;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const OnSchemaBreak = OnSchemaBreak$1;

//#endregion
export { ABI_RETURN_PREFIX, APP_DEPLOY_NOTE_DAPP, APP_PAGE_MAX_SIZE, DELETABLE_TEMPLATE_NAME, OnSchemaBreak, OnUpdate, UPDATABLE_TEMPLATE_NAME };
