// @algorandfoundation/algokit-utils/packages/testing/src/mockServer.js
//#region packages/testing/src/mockServer.ts
/** Environment variable names for external mock server URLs */
const EXTERNAL_URL_ENV_VARS = {
	algod: "MOCK_ALGOD_URL",
	indexer: "MOCK_INDEXER_URL",
	kmd: "MOCK_KMD_URL"
};
/** Default token used for mock server authentication */
const DEFAULT_TOKEN = "a".repeat(64);
/** Default ports for mock servers when running locally (matches algokit-polytest defaults) */
const MOCK_PORTS = {
	algod: { host: 8e3 },
	indexer: { host: 8002 },
	kmd: { host: 8001 }
};
/**
* Check if a server is reachable by performing a health check.
*
* @param url - The base URL of the server to check
* @param timeout - Maximum time to wait for the server to respond (default: 5000ms)
* @returns Promise resolving to true if server is healthy, false otherwise
*/
async function checkServerHealth(url, timeout = 5e3) {
	const healthUrl = `${url.replace(/\/$/, "")}/health`;
	const start = Date.now();
	while (Date.now() - start < timeout) try {
		const controller = new AbortController();
		const timeoutId = setTimeout(() => controller.abort(), 2e3);
		try {
			await fetch(healthUrl, { signal: controller.signal });
			clearTimeout(timeoutId);
			return true;
		} catch (error) {
			clearTimeout(timeoutId);
			if (error instanceof Error && error.name === "AbortError") {
				await new Promise((resolve) => setTimeout(resolve, 200));
				continue;
			}
			throw error;
		}
	} catch {
		await new Promise((resolve) => setTimeout(resolve, 200));
	}
	return false;
}
/**
* Get a mock server instance for the specified client type.
*
* Reads the appropriate environment variable (MOCK_ALGOD_URL, MOCK_INDEXER_URL, or MOCK_KMD_URL)
* and validates that the server is reachable before returning a MockServer instance.
*
* @param clientType - The type of client to get a mock server for ('algod', 'indexer', or 'kmd')
* @returns Promise resolving to a MockServer instance
* @throws Error if the environment variable is not set or the server is not reachable
*
* @example
* ```typescript
* const server = await getMockServer('algod')
* const client = new AlgodClient(DEFAULT_TOKEN, server.baseUrl)
* // ... run tests ...
* ```
*/
async function getMockServer(clientType) {
	const envVar = EXTERNAL_URL_ENV_VARS[clientType];
	const externalUrl = process.env[envVar];
	if (!externalUrl) throw new Error(`Environment variable ${envVar} is not set. Please start the mock server externally and set the URL. See the "Mock Server for Client Tests" section in README.md for local development setup.`);
	if (!await checkServerHealth(externalUrl)) throw new Error(`Mock ${clientType} server at ${externalUrl} is not reachable. Please ensure the server is running and accessible.`);
	return {
		baseUrl: externalUrl.replace(/\/$/, ""),
		clientType
	};
}
const TEST_ADDRESS = "25M5BT2DMMED3V6CWDEYKSNEFGPXX4QBIINCOICLXXRU3UGTSGRMF3MTOE";
const TEST_APP_ID = 718348254;
const TEST_APP_ID_WITH_BOXES = 742949200;
const TEST_BOX_NAME = "b64:cBbHBNV+zUy/Mz5IRhIrBLxr1on5wmidhXEavV+SasC8";
const TEST_ASSET_ID = 705457144;
const TEST_TXID = "VIXTUMAPT7NR4RB2WVOGMETW4QY43KIDA3HWDWWXS3UEDKGTEECQ";
const TEST_ROUND = 24099447;
//#endregion
exports.DEFAULT_TOKEN = DEFAULT_TOKEN;
exports.EXTERNAL_URL_ENV_VARS = EXTERNAL_URL_ENV_VARS;
exports.MOCK_PORTS = MOCK_PORTS;
exports.TEST_ADDRESS = TEST_ADDRESS;
exports.TEST_APP_ID = TEST_APP_ID;
exports.TEST_APP_ID_WITH_BOXES = TEST_APP_ID_WITH_BOXES;
exports.TEST_ASSET_ID = TEST_ASSET_ID;
exports.TEST_BOX_NAME = TEST_BOX_NAME;
exports.TEST_ROUND = TEST_ROUND;
exports.TEST_TXID = TEST_TXID;
exports.checkServerHealth = checkServerHealth;
exports.getMockServer = getMockServer;
