// @algorandfoundation/algokit-utils/packages/algo25/src/index.js
const require_hash = require('../../crypto/src/hash.js');
const require_english = require('./english.js');

//#region packages/algo25/src/index.ts
const FAIL_TO_DECODE_MNEMONIC_ERROR_MSG = "failed to decode mnemonic";
const NOT_IN_WORDS_LIST_ERROR_MSG = "the mnemonic contains a word that is not in the wordlist";
const SEED_BTYES_LENGTH = 32;
function toUint11Array(buffer8) {
	const buffer11 = [];
	let acc = 0;
	let accBits = 0;
	function add(octet) {
		acc |= octet << accBits;
		accBits += 8;
		if (accBits >= 11) {
			buffer11.push(acc & 2047);
			acc >>= 11;
			accBits -= 11;
		}
	}
	function flush() {
		if (accBits) buffer11.push(acc);
	}
	buffer8.forEach(add);
	flush();
	return buffer11;
}
function applyWords(nums) {
	return nums.map((n) => require_english.default[n]);
}
function computeChecksum(seed) {
	return applyWords(toUint11Array(require_hash.hash(seed)))[0];
}
/**
* mnemonicFromSeed converts a 32-byte key into a 25 word mnemonic. The generated mnemonic includes a checksum.
* Each word in the mnemonic represents 11 bits of data, and the last 11 bits are reserved for the checksum.
* @param seed - 32 bytes long seed
* @returns 25 words mnemonic
*/
function mnemonicFromSeed(seed) {
	if (seed.length !== SEED_BTYES_LENGTH) throw new RangeError(`Seed length must be ${SEED_BTYES_LENGTH}`);
	const words = applyWords(toUint11Array(seed));
	const checksumWord = computeChecksum(seed);
	return `${words.join(" ")} ${checksumWord}`;
}
function toUint8Array(buffer11) {
	const buffer8 = [];
	let acc = 0;
	let accBits = 0;
	function add(ui11) {
		acc |= ui11 << accBits;
		accBits += 11;
		while (accBits >= 8) {
			buffer8.push(acc & 255);
			acc >>= 8;
			accBits -= 8;
		}
	}
	function flush() {
		if (accBits) buffer8.push(acc);
	}
	buffer11.forEach(add);
	flush();
	return new Uint8Array(buffer8);
}
/**
* seedFromMnemonic converts a mnemonic generated using this library into the source key used to create it.
* It returns an error if the passed mnemonic has an incorrect checksum, if the number of words is unexpected, or if one
* of the passed words is not found in the words list.
* @param mnemonic - 25 words mnemonic
* @returns 32 bytes long seed
*/
function seedFromMnemonic(mnemonic) {
	const words = mnemonic.split(" ");
	const key = words.slice(0, 24);
	for (const w of key) if (require_english.default.indexOf(w) === -1) throw new Error(NOT_IN_WORDS_LIST_ERROR_MSG);
	const checksum = words[words.length - 1];
	let uint8Array = toUint8Array(key.map((word) => require_english.default.indexOf(word)));
	if (uint8Array.length !== 33) throw new Error(FAIL_TO_DECODE_MNEMONIC_ERROR_MSG);
	if (uint8Array[uint8Array.length - 1] !== 0) throw new Error(FAIL_TO_DECODE_MNEMONIC_ERROR_MSG);
	uint8Array = uint8Array.slice(0, uint8Array.length - 1);
	if (computeChecksum(uint8Array) === checksum) return uint8Array;
	throw new Error(FAIL_TO_DECODE_MNEMONIC_ERROR_MSG);
}
/**
* secretKeyToMnemonic takes an Algorand secret key and returns the corresponding mnemonic.
* @param sk - Algorand secret key
* @returns Secret key's associated mnemonic
*/
function secretKeyToMnemonic(sk) {
	return mnemonicFromSeed(sk.slice(0, SEED_BTYES_LENGTH));
}
/**
* mnemonicToMasterDerivationKey takes a mnemonic string and returns the corresponding master derivation key.
* @param mn - 25 words Algorand mnemonic
* @returns Uint8Array
* @throws error if fails to decode the mnemonic
*/
function mnemonicToMasterDerivationKey(mn) {
	return seedFromMnemonic(mn);
}
/**
* masterDerivationKeyToMnemonic takes a master derivation key and returns the corresponding mnemonic.
* @param mdk - Uint8Array
* @returns string mnemonic
*/
function masterDerivationKeyToMnemonic(mdk) {
	return mnemonicFromSeed(mdk);
}

//#endregion
exports.FAIL_TO_DECODE_MNEMONIC_ERROR_MSG = FAIL_TO_DECODE_MNEMONIC_ERROR_MSG;
exports.NOT_IN_WORDS_LIST_ERROR_MSG = NOT_IN_WORDS_LIST_ERROR_MSG;
exports.masterDerivationKeyToMnemonic = masterDerivationKeyToMnemonic;
exports.mnemonicFromSeed = mnemonicFromSeed;
exports.mnemonicToMasterDerivationKey = mnemonicToMasterDerivationKey;
exports.secretKeyToMnemonic = secretKeyToMnemonic;
exports.seedFromMnemonic = seedFromMnemonic;
