// @algorandfoundation/algokit-utils/transactions/app-call.mjs
import "../packages/common/src/constants.mjs";
import { getAddress, getApplicationAddress } from "../packages/common/src/address.mjs";
import { OnApplicationComplete } from "../packages/transact/src/transactions/app-call.mjs";
import { TransactionType } from "../packages/transact/src/transactions/transaction-type.mjs";
import { Transaction } from "../packages/transact/src/transactions/transaction.mjs";
import { calculateExtraProgramPages } from "../util.mjs";
import { AppManager } from "../app-manager.mjs";
import { buildTransactionCommonData } from "./common.mjs";
//#region src/transactions/app-call.ts
const buildAppCreate = async (params, appManager, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	const approvalProgram = typeof params.approvalProgram === "string" ? (await appManager.compileTeal(params.approvalProgram)).compiledBase64ToBytes : params.approvalProgram;
	const clearStateProgram = typeof params.clearStateProgram === "string" ? (await appManager.compileTeal(params.clearStateProgram)).compiledBase64ToBytes : params.clearStateProgram;
	const globalStateSchema = params.schema?.globalByteSlices !== void 0 || params.schema?.globalInts !== void 0 ? {
		numByteSlices: params.schema?.globalByteSlices ?? 0,
		numUints: params.schema?.globalInts ?? 0
	} : void 0;
	const localStateSchema = params.schema?.localByteSlices !== void 0 || params.schema?.localInts !== void 0 ? {
		numByteSlices: params.schema?.localByteSlices ?? 0,
		numUints: params.schema?.localInts ?? 0
	} : void 0;
	const extraProgramPages = params.extraProgramPages !== void 0 ? params.extraProgramPages : calculateExtraProgramPages(approvalProgram, clearStateProgram);
	const hasAccessReferences = params.accessReferences && params.accessReferences.length > 0;
	return new Transaction({
		...commonData,
		type: TransactionType.AppCall,
		appCall: {
			appId: 0n,
			onComplete: params.onComplete ?? OnApplicationComplete.NoOp,
			approvalProgram,
			clearStateProgram,
			globalStateSchema,
			localStateSchema,
			extraProgramPages,
			args: params.args,
			...hasAccessReferences ? { accessReferences: params.accessReferences } : {
				accountReferences: params.accountReferences?.map((a) => getAddress(a)),
				appReferences: params.appReferences,
				assetReferences: params.assetReferences,
				boxReferences: params.boxReferences?.map(AppManager.getBoxReference)
			},
			rejectVersion: params.rejectVersion
		}
	});
};
const buildAppUpdate = async (params, appManager, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	const approvalProgram = typeof params.approvalProgram === "string" ? (await appManager.compileTeal(params.approvalProgram)).compiledBase64ToBytes : params.approvalProgram;
	const clearStateProgram = typeof params.clearStateProgram === "string" ? (await appManager.compileTeal(params.clearStateProgram)).compiledBase64ToBytes : params.clearStateProgram;
	const hasAccessReferences = params.accessReferences && params.accessReferences.length > 0;
	return new Transaction({
		...commonData,
		type: TransactionType.AppCall,
		appCall: {
			appId: params.appId,
			onComplete: OnApplicationComplete.UpdateApplication,
			approvalProgram,
			clearStateProgram,
			args: params.args,
			...hasAccessReferences ? { accessReferences: params.accessReferences } : {
				accountReferences: params.accountReferences?.map((a) => getAddress(a)),
				appReferences: params.appReferences,
				assetReferences: params.assetReferences,
				boxReferences: params.boxReferences?.map(AppManager.getBoxReference)
			},
			rejectVersion: params.rejectVersion
		}
	});
};
const buildAppCall = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	const hasAccessReferences = params.accessReferences && params.accessReferences.length > 0;
	return new Transaction({
		...commonData,
		type: TransactionType.AppCall,
		appCall: {
			appId: params.appId,
			onComplete: params.onComplete ?? OnApplicationComplete.NoOp,
			args: params.args,
			...hasAccessReferences ? { accessReferences: params.accessReferences } : {
				accountReferences: params.accountReferences?.map((a) => getAddress(a)),
				appReferences: params.appReferences,
				assetReferences: params.assetReferences,
				boxReferences: params.boxReferences?.map(AppManager.getBoxReference)
			},
			rejectVersion: params.rejectVersion
		}
	});
};
/**
* Populate transaction-level resources for app call transactions
*/
function populateTransactionResources(transaction, resourcesAccessed, groupIndex) {
	if (transaction.type !== TransactionType.AppCall || transaction.appCall === void 0) return;
	if (resourcesAccessed.boxes || resourcesAccessed.extraBoxRefs) throw new Error("Unexpected boxes at the transaction level");
	if (resourcesAccessed.appLocals) throw new Error("Unexpected app locals at the transaction level");
	if (resourcesAccessed.assetHoldings) throw new Error("Unexpected asset holdings at the transaction level");
	let accountsCount = 0;
	let appsCount = 0;
	let assetsCount = 0;
	const boxesCount = transaction.appCall.boxReferences?.length ?? 0;
	if (resourcesAccessed.accounts) {
		transaction.appCall.accountReferences = transaction.appCall.accountReferences ?? [];
		for (const account of resourcesAccessed.accounts) {
			const address = account;
			if (!transaction.appCall.accountReferences.some((a) => a.equals(address))) transaction.appCall.accountReferences.push(address);
		}
		accountsCount = transaction.appCall.accountReferences.length;
	}
	if (resourcesAccessed.apps) {
		transaction.appCall.appReferences = transaction.appCall.appReferences ?? [];
		for (const appId of resourcesAccessed.apps) if (!transaction.appCall.appReferences.includes(appId)) transaction.appCall.appReferences.push(appId);
		appsCount = transaction.appCall.appReferences.length;
	}
	if (resourcesAccessed.assets) {
		transaction.appCall.assetReferences = transaction.appCall.assetReferences ?? [];
		for (const assetId of resourcesAccessed.assets) if (!transaction.appCall.assetReferences.includes(assetId)) transaction.appCall.assetReferences.push(assetId);
		assetsCount = transaction.appCall.assetReferences.length;
	}
	if (accountsCount + assetsCount + appsCount + boxesCount > 8) throw new Error(`Resource reference limit of 8 exceeded in transaction ${groupIndex}`);
}
var GroupResourceType = /* @__PURE__ */ function(GroupResourceType) {
	GroupResourceType[GroupResourceType["Account"] = 0] = "Account";
	GroupResourceType[GroupResourceType["App"] = 1] = "App";
	GroupResourceType[GroupResourceType["Asset"] = 2] = "Asset";
	GroupResourceType[GroupResourceType["Box"] = 3] = "Box";
	GroupResourceType[GroupResourceType["ExtraBoxRef"] = 4] = "ExtraBoxRef";
	GroupResourceType[GroupResourceType["AssetHolding"] = 5] = "AssetHolding";
	GroupResourceType[GroupResourceType["AppLocal"] = 6] = "AppLocal";
	return GroupResourceType;
}(GroupResourceType || {});
/**
* Populate group-level resources for app call transactions
*/
function populateGroupResources(transactions, groupResources) {
	let remainingAccounts = [...groupResources.accounts ?? []];
	let remainingApps = [...groupResources.apps ?? []];
	let remainingAssets = [...groupResources.assets ?? []];
	const remainingBoxes = [...groupResources.boxes ?? []];
	if (groupResources.appLocals) groupResources.appLocals.forEach((appLocal) => {
		populateGroupResource(transactions, {
			type: GroupResourceType.AppLocal,
			data: appLocal
		});
		remainingAccounts = remainingAccounts.filter((acc) => acc !== appLocal.address);
		remainingApps = remainingApps.filter((app) => app !== appLocal.appId);
	});
	if (groupResources.assetHoldings) groupResources.assetHoldings.forEach((assetHolding) => {
		populateGroupResource(transactions, {
			type: GroupResourceType.AssetHolding,
			data: assetHolding
		});
		remainingAccounts = remainingAccounts.filter((acc) => acc !== assetHolding.address);
		remainingAssets = remainingAssets.filter((asset) => asset !== assetHolding.assetId);
	});
	remainingAccounts.forEach((account) => {
		populateGroupResource(transactions, {
			type: GroupResourceType.Account,
			data: account
		});
	});
	remainingBoxes.forEach((boxRef) => {
		populateGroupResource(transactions, {
			type: GroupResourceType.Box,
			data: {
				appId: boxRef.appId,
				name: boxRef.name
			}
		});
		remainingApps = remainingApps.filter((app) => app !== boxRef.appId);
	});
	remainingAssets.forEach((asset) => {
		populateGroupResource(transactions, {
			type: GroupResourceType.Asset,
			data: asset
		});
	});
	remainingApps.forEach((app) => {
		populateGroupResource(transactions, {
			type: GroupResourceType.App,
			data: app
		});
	});
	if (groupResources.extraBoxRefs) for (let i = 0; i < groupResources.extraBoxRefs; i++) populateGroupResource(transactions, { type: GroupResourceType.ExtraBoxRef });
}
/**
* Helper function to check if an app call transaction is below resource limit
*/
function isAppCallBelowResourceLimit(txn) {
	if (txn.type !== TransactionType.AppCall) return false;
	if (txn.appCall?.accessReferences?.length) return false;
	const accountsCount = txn.appCall?.accountReferences?.length || 0;
	const assetsCount = txn.appCall?.assetReferences?.length || 0;
	const appsCount = txn.appCall?.appReferences?.length || 0;
	const boxesCount = txn.appCall?.boxReferences?.length || 0;
	return accountsCount + assetsCount + appsCount + boxesCount < 8;
}
/**
* Helper function to populate a specific resource into a transaction group
*/
function populateGroupResource(transactions, resource) {
	if (resource.type === GroupResourceType.AssetHolding || resource.type === GroupResourceType.AppLocal) {
		const address = resource.data.address;
		const groupIndex1 = transactions.findIndex((txn) => {
			if (!isAppCallBelowResourceLimit(txn)) return false;
			const appCall = txn.appCall;
			if (appCall.accountReferences?.some((a) => a.equals(address))) return true;
			if (appCall.appReferences) {
				for (const appId of appCall.appReferences) if (address === getApplicationAddress(appId)) return true;
			}
			if (txn.sender.equals(address)) return true;
			return false;
		});
		if (groupIndex1 !== -1) {
			const appCall = transactions[groupIndex1].appCall;
			if (resource.type === GroupResourceType.AssetHolding) {
				appCall.assetReferences = appCall.assetReferences ?? [];
				if (!appCall.assetReferences.includes(resource.data.assetId)) appCall.assetReferences.push(resource.data.assetId);
			} else {
				appCall.appReferences = appCall.appReferences ?? [];
				if (!appCall.appReferences.includes(resource.data.appId)) appCall.appReferences.push(resource.data.appId);
			}
			return;
		}
		const groupIndex2 = transactions.findIndex((txn) => {
			if (!isAppCallBelowResourceLimit(txn)) return false;
			const appCall = txn.appCall;
			if ((appCall.accountReferences?.length ?? 0) >= 8) return false;
			if (resource.type === GroupResourceType.AssetHolding) return appCall.assetReferences?.includes(resource.data.assetId) || false;
			else return appCall.appReferences?.includes(resource.data.appId) || appCall.appId === resource.data.appId;
		});
		if (groupIndex2 !== -1) {
			const appCall = transactions[groupIndex2].appCall;
			appCall.accountReferences = appCall.accountReferences ?? [];
			if (!appCall.accountReferences.includes(address)) appCall.accountReferences.push(address);
			return;
		}
	}
	if (resource.type === GroupResourceType.Box) {
		const groupIndex = transactions.findIndex((txn) => {
			if (!isAppCallBelowResourceLimit(txn)) return false;
			const appCall = txn.appCall;
			return appCall.appReferences?.includes(resource.data.appId) || appCall.appId === resource.data.appId;
		});
		if (groupIndex !== -1) {
			const appCall = transactions[groupIndex].appCall;
			appCall.boxReferences = appCall.boxReferences ?? [];
			if (!appCall.boxReferences.some((b) => b.appId === resource.data.appId && b.name.length === resource.data.name.length && b.name.every((byte, i) => byte === resource.data.name[i]))) appCall.boxReferences.push({
				appId: resource.data.appId,
				name: resource.data.name
			});
			return;
		}
	}
	const groupIndex = transactions.findIndex((txn) => {
		if (txn.type !== TransactionType.AppCall) return false;
		if (txn.appCall?.accessReferences?.length) return false;
		const appCall = txn.appCall;
		const accountsCount = appCall.accountReferences?.length ?? 0;
		const assetsCount = appCall.assetReferences?.length ?? 0;
		const appsCount = appCall.appReferences?.length ?? 0;
		const boxesCount = appCall.boxReferences?.length ?? 0;
		switch (resource.type) {
			case GroupResourceType.Account: return accountsCount + assetsCount + appsCount + boxesCount < 8;
			case GroupResourceType.AssetHolding:
			case GroupResourceType.AppLocal: return accountsCount + assetsCount + appsCount + boxesCount < 7;
			case GroupResourceType.Box: if (resource.data.appId !== 0n) return accountsCount + assetsCount + appsCount + boxesCount < 7;
			else return accountsCount + assetsCount + appsCount + boxesCount < 8;
			default: return accountsCount + assetsCount + appsCount + boxesCount < 8;
		}
	});
	if (groupIndex === -1) throw new Error("No more transactions below reference limit. Add another app call to the group.");
	const appCall = transactions[groupIndex].appCall;
	switch (resource.type) {
		case GroupResourceType.Account: {
			appCall.accountReferences = appCall.accountReferences ?? [];
			const address = resource.data;
			if (!appCall.accountReferences.some((a) => a.equals(address))) appCall.accountReferences.push(address);
			break;
		}
		case GroupResourceType.App:
			appCall.appReferences = appCall.appReferences ?? [];
			if (!appCall.appReferences.includes(resource.data)) appCall.appReferences.push(resource.data);
			break;
		case GroupResourceType.Box:
			appCall.boxReferences = appCall.boxReferences ?? [];
			if (!appCall.boxReferences.some((b) => b.appId === resource.data.appId && b.name.length === resource.data.name.length && b.name.every((byte, i) => byte === resource.data.name[i]))) appCall.boxReferences.push({
				appId: resource.data.appId,
				name: resource.data.name
			});
			if (resource.data.appId !== 0n) {
				appCall.appReferences = appCall.appReferences ?? [];
				if (!appCall.appReferences.includes(resource.data.appId)) appCall.appReferences.push(resource.data.appId);
			}
			break;
		case GroupResourceType.ExtraBoxRef:
			appCall.boxReferences = appCall.boxReferences ?? [];
			appCall.boxReferences.push({
				appId: 0n,
				name: new Uint8Array(0)
			});
			break;
		case GroupResourceType.AssetHolding: {
			appCall.assetReferences = appCall.assetReferences ?? [];
			if (!appCall.assetReferences.includes(resource.data.assetId)) appCall.assetReferences.push(resource.data.assetId);
			const address = resource.data.address;
			appCall.accountReferences = appCall.accountReferences ?? [];
			if (!appCall.accountReferences.some((a) => a.equals(address))) appCall.accountReferences.push(address);
			break;
		}
		case GroupResourceType.AppLocal: {
			appCall.appReferences = appCall.appReferences ?? [];
			if (!appCall.appReferences.includes(resource.data.appId)) appCall.appReferences.push(resource.data.appId);
			const address = resource.data.address;
			appCall.accountReferences = appCall.accountReferences ?? [];
			if (!appCall.accountReferences.some((a) => a.equals(address))) appCall.accountReferences.push(address);
			break;
		}
		case GroupResourceType.Asset:
			appCall.assetReferences = appCall.assetReferences ?? [];
			if (!appCall.assetReferences.includes(resource.data)) appCall.assetReferences.push(resource.data);
			break;
	}
}
//#endregion
export { buildAppCall, buildAppCreate, buildAppUpdate, populateGroupResources, populateTransactionResources };
