// @algorandfoundation/algokit-utils/transactions/asset-config.js
const require_address = require("../packages/common/src/address.js");
const require_transaction_type = require("../packages/transact/src/transactions/transaction-type.js");
const require_transaction = require("../packages/transact/src/transactions/transaction.js");
const require_common = require("./common.js");
//#region src/transactions/asset-config.ts
const buildAssetCreate = (params, suggestedParams, defaultValidityWindow) => {
	return new require_transaction.Transaction({
		...require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: require_transaction_type.TransactionType.AssetConfig,
		assetConfig: {
			assetId: 0n,
			total: params.total,
			decimals: params.decimals,
			defaultFrozen: params.defaultFrozen,
			assetName: params.assetName,
			unitName: params.unitName,
			url: params.url,
			metadataHash: require_common.ensureString(params.metadataHash),
			manager: require_address.getOptionalAddress(params.manager),
			reserve: require_address.getOptionalAddress(params.reserve),
			freeze: require_address.getOptionalAddress(params.freeze),
			clawback: require_address.getOptionalAddress(params.clawback)
		}
	});
};
const buildAssetConfig = (params, suggestedParams, defaultValidityWindow) => {
	return new require_transaction.Transaction({
		...require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: require_transaction_type.TransactionType.AssetConfig,
		assetConfig: {
			assetId: params.assetId,
			manager: require_address.getOptionalAddress(params.manager),
			reserve: require_address.getOptionalAddress(params.reserve),
			freeze: require_address.getOptionalAddress(params.freeze),
			clawback: require_address.getOptionalAddress(params.clawback)
		}
	});
};
const buildAssetFreeze = (params, suggestedParams, defaultValidityWindow) => {
	return new require_transaction.Transaction({
		...require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: require_transaction_type.TransactionType.AssetFreeze,
		assetFreeze: {
			assetId: params.assetId,
			freezeTarget: require_address.getAddress(params.freezeTarget),
			frozen: params.frozen
		}
	});
};
const buildAssetDestroy = (params, suggestedParams, defaultValidityWindow) => {
	return new require_transaction.Transaction({
		...require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: require_transaction_type.TransactionType.AssetConfig,
		assetConfig: { assetId: params.assetId }
	});
};
//#endregion
exports.buildAssetConfig = buildAssetConfig;
exports.buildAssetCreate = buildAssetCreate;
exports.buildAssetDestroy = buildAssetDestroy;
exports.buildAssetFreeze = buildAssetFreeze;
