// @algorandfoundation/algokit-utils/transactions/asset-transfer.mjs
import { getAddress, getOptionalAddress } from "../packages/common/src/address.mjs";
import { TransactionType } from "../packages/transact/src/transactions/transaction-type.mjs";
import { Transaction } from "../packages/transact/src/transactions/transaction.mjs";
import { buildTransactionCommonData } from "./common.mjs";
//#region src/transactions/asset-transfer.ts
const buildAssetTransfer = (params, suggestedParams, defaultValidityWindow) => {
	return new Transaction({
		...buildTransactionCommonData(params, suggestedParams, defaultValidityWindow),
		type: TransactionType.AssetTransfer,
		assetTransfer: {
			assetId: params.assetId,
			amount: params.amount,
			receiver: getAddress(params.receiver),
			assetSender: getOptionalAddress(params.clawbackTarget),
			closeRemainderTo: getOptionalAddress(params.closeAssetTo)
		}
	});
};
const buildAssetOptIn = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	return new Transaction({
		...commonData,
		type: TransactionType.AssetTransfer,
		assetTransfer: {
			assetId: params.assetId,
			amount: 0n,
			receiver: commonData.sender
		}
	});
};
const buildAssetOptOut = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	return new Transaction({
		...commonData,
		type: TransactionType.AssetTransfer,
		assetTransfer: {
			assetId: params.assetId,
			amount: 0n,
			receiver: commonData.sender,
			closeRemainderTo: getOptionalAddress(params.creator)
		}
	});
};
//#endregion
export { buildAssetOptIn, buildAssetOptOut, buildAssetTransfer };
