// @algorandfoundation/algokit-utils/transactions/fee-coverage.mjs
//#region src/transactions/fee-coverage.ts
let FeeDeltaType = /* @__PURE__ */ function(FeeDeltaType) {
	FeeDeltaType[FeeDeltaType["Deficit"] = 0] = "Deficit";
	FeeDeltaType[FeeDeltaType["Surplus"] = 1] = "Surplus";
	return FeeDeltaType;
}({});
const FeeDelta = {
	fromBigInt(value) {
		if (value > 0n) return {
			type: FeeDeltaType.Deficit,
			data: value
		};
		else if (value < 0n) return {
			type: FeeDeltaType.Surplus,
			data: -value
		};
	},
	toBigInt(delta) {
		return delta.type === FeeDeltaType.Deficit ? delta.data : -delta.data;
	},
	isDeficit(delta) {
		return delta.type === FeeDeltaType.Deficit;
	},
	isSurplus(delta) {
		return delta.type === FeeDeltaType.Surplus;
	},
	amount(delta) {
		return delta.data;
	},
	add(lhs, rhs) {
		return FeeDelta.fromBigInt(FeeDelta.toBigInt(lhs) + FeeDelta.toBigInt(rhs));
	}
};
var CoveredPriority = class CoveredPriority {
	getPriorityType() {
		return 0;
	}
	getDeficitAmount() {
		return 0n;
	}
	compare(other) {
		const typeDiff = this.getPriorityType() - other.getPriorityType();
		if (typeDiff !== 0) return typeDiff;
		return 0;
	}
	equals(other) {
		return other instanceof CoveredPriority;
	}
};
var ModifiableDeficitPriority = class ModifiableDeficitPriority {
	constructor(deficit) {
		this.deficit = deficit;
	}
	getPriorityType() {
		return 1;
	}
	getDeficitAmount() {
		return this.deficit;
	}
	compare(other) {
		const typeDiff = this.getPriorityType() - other.getPriorityType();
		if (typeDiff !== 0) return typeDiff;
		if (other instanceof ModifiableDeficitPriority) return Number(this.deficit - other.deficit);
		return 0;
	}
	equals(other) {
		return other instanceof ModifiableDeficitPriority && this.deficit === other.deficit;
	}
};
var ImmutableDeficitPriority = class ImmutableDeficitPriority {
	constructor(deficit) {
		this.deficit = deficit;
	}
	getPriorityType() {
		return 2;
	}
	getDeficitAmount() {
		return this.deficit;
	}
	compare(other) {
		const typeDiff = this.getPriorityType() - other.getPriorityType();
		if (typeDiff !== 0) return typeDiff;
		if (other instanceof ImmutableDeficitPriority) return Number(this.deficit - other.deficit);
		return 0;
	}
	equals(other) {
		return other instanceof ImmutableDeficitPriority && this.deficit === other.deficit;
	}
};
const FeePriority = {
	Covered: new CoveredPriority(),
	ModifiableDeficit(deficit) {
		return new ModifiableDeficitPriority(deficit);
	},
	ImmutableDeficit(deficit) {
		return new ImmutableDeficitPriority(deficit);
	}
};
function calculateInnerFeeDelta(innerTransactions, minTransactionFee = 1000n, acc) {
	if (!innerTransactions) return acc;
	return innerTransactions.reduceRight((acc, innerTxn) => {
		const recursiveDelta = calculateInnerFeeDelta(innerTxn.innerTxns, minTransactionFee, acc);
		const txnFeeDelta = FeeDelta.fromBigInt(minTransactionFee - (innerTxn.txn.txn.fee ?? 0n));
		const currentFeeDelta = FeeDelta.fromBigInt((recursiveDelta ? FeeDelta.toBigInt(recursiveDelta) : 0n) + (txnFeeDelta ? FeeDelta.toBigInt(txnFeeDelta) : 0n));
		if (currentFeeDelta && FeeDelta.isSurplus(currentFeeDelta)) return;
		return currentFeeDelta;
	}, acc);
}
//#endregion
export { FeeDelta, FeePriority, calculateInnerFeeDelta };
