// @algorandfoundation/algokit-utils/transactions/key-registration.js
const require_transaction_type = require("../packages/transact/src/transactions/transaction-type.js");
const require_transaction = require("../packages/transact/src/transactions/transaction.js");
const require_common = require("./common.js");
//#region src/transactions/key-registration.ts
const buildKeyReg = (params, suggestedParams, defaultValidityWindow) => {
	const commonData = require_common.buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	if ("voteKey" in params) return new require_transaction.Transaction({
		...commonData,
		type: require_transaction_type.TransactionType.KeyRegistration,
		keyRegistration: {
			voteKey: params.voteKey,
			selectionKey: params.selectionKey,
			voteFirst: params.voteFirst,
			voteLast: params.voteLast,
			voteKeyDilution: params.voteKeyDilution,
			nonParticipation: false,
			stateProofKey: params.stateProofKey
		}
	});
	else return new require_transaction.Transaction({
		...commonData,
		type: require_transaction_type.TransactionType.KeyRegistration,
		keyRegistration: { nonParticipation: params.preventAccountFromEverParticipatingAgain }
	});
};
//#endregion
exports.buildKeyReg = buildKeyReg;
