// @algorandfoundation/algokit-utils/transactions/method-call.mjs
import { Address, getAddress } from "../packages/common/src/address.mjs";
import { OnApplicationComplete } from "../packages/transact/src/transactions/app-call.mjs";
import { TransactionType } from "../packages/transact/src/transactions/transaction-type.mjs";
import { Transaction } from "../packages/transact/src/transactions/transaction.mjs";
import { calculateExtraProgramPages } from "../util.mjs";
import { AppManager } from "../app-manager.mjs";
import { buildTransactionCommonData } from "./common.mjs";
import { ABITupleType, ABIUintType } from "../packages/abi/src/abi-type.mjs";
import { argTypeIsReference, argTypeIsTransaction } from "../packages/abi/src/abi-method.mjs";
//#region src/transactions/method-call.ts
const ARGS_TUPLE_PACKING_THRESHOLD = 15;
function extractComposerTransactionsFromAppMethodCallParams(params, parentSigner) {
	const composerTransactions = new Array();
	const methodCallArgs = params.args;
	if (!methodCallArgs) return [];
	const currentSigner = params.signer ? "signer" in params.signer ? params.signer.signer : params.signer : parentSigner;
	for (let i = 0; i < methodCallArgs.length; i++) {
		const arg = methodCallArgs[i];
		if (arg === void 0) continue;
		if (isAbiValue(arg)) continue;
		if (isTransactionWithSignerArg(arg)) {
			composerTransactions.push({
				data: {
					txn: arg.txn,
					signer: arg.signer
				},
				type: "txn"
			});
			continue;
		}
		if (isAppCallMethodCallArg(arg)) {
			const nestedComposerTransactions = extractComposerTransactionsFromAppMethodCallParams(arg, currentSigner);
			composerTransactions.push(...nestedComposerTransactions);
			composerTransactions.push({
				data: {
					...arg,
					signer: arg.signer ?? currentSigner,
					args: processAppMethodCallArgs(arg.args)
				},
				type: "methodCall"
			});
			continue;
		}
		if (arg instanceof Promise) {
			composerTransactions.push({
				data: {
					txn: arg,
					signer: currentSigner
				},
				type: "asyncTxn"
			});
			continue;
		}
		composerTransactions.push({
			data: {
				txn: Promise.resolve(arg),
				signer: currentSigner
			},
			type: "asyncTxn"
		});
	}
	return composerTransactions;
}
function processAppMethodCallArgs(args) {
	if (args === void 0) return void 0;
	return args.map((arg) => {
		if (arg === void 0) return;
		else if (!isAbiValue(arg)) return;
		return arg;
	});
}
function isTransactionWithSignerArg(arg) {
	return typeof arg === "object" && arg !== void 0 && "txn" in arg && "signer" in arg;
}
function isAppCallMethodCallArg(arg) {
	return typeof arg === "object" && arg !== void 0 && "method" in arg;
}
const isAbiValue = (x) => {
	if (Array.isArray(x)) return x.length == 0 || x.every(isAbiValue);
	if (Object.getPrototypeOf(x) === Object.getPrototypeOf({})) return Object.values(x).every(isAbiValue);
	return typeof x === "bigint" || typeof x === "boolean" || typeof x === "number" || typeof x === "string" || x instanceof Uint8Array || x instanceof Address;
};
/**
* Prepares method arguments for ABI encoding by building reference arrays and
* replacing reference-type arguments (account, asset, application) with their indices.
*/
function prepareArgsForEncoding(sender, appId, method, methodArgs, accountReferences, appReferences, assetReferences) {
	const accounts = [...accountReferences ?? []];
	const assets = [...assetReferences ?? []];
	const apps = [...appReferences ?? []];
	return {
		accountReferences: accounts,
		appReferences: apps,
		assetReferences: assets,
		updatedArgs: methodArgs.map((arg, i) => {
			const argType = method.args[i].type;
			if (!argTypeIsReference(argType)) return arg;
			switch (argType) {
				case "account": {
					let addr;
					if (typeof arg === "string") addr = getAddress(arg);
					else if (arg instanceof Uint8Array) addr = new Address(arg);
					else throw new Error("Invalid value for account");
					if (sender.equals(addr)) return 0;
					const existing = accounts.findIndex((a) => a.equals(addr)) + 1;
					if (existing) return existing;
					accounts.push(addr);
					return accounts.length;
				}
				case "asset": {
					if (typeof arg !== "bigint") throw new Error("Invalid value for asset");
					const existing = assets.findIndex((a) => a === arg);
					if (existing === -1) {
						assets.push(arg);
						return assets.length - 1;
					}
					return existing;
				}
				case "application": {
					if (typeof arg !== "bigint") throw new Error("Invalid value for application");
					if (arg === appId) return 0;
					const existing = apps.findIndex((a) => a === arg) + 1;
					if (existing) return existing;
					apps.push(arg);
					return apps.length;
				}
			}
		})
	};
}
/**
* Encode ABI method arguments with tuple packing support
* Ports the logic from the Rust encode_method_arguments function
*/
function encodeMethodArguments(method, args) {
	const encodedArgs = new Array();
	encodedArgs.push(method.getSelector());
	const abiTypes = new Array();
	const abiValues = new Array();
	for (let i = 0; i < method.args.length; i++) {
		const methodArg = method.args[i];
		const argValue = args[i];
		if (argTypeIsTransaction(methodArg.type)) {} else if (argTypeIsReference(methodArg.type)) {
			const referenceType = methodArg.type;
			if (typeof argValue === "number") {
				abiTypes.push(new ABIUintType(8));
				abiValues.push(argValue);
			} else throw new Error(`Invalid reference value for ${referenceType}: ${argValue}`);
		} else if (argValue !== void 0) {
			abiTypes.push(methodArg.type);
			abiValues.push(argValue);
		}
	}
	if (abiValues.length !== abiTypes.length) throw new Error("Mismatch in length of non-transaction arguments");
	if (abiTypes.length > ARGS_TUPLE_PACKING_THRESHOLD) encodedArgs.push(...encodeArgsWithTuplePacking(abiTypes, abiValues));
	else encodedArgs.push(...encodeArgsIndividually(abiTypes, abiValues));
	return encodedArgs;
}
/**
* Encode individual ABI values
*/
function encodeArgsIndividually(abiTypes, abiValues) {
	const encodedArgs = [];
	for (let i = 0; i < abiTypes.length; i++) {
		const encoded = abiTypes[i].encode(abiValues[i]);
		encodedArgs.push(encoded);
	}
	return encodedArgs;
}
/**
* Encode ABI values with tuple packing for methods with many arguments
*/
function encodeArgsWithTuplePacking(abiTypes, abiValues) {
	const encodedArgs = [];
	const splitAt = ARGS_TUPLE_PACKING_THRESHOLD - 1;
	const firstAbiTypes = abiTypes.slice(0, splitAt);
	const firstAbiValues = abiValues.slice(0, splitAt);
	encodedArgs.push(...encodeArgsIndividually(firstAbiTypes, firstAbiValues));
	const remainingAbiTypes = abiTypes.slice(splitAt);
	const remainingAbiValues = abiValues.slice(splitAt);
	if (remainingAbiTypes.length > 0) {
		const tupleEncoded = new ABITupleType(remainingAbiTypes).encode(remainingAbiValues);
		encodedArgs.push(tupleEncoded);
	}
	return encodedArgs;
}
/**
* Builds encoded ABI method arguments and resolves reference arrays
*/
function buildMethodCallArgsAndReferences(params) {
	const { accountReferences, appReferences, assetReferences, updatedArgs } = prepareArgsForEncoding(params.sender, params.appId, params.method, params.args ?? [], params.accountReferences, params.appReferences, params.assetReferences);
	return {
		args: encodeMethodArguments(params.method, updatedArgs),
		accountReferences,
		appReferences,
		assetReferences
	};
}
const buildAppCreateMethodCall = async (params, appManager, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	const approvalProgram = typeof params.approvalProgram === "string" ? (await appManager.compileTeal(params.approvalProgram)).compiledBase64ToBytes : params.approvalProgram;
	const clearStateProgram = typeof params.clearStateProgram === "string" ? (await appManager.compileTeal(params.clearStateProgram)).compiledBase64ToBytes : params.clearStateProgram;
	const globalStateSchema = params.schema?.globalByteSlices !== void 0 || params.schema?.globalInts !== void 0 ? {
		numByteSlices: params.schema?.globalByteSlices ?? 0,
		numUints: params.schema?.globalInts ?? 0
	} : void 0;
	const localStateSchema = params.schema?.localByteSlices !== void 0 || params.schema?.localInts !== void 0 ? {
		numByteSlices: params.schema?.localByteSlices ?? 0,
		numUints: params.schema?.localInts ?? 0
	} : void 0;
	const extraProgramPages = params.extraProgramPages !== void 0 ? params.extraProgramPages : calculateExtraProgramPages(approvalProgram, clearStateProgram);
	const accountReferences = params.accountReferences?.map((a) => getAddress(a));
	const argsAndReferences = buildMethodCallArgsAndReferences({
		sender: commonData.sender,
		appId: 0n,
		method: params.method,
		args: params.args ?? [],
		accountReferences,
		appReferences: params.appReferences,
		assetReferences: params.assetReferences
	});
	const hasAccessReferences = params.accessReferences && params.accessReferences.length > 0;
	return new Transaction({
		...commonData,
		type: TransactionType.AppCall,
		appCall: {
			appId: 0n,
			onComplete: params.onComplete ?? OnApplicationComplete.NoOp,
			approvalProgram,
			clearStateProgram,
			globalStateSchema,
			localStateSchema,
			extraProgramPages,
			args: argsAndReferences.args,
			...hasAccessReferences ? { accessReferences: params.accessReferences } : {
				accountReferences: argsAndReferences.accountReferences,
				appReferences: argsAndReferences.appReferences,
				assetReferences: argsAndReferences.assetReferences,
				boxReferences: params.boxReferences?.map(AppManager.getBoxReference)
			},
			rejectVersion: params.rejectVersion
		}
	});
};
const buildAppUpdateMethodCall = async (params, appManager, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	const approvalProgram = typeof params.approvalProgram === "string" ? (await appManager.compileTeal(params.approvalProgram)).compiledBase64ToBytes : params.approvalProgram;
	const clearStateProgram = typeof params.clearStateProgram === "string" ? (await appManager.compileTeal(params.clearStateProgram)).compiledBase64ToBytes : params.clearStateProgram;
	const accountReferences = params.accountReferences?.map((a) => getAddress(a));
	const argsAndReferences = buildMethodCallArgsAndReferences({
		sender: commonData.sender,
		appId: params.appId,
		method: params.method,
		args: params.args ?? [],
		accountReferences,
		appReferences: params.appReferences,
		assetReferences: params.assetReferences
	});
	const hasAccessReferences = params.accessReferences && params.accessReferences.length > 0;
	return new Transaction({
		...commonData,
		type: TransactionType.AppCall,
		appCall: {
			appId: params.appId,
			onComplete: OnApplicationComplete.UpdateApplication,
			approvalProgram,
			clearStateProgram,
			args: argsAndReferences.args,
			...hasAccessReferences ? { accessReferences: params.accessReferences } : {
				accountReferences: argsAndReferences.accountReferences,
				appReferences: argsAndReferences.appReferences,
				assetReferences: argsAndReferences.assetReferences,
				boxReferences: params.boxReferences?.map(AppManager.getBoxReference)
			},
			rejectVersion: params.rejectVersion
		}
	});
};
const buildAppCallMethodCall = async (params, suggestedParams, defaultValidityWindow) => {
	const commonData = buildTransactionCommonData(params, suggestedParams, defaultValidityWindow);
	const accountReferences = params.accountReferences?.map((a) => getAddress(a));
	const argsAndReferences = buildMethodCallArgsAndReferences({
		sender: commonData.sender,
		appId: params.appId,
		method: params.method,
		args: params.args ?? [],
		accountReferences,
		appReferences: params.appReferences,
		assetReferences: params.assetReferences
	});
	const hasAccessReferences = params.accessReferences && params.accessReferences.length > 0;
	return new Transaction({
		...commonData,
		type: TransactionType.AppCall,
		appCall: {
			appId: params.appId,
			onComplete: params.onComplete ?? OnApplicationComplete.NoOp,
			args: argsAndReferences.args,
			...hasAccessReferences ? { accessReferences: params.accessReferences } : {
				accountReferences: argsAndReferences.accountReferences,
				appReferences: argsAndReferences.appReferences,
				assetReferences: argsAndReferences.assetReferences,
				boxReferences: params.boxReferences?.map(AppManager.getBoxReference)
			},
			rejectVersion: params.rejectVersion
		}
	});
};
//#endregion
export { buildAppCallMethodCall, buildAppCreateMethodCall, buildAppUpdateMethodCall, extractComposerTransactionsFromAppMethodCallParams, processAppMethodCallArgs };
