// @algorandfoundation/algokit-utils/common/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_array = require('../packages/common/src/array.js');
const require_constants = require('../packages/common/src/constants.js');
const require_address = require('../packages/common/src/address.js');
const require_codec = require('../packages/common/src/codecs/codec.js');
const require_address$1 = require('../packages/common/src/codecs/primitives/address.js');
const require_bigint = require('../packages/common/src/codecs/primitives/bigint.js');
const require_boolean = require('../packages/common/src/codecs/primitives/boolean.js');
const require_bytes = require('../packages/common/src/codecs/primitives/bytes.js');
const require_bytes_base64 = require('../packages/common/src/codecs/primitives/bytes-base64.js');
const require_fixed_bytes = require('../packages/common/src/codecs/primitives/fixed-bytes.js');
const require_number = require('../packages/common/src/codecs/primitives/number.js');
const require_wire = require('../packages/common/src/codecs/wire.js');
const require_string = require('../packages/common/src/codecs/primitives/string.js');
const require_array$1 = require('../packages/common/src/codecs/composite/array.js');
const require_record = require('../packages/common/src/codecs/composite/record.js');
const require_unknown = require('../packages/common/src/codecs/primitives/unknown.js');
const require_map = require('../packages/common/src/codecs/composite/map.js');
const require_array_model = require('../packages/common/src/codecs/models/array-model.js');
const require_object_model = require('../packages/common/src/codecs/models/object-model.js');
const require_primitive_model = require('../packages/common/src/codecs/models/primitive-model.js');
const require_encoding = require('../packages/common/src/encoding.js');
const require_json = require('../packages/common/src/json.js');
const require_msgpack = require('../packages/common/src/msgpack.js');
const require_sourcemap = require('../packages/common/src/sourcemap.js');

exports.ADDRESS_LENGTH = require_constants.ADDRESS_LENGTH;
exports.ALGORAND_ADDRESS_BYTE_LENGTH = require_address.ALGORAND_ADDRESS_BYTE_LENGTH;
exports.ALGORAND_ADDRESS_LENGTH = require_address.ALGORAND_ADDRESS_LENGTH;
exports.ALGORAND_CHECKSUM_BYTE_LENGTH = require_address.ALGORAND_CHECKSUM_BYTE_LENGTH;
exports.ALGORAND_ZERO_ADDRESS_STRING = require_address.ALGORAND_ZERO_ADDRESS_STRING;
exports.Address = require_address.Address;
exports.ArrayCodec = require_array$1.ArrayCodec;
exports.ArrayModelCodec = require_array_model.ArrayModelCodec;
exports.BOOL_FALSE_BYTE = require_constants.BOOL_FALSE_BYTE;
exports.BOOL_TRUE_BYTE = require_constants.BOOL_TRUE_BYTE;
exports.CHECKSUM_ADDRESS_ERROR_MSG = require_address.CHECKSUM_ADDRESS_ERROR_MSG;
exports.CHECKSUM_BYTE_LENGTH = require_constants.CHECKSUM_BYTE_LENGTH;
exports.Codec = require_codec.Codec;
exports.EMPTY_SIGNATURE = require_constants.EMPTY_SIGNATURE;
exports.HASH_BYTES_LENGTH = require_constants.HASH_BYTES_LENGTH;
exports.LENGTH_ENCODE_BYTE_SIZE = require_constants.LENGTH_ENCODE_BYTE_SIZE;
exports.MALFORMED_ADDRESS_ERROR_MSG = require_address.MALFORMED_ADDRESS_ERROR_MSG;
exports.MAX_ACCOUNT_REFERENCES = require_constants.MAX_ACCOUNT_REFERENCES;
exports.MAX_APP_ARGS = require_constants.MAX_APP_ARGS;
exports.MAX_APP_REFERENCES = require_constants.MAX_APP_REFERENCES;
exports.MAX_ARGS_SIZE = require_constants.MAX_ARGS_SIZE;
exports.MAX_ASSET_DECIMALS = require_constants.MAX_ASSET_DECIMALS;
exports.MAX_ASSET_NAME_LENGTH = require_constants.MAX_ASSET_NAME_LENGTH;
exports.MAX_ASSET_REFERENCES = require_constants.MAX_ASSET_REFERENCES;
exports.MAX_ASSET_UNIT_NAME_LENGTH = require_constants.MAX_ASSET_UNIT_NAME_LENGTH;
exports.MAX_ASSET_URL_LENGTH = require_constants.MAX_ASSET_URL_LENGTH;
exports.MAX_BOX_REFERENCES = require_constants.MAX_BOX_REFERENCES;
exports.MAX_EXTRA_PROGRAM_PAGES = require_constants.MAX_EXTRA_PROGRAM_PAGES;
exports.MAX_GLOBAL_STATE_KEYS = require_constants.MAX_GLOBAL_STATE_KEYS;
exports.MAX_LOCAL_STATE_KEYS = require_constants.MAX_LOCAL_STATE_KEYS;
exports.MAX_OVERALL_REFERENCES = require_constants.MAX_OVERALL_REFERENCES;
exports.MAX_TRANSACTION_GROUP_SIZE = require_constants.MAX_TRANSACTION_GROUP_SIZE;
exports.MULTISIG_DOMAIN_SEPARATOR = require_constants.MULTISIG_DOMAIN_SEPARATOR;
exports.MapCodec = require_map.MapCodec;
exports.ObjectModelCodec = require_object_model.ObjectModelCodec;
exports.PROGRAM_PAGE_SIZE = require_constants.PROGRAM_PAGE_SIZE;
exports.PUBLIC_KEY_BYTE_LENGTH = require_constants.PUBLIC_KEY_BYTE_LENGTH;
exports.PrimitiveModelCodec = require_primitive_model.PrimitiveModelCodec;
exports.ProgramSourceMap = require_sourcemap.ProgramSourceMap;
exports.RecordCodec = require_record.RecordCodec;
exports.SIGNATURE_BYTE_LENGTH = require_constants.SIGNATURE_BYTE_LENGTH;
exports.SIGNATURE_ENCODING_INCR = require_constants.SIGNATURE_ENCODING_INCR;
exports.TRANSACTION_DOMAIN_SEPARATOR = require_constants.TRANSACTION_DOMAIN_SEPARATOR;
exports.TRANSACTION_GROUP_DOMAIN_SEPARATOR = require_constants.TRANSACTION_GROUP_DOMAIN_SEPARATOR;
exports.TRANSACTION_ID_LENGTH = require_constants.TRANSACTION_ID_LENGTH;
exports.addressArrayCodec = require_array$1.addressArrayCodec;
exports.addressCodec = require_address$1.addressCodec;
exports.arrayEqual = require_array.arrayEqual;
exports.base64ToBytes = require_encoding.base64ToBytes;
exports.bigIntArrayCodec = require_array$1.bigIntArrayCodec;
exports.bigIntCodec = require_bigint.bigIntCodec;
exports.booleanArrayCodec = require_array$1.booleanArrayCodec;
exports.booleanCodec = require_boolean.booleanCodec;
exports.bytesArrayCodec = require_array$1.bytesArrayCodec;
exports.bytesBase64Codec = require_bytes_base64.bytesBase64Codec;
exports.bytesCodec = require_bytes.bytesCodec;
exports.bytesToBase64 = require_encoding.bytesToBase64;
exports.bytesToHex = require_encoding.bytesToHex;
exports.bytesToString = require_encoding.bytesToString;
exports.checksumFromPublicKey = require_address.checksumFromPublicKey;
exports.concatArrays = require_array.concatArrays;
exports.decodeAddress = require_address.decodeAddress;
exports.decodeMsgpack = require_msgpack.decodeMsgpack;
exports.encodeAddress = require_address.encodeAddress;
exports.encodeMsgpack = require_msgpack.encodeMsgpack;
exports.encodeUint64 = require_address.encodeUint64;
exports.fixedBytes1793Codec = require_fixed_bytes.fixedBytes1793Codec;
exports.fixedBytes32Codec = require_fixed_bytes.fixedBytes32Codec;
exports.fixedBytes64Codec = require_fixed_bytes.fixedBytes64Codec;
exports.getAddress = require_address.getAddress;
exports.getApplicationAddress = require_address.getApplicationAddress;
exports.getOptionalAddress = require_address.getOptionalAddress;
exports.hexToBytes = require_encoding.hexToBytes;
exports.isValidAddress = require_address.isValidAddress;
exports.normalizeWireObject = require_wire.normalizeWireObject;
exports.normalizeWireString = require_wire.normalizeWireString;
exports.numberArrayCodec = require_array$1.numberArrayCodec;
exports.numberCodec = require_number.numberCodec;
exports.parseJson = require_json.parseJson;
exports.stringArrayCodec = require_array$1.stringArrayCodec;
exports.stringCodec = require_string.stringCodec;
exports.stringToBytes = require_encoding.stringToBytes;
exports.stringifyJson = require_json.stringifyJson;
exports.unknownCodec = require_unknown.unknownCodec;
