// @algorandfoundation/algokit-utils/packages/algod_client/src/models/account-participation.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
//#region packages/algod_client/src/models/account-participation.ts
const AccountParticipationMeta = {
	name: "AccountParticipation",
	kind: "object",
	fields: [
		{
			name: "selectionParticipationKey",
			wireKey: "selection-participation-key",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "voteFirstValid",
			wireKey: "vote-first-valid",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "vote-key-dilution",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteLastValid",
			wireKey: "vote-last-valid",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "voteParticipationKey",
			wireKey: "vote-participation-key",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "stateProofKey",
			wireKey: "state-proof-key",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		}
	]
};
//#endregion
exports.AccountParticipationMeta = AccountParticipationMeta;
