// @algorandfoundation/algokit-utils/packages/algod_client/src/models/application-params.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayModelCodec } from "../../../common/src/codecs/models/array-model.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { ApplicationStateSchemaMeta } from "./application-state-schema.mjs";
import { TealKeyValueStoreMeta } from "./teal-key-value-store.mjs";
//#region packages/algod_client/src/models/application-params.ts
const ApplicationParamsMeta = {
	name: "ApplicationParams",
	kind: "object",
	fields: [
		{
			name: "creator",
			wireKey: "creator",
			optional: false,
			codec: addressCodec
		},
		{
			name: "approvalProgram",
			wireKey: "approval-program",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "clear-state-program",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "extraProgramPages",
			wireKey: "extra-program-pages",
			optional: true,
			codec: numberCodec
		},
		{
			name: "localStateSchema",
			wireKey: "local-state-schema",
			optional: true,
			codec: new ObjectModelCodec(ApplicationStateSchemaMeta)
		},
		{
			name: "globalStateSchema",
			wireKey: "global-state-schema",
			optional: true,
			codec: new ObjectModelCodec(ApplicationStateSchemaMeta)
		},
		{
			name: "globalState",
			wireKey: "global-state",
			optional: true,
			codec: new ArrayModelCodec(TealKeyValueStoreMeta)
		},
		{
			name: "version",
			wireKey: "version",
			optional: true,
			codec: numberCodec
		}
	]
};
//#endregion
export { ApplicationParamsMeta };
