// @algorandfoundation/algokit-utils/packages/algod_client/src/models/application-params.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_array_model = require("../../../common/src/codecs/models/array-model.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_application_state_schema = require("./application-state-schema.js");
const require_teal_key_value_store = require("./teal-key-value-store.js");
//#region packages/algod_client/src/models/application-params.ts
const ApplicationParamsMeta = {
	name: "ApplicationParams",
	kind: "object",
	fields: [
		{
			name: "creator",
			wireKey: "creator",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "approvalProgram",
			wireKey: "approval-program",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "clear-state-program",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "extraProgramPages",
			wireKey: "extra-program-pages",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "localStateSchema",
			wireKey: "local-state-schema",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_state_schema.ApplicationStateSchemaMeta)
		},
		{
			name: "globalStateSchema",
			wireKey: "global-state-schema",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_application_state_schema.ApplicationStateSchemaMeta)
		},
		{
			name: "globalState",
			wireKey: "global-state",
			optional: true,
			codec: new require_array_model.ArrayModelCodec(require_teal_key_value_store.TealKeyValueStoreMeta)
		},
		{
			name: "version",
			wireKey: "version",
			optional: true,
			codec: require_number.numberCodec
		}
	]
};
//#endregion
exports.ApplicationParamsMeta = ApplicationParamsMeta;
