// @algorandfoundation/algokit-utils/packages/algod_client/src/models/application-state-operation.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_avm_value = require("./avm-value.js");
//#region packages/algod_client/src/models/application-state-operation.ts
const ApplicationStateOperationMeta = {
	name: "ApplicationStateOperation",
	kind: "object",
	fields: [
		{
			name: "operation",
			wireKey: "operation",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "appStateType",
			wireKey: "app-state-type",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "key",
			wireKey: "key",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "newValue",
			wireKey: "new-value",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_avm_value.AvmValueMeta)
		},
		{
			name: "account",
			wireKey: "account",
			optional: true,
			codec: require_address.addressCodec
		}
	]
};
//#endregion
exports.ApplicationStateOperationMeta = ApplicationStateOperationMeta;
