// @algorandfoundation/algokit-utils/packages/algod_client/src/models/asset-params.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes32Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
//#region packages/algod_client/src/models/asset-params.ts
const AssetParamsMeta = {
	name: "AssetParams",
	kind: "object",
	fields: [
		{
			name: "clawback",
			wireKey: "clawback",
			optional: true,
			codec: stringCodec
		},
		{
			name: "creator",
			wireKey: "creator",
			optional: false,
			codec: stringCodec
		},
		{
			name: "decimals",
			wireKey: "decimals",
			optional: false,
			codec: numberCodec
		},
		{
			name: "defaultFrozen",
			wireKey: "default-frozen",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "freeze",
			wireKey: "freeze",
			optional: true,
			codec: stringCodec
		},
		{
			name: "manager",
			wireKey: "manager",
			optional: true,
			codec: stringCodec
		},
		{
			name: "metadataHash",
			wireKey: "metadata-hash",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "name",
			wireKey: "name",
			optional: true,
			codec: stringCodec
		},
		{
			name: "nameB64",
			wireKey: "name-b64",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "reserve",
			wireKey: "reserve",
			optional: true,
			codec: stringCodec
		},
		{
			name: "total",
			wireKey: "total",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "unitName",
			wireKey: "unit-name",
			optional: true,
			codec: stringCodec
		},
		{
			name: "unitNameB64",
			wireKey: "unit-name-b64",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "url",
			wireKey: "url",
			optional: true,
			codec: stringCodec
		},
		{
			name: "urlB64",
			wireKey: "url-b64",
			optional: true,
			codec: bytesCodec
		}
	]
};
//#endregion
export { AssetParamsMeta };
