// @algorandfoundation/algokit-utils/packages/algod_client/src/models/block.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_fixed_bytes = require("../../../common/src/codecs/primitives/fixed-bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_map = require("../../../common/src/codecs/composite/map.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_signed_transaction_meta = require("../../../transact/src/transactions/signed-transaction-meta.js");
//#region packages/algod_client/src/models/block.ts
const BlockEvalDeltaMeta = {
	name: "BlockEvalDelta",
	kind: "object",
	fields: [
		{
			name: "action",
			wireKey: "at",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "bytes",
			wireKey: "bs",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "uint",
			wireKey: "ui",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};
const BlockAppEvalDeltaMeta = {
	name: "BlockAppEvalDelta",
	kind: "object",
	fields: [
		{
			name: "globalDelta",
			wireKey: "gd",
			optional: true,
			codec: new require_map.MapCodec(require_bytes.bytesCodec, new require_object_model.ObjectModelCodec(BlockEvalDeltaMeta))
		},
		{
			name: "localDeltas",
			wireKey: "ld",
			optional: true,
			codec: new require_map.MapCodec(require_number.numberCodec, new require_map.MapCodec(require_bytes.bytesCodec, new require_object_model.ObjectModelCodec(BlockEvalDeltaMeta)))
		},
		{
			name: "innerTxns",
			wireKey: "itx",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(() => SignedTxnWithADMeta))
		},
		{
			name: "sharedAccounts",
			wireKey: "sa",
			optional: true,
			codec: require_array.addressArrayCodec
		},
		{
			name: "logs",
			wireKey: "lg",
			optional: true,
			codec: require_array.bytesArrayCodec
		}
	]
};
const BlockStateProofTrackingDataMeta = {
	name: "BlockStateProofTrackingData",
	kind: "object",
	fields: [
		{
			name: "stateProofVotersCommitment",
			wireKey: "v",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "stateProofOnlineTotalWeight",
			wireKey: "t",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "stateProofNextRound",
			wireKey: "n",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};
const ApplyDataMeta = {
	name: "SignedTxnInBlock",
	kind: "object",
	fields: [
		{
			name: "closingAmount",
			wireKey: "ca",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "assetClosingAmount",
			wireKey: "aca",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "senderRewards",
			wireKey: "rs",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "receiverRewards",
			wireKey: "rr",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "closeRewards",
			wireKey: "rc",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "evalDelta",
			wireKey: "dt",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(BlockAppEvalDeltaMeta)
		},
		{
			name: "configAsset",
			wireKey: "caid",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "applicationId",
			wireKey: "apid",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};
const SignedTxnWithADMeta = {
	name: "SignedTxnWithAD",
	kind: "object",
	fields: [{
		name: "signedTxn",
		flattened: true,
		optional: false,
		codec: new require_object_model.ObjectModelCodec(require_signed_transaction_meta.SignedTransactionMeta)
	}, {
		name: "applyData",
		flattened: true,
		optional: true,
		codec: new require_object_model.ObjectModelCodec(ApplyDataMeta)
	}]
};
const SignedTxnInBlockMeta = {
	name: "SignedTxnInBlock",
	kind: "object",
	fields: [
		{
			name: "signedTxn",
			flattened: true,
			optional: false,
			codec: new require_object_model.ObjectModelCodec(SignedTxnWithADMeta)
		},
		{
			name: "hasGenesisId",
			wireKey: "hgi",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "hasGenesisHash",
			wireKey: "hgh",
			optional: true,
			codec: require_boolean.booleanCodec
		}
	]
};
const ParticipationUpdatesMeta = {
	name: "ParticipationUpdates",
	kind: "object",
	fields: [{
		name: "expiredParticipationAccounts",
		wireKey: "partupdrmv",
		optional: false,
		codec: require_array.addressArrayCodec
	}, {
		name: "absentParticipationAccounts",
		wireKey: "partupdabs",
		optional: false,
		codec: require_array.addressArrayCodec
	}]
};
const TxnCommitmentsMeta = {
	name: "TxnCommitments",
	kind: "object",
	fields: [
		{
			name: "nativeSha512_256Commitment",
			wireKey: "txn",
			optional: false,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "sha256Commitment",
			wireKey: "txn256",
			optional: true,
			codec: require_fixed_bytes.fixedBytes32Codec
		},
		{
			name: "sha512Commitment",
			wireKey: "txn512",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		}
	]
};
const RewardStateMeta = {
	name: "RewardState",
	kind: "object",
	fields: [
		{
			name: "feeSink",
			wireKey: "fees",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "rewardsPool",
			wireKey: "rwd",
			optional: false,
			codec: require_address.addressCodec
		},
		{
			name: "rewardsLevel",
			wireKey: "earn",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsRate",
			wireKey: "rate",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsResidue",
			wireKey: "frac",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "rewardsRecalculationRound",
			wireKey: "rwcalr",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};
const UpgradeStateMeta = {
	name: "UpgradeState",
	kind: "object",
	fields: [
		{
			name: "currentProtocol",
			wireKey: "proto",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "nextProtocol",
			wireKey: "nextproto",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "nextProtocolApprovals",
			wireKey: "nextyes",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextProtocolVoteBefore",
			wireKey: "nextbefore",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "nextProtocolSwitchOn",
			wireKey: "nextswitch",
			optional: true,
			codec: require_bigint.bigIntCodec
		}
	]
};
const UpgradeVoteMeta = {
	name: "UpgradeVote",
	kind: "object",
	fields: [
		{
			name: "upgradePropose",
			wireKey: "upgradeprop",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "upgradeDelay",
			wireKey: "upgradedelay",
			optional: true,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "upgradeApprove",
			wireKey: "upgradeyes",
			optional: true,
			codec: require_boolean.booleanCodec
		}
	]
};
const BlockMeta = {
	name: "Block",
	kind: "object",
	fields: [{
		name: "header",
		flattened: true,
		optional: false,
		codec: new require_object_model.ObjectModelCodec({
			name: "BlockHeader",
			kind: "object",
			fields: [
				{
					name: "round",
					wireKey: "rnd",
					optional: false,
					codec: require_bigint.bigIntCodec
				},
				{
					name: "previousBlockHash",
					wireKey: "prev",
					optional: false,
					codec: require_fixed_bytes.fixedBytes32Codec
				},
				{
					name: "previousBlockHash512",
					wireKey: "prev512",
					optional: true,
					codec: require_fixed_bytes.fixedBytes64Codec
				},
				{
					name: "seed",
					wireKey: "seed",
					optional: false,
					codec: require_bytes.bytesCodec
				},
				{
					name: "txnCommitments",
					flattened: true,
					optional: false,
					codec: new require_object_model.ObjectModelCodec(TxnCommitmentsMeta)
				},
				{
					name: "timestamp",
					wireKey: "ts",
					optional: false,
					codec: require_bigint.bigIntCodec
				},
				{
					name: "genesisId",
					wireKey: "gen",
					optional: false,
					codec: require_string.stringCodec
				},
				{
					name: "genesisHash",
					wireKey: "gh",
					optional: false,
					codec: require_fixed_bytes.fixedBytes32Codec
				},
				{
					name: "proposer",
					wireKey: "prp",
					optional: true,
					codec: require_address.addressCodec
				},
				{
					name: "feesCollected",
					wireKey: "fc",
					optional: true,
					codec: require_bigint.bigIntCodec
				},
				{
					name: "bonus",
					wireKey: "bi",
					optional: true,
					codec: require_bigint.bigIntCodec
				},
				{
					name: "proposerPayout",
					wireKey: "pp",
					optional: true,
					codec: require_bigint.bigIntCodec
				},
				{
					name: "rewardState",
					flattened: true,
					optional: false,
					codec: new require_object_model.ObjectModelCodec(RewardStateMeta)
				},
				{
					name: "upgradeState",
					flattened: true,
					optional: false,
					codec: new require_object_model.ObjectModelCodec(UpgradeStateMeta)
				},
				{
					name: "upgradeVote",
					flattened: true,
					optional: true,
					codec: new require_object_model.ObjectModelCodec(UpgradeVoteMeta)
				},
				{
					name: "txnCounter",
					wireKey: "tc",
					optional: true,
					codec: require_bigint.bigIntCodec
				},
				{
					name: "stateProofTracking",
					wireKey: "spt",
					optional: true,
					codec: new require_map.MapCodec(require_number.numberCodec, new require_object_model.ObjectModelCodec(BlockStateProofTrackingDataMeta))
				},
				{
					name: "participationUpdates",
					flattened: true,
					optional: false,
					codec: new require_object_model.ObjectModelCodec(ParticipationUpdatesMeta)
				}
			]
		})
	}, {
		name: "payset",
		wireKey: "txns",
		optional: false,
		codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(SignedTxnInBlockMeta))
	}]
};
/**
* Custom codec for Block that populates genesis information on transactions after decoding.
*
* When blocks are returned from algod, transactions may not include the genesisId
* and genesisHash fields even though they are required for correct transaction ID calculation.
* The block contains `hasGenesisId` and `hasGenesisHash` flags that indicate whether these
* fields should be populated from the block header.
*
* This codec automatically populates these fields after decoding to ensure transaction IDs
* can be calculated correctly.
*/
var BlockCodec = class extends require_object_model.ObjectModelCodec {
	constructor() {
		super(BlockMeta);
	}
	fromEncoded(value, format) {
		const block = super.fromEncoded(value, format);
		const genesisId = block.header.genesisId;
		const genesisHash = block.header.genesisHash;
		for (const txnInBlock of block.payset ?? []) {
			const txn = txnInBlock.signedTxn.signedTxn.txn;
			if (txnInBlock.hasGenesisId && txn.genesisId === void 0) txn.genesisId = genesisId;
			if (txnInBlock.hasGenesisHash !== false && txn.genesisHash === void 0) txn.genesisHash = genesisHash;
		}
		return block;
	}
};
const blockCodec = new BlockCodec();
//#endregion
exports.blockCodec = blockCodec;
