// @algorandfoundation/algokit-utils/packages/algod_client/src/models/genesis.js
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_genesis_allocation = require("./genesis-allocation.js");
//#region packages/algod_client/src/models/genesis.ts
const GenesisMeta = {
	name: "Genesis",
	kind: "object",
	fields: [
		{
			name: "alloc",
			wireKey: "alloc",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_genesis_allocation.GenesisAllocationMeta))
		},
		{
			name: "comment",
			wireKey: "comment",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "devmode",
			wireKey: "devmode",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "fees",
			wireKey: "fees",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "network",
			wireKey: "network",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "proto",
			wireKey: "proto",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "rwd",
			wireKey: "rwd",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "timestamp",
			wireKey: "timestamp",
			optional: true,
			codec: require_number.numberCodec
		}
	]
};
//#endregion
exports.GenesisMeta = GenesisMeta;
