// @algorandfoundation/algokit-utils/packages/algod_client/src/models/genesis-allocation.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const GenesisAllocationMeta = {
	name: "GenesisAllocation",
	kind: "object",
	fields: [
		{
			name: "addr",
			wireKey: "addr",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "comment",
			wireKey: "comment",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "state",
			wireKey: "state",
			optional: false,
			codec: new require_object_model.ObjectModelCodec({
				name: "GenesisAllocationStateMeta",
				kind: "object",
				fields: [
					{
						name: "algo",
						wireKey: "algo",
						optional: false,
						codec: require_bigint.bigIntCodec
					},
					{
						name: "onl",
						wireKey: "onl",
						optional: false,
						codec: require_number.numberCodec
					},
					{
						name: "sel",
						wireKey: "sel",
						optional: true,
						codec: require_string.stringCodec
					},
					{
						name: "stprf",
						wireKey: "stprf",
						optional: true,
						codec: require_string.stringCodec
					},
					{
						name: "vote",
						wireKey: "vote",
						optional: true,
						codec: require_string.stringCodec
					},
					{
						name: "voteKd",
						wireKey: "voteKD",
						optional: true,
						codec: require_bigint.bigIntCodec
					},
					{
						name: "voteFst",
						wireKey: "voteFst",
						optional: true,
						codec: require_bigint.bigIntCodec
					},
					{
						name: "voteLst",
						wireKey: "voteLst",
						optional: true,
						codec: require_bigint.bigIntCodec
					}
				]
			})
		}
	]
};
//#endregion
exports.GenesisAllocationMeta = GenesisAllocationMeta;
