// @algorandfoundation/algokit-utils/packages/algod_client/src/models/ledger-state-delta.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { MapCodec } from "../../../common/src/codecs/composite/map.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { blockCodec } from "./block.mjs";
//#region packages/algod_client/src/models/ledger-state-delta.ts
const LedgerTealValueMeta = {
	name: "LedgerTealValue",
	kind: "object",
	fields: [
		{
			name: "type",
			wireKey: "tt",
			optional: false,
			codec: numberCodec
		},
		{
			name: "bytes",
			wireKey: "tb",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "uint",
			wireKey: "ui",
			optional: true,
			codec: bigIntCodec
		}
	]
};
const LedgerStateSchemaMeta = {
	name: "LedgerStateSchema",
	kind: "object",
	fields: [{
		name: "numUints",
		wireKey: "nui",
		optional: true,
		codec: bigIntCodec
	}, {
		name: "numByteSlices",
		wireKey: "nbs",
		optional: true,
		codec: bigIntCodec
	}]
};
const LedgerAppParamsMeta = {
	name: "LedgerAppParams",
	kind: "object",
	fields: [
		{
			name: "approvalProgram",
			wireKey: "approv",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "clearp",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "localStateSchema",
			wireKey: "lsch",
			optional: false,
			codec: new ObjectModelCodec(LedgerStateSchemaMeta)
		},
		{
			name: "globalStateSchema",
			wireKey: "gsch",
			optional: false,
			codec: new ObjectModelCodec(LedgerStateSchemaMeta)
		},
		{
			name: "extraProgramPages",
			wireKey: "epp",
			optional: false,
			codec: numberCodec
		},
		{
			name: "version",
			wireKey: "v",
			optional: true,
			codec: numberCodec
		},
		{
			name: "sizeSponsor",
			wireKey: "ss",
			optional: true,
			codec: addressCodec
		},
		{
			name: "globalState",
			wireKey: "gs",
			optional: true,
			codec: new MapCodec(bytesCodec, new ObjectModelCodec(LedgerTealValueMeta))
		}
	]
};
const LedgerAppLocalStateMeta = {
	name: "LedgerAppLocalState",
	kind: "object",
	fields: [{
		name: "schema",
		wireKey: "hsch",
		optional: false,
		codec: new ObjectModelCodec(LedgerStateSchemaMeta)
	}, {
		name: "keyValue",
		wireKey: "tkv",
		optional: true,
		codec: new MapCodec(bytesCodec, new ObjectModelCodec(LedgerTealValueMeta))
	}]
};
const LedgerAppLocalStateDeltaMeta = {
	name: "LedgerAppLocalStateDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: booleanCodec
	}, {
		name: "localState",
		wireKey: "LocalState",
		optional: true,
		codec: new ObjectModelCodec(LedgerAppLocalStateMeta)
	}]
};
const LedgerAppParamsDeltaMeta = {
	name: "LedgerAppParamsDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: booleanCodec
	}, {
		name: "params",
		wireKey: "Params",
		optional: true,
		codec: new ObjectModelCodec(LedgerAppParamsMeta)
	}]
};
const LedgerAppResourceRecordMeta = {
	name: "LedgerAppResourceRecord",
	kind: "object",
	fields: [
		{
			name: "appId",
			wireKey: "Aidx",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "address",
			wireKey: "Addr",
			optional: false,
			codec: addressCodec
		},
		{
			name: "params",
			wireKey: "Params",
			optional: false,
			codec: new ObjectModelCodec(LedgerAppParamsDeltaMeta)
		},
		{
			name: "state",
			wireKey: "State",
			optional: false,
			codec: new ObjectModelCodec(LedgerAppLocalStateDeltaMeta)
		}
	]
};
const LedgerAssetHoldingDeltaMeta = {
	name: "LedgerAssetHoldingDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: booleanCodec
	}, {
		name: "holding",
		wireKey: "Holding",
		optional: true,
		codec: new ObjectModelCodec({
			name: "LedgerAssetHolding",
			kind: "object",
			fields: [{
				name: "amount",
				wireKey: "a",
				optional: false,
				codec: bigIntCodec
			}, {
				name: "frozen",
				wireKey: "f",
				optional: false,
				codec: booleanCodec
			}]
		})
	}]
};
const LedgerAssetParamsDeltaMeta = {
	name: "LedgerAssetParamsDelta",
	kind: "object",
	fields: [{
		name: "deleted",
		wireKey: "Deleted",
		optional: false,
		codec: booleanCodec
	}, {
		name: "params",
		wireKey: "Params",
		optional: true,
		codec: new ObjectModelCodec({
			name: "LedgerAssetParams",
			kind: "object",
			fields: [
				{
					name: "total",
					wireKey: "t",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "decimals",
					wireKey: "dc",
					optional: false,
					codec: numberCodec
				},
				{
					name: "defaultFrozen",
					wireKey: "df",
					optional: false,
					codec: booleanCodec
				},
				{
					name: "unitName",
					wireKey: "un",
					optional: true,
					codec: stringCodec
				},
				{
					name: "assetName",
					wireKey: "an",
					optional: true,
					codec: stringCodec
				},
				{
					name: "url",
					wireKey: "au",
					optional: true,
					codec: stringCodec
				},
				{
					name: "metadataHash",
					wireKey: "am",
					optional: true,
					codec: bytesCodec
				},
				{
					name: "manager",
					wireKey: "m",
					optional: true,
					codec: addressCodec
				},
				{
					name: "reserve",
					wireKey: "r",
					optional: true,
					codec: addressCodec
				},
				{
					name: "freeze",
					wireKey: "f",
					optional: true,
					codec: addressCodec
				},
				{
					name: "clawback",
					wireKey: "c",
					optional: true,
					codec: addressCodec
				}
			]
		})
	}]
};
const LedgerAssetResourceRecordMeta = {
	name: "LedgerAssetResourceRecord",
	kind: "object",
	fields: [
		{
			name: "assetId",
			wireKey: "Aidx",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "address",
			wireKey: "Addr",
			optional: false,
			codec: addressCodec
		},
		{
			name: "params",
			wireKey: "Params",
			optional: false,
			codec: new ObjectModelCodec(LedgerAssetParamsDeltaMeta)
		},
		{
			name: "holding",
			wireKey: "Holding",
			optional: false,
			codec: new ObjectModelCodec(LedgerAssetHoldingDeltaMeta)
		}
	]
};
const LedgerVotingDataMeta = {
	name: "LedgerVotingData",
	kind: "object",
	fields: [
		{
			name: "voteId",
			wireKey: "VoteID",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "selectionId",
			wireKey: "SelectionID",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "stateProofId",
			wireKey: "StateProofID",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "voteFirstValid",
			wireKey: "VoteFirstValid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "voteLastValid",
			wireKey: "VoteLastValid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "VoteKeyDilution",
			optional: false,
			codec: bigIntCodec
		}
	]
};
const LedgerAccountDataMeta = {
	name: "LedgerAccountData",
	kind: "object",
	fields: [{
		name: "accountBaseData",
		flattened: true,
		optional: false,
		codec: new ObjectModelCodec({
			name: "LedgerAccountBaseData",
			kind: "object",
			fields: [
				{
					name: "status",
					wireKey: "Status",
					optional: false,
					codec: numberCodec
				},
				{
					name: "microAlgos",
					wireKey: "MicroAlgos",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "rewardsBase",
					wireKey: "RewardsBase",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "rewardedMicroAlgos",
					wireKey: "RewardedMicroAlgos",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "authAddress",
					wireKey: "AuthAddr",
					optional: false,
					codec: addressCodec
				},
				{
					name: "incentiveEligible",
					wireKey: "IncentiveEligible",
					optional: false,
					codec: booleanCodec
				},
				{
					name: "totalAppSchema",
					wireKey: "TotalAppSchema",
					optional: false,
					codec: new ObjectModelCodec(LedgerStateSchemaMeta)
				},
				{
					name: "totalExtraAppPages",
					wireKey: "TotalExtraAppPages",
					optional: false,
					codec: numberCodec
				},
				{
					name: "totalAppParams",
					wireKey: "TotalAppParams",
					optional: false,
					codec: numberCodec
				},
				{
					name: "totalAppLocalStates",
					wireKey: "TotalAppLocalStates",
					optional: false,
					codec: numberCodec
				},
				{
					name: "totalAssetParams",
					wireKey: "TotalAssetParams",
					optional: false,
					codec: numberCodec
				},
				{
					name: "totalAssets",
					wireKey: "TotalAssets",
					optional: false,
					codec: numberCodec
				},
				{
					name: "totalBoxes",
					wireKey: "TotalBoxes",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "totalBoxBytes",
					wireKey: "TotalBoxBytes",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "lastProposed",
					wireKey: "LastProposed",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "lastHeartbeat",
					wireKey: "LastHeartbeat",
					optional: false,
					codec: bigIntCodec
				}
			]
		})
	}, {
		name: "votingData",
		flattened: true,
		optional: false,
		codec: new ObjectModelCodec(LedgerVotingDataMeta)
	}]
};
const LedgerAccountDeltasMeta = {
	name: "LedgerAccountDeltas",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "Accts",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec({
				name: "LedgerBalanceRecord",
				kind: "object",
				fields: [{
					name: "address",
					wireKey: "Addr",
					optional: false,
					codec: addressCodec
				}, {
					name: "accountData",
					flattened: true,
					optional: false,
					codec: new ObjectModelCodec(LedgerAccountDataMeta)
				}]
			}))
		},
		{
			name: "appResources",
			wireKey: "AppResources",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(LedgerAppResourceRecordMeta))
		},
		{
			name: "assetResources",
			wireKey: "AssetResources",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(LedgerAssetResourceRecordMeta))
		}
	]
};
const LedgerKvValueDeltaMeta = {
	name: "LedgerKvValueDelta",
	kind: "object",
	fields: [{
		name: "data",
		wireKey: "Data",
		optional: true,
		codec: bytesCodec
	}, {
		name: "oldData",
		wireKey: "OldData",
		optional: true,
		codec: bytesCodec
	}]
};
const LedgerIncludedTransactionsMeta = {
	name: "LedgerIncludedTransactions",
	kind: "object",
	fields: [{
		name: "lastValid",
		wireKey: "LastValid",
		optional: false,
		codec: bigIntCodec
	}, {
		name: "intra",
		wireKey: "Intra",
		optional: false,
		codec: numberCodec
	}]
};
const LedgerModifiedCreatableMeta = {
	name: "LedgerModifiedCreatable",
	kind: "object",
	fields: [
		{
			name: "creatableType",
			wireKey: "Ctype",
			optional: false,
			codec: numberCodec
		},
		{
			name: "created",
			wireKey: "Created",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "creator",
			wireKey: "Creator",
			optional: false,
			codec: addressCodec
		},
		{
			name: "ndeltas",
			wireKey: "Ndeltas",
			optional: false,
			codec: numberCodec
		}
	]
};
const LedgerAlgoCountMeta = {
	name: "LedgerAlgoCount",
	kind: "object",
	fields: [{
		name: "money",
		wireKey: "mon",
		optional: false,
		codec: bigIntCodec
	}, {
		name: "rewardUnits",
		wireKey: "rwd",
		optional: false,
		codec: bigIntCodec
	}]
};
const LedgerAccountTotalsMeta = {
	name: "LedgerAccountTotals",
	kind: "object",
	fields: [
		{
			name: "online",
			wireKey: "online",
			optional: false,
			codec: new ObjectModelCodec(LedgerAlgoCountMeta)
		},
		{
			name: "offline",
			wireKey: "offline",
			optional: false,
			codec: new ObjectModelCodec(LedgerAlgoCountMeta)
		},
		{
			name: "notParticipating",
			wireKey: "notpart",
			optional: false,
			codec: new ObjectModelCodec(LedgerAlgoCountMeta)
		},
		{
			name: "rewardsLevel",
			wireKey: "rwdlvl",
			optional: false,
			codec: bigIntCodec
		}
	]
};
const LedgerStateDeltaMeta = {
	name: "LedgerStateDelta",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "Accts",
			optional: false,
			codec: new ObjectModelCodec(LedgerAccountDeltasMeta)
		},
		{
			name: "block",
			wireKey: "Hdr",
			optional: false,
			codec: blockCodec
		},
		{
			name: "stateProofNext",
			wireKey: "StateProofNext",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "prevTimestamp",
			wireKey: "PrevTimestamp",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "totals",
			wireKey: "Totals",
			optional: false,
			codec: new ObjectModelCodec(LedgerAccountTotalsMeta)
		},
		{
			name: "kvMods",
			wireKey: "KvMods",
			optional: true,
			codec: new MapCodec(bytesCodec, new ObjectModelCodec(LedgerKvValueDeltaMeta))
		},
		{
			name: "txIds",
			wireKey: "Txids",
			optional: true,
			codec: new MapCodec(bytesCodec, new ObjectModelCodec(LedgerIncludedTransactionsMeta))
		},
		{
			name: "creatables",
			wireKey: "Creatables",
			optional: true,
			codec: new MapCodec(numberCodec, new ObjectModelCodec(LedgerModifiedCreatableMeta))
		}
	]
};
//#endregion
export { LedgerStateDeltaMeta };
