// @algorandfoundation/algokit-utils/packages/algod_client/src/models/node-status-response.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
//#region packages/algod_client/src/models/node-status-response.ts
const NodeStatusResponseMeta = {
	name: "NodeStatusResponse",
	kind: "object",
	fields: [
		{
			name: "catchupTime",
			wireKey: "catchup-time",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "lastRound",
			wireKey: "last-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "lastVersion",
			wireKey: "last-version",
			optional: false,
			codec: stringCodec
		},
		{
			name: "nextVersion",
			wireKey: "next-version",
			optional: false,
			codec: stringCodec
		},
		{
			name: "nextVersionRound",
			wireKey: "next-version-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "nextVersionSupported",
			wireKey: "next-version-supported",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "stoppedAtUnsupportedRound",
			wireKey: "stopped-at-unsupported-round",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "timeSinceLastRound",
			wireKey: "time-since-last-round",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "lastCatchpoint",
			wireKey: "last-catchpoint",
			optional: true,
			codec: stringCodec
		},
		{
			name: "catchpoint",
			wireKey: "catchpoint",
			optional: true,
			codec: stringCodec
		},
		{
			name: "catchpointTotalAccounts",
			wireKey: "catchpoint-total-accounts",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointProcessedAccounts",
			wireKey: "catchpoint-processed-accounts",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointVerifiedAccounts",
			wireKey: "catchpoint-verified-accounts",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointTotalKvs",
			wireKey: "catchpoint-total-kvs",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointProcessedKvs",
			wireKey: "catchpoint-processed-kvs",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointVerifiedKvs",
			wireKey: "catchpoint-verified-kvs",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointTotalBlocks",
			wireKey: "catchpoint-total-blocks",
			optional: true,
			codec: numberCodec
		},
		{
			name: "catchpointAcquiredBlocks",
			wireKey: "catchpoint-acquired-blocks",
			optional: true,
			codec: numberCodec
		},
		{
			name: "upgradeDelay",
			wireKey: "upgrade-delay",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "upgradeNodeVote",
			wireKey: "upgrade-node-vote",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "upgradeVotesRequired",
			wireKey: "upgrade-votes-required",
			optional: true,
			codec: numberCodec
		},
		{
			name: "upgradeVotes",
			wireKey: "upgrade-votes",
			optional: true,
			codec: numberCodec
		},
		{
			name: "upgradeYesVotes",
			wireKey: "upgrade-yes-votes",
			optional: true,
			codec: numberCodec
		},
		{
			name: "upgradeNoVotes",
			wireKey: "upgrade-no-votes",
			optional: true,
			codec: numberCodec
		},
		{
			name: "upgradeNextProtocolVoteBefore",
			wireKey: "upgrade-next-protocol-vote-before",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "upgradeVoteRounds",
			wireKey: "upgrade-vote-rounds",
			optional: true,
			codec: numberCodec
		}
	]
};
//#endregion
export { NodeStatusResponseMeta };
