// @algorandfoundation/algokit-utils/packages/algod_client/src/models/pending-transaction-response.mjs
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec, bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ArrayModelCodec } from "../../../common/src/codecs/models/array-model.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { SignedTransactionMeta } from "../../../transact/src/transactions/signed-transaction-meta.mjs";
import { StateDeltaMeta } from "./state-delta.mjs";
import { AccountStateDeltaMeta } from "./account-state-delta.mjs";
//#region packages/algod_client/src/models/pending-transaction-response.ts
const PendingTransactionResponseMeta = {
	name: "PendingTransactionResponse",
	kind: "object",
	fields: [
		{
			name: "assetId",
			wireKey: "asset-index",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "appId",
			wireKey: "application-index",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "closeRewards",
			wireKey: "close-rewards",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "closingAmount",
			wireKey: "closing-amount",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "assetClosingAmount",
			wireKey: "asset-closing-amount",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "confirmedRound",
			wireKey: "confirmed-round",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "poolError",
			wireKey: "pool-error",
			optional: false,
			codec: stringCodec
		},
		{
			name: "receiverRewards",
			wireKey: "receiver-rewards",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "senderRewards",
			wireKey: "sender-rewards",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "localStateDelta",
			wireKey: "local-state-delta",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(AccountStateDeltaMeta))
		},
		{
			name: "globalStateDelta",
			wireKey: "global-state-delta",
			optional: true,
			codec: new ArrayModelCodec(StateDeltaMeta)
		},
		{
			name: "logs",
			wireKey: "logs",
			optional: true,
			codec: bytesArrayCodec
		},
		{
			name: "innerTxns",
			wireKey: "inner-txns",
			optional: true,
			codec: new ArrayCodec(new ObjectModelCodec(() => PendingTransactionResponseMeta))
		},
		{
			name: "txn",
			wireKey: "txn",
			optional: false,
			codec: new ObjectModelCodec(SignedTransactionMeta)
		}
	]
};
//#endregion
export { PendingTransactionResponseMeta };
