// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulation-transaction-exec-trace.js
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_simulation_opcode_trace_unit = require("./simulation-opcode-trace-unit.js");
//#region packages/algod_client/src/models/simulation-transaction-exec-trace.ts
const SimulationTransactionExecTraceMeta = {
	name: "SimulationTransactionExecTrace",
	kind: "object",
	fields: [
		{
			name: "approvalProgramTrace",
			wireKey: "approval-program-trace",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_simulation_opcode_trace_unit.SimulationOpcodeTraceUnitMeta))
		},
		{
			name: "approvalProgramHash",
			wireKey: "approval-program-hash",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "clearStateProgramTrace",
			wireKey: "clear-state-program-trace",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_simulation_opcode_trace_unit.SimulationOpcodeTraceUnitMeta))
		},
		{
			name: "clearStateProgramHash",
			wireKey: "clear-state-program-hash",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "clearStateRollback",
			wireKey: "clear-state-rollback",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "clearStateRollbackError",
			wireKey: "clear-state-rollback-error",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "logicSigTrace",
			wireKey: "logic-sig-trace",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_simulation_opcode_trace_unit.SimulationOpcodeTraceUnitMeta))
		},
		{
			name: "logicSigHash",
			wireKey: "logic-sig-hash",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "innerTrace",
			wireKey: "inner-trace",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(() => SimulationTransactionExecTraceMeta))
		}
	]
};
//#endregion
exports.SimulationTransactionExecTraceMeta = SimulationTransactionExecTraceMeta;
