// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-transaction-result.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { PendingTransactionResponseMeta } from "./pending-transaction-response.mjs";
import { SimulateUnnamedResourcesAccessedMeta } from "./simulate-unnamed-resources-accessed.mjs";
import { SimulationTransactionExecTraceMeta } from "./simulation-transaction-exec-trace.mjs";
//#region packages/algod_client/src/models/simulate-transaction-result.ts
const SimulateTransactionResultMeta = {
	name: "SimulateTransactionResult",
	kind: "object",
	fields: [
		{
			name: "txnResult",
			wireKey: "txn-result",
			optional: false,
			codec: new ObjectModelCodec(PendingTransactionResponseMeta)
		},
		{
			name: "appBudgetConsumed",
			wireKey: "app-budget-consumed",
			optional: true,
			codec: numberCodec
		},
		{
			name: "logicSigBudgetConsumed",
			wireKey: "logic-sig-budget-consumed",
			optional: true,
			codec: numberCodec
		},
		{
			name: "execTrace",
			wireKey: "exec-trace",
			optional: true,
			codec: new ObjectModelCodec(SimulationTransactionExecTraceMeta)
		},
		{
			name: "unnamedResourcesAccessed",
			wireKey: "unnamed-resources-accessed",
			optional: true,
			codec: new ObjectModelCodec(SimulateUnnamedResourcesAccessedMeta)
		},
		{
			name: "fixedSigner",
			wireKey: "fixed-signer",
			optional: true,
			codec: addressCodec
		}
	]
};
//#endregion
export { SimulateTransactionResultMeta };
