// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-unnamed-resources-accessed.js
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_reference_types_meta = require("../../../transact/src/transactions/reference-types-meta.js");
//#region packages/algod_client/src/models/simulate-unnamed-resources-accessed.ts
const SimulateUnnamedResourcesAccessedMeta = {
	name: "SimulateUnnamedResourcesAccessed",
	kind: "object",
	fields: [
		{
			name: "accounts",
			wireKey: "accounts",
			optional: true,
			codec: require_array.addressArrayCodec
		},
		{
			name: "assets",
			wireKey: "assets",
			optional: true,
			codec: require_array.bigIntArrayCodec
		},
		{
			name: "apps",
			wireKey: "apps",
			optional: true,
			codec: require_array.bigIntArrayCodec
		},
		{
			name: "boxes",
			wireKey: "boxes",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_reference_types_meta.BoxReferenceMeta))
		},
		{
			name: "extraBoxRefs",
			wireKey: "extra-box-refs",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "assetHoldings",
			wireKey: "asset-holdings",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_reference_types_meta.HoldingReferenceMeta))
		},
		{
			name: "appLocals",
			wireKey: "app-locals",
			optional: true,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_reference_types_meta.LocalsReferenceMeta))
		}
	]
};
//#endregion
exports.SimulateUnnamedResourcesAccessedMeta = SimulateUnnamedResourcesAccessedMeta;
