// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulation-eval-overrides.mjs
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
//#region packages/algod_client/src/models/simulation-eval-overrides.ts
const SimulationEvalOverridesMeta = {
	name: "SimulationEvalOverrides",
	kind: "object",
	fields: [
		{
			name: "allowEmptySignatures",
			wireKey: "allow-empty-signatures",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "allowUnnamedResources",
			wireKey: "allow-unnamed-resources",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "maxLogCalls",
			wireKey: "max-log-calls",
			optional: true,
			codec: numberCodec
		},
		{
			name: "maxLogSize",
			wireKey: "max-log-size",
			optional: true,
			codec: numberCodec
		},
		{
			name: "extraOpcodeBudget",
			wireKey: "extra-opcode-budget",
			optional: true,
			codec: numberCodec
		},
		{
			name: "fixSigners",
			wireKey: "fix-signers",
			optional: true,
			codec: booleanCodec
		}
	]
};
//#endregion
export { SimulationEvalOverridesMeta };
