// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulation-eval-overrides.js
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
//#region packages/algod_client/src/models/simulation-eval-overrides.ts
const SimulationEvalOverridesMeta = {
	name: "SimulationEvalOverrides",
	kind: "object",
	fields: [
		{
			name: "allowEmptySignatures",
			wireKey: "allow-empty-signatures",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "allowUnnamedResources",
			wireKey: "allow-unnamed-resources",
			optional: true,
			codec: require_boolean.booleanCodec
		},
		{
			name: "maxLogCalls",
			wireKey: "max-log-calls",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "maxLogSize",
			wireKey: "max-log-size",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "extraOpcodeBudget",
			wireKey: "extra-opcode-budget",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "fixSigners",
			wireKey: "fix-signers",
			optional: true,
			codec: require_boolean.booleanCodec
		}
	]
};
//#endregion
exports.SimulationEvalOverridesMeta = SimulationEvalOverridesMeta;
