// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-transaction-result.js
const require_address = require("../../../common/src/codecs/primitives/address.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_pending_transaction_response = require("./pending-transaction-response.js");
const require_simulate_unnamed_resources_accessed = require("./simulate-unnamed-resources-accessed.js");
const require_simulation_transaction_exec_trace = require("./simulation-transaction-exec-trace.js");
//#region packages/algod_client/src/models/simulate-transaction-result.ts
const SimulateTransactionResultMeta = {
	name: "SimulateTransactionResult",
	kind: "object",
	fields: [
		{
			name: "txnResult",
			wireKey: "txn-result",
			optional: false,
			codec: new require_object_model.ObjectModelCodec(require_pending_transaction_response.PendingTransactionResponseMeta)
		},
		{
			name: "appBudgetConsumed",
			wireKey: "app-budget-consumed",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "logicSigBudgetConsumed",
			wireKey: "logic-sig-budget-consumed",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "execTrace",
			wireKey: "exec-trace",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulation_transaction_exec_trace.SimulationTransactionExecTraceMeta)
		},
		{
			name: "unnamedResourcesAccessed",
			wireKey: "unnamed-resources-accessed",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulate_unnamed_resources_accessed.SimulateUnnamedResourcesAccessedMeta)
		},
		{
			name: "fixedSigner",
			wireKey: "fixed-signer",
			optional: true,
			codec: require_address.addressCodec
		}
	]
};
//#endregion
exports.SimulateTransactionResultMeta = SimulateTransactionResultMeta;
