// @algorandfoundation/algokit-utils/packages/algod_client/src/models/simulate-transaction-group-result.js
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_simulate_unnamed_resources_accessed = require("./simulate-unnamed-resources-accessed.js");
const require_simulate_transaction_result = require("./simulate-transaction-result.js");
//#region packages/algod_client/src/models/simulate-transaction-group-result.ts
const SimulateTransactionGroupResultMeta = {
	name: "SimulateTransactionGroupResult",
	kind: "object",
	fields: [
		{
			name: "txnResults",
			wireKey: "txn-results",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec(require_simulate_transaction_result.SimulateTransactionResultMeta))
		},
		{
			name: "failureMessage",
			wireKey: "failure-message",
			optional: true,
			codec: require_string.stringCodec
		},
		{
			name: "failedAt",
			wireKey: "failed-at",
			optional: true,
			codec: require_array.numberArrayCodec
		},
		{
			name: "appBudgetAdded",
			wireKey: "app-budget-added",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "appBudgetConsumed",
			wireKey: "app-budget-consumed",
			optional: true,
			codec: require_number.numberCodec
		},
		{
			name: "unnamedResourcesAccessed",
			wireKey: "unnamed-resources-accessed",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_simulate_unnamed_resources_accessed.SimulateUnnamedResourcesAccessedMeta)
		}
	]
};
//#endregion
exports.SimulateTransactionGroupResultMeta = SimulateTransactionGroupResultMeta;
