// @algorandfoundation/algokit-utils/packages/algod_client/src/models/state-proof-message.js
const require_bigint = require("../../../common/src/codecs/primitives/bigint.js");
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
//#region packages/algod_client/src/models/state-proof-message.ts
const StateProofMessageMeta = {
	name: "StateProofMessage",
	kind: "object",
	fields: [
		{
			name: "blockHeadersCommitment",
			wireKey: "BlockHeadersCommitment",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "votersCommitment",
			wireKey: "VotersCommitment",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "lnProvenWeight",
			wireKey: "LnProvenWeight",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "firstAttestedRound",
			wireKey: "FirstAttestedRound",
			optional: false,
			codec: require_bigint.bigIntCodec
		},
		{
			name: "lastAttestedRound",
			wireKey: "LastAttestedRound",
			optional: false,
			codec: require_bigint.bigIntCodec
		}
	]
};
//#endregion
exports.StateProofMessageMeta = StateProofMessageMeta;
