// @algorandfoundation/algokit-utils/packages/algod_client/src/models/transaction-proof.js
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
//#region packages/algod_client/src/models/transaction-proof.ts
const TransactionProofMeta = {
	name: "TransactionProof",
	kind: "object",
	fields: [
		{
			name: "proof",
			wireKey: "proof",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "stibhash",
			wireKey: "stibhash",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "treedepth",
			wireKey: "treedepth",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "idx",
			wireKey: "idx",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "hashtype",
			wireKey: "hashtype",
			optional: false,
			codec: require_string.stringCodec
		}
	]
};
//#endregion
exports.TransactionProofMeta = TransactionProofMeta;
