// @algorandfoundation/algokit-utils/packages/common/src/codecs/primitives/address.js
require("../../../../../_virtual/_rolldown/runtime.js");
require("../../constants.js");
const require_address = require("../../address.js");
const require_codec = require("../codec.js");
let buffer = require("buffer");
//#region packages/common/src/codecs/primitives/address.ts
var AddressCodec = class extends require_codec.Codec {
	defaultValue() {
		return require_address.Address.zeroAddress();
	}
	toEncoded(value, format) {
		if (format === "json") return value.toString();
		return value.publicKey;
	}
	fromEncoded(value, _format) {
		if (typeof value === "string") return require_address.Address.fromString(value);
		if (value instanceof Uint8Array) {
			if (value.length === 32) return new require_address.Address(value);
			else if (value.length === 58) return require_address.Address.fromString(buffer.Buffer.from(value).toString("utf-8"));
		}
		throw new Error(`AddressCodec cannot decode address from ${typeof value}`);
	}
	isDefaultValue(value) {
		return value.equals(this.defaultValue());
	}
};
const addressCodec = new AddressCodec();
//#endregion
exports.addressCodec = addressCodec;
