// @algorandfoundation/algokit-utils/packages/common/src/codecs/primitives/fixed-bytes.js
require("../../../../../_virtual/_rolldown/runtime.js");
const require_codec = require("../codec.js");
let buffer = require("buffer");
//#region packages/common/src/codecs/primitives/fixed-bytes.ts
var FixedBytesCodec = class extends require_codec.Codec {
	constructor(length) {
		super();
		this.length = length;
	}
	defaultValue() {
		return new Uint8Array(this.length);
	}
	toEncoded(value, format) {
		if (format === "json") return buffer.Buffer.from(value).toString("base64");
		return value;
	}
	fromEncoded(value, _format) {
		if (value instanceof Uint8Array) return value;
		if (typeof value === "string") return new Uint8Array(buffer.Buffer.from(value, "base64"));
		throw new Error(`Cannot decode fixed ${this.length} bytes from ${typeof value}`);
	}
	isDefaultValue(value) {
		if (value.byteLength !== this.length) return false;
		return value.every((byte) => byte === 0);
	}
};
const fixedBytes32Codec = new FixedBytesCodec(32);
const fixedBytes64Codec = new FixedBytesCodec(64);
const fixedBytes1793Codec = new FixedBytesCodec(1793);
//#endregion
exports.fixedBytes1793Codec = fixedBytes1793Codec;
exports.fixedBytes32Codec = fixedBytes32Codec;
exports.fixedBytes64Codec = fixedBytes64Codec;
