// @algorandfoundation/algokit-utils/packages/common/src/codecs/primitives/unknown.js
require("../../constants.js");
const require_codec = require("../codec.js");
const require_bigint = require("./bigint.js");
const require_boolean = require("./boolean.js");
const require_number = require("./number.js");
const require_string = require("./string.js");
const require_array = require("../composite/array.js");
const require_record = require("../composite/record.js");
//#region packages/common/src/codecs/primitives/unknown.ts
/**
* Unknown codec - passthrough for unknown/any types
* Converts Maps with Uint8Array keys to objects with string keys recursively
*/
var UnknownCodec = class extends require_codec.Codec {
	textDecoder = new TextDecoder("utf-8", { fatal: true });
	recordCodec;
	arrayCodec;
	constructor() {
		super();
		this.recordCodec = new require_record.RecordCodec(this);
		this.arrayCodec = new require_array.ArrayCodec(this);
	}
	defaultValue() {}
	toEncoded(_value, _format) {
		throw new Error("UnknownCodec does not support encoding");
	}
	fromEncoded(value, format) {
		return this.processValue(value, format);
	}
	maybeDecodeAsUtf8(bytes) {
		try {
			if ((bytes.length === 58 || bytes.length === 32 || bytes.length === 64) && bytes.every((byte) => byte === 0)) return bytes;
			return this.textDecoder.decode(bytes);
		} catch {
			return bytes;
		}
	}
	/**
	* Recursively process unknown values, using various codecs where appropriate
	*/
	processValue(value, format) {
		if (value === null || value === void 0) return value;
		if (value instanceof Uint8Array) return this.maybeDecodeAsUtf8(value);
		if (typeof value === "bigint") return require_bigint.bigIntCodec.decode(value, format);
		if (typeof value === "number") return require_number.numberCodec.decode(value, format);
		if (typeof value === "string") return require_string.stringCodec.decode(value, format);
		if (typeof value === "boolean") return require_boolean.booleanCodec.decode(value, format);
		if (Array.isArray(value)) return this.arrayCodec.decode(value, format);
		if (value instanceof Map || typeof value === "object") return this.recordCodec.decode(value, format);
		return value;
	}
};
const unknownCodec = new UnknownCodec();
//#endregion
exports.unknownCodec = unknownCodec;
