// @algorandfoundation/algokit-utils/testing/account.js
const require_address = require('../packages/common/src/address.js');
const require_config = require('../config.js');
const require_algorand_client = require('../algorand-client.js');

//#region src/testing/account.ts
async function getTestAccount({ suppressLog, initialFunds, accountGetter }, algodOrAlgorandClient, kmd) {
	const algorand = algodOrAlgorandClient instanceof require_algorand_client.AlgorandClient ? algodOrAlgorandClient : require_algorand_client.AlgorandClient.fromClients({
		algod: algodOrAlgorandClient,
		kmd
	});
	const account = accountGetter ? await accountGetter(algorand) : algorand.account.random();
	require_config.Config.getLogger(suppressLog).info(`New test account created with address '${account.addr}'.`);
	const dispenser = await algorand.account.dispenserFromEnvironment();
	await algorand.send.payment({
		sender: dispenser,
		receiver: account.addr,
		amount: initialFunds,
		note: "Funding test account",
		suppressLog
	});
	const accountInfo = await algorand.account.getInformation(account.addr);
	require_config.Config.getLogger(suppressLog).info("Test account funded; account balance: %d µALGO", accountInfo.balance.microAlgo);
	algorand.setSignerFromAccount(account);
	const address = require_address.Address.fromString(account.addr.toString());
	for (const key of Object.keys(account)) if (!(key in address)) address[key] = account[key];
	return address;
}

//#endregion
exports.getTestAccount = getTestAccount;
