// @algorandfoundation/algokit-utils/packages/common/src/codecs/composite/record.mjs
import { Codec } from "../codec.mjs";
import { Buffer } from "buffer";
//#region packages/common/src/codecs/composite/record.ts
/**
* Record codec - for string-keyed objects with homogeneous values
*/
var RecordCodec = class extends Codec {
	constructor(valueCodec) {
		super();
		this.valueCodec = valueCodec;
	}
	defaultValue() {
		return {};
	}
	toEncoded(value, format) {
		const result = {};
		for (const [key, val] of Object.entries(value)) result[key] = this.valueCodec.encode(val, format);
		return result;
	}
	fromEncoded(value, format) {
		const result = {};
		if (value instanceof Map) for (const [_key, val] of value.entries()) {
			const keyType = typeof _key;
			if (keyType === "number" || keyType === "bigint") throw new Error(`RecordCodec received a non-string key of type ${keyType}`);
			const key = _key instanceof Uint8Array ? Buffer.from(_key).toString("utf-8") : String(_key);
			result[key] = this.valueCodec.decode(val, format);
		}
		else for (const [key, val] of Object.entries(value)) result[key] = this.valueCodec.decode(val, format);
		return result;
	}
	isDefaultValue(value) {
		return Object.keys(value).length === 0;
	}
};
//#endregion
export { RecordCodec };
