// @algorandfoundation/algokit-utils/packages/abi/src/abi-method.mjs
import { ABIAddressType, ABIStructType, ABIType, ABIUintType, parseTupleContent } from "./abi-type.mjs";
import { sha512_256 } from "@noble/hashes/sha2.js";

//#region packages/abi/src/abi-method.ts
let ABITransactionType = /* @__PURE__ */ function(ABITransactionType) {
	ABITransactionType["Txn"] = "txn";
	ABITransactionType["Payment"] = "pay";
	ABITransactionType["KeyRegistration"] = "keyreg";
	ABITransactionType["AssetConfig"] = "acfg";
	ABITransactionType["AssetTransfer"] = "axfer";
	ABITransactionType["AssetFreeze"] = "afrz";
	ABITransactionType["AppCall"] = "appl";
	return ABITransactionType;
}({});
let ABIReferenceType = /* @__PURE__ */ function(ABIReferenceType) {
	ABIReferenceType["Account"] = "account";
	ABIReferenceType["Application"] = "application";
	ABIReferenceType["Asset"] = "asset";
	return ABIReferenceType;
}({});
var ABIMethod = class ABIMethod {
	name;
	description;
	args;
	returns;
	events;
	readonly;
	constructor(params) {
		this.name = params.name;
		this.description = params.description;
		this.args = params.args;
		this.returns = params.returns;
		this.events = params.events;
		this.readonly = params.readonly;
	}
	/**
	* Returns the signature of this ABI method.
	* @returns The signature, e.g. `my_method(unit64,string)bytes`
	*/
	getSignature() {
		const args = this.args.map((arg) => {
			if (argTypeIsTransaction(arg.type) || argTypeIsReference(arg.type)) return arg.type;
			return arg.type.name;
		}).join(",");
		return `${this.name}(${args})${this.returns.type === "void" ? "void" : this.returns.type.name}`;
	}
	/**
	* Returns the method selector of this ABI method.
	* @returns The 4-byte method selector
	*/
	getSelector() {
		const hash = sha512_256(new TextEncoder().encode(this.getSignature()));
		return new Uint8Array(hash.slice(0, 4));
	}
	/**
	* Returns the ABI method object for a given method signature.
	* @param signature The method signature
	* e.g. `my_method(unit64,string)bytes`
	* @returns The `ABIMethod`
	*/
	static fromSignature(signature) {
		const argsStart = signature.indexOf("(");
		if (argsStart === -1) throw new Error(`Invalid method signature: ${signature}`);
		let argsEnd = -1;
		let depth = 0;
		for (let i = argsStart; i < signature.length; i++) {
			const char = signature[i];
			if (char === "(") depth += 1;
			else if (char === ")") {
				if (depth === 0) break;
				depth -= 1;
				if (depth === 0) {
					argsEnd = i;
					break;
				}
			}
		}
		if (argsEnd === -1) throw new Error(`Invalid method signature: ${signature}`);
		const name = signature.slice(0, argsStart);
		const args = parseTupleContent(signature.slice(argsStart + 1, argsEnd)).map((n) => {
			if (argTypeIsTransaction(n) || argTypeIsReference(n)) return { type: n };
			return { type: ABIType.from(n) };
		});
		const returnType = signature.slice(argsEnd + 1);
		return new ABIMethod({
			name,
			args,
			returns: { type: returnType === "void" ? "void" : ABIType.from(returnType) }
		});
	}
};
/**
* Returns the ABI method object for a given method name or signature and ARC-56 app spec.
* @param methodNameOrSignature The method name or method signature to call if an ABI call is being emitted.
* e.g. `my_method` or `my_method(unit64,string)bytes`
* @param appSpec The app spec for the app
* @returns The `ABIMethod`
*/
function getABIMethod(methodNameOrSignature, appSpec) {
	if (!methodNameOrSignature.includes("(")) {
		const methods = appSpec.methods.filter((m) => m.name === methodNameOrSignature);
		if (methods.length === 0) throw new Error(`Unable to find method ${methodNameOrSignature} in ${appSpec.name} app.`);
		if (methods.length > 1) throw new Error(`Received a call to method ${methodNameOrSignature} in contract ${appSpec.name}, but this resolved to multiple methods; please pass in an ABI signature instead: ${appSpec.methods.map((m) => getArc56MethodSignature(m)).join(", ")}`);
		return arc56MethodToABIMethod(methods[0], appSpec);
	} else {
		const method = appSpec.methods.find((m) => getArc56MethodSignature(m) === methodNameOrSignature);
		if (!method) throw new Error(`Unable to find method ${methodNameOrSignature} in ${appSpec.name} app.`);
		return arc56MethodToABIMethod(method, appSpec);
	}
}
function arc56MethodToABIMethod(method, appSpec) {
	if (typeof method.name !== "string" || typeof method.returns !== "object" || !Array.isArray(method.args)) throw new Error("Invalid ABIMethod parameters");
	const args = method.args.map(({ type, name, desc, struct, defaultValue }) => {
		const convertedDefaultValue = defaultValue ? {
			data: defaultValue.data,
			source: defaultValue.source,
			type: defaultValue.type ? isAVMType(defaultValue.type) ? defaultValue.type : ABIType.from(defaultValue.type) : void 0
		} : void 0;
		if (argTypeIsTransaction(type) || argTypeIsReference(type)) return {
			type,
			name,
			description: desc,
			defaultValue: convertedDefaultValue
		};
		if (struct) return {
			type: ABIStructType.fromStruct(struct, appSpec.structs),
			name,
			description: desc,
			defaultValue: convertedDefaultValue
		};
		return {
			type: ABIType.from(type),
			name,
			description: desc,
			defaultValue: convertedDefaultValue
		};
	});
	const returns = {
		type: method.returns.type === "void" ? "void" : method.returns.struct ? ABIStructType.fromStruct(method.returns.struct, appSpec.structs) : ABIType.from(method.returns.type),
		description: method.returns.desc
	};
	return new ABIMethod({
		name: method.name,
		description: method.desc,
		args,
		returns,
		events: method.events,
		readonly: method.readonly
	});
}
function argTypeIsTransaction(type) {
	return typeof type === "string" && (type === ABITransactionType.Txn || type === ABITransactionType.Payment || type === ABITransactionType.KeyRegistration || type === ABITransactionType.AssetConfig || type === ABITransactionType.AssetTransfer || type === ABITransactionType.AssetFreeze || type === ABITransactionType.AppCall);
}
function argTypeIsReference(type) {
	return typeof type === "string" && (type === ABIReferenceType.Account || type === ABIReferenceType.Application || type === ABIReferenceType.Asset);
}
function argTypeIsAbiType(type) {
	return !argTypeIsTransaction(type) && !argTypeIsReference(type);
}
function getArc56MethodSignature(method) {
	const args = method.args.map((arg) => arg.type).join(",");
	return `${method.name}(${args})${method.returns.type}`;
}
function decodeAVMValue(avmType, bytes) {
	switch (avmType) {
		case "AVMString": return Buffer.from(bytes).toString("utf-8");
		case "AVMBytes": return bytes;
		case "AVMUint64": return ABIType.from("uint64").decode(bytes);
	}
}
function encodeAVMValue(avmType, value) {
	switch (avmType) {
		case "AVMString": return ABIType.from("string").encode(value);
		case "AVMBytes":
			if (typeof value === "string") return Buffer.from(value, "utf-8");
			if (typeof value !== "object" || !(value instanceof Uint8Array)) throw new Error(`Expected bytes value for AVMBytes, but got ${value}`);
			return value;
		case "AVMUint64": return ABIType.from("uint64").encode(value);
	}
}
function isAVMType(type) {
	return typeof type === "string" && (type === "AVMString" || type === "AVMBytes" || type === "AVMUint64");
}
function getABIDecodedValue(type, bytes) {
	if (type === ABIReferenceType.Asset || type === ABIReferenceType.Application) return new ABIUintType(64).decode(bytes);
	else if (type === ABIReferenceType.Account) return new ABIAddressType().decode(bytes);
	else if (isAVMType(type)) return decodeAVMValue(type, bytes);
	return type.decode(bytes);
}
function getABIEncodedValue(type, value) {
	return isAVMType(type) ? encodeAVMValue(type, value) : type.encode(value);
}

//#endregion
export { ABIMethod, ABIReferenceType, ABITransactionType, arc56MethodToABIMethod, argTypeIsAbiType, argTypeIsReference, argTypeIsTransaction, decodeAVMValue, getABIDecodedValue, getABIEncodedValue, getABIMethod, isAVMType };
