// @algorandfoundation/algokit-utils/packages/abi/src/arc56-contract.mjs
import { ABIStructType, ABIType } from "./abi-type.mjs";

//#region packages/abi/src/arc56-contract.ts
/** ARC-56 spec */
/** Storage Helper Functions */
/**
* Resolves a storage type string (ABI type, AVM type, or struct name) into an ABIType or AVMType
* @param typeStr The type string to resolve
* @param structs The structs defined in the contract
* @returns The parsed ABIType or AVMType string
*/
function resolveStorageType(typeStr, structs) {
	if (typeStr === "AVMBytes" || typeStr === "AVMString" || typeStr === "AVMUint64") return typeStr;
	if (structs[typeStr]) return ABIStructType.fromStruct(typeStr, structs);
	try {
		return ABIType.from(typeStr);
	} catch (error) {
		throw new Error(`Failed to parse storage type '${typeStr}': ${error}`);
	}
}
/**
* Converts a StorageKey to an ABIStorageKey with parsed types
* @param storageKey The storage key to convert
* @param structs The structs defined in the contract
* @returns The converted ABIStorageKey
*/
function convertStorageKey(storageKey, structs) {
	const keyType = resolveStorageType(storageKey.keyType, structs);
	const valueType = resolveStorageType(storageKey.valueType, structs);
	return {
		key: storageKey.key,
		keyType,
		valueType,
		desc: storageKey.desc
	};
}
/**
* Converts a StorageMap to an ABIStorageMap with parsed types
* @param storageMap The storage map to convert
* @param structs The structs defined in the contract
* @returns The converted ABIStorageMap
*/
function convertStorageMap(storageMap, structs) {
	return {
		keyType: resolveStorageType(storageMap.keyType, structs),
		valueType: resolveStorageType(storageMap.valueType, structs),
		desc: storageMap.desc,
		prefix: storageMap.prefix
	};
}
/**
* Get a specific global storage key with parsed ABI types
* @param contract The ARC-56 contract
* @param keyName The name of the storage key
* @returns The ABIStorageKey with parsed types
* @throws Error if the key is not found
*/
function getGlobalABIStorageKey(contract, keyName) {
	const storageKey = contract.state.keys.global[keyName];
	if (!storageKey) throw new Error(`Global storage key '${keyName}' not found in contract '${contract.name}'`);
	return convertStorageKey(storageKey, contract.structs);
}
/**
* Get a specific local storage key with parsed ABI types
* @param contract The ARC-56 contract
* @param keyName The name of the storage key
* @returns The ABIStorageKey with parsed types
* @throws Error if the key is not found
*/
function getLocalABIStorageKey(contract, keyName) {
	const storageKey = contract.state.keys.local[keyName];
	if (!storageKey) throw new Error(`Local storage key '${keyName}' not found in contract '${contract.name}'`);
	return convertStorageKey(storageKey, contract.structs);
}
/**
* Get a specific box storage key with parsed ABI types
* @param contract The ARC-56 contract
* @param keyName The name of the storage key
* @returns The ABIStorageKey with parsed types
* @throws Error if the key is not found
*/
function getBoxABIStorageKey(contract, keyName) {
	const storageKey = contract.state.keys.box[keyName];
	if (!storageKey) throw new Error(`Box storage key '${keyName}' not found in contract '${contract.name}'`);
	return convertStorageKey(storageKey, contract.structs);
}
/**
* Get all global storage keys with parsed ABI types
* @param contract The ARC-56 contract
* @returns A record of storage key names to ABIStorageKey objects
*/
function getGlobalABIStorageKeys(contract) {
	const result = {};
	for (const [name, storageKey] of Object.entries(contract.state.keys.global)) result[name] = convertStorageKey(storageKey, contract.structs);
	return result;
}
/**
* Get all local storage keys with parsed ABI types
* @param contract The ARC-56 contract
* @returns A record of storage key names to ABIStorageKey objects
*/
function getLocalABIStorageKeys(contract) {
	const result = {};
	for (const [name, storageKey] of Object.entries(contract.state.keys.local)) result[name] = convertStorageKey(storageKey, contract.structs);
	return result;
}
/**
* Get all box storage keys with parsed ABI types
* @param contract The ARC-56 contract
* @returns A record of storage key names to ABIStorageKey objects
*/
function getBoxABIStorageKeys(contract) {
	const result = {};
	for (const [name, storageKey] of Object.entries(contract.state.keys.box)) result[name] = convertStorageKey(storageKey, contract.structs);
	return result;
}
/**
* Get a specific global storage map with parsed ABI types
* @param contract The ARC-56 contract
* @param mapName The name of the storage map
* @returns The ABIStorageMap with parsed types
* @throws Error if the map is not found
*/
function getGlobalABIStorageMap(contract, mapName) {
	const storageMap = contract.state.maps.global[mapName];
	if (!storageMap) throw new Error(`Global storage map '${mapName}' not found in contract '${contract.name}'`);
	return convertStorageMap(storageMap, contract.structs);
}
/**
* Get a specific local storage map with parsed ABI types
* @param contract The ARC-56 contract
* @param mapName The name of the storage map
* @returns The ABIStorageMap with parsed types
* @throws Error if the map is not found
*/
function getLocalABIStorageMap(contract, mapName) {
	const storageMap = contract.state.maps.local[mapName];
	if (!storageMap) throw new Error(`Local storage map '${mapName}' not found in contract '${contract.name}'`);
	return convertStorageMap(storageMap, contract.structs);
}
/**
* Get a specific box storage map with parsed ABI types
* @param contract The ARC-56 contract
* @param mapName The name of the storage map
* @returns The ABIStorageMap with parsed types
* @throws Error if the map is not found
*/
function getBoxABIStorageMap(contract, mapName) {
	const storageMap = contract.state.maps.box[mapName];
	if (!storageMap) throw new Error(`Box storage map '${mapName}' not found in contract '${contract.name}'`);
	return convertStorageMap(storageMap, contract.structs);
}
/**
* Get all global storage maps with parsed ABI types
* @param contract The ARC-56 contract
* @returns A record of storage map names to ABIStorageMap objects
*/
function getGlobalABIStorageMaps(contract) {
	const result = {};
	for (const [name, storageMap] of Object.entries(contract.state.maps.global)) result[name] = convertStorageMap(storageMap, contract.structs);
	return result;
}
/**
* Get all local storage maps with parsed ABI types
* @param contract The ARC-56 contract
* @returns A record of storage map names to ABIStorageMap objects
*/
function getLocalABIStorageMaps(contract) {
	const result = {};
	for (const [name, storageMap] of Object.entries(contract.state.maps.local)) result[name] = convertStorageMap(storageMap, contract.structs);
	return result;
}
/**
* Get all box storage maps with parsed ABI types
* @param contract The ARC-56 contract
* @returns A record of storage map names to ABIStorageMap objects
*/
function getBoxABIStorageMaps(contract) {
	const result = {};
	for (const [name, storageMap] of Object.entries(contract.state.maps.box)) result[name] = convertStorageMap(storageMap, contract.structs);
	return result;
}

//#endregion
export { getBoxABIStorageKey, getBoxABIStorageKeys, getBoxABIStorageMap, getBoxABIStorageMaps, getGlobalABIStorageKey, getGlobalABIStorageKeys, getGlobalABIStorageMap, getGlobalABIStorageMaps, getLocalABIStorageKey, getLocalABIStorageKeys, getLocalABIStorageMap, getLocalABIStorageMaps };
