// @algorandfoundation/xhd-wallet-api/dist/x.hd.wallet.api.crypto.js
import { crypto_core_ed25519_scalar_add, crypto_core_ed25519_scalar_mul, crypto_core_ed25519_scalar_reduce, crypto_hash_sha512, crypto_scalarmult_ed25519_base_noclamp, crypto_sign_verify_detached, crypto_sign_ed25519_pk_to_curve25519, crypto_scalarmult, crypto_generichash, } from './sumo.facade.js';
import * as msgpack from "algo-msgpack-with-bigint";
import Ajv from "ajv";
//@ts-expect-error, we handle this with ts-alias
import { deriveChildNodePrivate } from './bip32-ed25519.js';
/**
 *
 */
export var KeyContext;
(function (KeyContext) {
    KeyContext[KeyContext["Address"] = 0] = "Address";
    KeyContext[KeyContext["Identity"] = 1] = "Identity";
})(KeyContext || (KeyContext = {}));
export var BIP32DerivationType;
(function (BIP32DerivationType) {
    // standard Ed25519 bip32 derivations based of: https://acrobat.adobe.com/id/urn:aaid:sc:EU:04fe29b0-ea1a-478b-a886-9bb558a5242a
    // Defines 32 bits to be zeroed from each derived zL
    BIP32DerivationType[BIP32DerivationType["Khovratovich"] = 32] = "Khovratovich";
    // Derivations based on Peikert's ammendments to the original BIP32-Ed25519
    // Picking only 9 bits to be zeroed from each derived zL
    BIP32DerivationType[BIP32DerivationType["Peikert"] = 9] = "Peikert";
})(BIP32DerivationType || (BIP32DerivationType = {}));
export var Encoding;
(function (Encoding) {
    Encoding["MSGPACK"] = "msgpack";
    Encoding["BASE64"] = "base64";
    Encoding["NONE"] = "none";
})(Encoding || (Encoding = {}));
export const harden = (num) => 0x80_00_00_00 + num;
function GetBIP44PathFromContext(context, account, key_index) {
    switch (context) {
        case KeyContext.Address:
            return [harden(44), harden(283), harden(account), 0, key_index];
        case KeyContext.Identity:
            return [harden(44), harden(0), harden(account), 0, key_index];
        default:
            throw Error("Invalid context");
    }
}
export const ERROR_BAD_DATA = Error("Invalid Data");
export const ERROR_TAGS_FOUND = Error("Transactions tags found");
export class XHDWalletAPI {
    constructor() { }
    /**
     * Derives a child key from the root key based on BIP44 path
     *
     * @param rootKey - root key in extended format (kL, kR, c). It should be 96 bytes long
     * @param bip44Path - BIP44 path (m / purpose' / coin_type' / account' / change / address_index). The ' indicates that the value is hardened
     * @param isPrivate  - if true, return the private key, otherwise return the public key
     * @returns - The extended private key (kL, kR, chainCode) or the extended public key (pub, chainCode)
     */
    async deriveKey(rootKey, bip44Path, isPrivate = true, derivationType) {
        // Pick `g`, which is amount of bits zeroed from each derived node
        const g = derivationType === BIP32DerivationType.Peikert ? 9 : 32;
        for (let i = 0; i < bip44Path.length; i++) {
            rootKey = await deriveChildNodePrivate(rootKey, bip44Path[i], g);
        }
        if (isPrivate)
            return rootKey;
        // extended public key
        // [public] [nodeCC]
        return new Uint8Array(Buffer.concat([crypto_scalarmult_ed25519_base_noclamp(rootKey.subarray(0, 32)), rootKey.subarray(64, 96)]));
    }
    /**
     *
     *
     * @param context - context of the key (i.e Address, Identity)
     * @param account - account number. This value will be hardened as part of BIP44
     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
     * @returns - public key 32 bytes
     */
    async keyGen(rootKey, context, account, keyIndex, derivationType = BIP32DerivationType.Peikert) {
        const bip44Path = GetBIP44PathFromContext(context, account, keyIndex);
        const extendedKey = await this.deriveKey(rootKey, bip44Path, false, derivationType);
        return extendedKey.subarray(0, 32); // only public key
    }
    /**
     * Raw Signing function called by signData and signTransaction
     *
     * Ref: https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.6
     *
     * Edwards-Curve Digital Signature Algorithm (EdDSA)
     *
     * @param bip44Path
     * - BIP44 path (m / purpose' / coin_type' / account' / change / address_index)
     * @param data
     * - data to be signed in raw bytes
     *
     * @returns
     * - signature holding R and S, totally 64 bytes
     */
    async rawSign(rootKey, bip44Path, data, derivationType) {
        const raw = await this.deriveKey(rootKey, bip44Path, true, derivationType);
        const scalar = raw.slice(0, 32);
        const kR = raw.slice(32, 64);
        // \(1): pubKey = scalar * G (base point, no clamp)
        const publicKey = crypto_scalarmult_ed25519_base_noclamp(scalar);
        // \(2): h = hash(c || msg) mod q
        const r = crypto_core_ed25519_scalar_reduce(crypto_hash_sha512(Buffer.concat([kR, data])));
        // \(4):  R = r * G (base point, no clamp)
        const R = crypto_scalarmult_ed25519_base_noclamp(r);
        // h = hash(R || pubKey || msg) mod q
        let h = crypto_core_ed25519_scalar_reduce(crypto_hash_sha512(Buffer.concat([R, publicKey, data])));
        // \(5): S = (r + h * k) mod q
        const S = crypto_core_ed25519_scalar_add(r, crypto_core_ed25519_scalar_mul(h, scalar));
        return Buffer.concat([R, S]);
    }
    /**
     * Ref: https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.6
     *
     *  Edwards-Curve Digital Signature Algorithm (EdDSA)
     *
     * @param context - context of the key (i.e Address, Identity)
     * @param account - account number. This value will be hardened as part of BIP44
     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
     * @param data - data to be signed in raw bytes
     * @param metadata - metadata object that describes how `data` was encoded and what schema to use to validate against
     * @param derivationType
     * - BIP32 derivation type, defines if it's standard Ed25519 or Peikert's ammendment to BIP32-Ed25519
     *
     * @returns - signature holding R and S, totally 64 bytes
     * */
    async signData(rootKey, context, account, keyIndex, data, metadata, derivationType = BIP32DerivationType.Peikert) {
        // validate data
        const result = this.validateData(data, metadata);
        if (result instanceof Error) { // decoding errors
            throw result;
        }
        if (!result) { // failed schema validation
            throw ERROR_BAD_DATA;
        }
        const bip44Path = GetBIP44PathFromContext(context, account, keyIndex);
        return await this.rawSign(rootKey, bip44Path, data, derivationType);
    }
    /**
     * Sign Algorand transaction
     * @param context
     * - context of the key (i.e Address, Identity)
     * @param account
     * - account number. This value will be hardened as part of BIP44
     * @param keyIndex
     * - key index. This value will be a SOFT derivation as part of BIP44.
     * @param prefixEncodedTx
     * - Encoded transaction object
     * @param derivationType
     * - BIP32 derivation type, defines if it's standard Ed25519 or Peikert's ammendment to BIP32-Ed25519
     *
     * @returns sig
     * - Raw bytes signature
     */
    async signAlgoTransaction(rootKey, context, account, keyIndex, prefixEncodedTx, derivationType = BIP32DerivationType.Peikert) {
        const bip44Path = GetBIP44PathFromContext(context, account, keyIndex);
        const sig = await this.rawSign(rootKey, bip44Path, prefixEncodedTx, derivationType);
        return sig;
    }
    /**
     * SAMPLE IMPLEMENTATION to show how to validate data with encoding and schema, using base64 as an example
     *
     * @param message
     * @param metadata
     * @returns
     */
    validateData(message, metadata) {
        // Check that decoded doesn't include the following prefixes: TX, MX, progData, Program
        // These prefixes are reserved for the protocol
        if (this.hasAlgorandTags(message)) {
            return ERROR_TAGS_FOUND;
        }
        let decoded;
        switch (metadata.encoding) {
            case Encoding.BASE64:
                decoded = new Uint8Array(Buffer.from(Buffer.from(message).toString(), 'base64'));
                break;
            case Encoding.MSGPACK:
                decoded = msgpack.decode(message);
                break;
            case Encoding.NONE:
                decoded = message;
                break;
            default:
                throw Error("Invalid encoding");
        }
        // validate with schema
        //@ts-expect-error, this is constructable
        const ajv = new Ajv();
        const validate = ajv.compile(metadata.schema);
        const valid = validate(decoded);
        if (!valid)
            console.log(ajv.errors);
        return valid;
    }
    /**
     * Detect if the message has Algorand protocol specific tags
     *
     * @param message - raw bytes of the message
     * @returns - true if message has Algorand protocol specific tags, false otherwise
     */
    hasAlgorandTags(message) {
        // Check that decoded doesn't include the following prefixes
        // Prefixes taken from go-algorand node software code
        // https://github.com/algorand/go-algorand/blob/master/protocol/hash.go
        const prefixes = [
            "appID", "arc", "aB", "aD", "aO", "aP", "aS", "AS", "B256", "BH", "BR", "CR", "GE", "KP", "MA", "MB",
            "MX", "NIC", "NIR", "NIV", "NPR", "OT1", "OT2", "PF", "PL", "Program", "ProgData", "PS", "PK", "SD",
            "SpecialAddr", "STIB", "spc", "spm", "spp", "sps", "spv", "TE", "TG", "TL", "TX", "VO"
        ];
        for (const prefix of prefixes) {
            if (Buffer.from(message.subarray(0, prefix.length)).toString("ascii") === prefix) {
                return true;
            }
        }
        return false;
    }
    /**
     * Wrapper around libsodium basica signature verification
     *
     * Any lib or system that can verify EdDSA signatures can be used
     *
     * @param signature - raw 64 bytes signature (R, S)
     * @param message - raw bytes of the message
     * @param publicKey - raw 32 bytes public key (x,y)
     * @returns true if signature is valid, false otherwise
     */
    async verifyWithPublicKey(signature, message, publicKey) {
        return crypto_sign_verify_detached(signature, message, publicKey);
    }
    /**
     * Function to perform ECDH against a provided public key
     *
     * ECDH reference link: https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman
     *
     * It creates a shared secret between two parties. Each party only needs to be aware of the other's public key.
     * This symmetric secret can be used to derive a symmetric key for encryption and decryption. Creating a private channel between the two parties.
     *
     * @param context - context of the key (i.e Address, Identity)
     * @param account - account number. This value will be hardened as part of BIP44
     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
     * @param otherPartyPub - raw 32 bytes public key of the other party
     * @param meFirst - defines the order in which the keys will be considered for the shared secret. If true, our key will be used first, otherwise the other party's key will be used first
     * @returns - raw 32 bytes shared secret
     */
    async ECDH(rootKey, context, account, keyIndex, otherPartyPub, meFirst, derivationType = BIP32DerivationType.Peikert) {
        const bip44Path = GetBIP44PathFromContext(context, account, keyIndex);
        const childKey = await this.deriveKey(rootKey, bip44Path, true, derivationType);
        const scalar = childKey.slice(0, 32);
        // our public key is derived from the private key
        const ourPub = crypto_scalarmult_ed25519_base_noclamp(scalar);
        // convert from ed25519 to curve25519
        const ourPubCurve25519 = crypto_sign_ed25519_pk_to_curve25519(ourPub);
        const otherPartyPubCurve25519 = crypto_sign_ed25519_pk_to_curve25519(otherPartyPub);
        // find common point
        const sharedPoint = crypto_scalarmult(scalar, otherPartyPubCurve25519);
        let concatenation;
        if (meFirst) {
            concatenation = Buffer.concat([sharedPoint, ourPubCurve25519, otherPartyPubCurve25519]);
        }
        else {
            concatenation = Buffer.concat([sharedPoint, otherPartyPubCurve25519, ourPubCurve25519]);
        }
        return crypto_generichash(32, new Uint8Array(concatenation));
    }
}
