// @algorandfoundation/algokit-utils/amount.mjs
//#region src/amount.ts
/** Wrapper class to ensure safe, explicit conversion between µAlgo, Algo and numbers */
var AlgoAmount = class AlgoAmount {
	amountInMicroAlgo;
	/** Return the amount as a number in µAlgo */
	get microAlgos() {
		return this.amountInMicroAlgo;
	}
	/** Return the amount as a number in µAlgo */
	get microAlgo() {
		return this.amountInMicroAlgo;
	}
	/** Return the amount as a number in Algo */
	get algos() {
		return microalgosToAlgos(Number(this.amountInMicroAlgo));
	}
	/** Return the amount as a number in Algo */
	get algo() {
		return microalgosToAlgos(Number(this.amountInMicroAlgo));
	}
	/**
	* Create a new `AlgoAmount` instance.
	*
	* @param amount - An object specifying the amount in Algo or µALGO. Use the key 'algo' for Algo amounts and 'microAlgo' for µALGO.
	* @returns A new instance of `AlgoAmount` representing the specified amount.
	* @example
	* ```typescript
	* const amount = new AlgoAmount({ algo: 5 });
	* ```
	*/
	constructor(amount) {
		this.amountInMicroAlgo = "microAlgos" in amount ? BigInt(amount.microAlgos) : "microAlgo" in amount ? BigInt(amount.microAlgo) : "algos" in amount ? BigInt(algosToMicroalgos(Number(amount.algos))) : BigInt(algosToMicroalgos(Number(amount.algo)));
	}
	toString() {
		return `${this.microAlgo.toLocaleString("en-US")} µALGO`;
	}
	/** valueOf allows you to use `AlgoAmount` in comparison operations such as `<` and `>=` etc.,
	* but it's not recommended to use this to convert to a number, it's much safer to explicitly call
	* the algos or microAlgos properties
	*/
	valueOf() {
		return Number(this.microAlgo);
	}
	/** Create a `AlgoAmount` object representing the given number of Algo */
	static Algos(amount) {
		return new AlgoAmount({ algos: amount });
	}
	/** Create a `AlgoAmount` object representing the given number of Algo */
	static Algo(amount) {
		return new AlgoAmount({ algos: amount });
	}
	/** Create a `AlgoAmount` object representing the given number of µAlgo */
	static MicroAlgos(amount) {
		return new AlgoAmount({ microAlgos: amount });
	}
	/** Create a `AlgoAmount` object representing the given number of µAlgo */
	static MicroAlgo(amount) {
		return new AlgoAmount({ microAlgos: amount });
	}
};
const MICROALGOS_TO_ALGOS_RATIO = 1e6;
const INVALID_MICROALGOS_ERROR_MSG = "Microalgos should be positive and less than 2^53 - 1.";
/**
* microalgosToAlgos converts microalgos to algos
* @param microalgos - number
* @returns number
*/
function microalgosToAlgos(microalgos) {
	if (microalgos < 0 || !Number.isSafeInteger(microalgos)) throw new Error(INVALID_MICROALGOS_ERROR_MSG);
	return microalgos / MICROALGOS_TO_ALGOS_RATIO;
}
/**
* algosToMicroalgos converts algos to microalgos
* @param algos - number
* @returns number
*/
function algosToMicroalgos(algos) {
	const microalgos = algos * MICROALGOS_TO_ALGOS_RATIO;
	return Math.round(microalgos);
}
Number.prototype.microAlgos = function() {
	return AlgoAmount.MicroAlgo(this);
};
Number.prototype.algos = function() {
	return AlgoAmount.Algo(this);
};
Number.prototype.microAlgo = function() {
	return AlgoAmount.MicroAlgo(this);
};
Number.prototype.algo = function() {
	return AlgoAmount.Algo(this);
};
BigInt.prototype.microAlgo = function() {
	return AlgoAmount.MicroAlgo(this);
};
BigInt.prototype.algo = function() {
	return AlgoAmount.Algo(this);
};
/** Returns an amount of Algo using AlgoAmount
* @param algos The amount of Algo
* @example
* {@includeCode ./amount.spec.ts#example-algos}
* @see {@link AlgoAmount}
*/
const algos = (algos) => {
	return AlgoAmount.Algo(algos);
};
/** Returns an amount of Algo using AlgoAmount
* @param algos The amount of Algo
* @example
* {@includeCode ./amount.spec.ts#example-algo}
*/
const algo = (algos) => {
	return AlgoAmount.Algo(algos);
};
/** Returns an amount of µAlgo using AlgoAmount
* @param microAlgos The amount of µAlgo
* @example
* {@includeCode ./amount.spec.ts#example-microAlgos}
* @see {@link algos} for Algo denomination
*/
const microAlgos = (microAlgos) => {
	return AlgoAmount.MicroAlgo(microAlgos);
};
/** Returns an amount of µAlgo using AlgoAmount
* @param microAlgos The amount of µAlgo
* @example
* {@includeCode ./amount.spec.ts#example-microAlgo}
*/
const microAlgo = (microAlgos) => {
	return AlgoAmount.MicroAlgo(microAlgos);
};
/** Returns an amount of µAlgo to cover standard fees for the given number of transactions using AlgoAmount
* @param numberOfTransactions The of standard transaction fees to return the amount of Algo
* @example
* {@includeCode ./amount.spec.ts#example-transactionFees}
*/
const transactionFees = (numberOfTransactions) => {
	return AlgoAmount.MicroAlgo(BigInt(numberOfTransactions) * ALGORAND_MIN_TX_FEE.microAlgo);
};
const ALGORAND_MIN_TX_FEE = AlgoAmount.MicroAlgo(1e3);
//#endregion
export { ALGORAND_MIN_TX_FEE, AlgoAmount, algo, algos, microAlgo, microAlgos, transactionFees };
