// @algorandfoundation/algokit-utils/packages/algod_client/src/apis/api-service.mjs
import { concatArrays } from "../../../common/src/array.mjs";
import { decodeSignedTransaction } from "../../../transact/src/transactions/signed-transaction.mjs";
import { PendingTransactionResponseMeta } from "../models/pending-transaction-response.mjs";
import { decodeJson, decodeMsgpack, encodeMsgpack } from "../core/model-runtime.mjs";
import { SimulateResponseMeta } from "../models/simulate-response.mjs";
import { GenesisMeta } from "../models/genesis.mjs";
import { LedgerStateDeltaMeta } from "../models/ledger-state-delta.mjs";
import { ApplicationMeta } from "../models/application.mjs";
import { AssetMeta } from "../models/asset.mjs";
import { AccountMeta } from "../models/account.mjs";
import { SimulateRequestMeta } from "../models/simulate-request.mjs";
import { BoxMeta } from "../models/box.mjs";
import { VersionMeta } from "../models/version.mjs";
import { StateProofMeta } from "../models/state-proof.mjs";
import { LightBlockHeaderProofMeta } from "../models/light-block-header-proof.mjs";
import { TransactionProofMeta } from "../models/transaction-proof.mjs";
import { GetBlockTimeStampOffsetResponseMeta } from "../models/get-block-time-stamp-offset-response.mjs";
import { GetSyncRoundResponseMeta } from "../models/get-sync-round-response.mjs";
import { TransactionGroupLedgerStateDeltasForRoundResponseMeta } from "../models/transaction-group-ledger-state-deltas-for-round-response.mjs";
import { AccountAssetResponseMeta } from "../models/account-asset-response.mjs";
import { AccountApplicationResponseMeta } from "../models/account-application-response.mjs";
import { BlockResponseMeta } from "../models/block-response.mjs";
import { BlockTxidsResponseMeta } from "../models/block-txids-response.mjs";
import { BlockHashResponseMeta } from "../models/block-hash-response.mjs";
import { NodeStatusResponseMeta } from "../models/node-status-response.mjs";
import { PendingTransactionsResponseMeta } from "../models/pending-transactions-response.mjs";
import { PostTransactionsResponseMeta } from "../models/post-transactions-response.mjs";
import { SupplyResponseMeta } from "../models/supply-response.mjs";
import { TransactionParametersResponseMeta } from "../models/transaction-parameters-response.mjs";
import { BoxesResponseMeta } from "../models/boxes-response.mjs";
import { CompileResponseMeta } from "../models/compile-response.mjs";
import { DisassembleResponseMeta } from "../models/disassemble-response.mjs";

//#region packages/algod_client/src/apis/api-service.ts
var AlgodApi = class {
	constructor(httpRequest) {
		this.httpRequest = httpRequest;
	}
	mimeTypeFor(format) {
		return format === "json" ? "application/json" : format === "msgpack" ? "application/msgpack" : "text/plain";
	}
	/**
	* Given a specific account public key and application ID, this call returns the account's application local state and global state (AppLocalState and AppParams, if either exists). Global state will only be returned if the provided address is the application's creator.
	*/
	async accountApplicationInformation(address, applicationId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{address}/applications/{application-id}",
			path: {
				address,
				"application-id": applicationId
			},
			query: {},
			headers,
			body: void 0
		}), AccountApplicationResponseMeta);
	}
	/**
	* Given a specific account public key and asset ID, this call returns the account's asset holding and asset parameters (if either exist). Asset parameters will only be returned if the provided address is the asset's creator.
	*/
	async accountAssetInformation(address, assetId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{address}/assets/{asset-id}",
			path: {
				address,
				"asset-id": assetId
			},
			query: {},
			headers,
			body: void 0
		}), AccountAssetResponseMeta);
	}
	/**
	* Given a specific account public key, this call returns the account's status, balance and spendable amounts
	*/
	async accountInformation(address, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{address}",
			path: { address },
			query: { exclude: params?.exclude },
			headers,
			body: void 0
		}), AccountMeta);
	}
	/**
	* Given an application ID and box name, it returns the round, box name, and value (each base64 encoded). Box names must be in the goal app call arg encoding form 'encoding:value'. For ints, use the form 'int:1234'. For raw bytes, use the form 'b64:A=='. For printable strings, use the form 'str:hello'. For addresses, use the form 'addr:XYZ...'.
	*/
	async _applicationBoxByName(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/box",
			path: { "application-id": applicationId },
			query: { name: params?.name },
			headers,
			body: void 0
		}), BoxMeta);
	}
	/**
	* Given an application ID, return all Box names. No particular ordering is guaranteed. Request fails when client or server-side configured limits prevent returning all Box names.
	*/
	async applicationBoxes(applicationId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}/boxes",
			path: { "application-id": applicationId },
			query: { max: params?.max },
			headers,
			body: void 0
		}), BoxesResponseMeta);
	}
	/**
	* Given a application ID, it returns application information including creator, approval and clear programs, global and local schemas, and global state.
	*/
	async applicationById(applicationId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/applications/{application-id}",
			path: { "application-id": applicationId },
			query: {},
			headers,
			body: void 0
		}), ApplicationMeta);
	}
	/**
	* Given a asset ID, it returns asset information including creator, name, total supply and special addresses.
	*/
	async assetById(assetId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/assets/{asset-id}",
			path: { "asset-id": assetId },
			query: {},
			headers,
			body: void 0
		}), AssetMeta);
	}
	async block(round, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round}",
			path: { round },
			query: {
				"header-only": params?.headerOnly,
				format: "msgpack"
			},
			headers,
			body: void 0
		}), BlockResponseMeta);
	}
	async blockHash(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round}/hash",
			path: { round },
			query: {},
			headers,
			body: void 0
		}), BlockHashResponseMeta);
	}
	/**
	* Gets the current timestamp offset.
	*/
	async blockTimeStampOffset() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/devmode/blocks/offset",
			path: {},
			query: {},
			headers,
			body: void 0
		}), GetBlockTimeStampOffsetResponseMeta);
	}
	async blockTxIds(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round}/txids",
			path: { round },
			query: {},
			headers,
			body: void 0
		}), BlockTxidsResponseMeta);
	}
	/**
	* Returns the entire genesis file in json.
	*/
	async genesis() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/genesis",
			path: {},
			query: {},
			headers,
			body: void 0
		}), GenesisMeta);
	}
	async healthCheck() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		await this.httpRequest.request({
			method: "GET",
			url: "/health",
			path: {},
			query: {},
			headers,
			body: void 0
		});
	}
	/**
	* Get ledger deltas for a round.
	*/
	async ledgerStateDelta(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/deltas/{round}",
			path: { round },
			query: { format: "msgpack" },
			headers,
			body: void 0
		}), LedgerStateDeltaMeta);
	}
	/**
	* Get a ledger delta for a given transaction group.
	*/
	async ledgerStateDeltaForTransactionGroup(id) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/deltas/txn/group/{id}",
			path: { id },
			query: { format: "msgpack" },
			headers,
			body: void 0
		}), LedgerStateDeltaMeta);
	}
	async lightBlockHeaderProof(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round}/lightheader/proof",
			path: { round },
			query: {},
			headers,
			body: void 0
		}), LightBlockHeaderProofMeta);
	}
	/**
	* Given a transaction ID of a recently submitted transaction, it returns information about it.  There are several cases when this might succeed:
	* - transaction committed (committed round > 0)
	* - transaction still in the pool (committed round = 0, pool error = "")
	* - transaction removed from pool due to error (committed round = 0, pool error != "")
	* Or the transaction may have happened sufficiently long ago that the node no longer remembers it, and this will return an error.
	*/
	async pendingTransactionInformation(txId) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions/pending/{txid}",
			path: { txid: txId },
			query: { format: "msgpack" },
			headers,
			body: void 0
		}), PendingTransactionResponseMeta);
	}
	/**
	* Get the list of pending transactions, sorted by priority, in decreasing order, truncated at the end at MAX. If MAX = 0, returns all pending transactions.
	*/
	async pendingTransactions(params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions/pending",
			path: {},
			query: {
				max: params?.max,
				format: "msgpack"
			},
			headers,
			body: void 0
		}), PendingTransactionsResponseMeta);
	}
	/**
	* Get the list of pending transactions by address, sorted by priority, in decreasing order, truncated at the end at MAX. If MAX = 0, returns all pending transactions.
	*/
	async pendingTransactionsByAddress(address, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/accounts/{address}/transactions/pending",
			path: { address },
			query: {
				max: params?.max,
				format: "msgpack"
			},
			headers,
			body: void 0
		}), PendingTransactionsResponseMeta);
	}
	async _rawTransaction(body) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		const serializedBody = body ?? void 0;
		headers["Content-Type"] = "application/x-binary";
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v2/transactions",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), PostTransactionsResponseMeta);
	}
	async ready() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		await this.httpRequest.request({
			method: "GET",
			url: "/ready",
			path: {},
			query: {},
			headers,
			body: void 0
		});
	}
	/**
	* Sets the timestamp offset (seconds) for blocks in dev mode. Providing an offset of 0 will unset this value and try to use the real clock for the timestamp.
	*/
	async setBlockTimeStampOffset(offset) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		await this.httpRequest.request({
			method: "POST",
			url: "/v2/devmode/blocks/offset/{offset}",
			path: { offset },
			query: {},
			headers,
			body: void 0
		});
	}
	/**
	* Sets the minimum sync round on the ledger.
	*/
	async setSyncRound(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		await this.httpRequest.request({
			method: "POST",
			url: "/v2/ledger/sync/{round}",
			path: { round },
			query: {},
			headers,
			body: void 0
		});
	}
	async simulateTransactions(body) {
		const headers = {};
		const responseFormat = "msgpack";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = SimulateRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? encodeMsgpack(body, bodyMeta) : void 0;
		return decodeMsgpack(await this.httpRequest.request({
			method: "POST",
			url: "/v2/transactions/simulate",
			path: {},
			query: { format: "msgpack" },
			headers,
			body: serializedBody
		}), SimulateResponseMeta);
	}
	async stateProof(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/stateproofs/{round}",
			path: { round },
			query: {},
			headers,
			body: void 0
		}), StateProofMeta);
	}
	async status() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/status",
			path: {},
			query: {},
			headers,
			body: void 0
		}), NodeStatusResponseMeta);
	}
	/**
	* Waits for a block to appear after round {round} and returns the node's status at the time. There is a 1 minute timeout, when reached the current status is returned regardless of whether or not it is the round after the given round.
	*/
	async statusAfterBlock(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/status/wait-for-block-after/{round}",
			path: { round },
			query: {},
			headers,
			body: void 0
		}), NodeStatusResponseMeta);
	}
	async supply() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/ledger/supply",
			path: {},
			query: {},
			headers,
			body: void 0
		}), SupplyResponseMeta);
	}
	/**
	* Gets the minimum sync round for the ledger.
	*/
	async syncRound() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/ledger/sync",
			path: {},
			query: {},
			headers,
			body: void 0
		}), GetSyncRoundResponseMeta);
	}
	/**
	* Given TEAL source code in plain text, return base64 encoded program bytes and base32 SHA512_256 hash of program bytes (Address style). This endpoint is only enabled when a node's configuration file sets EnableDeveloperAPI to true.
	*/
	async tealCompile(body, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		const mediaType = this.mimeTypeFor("text");
		if (mediaType) headers["Content-Type"] = mediaType;
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v2/teal/compile",
			path: {},
			query: { sourcemap: params?.sourcemap },
			headers,
			body
		}), CompileResponseMeta);
	}
	/**
	* Given the program bytes, return the TEAL source code in plain text. This endpoint is only enabled when a node's configuration file sets EnableDeveloperAPI to true.
	*/
	async tealDisassemble(body) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		const serializedBody = body ?? void 0;
		headers["Content-Type"] = "application/x-binary";
		return decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v2/teal/disassemble",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), DisassembleResponseMeta);
	}
	/**
	* Get ledger deltas for transaction groups in a given round.
	*/
	async transactionGroupLedgerStateDeltasForRound(round) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("msgpack");
		return decodeMsgpack(await this.httpRequest.request({
			method: "GET",
			url: "/v2/deltas/{round}/txn/group",
			path: { round },
			query: { format: "msgpack" },
			headers,
			body: void 0
		}), TransactionGroupLedgerStateDeltasForRoundResponseMeta);
	}
	async _transactionParams() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/transactions/params",
			path: {},
			query: {},
			headers,
			body: void 0
		}), TransactionParametersResponseMeta);
	}
	async transactionProof(round, txId, params) {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v2/blocks/{round}/transactions/{txid}/proof",
			path: {
				round,
				txid: txId
			},
			query: { hashtype: params?.hashtype },
			headers,
			body: void 0
		}), TransactionProofMeta);
	}
	/**
	* Unset the ledger sync round.
	*/
	async unsetSyncRound() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		await this.httpRequest.request({
			method: "DELETE",
			url: "/v2/ledger/sync",
			path: {},
			query: {},
			headers,
			body: void 0
		});
	}
	/**
	* Retrieves the supported API versions, binary build versions, and genesis information.
	*/
	async version() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/versions",
			path: {},
			query: {},
			headers,
			body: void 0
		}), VersionMeta);
	}
	/**
	* Send a signed transaction or array of signed transactions to the network.
	*/
	async sendRawTransaction(stxOrStxs) {
		let rawTransactions = stxOrStxs;
		if (Array.isArray(stxOrStxs)) {
			if (!stxOrStxs.every((a) => a instanceof Uint8Array)) throw new Error("Array elements must be byte arrays");
			rawTransactions = concatArrays(...stxOrStxs);
		} else if (!(rawTransactions instanceof Uint8Array)) throw new Error("Argument must be byte array");
		return this._rawTransaction(rawTransactions);
	}
	/**
	* Given an application ID and box name, it returns the round, box name, and value.
	*/
	async applicationBoxByName(applicationId, boxName) {
		const name = `b64:${Buffer.from(boxName).toString("base64")}`;
		return this._applicationBoxByName(applicationId, { name });
	}
	/**
	* Returns the common needed parameters for a new transaction.
	*/
	async suggestedParams() {
		const txnParams = await this._transactionParams();
		return {
			flatFee: false,
			fee: txnParams.fee,
			firstValid: txnParams.lastRound,
			lastValid: txnParams.lastRound + 1000n,
			genesisHash: txnParams.genesisHash,
			genesisId: txnParams.genesisId,
			minFee: txnParams.minFee,
			consensusVersion: txnParams.consensusVersion
		};
	}
	/**
	* Returns the common needed parameters for a new transaction.
	*/
	async transactionParams() {
		return await this.suggestedParams();
	}
	/**
	* Simulate an encoded signed transaction or array of encoded signed transactions.
	*/
	async simulateRawTransactions(stxOrStxs) {
		const txns = Array.isArray(stxOrStxs) ? stxOrStxs.map((stxn) => decodeSignedTransaction(stxn)) : [decodeSignedTransaction(stxOrStxs)];
		return this.simulateTransactions({ txnGroups: [{ txns }] });
	}
};

//#endregion
export { AlgodApi };
