// @algorandfoundation/algokit-utils/types/app.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_app = require("../app.js");
//#region src/types/app.ts
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const UPDATABLE_TEMPLATE_NAME = require_app.UPDATABLE_TEMPLATE_NAME;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const DELETABLE_TEMPLATE_NAME = require_app.DELETABLE_TEMPLATE_NAME;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const APP_DEPLOY_NOTE_DAPP = require_app.APP_DEPLOY_NOTE_DAPP;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const APP_PAGE_MAX_SIZE = require_app.APP_PAGE_MAX_SIZE;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const ABI_RETURN_PREFIX = require_app.ABI_RETURN_PREFIX;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const OnUpdate = require_app.OnUpdate;
/** @deprecated Import from `@algorandfoundation/algokit-utils/app` instead */
const OnSchemaBreak = require_app.OnSchemaBreak;
//#endregion
exports.ABI_RETURN_PREFIX = ABI_RETURN_PREFIX;
exports.APP_DEPLOY_NOTE_DAPP = APP_DEPLOY_NOTE_DAPP;
exports.APP_PAGE_MAX_SIZE = APP_PAGE_MAX_SIZE;
exports.DELETABLE_TEMPLATE_NAME = DELETABLE_TEMPLATE_NAME;
exports.OnSchemaBreak = OnSchemaBreak;
exports.OnUpdate = OnUpdate;
exports.UPDATABLE_TEMPLATE_NAME = UPDATABLE_TEMPLATE_NAME;
