// @algorandfoundation/algokit-utils/packages/kmd_client/src/models/sign-multisig-txn-request.mjs
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { MultisigSigMeta } from "./multisig-sig.mjs";
//#region packages/kmd_client/src/models/sign-multisig-txn-request.ts
const SignMultisigTxnRequestMeta = {
	name: "SignMultisigTxnRequest",
	kind: "object",
	fields: [
		{
			name: "partialMultisig",
			wireKey: "partial_multisig",
			optional: true,
			codec: new ObjectModelCodec(MultisigSigMeta)
		},
		{
			name: "publicKey",
			wireKey: "public_key",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "signer",
			wireKey: "signer",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "transaction",
			wireKey: "transaction",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "walletHandleToken",
			wireKey: "wallet_handle_token",
			optional: false,
			codec: stringCodec
		},
		{
			name: "walletPassword",
			wireKey: "wallet_password",
			optional: true,
			codec: stringCodec
		}
	]
};
//#endregion
export { SignMultisigTxnRequestMeta };
