// @algorandfoundation/algokit-utils/packages/kmd_client/src/models/sign-program-multisig-request.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { MultisigSigMeta } from "./multisig-sig.mjs";
//#region packages/kmd_client/src/models/sign-program-multisig-request.ts
const SignProgramMultisigRequestMeta = {
	name: "SignProgramMultisigRequest",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "address",
			optional: false,
			codec: addressCodec
		},
		{
			name: "program",
			wireKey: "data",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "partialMultisig",
			wireKey: "partial_multisig",
			optional: true,
			codec: new ObjectModelCodec(MultisigSigMeta)
		},
		{
			name: "publicKey",
			wireKey: "public_key",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "useLegacyMsig",
			wireKey: "use_legacy_msig",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "walletHandleToken",
			wireKey: "wallet_handle_token",
			optional: false,
			codec: stringCodec
		},
		{
			name: "walletPassword",
			wireKey: "wallet_password",
			optional: true,
			codec: stringCodec
		}
	]
};
//#endregion
export { SignProgramMultisigRequestMeta };
