// @algorandfoundation/algokit-utils/packages/kmd_client/src/models/sign-multisig-txn-request.js
const require_bytes = require("../../../common/src/codecs/primitives/bytes.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_object_model = require("../../../common/src/codecs/models/object-model.js");
const require_multisig_sig = require("./multisig-sig.js");
//#region packages/kmd_client/src/models/sign-multisig-txn-request.ts
const SignMultisigTxnRequestMeta = {
	name: "SignMultisigTxnRequest",
	kind: "object",
	fields: [
		{
			name: "partialMultisig",
			wireKey: "partial_multisig",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(require_multisig_sig.MultisigSigMeta)
		},
		{
			name: "publicKey",
			wireKey: "public_key",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "signer",
			wireKey: "signer",
			optional: true,
			codec: require_bytes.bytesCodec
		},
		{
			name: "transaction",
			wireKey: "transaction",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "walletHandleToken",
			wireKey: "wallet_handle_token",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "walletPassword",
			wireKey: "wallet_password",
			optional: true,
			codec: require_string.stringCodec
		}
	]
};
//#endregion
exports.SignMultisigTxnRequestMeta = SignMultisigTxnRequestMeta;
