// @algorandfoundation/algokit-utils/packages/kmd_client/src/models/wallet.mjs
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { ArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { PrimitiveModelCodec } from "../../../common/src/codecs/models/primitive-model.mjs";
import { TxTypeMeta } from "./tx-type.mjs";
//#region packages/kmd_client/src/models/wallet.ts
const WalletMeta = {
	name: "Wallet",
	kind: "object",
	fields: [
		{
			name: "driverName",
			wireKey: "driver_name",
			optional: false,
			codec: stringCodec
		},
		{
			name: "driverVersion",
			wireKey: "driver_version",
			optional: false,
			codec: numberCodec
		},
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: stringCodec
		},
		{
			name: "mnemonicUx",
			wireKey: "mnemonic_ux",
			optional: false,
			codec: booleanCodec
		},
		{
			name: "name",
			wireKey: "name",
			optional: false,
			codec: stringCodec
		},
		{
			name: "supportedTxs",
			wireKey: "supported_txs",
			optional: false,
			codec: new ArrayCodec(new PrimitiveModelCodec(TxTypeMeta))
		}
	]
};
//#endregion
export { WalletMeta };
