// @algorandfoundation/algokit-utils/packages/kmd_client/src/models/wallet.js
const require_boolean = require("../../../common/src/codecs/primitives/boolean.js");
const require_number = require("../../../common/src/codecs/primitives/number.js");
const require_string = require("../../../common/src/codecs/primitives/string.js");
const require_array = require("../../../common/src/codecs/composite/array.js");
const require_primitive_model = require("../../../common/src/codecs/models/primitive-model.js");
const require_tx_type = require("./tx-type.js");
//#region packages/kmd_client/src/models/wallet.ts
const WalletMeta = {
	name: "Wallet",
	kind: "object",
	fields: [
		{
			name: "driverName",
			wireKey: "driver_name",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "driverVersion",
			wireKey: "driver_version",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "id",
			wireKey: "id",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "mnemonicUx",
			wireKey: "mnemonic_ux",
			optional: false,
			codec: require_boolean.booleanCodec
		},
		{
			name: "name",
			wireKey: "name",
			optional: false,
			codec: require_string.stringCodec
		},
		{
			name: "supportedTxs",
			wireKey: "supported_txs",
			optional: false,
			codec: new require_array.ArrayCodec(new require_primitive_model.PrimitiveModelCodec(require_tx_type.TxTypeMeta))
		}
	]
};
//#endregion
exports.WalletMeta = WalletMeta;
