// @algorandfoundation/algokit-utils/packages/transact/src/signer.js
const require_array = require("../../common/src/array.js");
const require_address = require("../../common/src/address.js");
const require_transaction = require("./transactions/transaction.js");
const require_signed_transaction = require("./transactions/signed-transaction.js");
//#region packages/transact/src/signer.ts
const SIGN_BYTES_PREFIX = Uint8Array.from([77, 88]);
/**
* Generate type-safe domain-separated signer callbacks given an ed25519 pubkey and a signing callback
* @param args - The arguments for generating signers
* @param args.ed25519Pubkey - The ed25519 public key used for signing
* @param args.sendingAddress - The address that will be used as the sender of transactions. If not provided, defaults to the address derived from the ed25519 public key. This is useful when signing for a rekeyed account where the sending address differs from the auth address.
* @param args.rawEd25519Signer - A callback function that signs raw bytes using the ed25519 private key
* @returns An object containing the sending address and various signer functions
*/
function generateAddressWithSigners(args) {
	const { ed25519Pubkey, rawEd25519Signer } = args;
	const authAddr = new require_address.Address(ed25519Pubkey);
	const sendingAddress = args.sendingAddress ?? authAddr;
	const signer = async (txnGroup, indexesToSign) => {
		const stxns = [];
		for (const index of indexesToSign) {
			const txn = txnGroup[index];
			const stxn = {
				txn,
				sig: await rawEd25519Signer(require_transaction.encodeTransaction(txn))
			};
			if (!txn.sender.equals(sendingAddress)) stxn.authAddress = authAddr;
			stxns.push(stxn);
		}
		return stxns.map((stxn) => {
			require_signed_transaction.validateSignedTransaction(stxn);
			return require_signed_transaction.encodeSignedTransaction(stxn);
		});
	};
	const lsigSigner = async (lsig, msig) => {
		return {
			sig: await rawEd25519Signer(lsig.bytesToSignForDelegation(msig)),
			addr: sendingAddress
		};
	};
	const programDataSigner = async (data, lsig) => {
		return await rawEd25519Signer(lsig.programDataToSign(data));
	};
	const mxBytesSigner = async (bytes) => {
		return await rawEd25519Signer(require_array.concatArrays(SIGN_BYTES_PREFIX, bytes));
	};
	return {
		addr: sendingAddress,
		signer,
		lsigSigner,
		programDataSigner,
		mxBytesSigner
	};
}
/**
* Create a makeEmptyTransactionSigner that does not specify any signer or
* signing capabilities. This should only be used to simulate transactions.
*/
function makeEmptyTransactionSigner() {
	return (txnGroup, indexesToSign) => {
		const unsigned = [];
		for (const index of indexesToSign) {
			const stxn = {
				txn: txnGroup[index],
				sig: new Uint8Array(64).fill(0)
			};
			require_signed_transaction.validateSignedTransaction(stxn);
			unsigned.push(require_signed_transaction.encodeSignedTransaction(stxn));
		}
		return Promise.resolve(unsigned);
	};
}
const bytesForSigning = {
	transaction(txn) {
		return require_transaction.encodeTransaction(txn);
	},
	mxBytes(bytes) {
		return require_array.concatArrays(SIGN_BYTES_PREFIX, bytes);
	},
	lsigForDelegation(lsig, msig) {
		return lsig.bytesToSignForDelegation(msig);
	},
	programData(lsig, data) {
		return lsig.programDataToSign(data);
	}
};
//#endregion
exports.bytesForSigning = bytesForSigning;
exports.generateAddressWithSigners = generateAddressWithSigners;
exports.makeEmptyTransactionSigner = makeEmptyTransactionSigner;
