// @algorandfoundation/algokit-utils/packages/transact/src/transactions/common.mjs
//#region packages/transact/src/transactions/common.ts
let TransactionValidationErrorType = /* @__PURE__ */ function(TransactionValidationErrorType) {
	TransactionValidationErrorType[TransactionValidationErrorType["RequiredField"] = 0] = "RequiredField";
	TransactionValidationErrorType[TransactionValidationErrorType["FieldTooLong"] = 1] = "FieldTooLong";
	TransactionValidationErrorType[TransactionValidationErrorType["ImmutableField"] = 2] = "ImmutableField";
	TransactionValidationErrorType[TransactionValidationErrorType["ZeroValueField"] = 3] = "ZeroValueField";
	TransactionValidationErrorType[TransactionValidationErrorType["ArbitraryConstraint"] = 4] = "ArbitraryConstraint";
	return TransactionValidationErrorType;
}({});
function getValidationErrorMessage(validationError) {
	switch (validationError.type) {
		case TransactionValidationErrorType.RequiredField: return `${validationError.data} is required`;
		case TransactionValidationErrorType.FieldTooLong: return `${validationError.data.field} cannot exceed ${validationError.data.max} ${validationError.data.unit}, got ${validationError.data.actual}`;
		case TransactionValidationErrorType.ImmutableField: return `${validationError.data} is immutable and cannot be changed`;
		case TransactionValidationErrorType.ZeroValueField: return `${validationError.data} must not be 0`;
		case TransactionValidationErrorType.ArbitraryConstraint: return validationError.data;
	}
}

//#endregion
export { TransactionValidationErrorType, getValidationErrorMessage };
