// @algorandfoundation/algokit-utils/packages/transact/src/transactions/reference-types-meta.js
const require_address = require('../../../common/src/codecs/primitives/address.js');
const require_bigint = require('../../../common/src/codecs/primitives/bigint.js');
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');

//#region packages/transact/src/transactions/reference-types-meta.ts
/**
* Metadata for BoxReference
*
* Maps wire format (app, name) to BoxReference (appId, name)
*/
const BoxReferenceMeta = {
	name: "BoxReference",
	kind: "object",
	fields: [{
		name: "appId",
		wireKey: "app",
		optional: false,
		codec: require_bigint.bigIntCodec
	}, {
		name: "name",
		wireKey: "name",
		optional: false,
		codec: require_bytes.bytesCodec
	}]
};
/**
* Metadata for HoldingReference
*
* Maps wire format (account, asset) to HoldingReference (address, assetId)
*/
const HoldingReferenceMeta = {
	name: "HoldingReference",
	kind: "object",
	fields: [{
		name: "address",
		wireKey: "account",
		optional: false,
		codec: require_address.addressCodec
	}, {
		name: "assetId",
		wireKey: "asset",
		optional: false,
		codec: require_bigint.bigIntCodec
	}]
};
/**
* Metadata for LocalsReference
*
* Maps wire format (account, app) to LocalsReference (address, appId)
*/
const LocalsReferenceMeta = {
	name: "LocalsReference",
	kind: "object",
	fields: [{
		name: "address",
		wireKey: "account",
		optional: false,
		codec: require_address.addressCodec
	}, {
		name: "appId",
		wireKey: "app",
		optional: false,
		codec: require_bigint.bigIntCodec
	}]
};

//#endregion
exports.BoxReferenceMeta = BoxReferenceMeta;
exports.HoldingReferenceMeta = HoldingReferenceMeta;
exports.LocalsReferenceMeta = LocalsReferenceMeta;
