// @algorandfoundation/algokit-utils/packages/transact/src/transactions/signed-transaction-meta.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes32Codec, fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { ArrayCodec, bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { transactionCodec } from "./transaction.mjs";

//#region packages/transact/src/transactions/signed-transaction-meta.ts
const MultisigSignatureMeta = {
	name: "MultisigSignature",
	kind: "object",
	fields: [
		{
			name: "version",
			wireKey: "v",
			optional: false,
			codec: numberCodec
		},
		{
			name: "threshold",
			wireKey: "thr",
			optional: false,
			codec: numberCodec
		},
		{
			name: "subsigs",
			wireKey: "subsig",
			optional: false,
			codec: new ArrayCodec(new ObjectModelCodec({
				name: "MultisigSubsignature",
				kind: "object",
				fields: [{
					name: "publicKey",
					wireKey: "pk",
					optional: false,
					codec: fixedBytes32Codec
				}, {
					name: "sig",
					wireKey: "s",
					optional: true,
					codec: fixedBytes64Codec
				}]
			}))
		}
	]
};
const LogicSigSignatureMeta = {
	name: "LogicSignature",
	kind: "object",
	fields: [
		{
			name: "logic",
			wireKey: "l",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "args",
			wireKey: "arg",
			optional: true,
			codec: bytesArrayCodec
		},
		{
			name: "sig",
			wireKey: "sig",
			optional: true,
			codec: fixedBytes64Codec
		},
		{
			name: "msig",
			wireKey: "msig",
			optional: true,
			codec: new ObjectModelCodec(MultisigSignatureMeta)
		},
		{
			name: "lmsig",
			wireKey: "lmsig",
			optional: true,
			codec: new ObjectModelCodec(MultisigSignatureMeta)
		}
	]
};
const multisigSignatureCodec = new ObjectModelCodec(MultisigSignatureMeta);
const logicSigSignatureCodec = new ObjectModelCodec(LogicSigSignatureMeta);
/**
* Metadata for SignedTransaction
*/
const SignedTransactionMeta = {
	name: "SignedTransaction",
	kind: "object",
	fields: [
		{
			name: "txn",
			wireKey: "txn",
			optional: false,
			codec: transactionCodec
		},
		{
			name: "sig",
			wireKey: "sig",
			optional: true,
			codec: fixedBytes64Codec
		},
		{
			name: "msig",
			wireKey: "msig",
			optional: true,
			codec: multisigSignatureCodec
		},
		{
			name: "lsig",
			wireKey: "lsig",
			optional: true,
			codec: logicSigSignatureCodec
		},
		{
			name: "authAddress",
			wireKey: "sgnr",
			optional: true,
			codec: addressCodec
		}
	]
};
const signedTransactionCodec = new ObjectModelCodec(SignedTransactionMeta);

//#endregion
export { SignedTransactionMeta, logicSigSignatureCodec, signedTransactionCodec };
