// @algorandfoundation/algokit-utils/packages/transact/src/transactions/signed-transaction-meta.js
const require_address = require('../../../common/src/codecs/primitives/address.js');
const require_bytes = require('../../../common/src/codecs/primitives/bytes.js');
const require_fixed_bytes = require('../../../common/src/codecs/primitives/fixed-bytes.js');
const require_number = require('../../../common/src/codecs/primitives/number.js');
const require_array = require('../../../common/src/codecs/composite/array.js');
const require_object_model = require('../../../common/src/codecs/models/object-model.js');
const require_transaction = require('./transaction.js');

//#region packages/transact/src/transactions/signed-transaction-meta.ts
const MultisigSignatureMeta = {
	name: "MultisigSignature",
	kind: "object",
	fields: [
		{
			name: "version",
			wireKey: "v",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "threshold",
			wireKey: "thr",
			optional: false,
			codec: require_number.numberCodec
		},
		{
			name: "subsigs",
			wireKey: "subsig",
			optional: false,
			codec: new require_array.ArrayCodec(new require_object_model.ObjectModelCodec({
				name: "MultisigSubsignature",
				kind: "object",
				fields: [{
					name: "publicKey",
					wireKey: "pk",
					optional: false,
					codec: require_fixed_bytes.fixedBytes32Codec
				}, {
					name: "sig",
					wireKey: "s",
					optional: true,
					codec: require_fixed_bytes.fixedBytes64Codec
				}]
			}))
		}
	]
};
const LogicSigSignatureMeta = {
	name: "LogicSignature",
	kind: "object",
	fields: [
		{
			name: "logic",
			wireKey: "l",
			optional: false,
			codec: require_bytes.bytesCodec
		},
		{
			name: "args",
			wireKey: "arg",
			optional: true,
			codec: require_array.bytesArrayCodec
		},
		{
			name: "sig",
			wireKey: "sig",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		},
		{
			name: "msig",
			wireKey: "msig",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(MultisigSignatureMeta)
		},
		{
			name: "lmsig",
			wireKey: "lmsig",
			optional: true,
			codec: new require_object_model.ObjectModelCodec(MultisigSignatureMeta)
		}
	]
};
const multisigSignatureCodec = new require_object_model.ObjectModelCodec(MultisigSignatureMeta);
const logicSigSignatureCodec = new require_object_model.ObjectModelCodec(LogicSigSignatureMeta);
/**
* Metadata for SignedTransaction
*/
const SignedTransactionMeta = {
	name: "SignedTransaction",
	kind: "object",
	fields: [
		{
			name: "txn",
			wireKey: "txn",
			optional: false,
			codec: require_transaction.transactionCodec
		},
		{
			name: "sig",
			wireKey: "sig",
			optional: true,
			codec: require_fixed_bytes.fixedBytes64Codec
		},
		{
			name: "msig",
			wireKey: "msig",
			optional: true,
			codec: multisigSignatureCodec
		},
		{
			name: "lsig",
			wireKey: "lsig",
			optional: true,
			codec: logicSigSignatureCodec
		},
		{
			name: "authAddress",
			wireKey: "sgnr",
			optional: true,
			codec: require_address.addressCodec
		}
	]
};
const signedTransactionCodec = new require_object_model.ObjectModelCodec(SignedTransactionMeta);

//#endregion
exports.SignedTransactionMeta = SignedTransactionMeta;
exports.logicSigSignatureCodec = logicSigSignatureCodec;
exports.signedTransactionCodec = signedTransactionCodec;
