// @algorandfoundation/algokit-utils/packages/transact/src/transactions/transaction-meta.mjs
import { Address } from "../../../common/src/address.mjs";
import { Codec } from "../../../common/src/codecs/codec.mjs";
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { booleanCodec } from "../../../common/src/codecs/primitives/boolean.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
import { fixedBytes1793Codec, fixedBytes32Codec, fixedBytes64Codec } from "../../../common/src/codecs/primitives/fixed-bytes.mjs";
import { numberCodec } from "../../../common/src/codecs/primitives/number.mjs";
import { normalizeWireObject, normalizeWireString } from "../../../common/src/codecs/wire.mjs";
import { stringCodec } from "../../../common/src/codecs/primitives/string.mjs";
import { addressArrayCodec, bigIntArrayCodec, bytesArrayCodec } from "../../../common/src/codecs/composite/array.mjs";
import { MapCodec } from "../../../common/src/codecs/composite/map.mjs";
import { ObjectModelCodec } from "../../../common/src/codecs/models/object-model.mjs";
import { TransactionType } from "./transaction-type.mjs";

//#region packages/transact/src/transactions/transaction-meta.ts
var TransactionTypeCodec = class extends Codec {
	defaultValue() {
		return TransactionType.Unknown;
	}
	toEncoded(value, _format) {
		switch (value) {
			case TransactionType.Payment: return "pay";
			case TransactionType.AssetTransfer: return "axfer";
			case TransactionType.AssetFreeze: return "afrz";
			case TransactionType.AssetConfig: return "acfg";
			case TransactionType.KeyRegistration: return "keyreg";
			case TransactionType.AppCall: return "appl";
			case TransactionType.StateProof: return "stpf";
			case TransactionType.Heartbeat: return "hb";
			default: return "unknown";
		}
	}
	fromEncoded(value, _format) {
		switch (normalizeWireString(value)) {
			case "pay": return TransactionType.Payment;
			case "axfer": return TransactionType.AssetTransfer;
			case "afrz": return TransactionType.AssetFreeze;
			case "acfg": return TransactionType.AssetConfig;
			case "keyreg": return TransactionType.KeyRegistration;
			case "appl": return TransactionType.AppCall;
			case "stpf": return TransactionType.StateProof;
			case "hb": return TransactionType.Heartbeat;
			default: return TransactionType.Unknown;
		}
	}
};
var TransactionDataCodec = class extends Codec {
	transactionDataCodec;
	constructor(transactionType, transactionTypeDataMetadata) {
		super();
		this.transactionType = transactionType;
		this.transactionDataCodec = new ObjectModelCodec(transactionTypeDataMetadata);
	}
	defaultValue() {}
	toEncoded(value, format) {
		if (!value) throw new Error("Transaction data is missing");
		return this.transactionDataCodec.encode(value, format);
	}
	fromEncoded(value, format) {
		if (value.type === void 0) throw new Error("Transaction is missing type field");
		if (normalizeWireString(value.type) === this.transactionType.toString()) return this.transactionDataCodec.decode(value, format);
	}
};
var AssetConfigDataCodec = class extends Codec {
	assetParamsCodec = new ObjectModelCodec(AssetParamsMeta);
	defaultValue() {}
	toEncoded(value, format) {
		const result = {};
		if (!value) throw new Error("Transaction is not an asset config");
		const { assetId, ...assetParams } = value;
		const encodedAssetId = bigIntCodec.encodeOptional(assetId, format);
		if (encodedAssetId !== void 0) result.caid = encodedAssetId;
		const encodedParams = this.assetParamsCodec.encodeOptional(assetParams, format);
		if (encodedParams && Object.keys(encodedParams).length > 0) result.apar = encodedParams;
		return result;
	}
	fromEncoded(value, format) {
		if (value.type === void 0) throw new Error("Transaction is missing type field");
		const type = normalizeWireString(value.type);
		const caid = value.caid;
		const apar = value.apar;
		if (type !== TransactionType.AssetConfig || caid === void 0 && !apar) return;
		return {
			assetId: bigIntCodec.decode(caid, format),
			...this.assetParamsCodec.decode(apar, format)
		};
	}
	isDefaultValue(value) {
		return value === void 0;
	}
};
var AppCallDataCodec = class extends Codec {
	appCallFieldsCodec = new ObjectModelCodec(AppCallTransactionFieldsMeta);
	defaultValue() {}
	encodeResourceReferences(appId, resourceReferences, format) {
		if (!resourceReferences || resourceReferences.length === 0) return void 0;
		const accessList = [];
		const ensure = (target) => {
			for (let idx = 0; idx < accessList.length; idx++) {
				const entry = accessList[idx];
				const matchesAddress = !entry.d && !target.address || entry.d && target.address && addressCodec.decode(entry.d, format).equals(target.address);
				const matchesAssetId = entry.s === void 0 && target.assetId === void 0 || entry.s !== void 0 && target.assetId !== void 0 && bigIntCodec.decode(entry.s, format) === target.assetId;
				const matchesAppId = entry.p === void 0 && target.appId === void 0 || entry.p !== void 0 && target.appId !== void 0 && bigIntCodec.decode(entry.p, format) === target.appId;
				if (matchesAddress && matchesAssetId && matchesAppId) return idx + 1;
			}
			if (target.address && !target.address.equals(Address.zeroAddress())) accessList.push({ d: addressCodec.encodeOptional(target.address, format) });
			if (target.assetId !== void 0) accessList.push({ s: bigIntCodec.encodeOptional(target.assetId, format) });
			if (target.appId !== void 0) accessList.push({ p: bigIntCodec.encodeOptional(target.appId, format) });
			return accessList.length;
		};
		for (const resourceRef of resourceReferences) {
			if (resourceRef.address || resourceRef.assetId !== void 0 || resourceRef.appId !== void 0) {
				ensure(resourceRef);
				continue;
			}
			if (resourceRef.holding) {
				const holding = resourceRef.holding;
				let addressIndex = 0;
				if (holding.address && !holding.address.equals(Address.zeroAddress())) addressIndex = ensure({ address: holding.address });
				const assetIndex = ensure({ assetId: holding.assetId });
				accessList.push({ h: {
					d: numberCodec.encodeOptional(addressIndex, format),
					s: numberCodec.encodeOptional(assetIndex, format)
				} });
				continue;
			}
			if (resourceRef.locals) {
				const locals = resourceRef.locals;
				let addressIndex = 0;
				if (locals.address && !locals.address.equals(Address.zeroAddress())) addressIndex = ensure({ address: locals.address });
				let appIndex = 0;
				if (locals.appId && locals.appId !== appId) appIndex = ensure({ appId: locals.appId });
				if (addressIndex !== 0 || appIndex !== 0) accessList.push({ l: {
					d: numberCodec.encodeOptional(addressIndex, format),
					p: numberCodec.encodeOptional(appIndex, format)
				} });
				continue;
			}
			if (resourceRef.box) {
				const box = resourceRef.box;
				let appIndex = 0;
				if (box.appId && box.appId !== appId) appIndex = ensure({ appId: box.appId });
				accessList.push({ b: {
					i: numberCodec.encodeOptional(appIndex, format),
					n: bytesCodec.encodeOptional(box.name, format)
				} });
				continue;
			}
		}
		return accessList.length > 0 ? accessList : void 0;
	}
	decodeResourceReferences(_wireResourceReferences, format) {
		if (!_wireResourceReferences || _wireResourceReferences.length === 0) return [];
		const result = [];
		const wireResourceReferences = _wireResourceReferences.map((ref) => normalizeWireObject(ref));
		for (const ref of wireResourceReferences) {
			const d = ref.d;
			const s = ref.s;
			const p = ref.p;
			const _h = ref.h;
			const h = _h ? normalizeWireObject(_h) : void 0;
			const _l = ref.l;
			const l = _l ? normalizeWireObject(_l) : void 0;
			const _b = ref.b;
			const b = _b ? normalizeWireObject(_b) : void 0;
			if (d) {
				result.push({ address: addressCodec.decode(d, format) });
				continue;
			}
			if (s !== void 0) {
				result.push({ assetId: bigIntCodec.decode(s, format) });
				continue;
			}
			if (p !== void 0) {
				result.push({ appId: bigIntCodec.decode(p, format) });
				continue;
			}
			if (h) {
				const addrIdx = h.d ?? 0;
				const assetIdx = h.s;
				if (assetIdx === void 0) throw new Error("Access list holding reference is missing asset index");
				const holdingAddress = addrIdx === 0 ? Address.zeroAddress() : addressCodec.decode(wireResourceReferences[addrIdx - 1].d, format);
				const holdingAssetId = wireResourceReferences[assetIdx - 1].s;
				result.push({ holding: {
					assetId: bigIntCodec.decode(holdingAssetId, format),
					address: holdingAddress
				} });
				continue;
			}
			if (l) {
				const addrIdx = l.d ?? 0;
				const appIdx = l.p ?? 0;
				const localsAddress = addrIdx === 0 ? Address.zeroAddress() : addressCodec.decode(wireResourceReferences[addrIdx - 1].d, format);
				const localsAppId = appIdx === 0 ? 0n : wireResourceReferences[appIdx - 1].p;
				result.push({ locals: {
					appId: bigIntCodec.decode(localsAppId, format),
					address: localsAddress
				} });
				continue;
			}
			if (b) {
				const boxAppIdx = b.i ?? 0;
				const name = b.n;
				if (!name) throw new Error("Access list box reference is missing name");
				const boxAppId = boxAppIdx === 0 ? 0n : wireResourceReferences[boxAppIdx - 1].p;
				result.push({ box: {
					appId: bigIntCodec.decode(boxAppId, format),
					name: bytesCodec.decode(name, format)
				} });
				continue;
			}
		}
		return result;
	}
	encodeBoxReferences(appId, appReferences, boxReferences, format) {
		if (!boxReferences || boxReferences.length === 0) return void 0;
		const appRefs = appReferences ?? [];
		return boxReferences.map((box) => {
			const isCurrentApp = box.appId === 0n || box.appId === appId;
			const index = isCurrentApp ? 0 : appRefs.indexOf(box.appId) + 1;
			if (index === 0 && !isCurrentApp) throw new Error(`Box ref with appId ${box.appId} not in appReferences`);
			return {
				i: numberCodec.encodeOptional(index, format),
				n: bytesCodec.encodeOptional(box.name, format)
			};
		});
	}
	decodeBoxReferences(appReferences, wireBoxReferences, format) {
		if (!wireBoxReferences || wireBoxReferences.length === 0) return [];
		return wireBoxReferences.map((_box) => {
			const box = normalizeWireObject(_box);
			const boxIndex = box.i;
			const boxName = box.n;
			const index = boxIndex ?? 0;
			let appId;
			if (index === 0) appId = 0n;
			else {
				const foreignAppId = appReferences?.[index - 1];
				if (foreignAppId === void 0) throw new Error(`Failed to find the app reference at index ${index - 1}`);
				appId = bigIntCodec.decode(foreignAppId, format);
			}
			return {
				appId,
				name: bytesCodec.decode(boxName, format)
			};
		});
	}
	toEncoded(value, format) {
		const result = {};
		if (!value) throw new Error("Transaction is not an app call");
		const encodedParams = this.appCallFieldsCodec.encodeOptional(value, format);
		if (encodedParams && Object.keys(encodedParams).length > 0) Object.assign(result, encodedParams);
		const wireBoxReferences = this.encodeBoxReferences(value.appId, value.appReferences, value.boxReferences, format);
		if (wireBoxReferences) result.apbx = wireBoxReferences;
		const wireAccessReferences = this.encodeResourceReferences(value.appId, value.accessReferences, format);
		if (wireAccessReferences) result.al = wireAccessReferences;
		return result;
	}
	fromEncoded(value, format) {
		if (value.type === void 0) throw new Error("Transaction is missing type field");
		if (normalizeWireString(value.type) !== TransactionType.AppCall) return;
		const appReferences = value.apfa;
		const wireBoxReferences = value.apbx;
		const wireAccessReferences = value.al;
		const boxReferences = this.decodeBoxReferences(appReferences?.map((ar) => bigIntCodec.decode(ar, format)), wireBoxReferences, format);
		const accessReferences = this.decodeResourceReferences(wireAccessReferences, format);
		return {
			...this.appCallFieldsCodec.decode(value, format),
			...accessReferences.length > 0 && { accessReferences },
			...boxReferences.length > 0 && { boxReferences }
		};
	}
	isDefaultValue(value) {
		return value === void 0;
	}
};
const PaymentTransactionFieldsMeta = {
	name: "PaymentTransactionFields",
	kind: "object",
	fields: [
		{
			name: "amount",
			wireKey: "amt",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "receiver",
			wireKey: "rcv",
			optional: false,
			codec: addressCodec
		},
		{
			name: "closeRemainderTo",
			wireKey: "close",
			optional: true,
			codec: addressCodec
		}
	]
};
const AssetTransferTransactionFieldsMeta = {
	name: "AssetTransferTransactionFields",
	kind: "object",
	fields: [
		{
			name: "assetId",
			wireKey: "xaid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "amount",
			wireKey: "aamt",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "receiver",
			wireKey: "arcv",
			optional: false,
			codec: addressCodec
		},
		{
			name: "assetSender",
			wireKey: "asnd",
			optional: true,
			codec: addressCodec
		},
		{
			name: "closeRemainderTo",
			wireKey: "aclose",
			optional: true,
			codec: addressCodec
		}
	]
};
const AssetFreezeTransactionFieldsMeta = {
	name: "AssetFreezeTransactionFields",
	kind: "object",
	fields: [
		{
			name: "assetId",
			wireKey: "faid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "freezeTarget",
			wireKey: "fadd",
			optional: false,
			codec: addressCodec
		},
		{
			name: "frozen",
			wireKey: "afrz",
			optional: false,
			codec: booleanCodec
		}
	]
};
const KeyRegistrationTransactionFieldsMeta = {
	name: "KeyRegistrationTransactionFields",
	kind: "object",
	fields: [
		{
			name: "voteKey",
			wireKey: "votekey",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "selectionKey",
			wireKey: "selkey",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "stateProofKey",
			wireKey: "sprfkey",
			optional: true,
			codec: fixedBytes64Codec
		},
		{
			name: "voteFirst",
			wireKey: "votefst",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteLast",
			wireKey: "votelst",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "voteKeyDilution",
			wireKey: "votekd",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "nonParticipation",
			wireKey: "nonpart",
			optional: true,
			codec: booleanCodec
		}
	]
};
const AssetParamsMeta = {
	name: "AssetParams",
	kind: "object",
	fields: [
		{
			name: "total",
			wireKey: "t",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "decimals",
			wireKey: "dc",
			optional: true,
			codec: numberCodec
		},
		{
			name: "defaultFrozen",
			wireKey: "df",
			optional: true,
			codec: booleanCodec
		},
		{
			name: "unitName",
			wireKey: "un",
			optional: true,
			codec: stringCodec
		},
		{
			name: "assetName",
			wireKey: "an",
			optional: true,
			codec: stringCodec
		},
		{
			name: "url",
			wireKey: "au",
			optional: true,
			codec: stringCodec
		},
		{
			name: "metadataHash",
			wireKey: "am",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "manager",
			wireKey: "m",
			optional: true,
			codec: addressCodec
		},
		{
			name: "reserve",
			wireKey: "r",
			optional: true,
			codec: addressCodec
		},
		{
			name: "freeze",
			wireKey: "f",
			optional: true,
			codec: addressCodec
		},
		{
			name: "clawback",
			wireKey: "c",
			optional: true,
			codec: addressCodec
		}
	]
};
const StateSchemaMeta = {
	name: "StateSchema",
	kind: "object",
	fields: [{
		name: "numUints",
		wireKey: "nui",
		optional: false,
		codec: numberCodec
	}, {
		name: "numByteSlices",
		wireKey: "nbs",
		optional: false,
		codec: numberCodec
	}]
};
const AppCallTransactionFieldsMeta = {
	name: "AppCallTransactionFields",
	kind: "object",
	fields: [
		{
			name: "appId",
			wireKey: "apid",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "onComplete",
			wireKey: "apan",
			optional: false,
			codec: numberCodec
		},
		{
			name: "approvalProgram",
			wireKey: "apap",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "clearStateProgram",
			wireKey: "apsu",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "globalStateSchema",
			wireKey: "apgs",
			optional: true,
			codec: new ObjectModelCodec(StateSchemaMeta)
		},
		{
			name: "localStateSchema",
			wireKey: "apls",
			optional: true,
			codec: new ObjectModelCodec(StateSchemaMeta)
		},
		{
			name: "extraProgramPages",
			wireKey: "apep",
			optional: true,
			codec: numberCodec
		},
		{
			name: "args",
			wireKey: "apaa",
			optional: true,
			codec: bytesArrayCodec
		},
		{
			name: "accountReferences",
			wireKey: "apat",
			optional: true,
			codec: addressArrayCodec
		},
		{
			name: "appReferences",
			wireKey: "apfa",
			optional: true,
			codec: bigIntArrayCodec
		},
		{
			name: "assetReferences",
			wireKey: "apas",
			optional: true,
			codec: bigIntArrayCodec
		},
		{
			name: "rejectVersion",
			wireKey: "aprv",
			optional: true,
			codec: numberCodec
		}
	]
};
const MerkleArrayProofMeta = {
	name: "MerkleArrayProof",
	kind: "object",
	fields: [
		{
			name: "path",
			wireKey: "pth",
			optional: false,
			codec: bytesArrayCodec
		},
		{
			name: "hashFactory",
			wireKey: "hsh",
			optional: false,
			codec: new ObjectModelCodec({
				name: "HashFactory",
				kind: "object",
				fields: [{
					name: "hashType",
					wireKey: "t",
					optional: false,
					codec: numberCodec
				}]
			})
		},
		{
			name: "treeDepth",
			wireKey: "td",
			optional: false,
			codec: numberCodec
		}
	]
};
const FalconVerifierMeta = {
	name: "FalconVerifier",
	kind: "object",
	fields: [{
		name: "publicKey",
		wireKey: "k",
		optional: false,
		codec: fixedBytes1793Codec
	}]
};
const SigslotCommitMeta = {
	name: "SigslotCommit",
	kind: "object",
	fields: [{
		name: "sig",
		wireKey: "s",
		optional: false,
		codec: new ObjectModelCodec({
			name: "FalconSignatureStruct",
			kind: "object",
			fields: [
				{
					name: "signature",
					wireKey: "sig",
					optional: false,
					codec: bytesCodec
				},
				{
					name: "vectorCommitmentIndex",
					wireKey: "idx",
					optional: false,
					codec: bigIntCodec
				},
				{
					name: "proof",
					wireKey: "prf",
					optional: false,
					codec: new ObjectModelCodec(MerkleArrayProofMeta)
				},
				{
					name: "verifyingKey",
					wireKey: "vkey",
					optional: false,
					codec: new ObjectModelCodec(FalconVerifierMeta)
				}
			]
		})
	}, {
		name: "lowerSigWeight",
		wireKey: "l",
		optional: false,
		codec: bigIntCodec
	}]
};
const ParticipantMeta = {
	name: "Participant",
	kind: "object",
	fields: [{
		name: "verifier",
		wireKey: "p",
		optional: false,
		codec: new ObjectModelCodec({
			name: "MerkleSignatureVerifier",
			kind: "object",
			fields: [{
				name: "commitment",
				wireKey: "cmt",
				optional: false,
				codec: fixedBytes64Codec
			}, {
				name: "keyLifetime",
				wireKey: "lf",
				optional: false,
				codec: bigIntCodec
			}]
		})
	}, {
		name: "weight",
		wireKey: "w",
		optional: false,
		codec: bigIntCodec
	}]
};
const RevealMeta = {
	name: "Reveal",
	kind: "object",
	fields: [{
		name: "sigslot",
		wireKey: "s",
		optional: false,
		codec: new ObjectModelCodec(SigslotCommitMeta)
	}, {
		name: "participant",
		wireKey: "p",
		optional: false,
		codec: new ObjectModelCodec(ParticipantMeta)
	}]
};
const StateProofMeta = {
	name: "StateProof",
	kind: "object",
	fields: [
		{
			name: "sigCommit",
			wireKey: "c",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "signedWeight",
			wireKey: "w",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "sigProofs",
			wireKey: "S",
			optional: false,
			codec: new ObjectModelCodec(MerkleArrayProofMeta)
		},
		{
			name: "partProofs",
			wireKey: "P",
			optional: false,
			codec: new ObjectModelCodec(MerkleArrayProofMeta)
		},
		{
			name: "merkleSignatureSaltVersion",
			wireKey: "v",
			optional: false,
			codec: numberCodec
		},
		{
			name: "reveals",
			wireKey: "r",
			optional: false,
			codec: new MapCodec(bigIntCodec, new ObjectModelCodec(RevealMeta))
		},
		{
			name: "positionsToReveal",
			wireKey: "pr",
			optional: false,
			codec: bigIntArrayCodec
		}
	]
};
const StateProofMessageMeta = {
	name: "StateProofMessage",
	kind: "object",
	fields: [
		{
			name: "blockHeadersCommitment",
			wireKey: "b",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "votersCommitment",
			wireKey: "v",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "lnProvenWeight",
			wireKey: "P",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "firstAttestedRound",
			wireKey: "f",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "lastAttestedRound",
			wireKey: "l",
			optional: false,
			codec: bigIntCodec
		}
	]
};
const StateProofTransactionFieldsMeta = {
	name: "StateProofTransactionFields",
	kind: "object",
	fields: [
		{
			name: "stateProofType",
			wireKey: "sptype",
			optional: false,
			codec: numberCodec
		},
		{
			name: "stateProof",
			wireKey: "sp",
			optional: true,
			codec: new ObjectModelCodec(StateProofMeta)
		},
		{
			name: "message",
			wireKey: "spmsg",
			optional: true,
			codec: new ObjectModelCodec(StateProofMessageMeta)
		}
	]
};
const HeartbeatTransactionFieldsMeta = {
	name: "HeartbeatTransactionFields",
	kind: "object",
	fields: [
		{
			name: "address",
			wireKey: "a",
			optional: false,
			codec: addressCodec
		},
		{
			name: "proof",
			wireKey: "prf",
			optional: false,
			codec: new ObjectModelCodec({
				name: "HeartbeatProof",
				kind: "object",
				fields: [
					{
						name: "sig",
						wireKey: "s",
						optional: false,
						codec: fixedBytes64Codec
					},
					{
						name: "pk",
						wireKey: "p",
						optional: false,
						codec: fixedBytes32Codec
					},
					{
						name: "pk2",
						wireKey: "p2",
						optional: false,
						codec: fixedBytes32Codec
					},
					{
						name: "pk1Sig",
						wireKey: "p1s",
						optional: false,
						codec: fixedBytes64Codec
					},
					{
						name: "pk2Sig",
						wireKey: "p2s",
						optional: false,
						codec: fixedBytes64Codec
					}
				]
			})
		},
		{
			name: "seed",
			wireKey: "sd",
			optional: false,
			codec: bytesCodec
		},
		{
			name: "voteId",
			wireKey: "vid",
			optional: false,
			codec: fixedBytes32Codec
		},
		{
			name: "keyDilution",
			wireKey: "kd",
			optional: false,
			codec: bigIntCodec
		}
	]
};
/**
* Metadata for TransactionParams
*/
const TransactionParamsMeta = {
	name: "TransactionParams",
	kind: "object",
	fields: [
		{
			name: "type",
			wireKey: "type",
			optional: false,
			codec: new TransactionTypeCodec()
		},
		{
			name: "sender",
			wireKey: "snd",
			optional: false,
			codec: addressCodec
		},
		{
			name: "fee",
			wireKey: "fee",
			optional: true,
			codec: bigIntCodec
		},
		{
			name: "firstValid",
			wireKey: "fv",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "lastValid",
			wireKey: "lv",
			optional: false,
			codec: bigIntCodec
		},
		{
			name: "genesisHash",
			wireKey: "gh",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "genesisId",
			wireKey: "gen",
			optional: true,
			codec: stringCodec
		},
		{
			name: "note",
			wireKey: "note",
			optional: true,
			codec: bytesCodec
		},
		{
			name: "rekeyTo",
			wireKey: "rekey",
			optional: true,
			codec: addressCodec
		},
		{
			name: "lease",
			wireKey: "lx",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "group",
			wireKey: "grp",
			optional: true,
			codec: fixedBytes32Codec
		},
		{
			name: "payment",
			flattened: true,
			optional: true,
			codec: new TransactionDataCodec(TransactionType.Payment, PaymentTransactionFieldsMeta)
		},
		{
			name: "assetTransfer",
			flattened: true,
			optional: true,
			codec: new TransactionDataCodec(TransactionType.AssetTransfer, AssetTransferTransactionFieldsMeta)
		},
		{
			name: "assetFreeze",
			flattened: true,
			optional: true,
			codec: new TransactionDataCodec(TransactionType.AssetFreeze, AssetFreezeTransactionFieldsMeta)
		},
		{
			name: "keyRegistration",
			flattened: true,
			optional: true,
			codec: new TransactionDataCodec(TransactionType.KeyRegistration, KeyRegistrationTransactionFieldsMeta)
		},
		{
			name: "assetConfig",
			flattened: true,
			optional: true,
			codec: new AssetConfigDataCodec()
		},
		{
			name: "heartbeat",
			wireKey: "hb",
			optional: true,
			codec: new ObjectModelCodec(HeartbeatTransactionFieldsMeta)
		},
		{
			name: "appCall",
			flattened: true,
			optional: true,
			codec: new AppCallDataCodec()
		},
		{
			name: "stateProof",
			flattened: true,
			optional: true,
			codec: new TransactionDataCodec(TransactionType.StateProof, StateProofTransactionFieldsMeta)
		}
	]
};
const transactionParamsCodec = new ObjectModelCodec(TransactionParamsMeta);

//#endregion
export { TransactionParamsMeta, transactionParamsCodec };
