// @algorandfoundation/algokit-utils/packages/transact/src/transactions/transaction.js
const require_runtime = require('../../../../_virtual/_rolldown/runtime.js');
const require_array = require('../../../common/src/array.js');
const require_constants = require('../../../common/src/constants.js');
const require_address = require('../../../common/src/address.js');
const require_codec = require('../../../common/src/codecs/codec.js');
const require_msgpack = require('../../../common/src/msgpack.js');
const require_hash = require('../../../crypto/src/hash.js');
const require_common = require('./common.js');
const require_app_call = require('./app-call.js');
const require_asset_config = require('./asset-config.js');
const require_asset_freeze = require('./asset-freeze.js');
const require_asset_transfer = require('./asset-transfer.js');
const require_key_registration = require('./key-registration.js');
const require_transaction_type = require('./transaction-type.js');
const require_transaction_meta = require('./transaction-meta.js');
let hi_base32 = require("hi-base32");
hi_base32 = require_runtime.__toESM(hi_base32);

//#region packages/transact/src/transactions/transaction.ts
/** Symbol used for instanceof checks across packages (CJS/ESM) */
const TXN_SYMBOL = Symbol.for("algokit_transact:Transaction");
/**
* Represents a complete Algorand transaction.
*/
var Transaction = class {
	/** @internal */
	[TXN_SYMBOL];
	/**
	* The type of transaction
	*/
	type;
	/**
	* The account that authorized the transaction.
	*
	* Fees are deducted from this account.
	*/
	sender;
	/**
	* Optional transaction fee in microALGO.
	*
	* When not set, the fee will be interpreted as 0 by the network.
	*/
	fee;
	/**
	* First round for when the transaction is valid.
	*/
	firstValid;
	/**
	* Last round for when the transaction is valid.
	*
	* After this round, the transaction will be expired.
	*/
	lastValid;
	/**
	* Hash of the genesis block of the network.
	*
	* Used to identify which network the transaction is for.
	*/
	genesisHash;
	/**
	* Genesis ID of the network.
	*
	* A human-readable string used alongside genesis hash to identify the network.
	*/
	genesisId;
	/**
	* Optional user-defined note field.
	*
	* Can contain arbitrary data up to 1KB in size.
	*/
	note;
	/**
	* Optional authorized account for future transactions.
	*
	* If set, only this account will be used for transaction authorization going forward.
	* Reverting back control to the original address must be done by setting this field to
	* the original address.
	*/
	rekeyTo;
	/**
	* Optional lease value to enforce mutual transaction exclusion.
	*
	* When a transaction with a non-empty lease field is confirmed, the lease is acquired.
	* A lease X is acquired by the sender, generating the (sender, X) lease.
	* The lease is kept active until the last_valid round of the transaction has elapsed.
	* No other transaction sent by the same sender can be confirmed until the lease expires.
	*/
	lease;
	/**
	* Optional group ID for atomic transaction grouping.
	*
	* Transactions with the same group ID must execute together or not at all.
	*/
	group;
	/**
	* Payment specific fields
	*/
	payment;
	/**
	* Asset transfer specific fields
	*/
	assetTransfer;
	/**
	* Asset config specific fields
	*/
	assetConfig;
	/**
	* App call specific fields
	*/
	appCall;
	/**
	* Key registration specific fields
	*/
	keyRegistration;
	/**
	* Asset freeze specific fields
	*/
	assetFreeze;
	/**
	* Heartbeat specific fields
	*/
	heartbeat;
	/**
	* State proof specific fields
	*/
	stateProof;
	constructor(params) {
		this[TXN_SYMBOL] = true;
		this.type = params.type;
		this.sender = params.sender;
		this.fee = params.fee;
		this.firstValid = params.firstValid;
		this.lastValid = params.lastValid;
		this.genesisHash = params.genesisHash;
		this.genesisId = params.genesisId;
		this.note = params.note;
		this.rekeyTo = params.rekeyTo;
		this.lease = params.lease;
		this.group = params.group;
		this.payment = params.payment;
		this.assetTransfer = params.assetTransfer;
		this.assetConfig = params.assetConfig;
		this.appCall = params.appCall;
		this.keyRegistration = params.keyRegistration;
		this.assetFreeze = params.assetFreeze;
		this.heartbeat = params.heartbeat;
		this.stateProof = params.stateProof;
	}
	rawTxId() {
		if (this.genesisHash === void 0) throw new Error("Cannot compute transaction id without genesis hash");
		return require_hash.hash(encodeTransaction(this));
	}
	/**
	* Get the transaction ID as a base32-encoded string.
	*/
	txId() {
		const rawTxId = this.rawTxId();
		return hi_base32.default.encode(rawTxId).slice(0, require_constants.TRANSACTION_ID_LENGTH);
	}
};
Object.defineProperty(Transaction, Symbol.hasInstance, { value: function(obj) {
	if (obj instanceof Object && Object.getPrototypeOf(obj) === Transaction.prototype) return true;
	return Boolean(obj && typeof obj === "object" && TXN_SYMBOL in obj && obj[TXN_SYMBOL]);
} });
/**
* Codec for Transaction class.
* Handles encoding/decoding between Transaction class instances and wire format.
*/
var TransactionCodec = class extends require_codec.Codec {
	defaultValue() {
		return new Transaction({
			type: require_transaction_type.TransactionType.Unknown,
			sender: require_address.Address.zeroAddress(),
			firstValid: 0n,
			lastValid: 0n
		});
	}
	toEncoded(value, format) {
		return require_transaction_meta.transactionParamsCodec.encode({ ...value }, format);
	}
	fromEncoded(value, format) {
		return new Transaction(require_transaction_meta.transactionParamsCodec.decode(value, format));
	}
	isDefaultValue(_) {
		return false;
	}
};
const transactionCodec = new TransactionCodec();
/**
* Get the transaction type from the encoded transaction.
* This is particularly useful when decoding a transaction that has an unknown type
*/
function getEncodedTransactionType(encoded_transaction) {
	return decodeTransaction(encoded_transaction).type;
}
/**
* Encode the transaction with the domain separation (e.g. "TX") prefix
*
* @param transaction - The transaction to encode
* @returns The MsgPack encoded bytes or an error if encoding fails.
*/
function encodeTransaction(transaction) {
	const rawBytes = encodeTransactionRaw(transaction);
	return require_array.concatArrays(new TextEncoder().encode(require_constants.TRANSACTION_DOMAIN_SEPARATOR), rawBytes);
}
/**
* Encode transactions with the domain separation (e.g. "TX") prefix
*
* @param transactions - A collection of transactions to encode
* @returns A collection of MsgPack encoded bytes or an error if encoding fails.
*/
function encodeTransactions(transactions) {
	return transactions.map((tx) => encodeTransaction(tx));
}
/**
* Validate a transaction
*/
function validateTransaction(transaction) {
	if ([
		transaction.payment,
		transaction.assetTransfer,
		transaction.assetConfig,
		transaction.appCall,
		transaction.keyRegistration,
		transaction.assetFreeze,
		transaction.heartbeat,
		transaction.stateProof
	].filter((field) => field !== void 0).length > 1) throw new Error("Multiple transaction type specific fields set");
	let typeName = "Transaction";
	const errors = new Array();
	if (transaction.assetTransfer) {
		typeName = "Asset transfer";
		errors.push(...require_asset_transfer.validateAssetTransferTransaction(transaction.assetTransfer));
	} else if (transaction.assetConfig) {
		typeName = "Asset config";
		errors.push(...require_asset_config.validateAssetConfigTransaction(transaction.assetConfig));
	} else if (transaction.appCall) {
		typeName = "App call";
		errors.push(...require_app_call.validateAppCallTransaction(transaction.appCall));
	} else if (transaction.keyRegistration) {
		typeName = "Key registration";
		errors.push(...require_key_registration.validateKeyRegistrationTransaction(transaction.keyRegistration));
	} else if (transaction.assetFreeze) {
		typeName = "Asset freeze";
		errors.push(...require_asset_freeze.validateAssetFreezeTransaction(transaction.assetFreeze));
	}
	if (errors.length > 0) {
		const errorMessages = errors.map((e) => require_common.getValidationErrorMessage(e));
		throw new Error(`${typeName} validation failed: ${errorMessages.join("\n")}`);
	}
}
/**
* Encode the transaction without the domain separation (e.g. "TX") prefix
* This is useful for encoding the transaction for signing with tools that automatically add "TX" prefix to the transaction bytes.
*/
function encodeTransactionRaw(transaction) {
	return require_msgpack.encodeMsgpack(transactionCodec.encode(transaction, "msgpack"));
}
/**
* Decodes MsgPack bytes into a transaction.
*
* # Parameters
* * `encoded_transaction` - MsgPack encoded bytes representing a transaction.
*
* # Returns
* A decoded transaction or an error if decoding fails.
*/
function decodeTransaction(encoded_transaction) {
	if (encoded_transaction.length === 0) throw new Error("attempted to decode 0 bytes");
	const prefixBytes = new TextEncoder().encode(require_constants.TRANSACTION_DOMAIN_SEPARATOR);
	let hasPrefix = true;
	if (encoded_transaction.length < prefixBytes.length) hasPrefix = false;
	else for (let i = 0; i < prefixBytes.length; i++) if (encoded_transaction[i] !== prefixBytes[i]) {
		hasPrefix = false;
		break;
	}
	const decodedData = require_msgpack.decodeMsgpack(hasPrefix ? encoded_transaction.slice(prefixBytes.length) : encoded_transaction);
	return transactionCodec.decode(decodedData, "msgpack");
}
/**
* Decodes a collection of MsgPack bytes into a transaction collection.
*
* # Parameters
* * `encoded_transaction` - A collection of MsgPack encoded bytes, each representing a transaction.
*
* # Returns
* A collection of decoded transactions or an error if decoding fails.
*/
function decodeTransactions(encoded_transactions) {
	return encoded_transactions.map((et) => decodeTransaction(et));
}
/**
* Return the size of the transaction in bytes as if it was already signed and encoded.
* This is useful for estimating the fee for the transaction.
*/
function estimateTransactionSize(transaction) {
	const encoded = encodeTransactionRaw(transaction);
	return BigInt(encoded.length + require_constants.SIGNATURE_ENCODING_INCR);
}
/**
* Groups a collection of transactions by calculating and assigning the group to each transaction.
*/
function groupTransactions(transactions) {
	const group = computeGroup(transactions);
	return transactions.map((tx) => new Transaction({
		...tx,
		group
	}));
}
function assignFee(transaction, feeParams) {
	const fee = calculateFee(transaction, feeParams);
	return new Transaction({
		...transaction,
		fee
	});
}
function computeGroup(transactions) {
	if (transactions.length === 0) throw new Error("Transaction group size cannot be 0");
	if (transactions.length > require_constants.MAX_TRANSACTION_GROUP_SIZE) throw new Error(`Transaction group size exceeds the max limit of ${require_constants.MAX_TRANSACTION_GROUP_SIZE}`);
	const txHashes = transactions.map((tx) => {
		if (tx.group) throw new Error("Transactions must not already be grouped");
		return require_hash.hash(encodeTransaction(tx));
	});
	return require_hash.hash(require_array.concatArrays(new TextEncoder().encode(require_constants.TRANSACTION_GROUP_DOMAIN_SEPARATOR), require_msgpack.encodeMsgpack({ txlist: txHashes })));
}
function calculateFee(transaction, feeParams) {
	let calculatedFee = 0n;
	if (feeParams.feePerByte > 0n) {
		const estimatedSize = estimateTransactionSize(transaction);
		calculatedFee = feeParams.feePerByte * BigInt(estimatedSize);
	}
	if (calculatedFee < feeParams.minFee) calculatedFee = feeParams.minFee;
	if (feeParams.extraFee) calculatedFee += feeParams.extraFee;
	if (feeParams.maxFee && calculatedFee > feeParams.maxFee) throw new Error(`Transaction fee ${calculatedFee} µALGO is greater than maxFee ${feeParams.maxFee} µALGO`);
	return calculatedFee;
}

//#endregion
exports.Transaction = Transaction;
exports.assignFee = assignFee;
exports.calculateFee = calculateFee;
exports.decodeTransaction = decodeTransaction;
exports.decodeTransactions = decodeTransactions;
exports.encodeTransaction = encodeTransaction;
exports.encodeTransactionRaw = encodeTransactionRaw;
exports.encodeTransactions = encodeTransactions;
exports.estimateTransactionSize = estimateTransactionSize;
exports.getEncodedTransactionType = getEncodedTransactionType;
exports.groupTransactions = groupTransactions;
exports.transactionCodec = transactionCodec;
exports.validateTransaction = validateTransaction;
