// @x402/avm/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ALGOKIT_SIGNER: () => ALGOKIT_SIGNER,
  ALGORAND_ADDRESS_LENGTH: () => import_common3.ALGORAND_ADDRESS_LENGTH,
  ALGORAND_MAINNET_CAIP2: () => ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH: () => ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_MIN_TX_FEE: () => import_amount2.ALGORAND_MIN_TX_FEE,
  ALGORAND_TESTNET_CAIP2: () => ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH: () => ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS: () => CAIP2_NETWORKS,
  ExactAvmScheme: () => ExactAvmScheme,
  MAX_REASONABLE_FEE_PER_TXN: () => MAX_REASONABLE_FEE_PER_TXN,
  USDC_CONFIG: () => USDC_CONFIG,
  USDC_DECIMALS: () => USDC_DECIMALS,
  USDC_MAINNET_ASA_ID: () => USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID: () => USDC_TESTNET_ASA_ID,
  convertFromTokenAmount: () => convertFromTokenAmount,
  convertToTokenAmount: () => import_utils.convertToTokenAmount,
  decodeSignedTransaction: () => decodeSignedTransaction,
  decodeTransaction: () => decodeTransaction,
  decodeUnsignedTransaction: () => decodeUnsignedTransaction,
  encodeTransaction: () => encodeTransaction,
  getAlgokitSigner: () => getAlgokitSigner,
  getGenesisHashFromTransaction: () => getGenesisHashFromTransaction,
  getNetworkFromCaip2: () => getNetworkFromCaip2,
  getSenderFromTransaction: () => getSenderFromTransaction,
  getTransactionId: () => getTransactionId,
  hasSignature: () => hasSignature,
  isAlgorandNetwork: () => isAlgorandNetwork,
  isAvmSignerWallet: () => isAvmSignerWallet,
  isExactAvmPayload: () => isExactAvmPayload,
  isTestnetNetwork: () => isTestnetNetwork,
  isValidAlgorandAddress: () => isValidAlgorandAddress,
  maxReasonableGroupFee: () => maxReasonableGroupFee,
  toClientAvmSigner: () => toClientAvmSigner,
  toFacilitatorAvmSigner: () => toFacilitatorAvmSigner,
  validateGroupId: () => validateGroupId
});
module.exports = __toCommonJS(src_exports);

// src/exact/client/scheme.ts
var import_algorand_client = require("@algorandfoundation/algokit-utils/algorand-client");
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_amount = require("@algorandfoundation/algokit-utils/amount");

// src/utils.ts
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var CAIP2_NETWORKS = [ALGORAND_MAINNET_CAIP2, ALGORAND_TESTNET_CAIP2];
var ALGORAND_MAINNET_GENESIS_HASH = "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_GENESIS_HASH = "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var MAX_REASONABLE_FEE_PER_TXN = 5e3;
function maxReasonableGroupFee(groupSize) {
  return MAX_REASONABLE_FEE_PER_TXN * groupSize;
}

// src/utils.ts
var import_utils = require("@x402/core/utils");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function decodeSignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return (0, import_transact.decodeSignedTransaction)(bytes);
}
function decodeUnsignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return (0, import_transact.decodeTransaction)(bytes);
}
function isValidAlgorandAddress(address) {
  return (0, import_common.isValidAddress)(address);
}
function getSenderFromTransaction(txnBytes, isSigned = true) {
  if (isSigned) {
    const signedTxn = (0, import_transact.decodeSignedTransaction)(txnBytes);
    return signedTxn.txn.sender.toString();
  }
  const txn = (0, import_transact.decodeTransaction)(txnBytes);
  return txn.sender.toString();
}
function convertFromTokenAmount(atomicAmount, decimals) {
  const amount = BigInt(atomicAmount);
  const divisor = BigInt(10 ** decimals);
  const intPart = amount / divisor;
  const decPart = amount % divisor;
  if (decPart === BigInt(0)) {
    return intPart.toString();
  }
  const decStr = decPart.toString().padStart(decimals, "0");
  const trimmedDec = decStr.replace(/0+$/, "");
  return `${intPart}.${trimmedDec}`;
}
function getNetworkFromCaip2(caip2) {
  if (!caip2.startsWith("algorand:")) {
    return null;
  }
  const genesisHash = caip2.slice("algorand:".length);
  if (genesisHash === ALGORAND_MAINNET_GENESIS_HASH) {
    return "mainnet";
  }
  if (genesisHash === ALGORAND_TESTNET_GENESIS_HASH) {
    return "testnet";
  }
  return null;
}
function isAlgorandNetwork(network) {
  return network.startsWith("algorand:");
}
function isTestnetNetwork(network) {
  return network === ALGORAND_TESTNET_CAIP2;
}
function getGenesisHashFromTransaction(txn) {
  if (!txn.genesisHash) {
    throw new Error("Transaction does not have a genesis hash");
  }
  return Buffer.from(txn.genesisHash).toString("base64");
}
function validateGroupId(txns) {
  if (txns.length <= 1) {
    return true;
  }
  let expectedGroupId = null;
  for (const txnBytes of txns) {
    const txn = (0, import_transact.decodeTransaction)(txnBytes);
    const groupId = txn.group ? Buffer.from(txn.group).toString("base64") : null;
    if (expectedGroupId === null) {
      expectedGroupId = groupId;
    } else if (groupId !== expectedGroupId) {
      return false;
    }
  }
  return true;
}
function getTransactionId(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.txn.txId();
}
function hasSignature(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

// src/exact/client/scheme.ts
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   *
   * Constructs an atomic transaction group with:
   * - Optional fee payer transaction (if feePayer specified in requirements.extra)
   * - ASA transfer transaction to payTo address
   *
   * Uses TransactionComposer for automatic suggested params, group ID assignment,
   * and fee pooling. For sponsored (gasless) transactions, exact fees are calculated
   * from actual encoded transaction sizes using the protocol fee formula:
   *   fee = max(fee_per_byte × txn_size_in_bytes, min_fee)
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const { amount, asset, payTo, network, extra } = paymentRequirements;
    const algorandClient = this.getAlgorandClient(network);
    const assetId = this.getAssetId(asset, network);
    const feePayer = extra?.feePayer;
    let paymentIndex = 0;
    const emptySigner = (0, import_transact2.makeEmptyTransactionSigner)();
    const composer = algorandClient.newGroup();
    if (feePayer) {
      composer.addPayment({
        sender: feePayer,
        receiver: feePayer,
        amount: (0, import_amount.microAlgo)(0),
        note: `x402-fee-payer-${Date.now()}`,
        signer: emptySigner
      });
      paymentIndex = 1;
    }
    composer.addAssetTransfer({
      sender: this.signer.address,
      receiver: payTo,
      assetId: BigInt(assetId),
      amount: BigInt(amount),
      staticFee: feePayer ? (0, import_amount.microAlgo)(0) : void 0,
      // 0 fee when fee payer covers
      note: `x402-payment-v${x402Version}-${Date.now()}`,
      signer: emptySigner
    });
    const built = await composer.build();
    let transactions = built.transactions.map((tws) => tws.txn);
    if (feePayer) {
      const sp = await algorandClient.getSuggestedParams();
      const feePerByte = Number(sp.fee);
      const minFee = Number(sp.minFee);
      let totalGroupFee = BigInt(0);
      for (const txn of transactions) {
        const txnSize = (0, import_transact2.encodeTransactionRaw)(txn).length;
        const txnFee = feePerByte > 0 ? Math.max(feePerByte * txnSize, minFee) : minFee;
        totalGroupFee += BigInt(txnFee);
      }
      const ungrouped = transactions.map((txn, i) => {
        const fee = i === 0 ? totalGroupFee : BigInt(0);
        return new import_transact2.Transaction({ ...txn, fee, group: void 0 });
      });
      transactions = (0, import_transact2.groupTransactions)(ungrouped);
    }
    const encodedTxns = transactions.map((txn) => (0, import_transact2.encodeTransactionRaw)(txn));
    const clientIndexes = transactions.map((txn, i) => txn.sender.toString() === this.signer.address ? i : -1).filter((i) => i !== -1);
    const signedTxns = await this.signer.signTransactions(encodedTxns, clientIndexes);
    const paymentGroup = encodedTxns.map((txnBytes, i) => {
      const signedTxn = signedTxns[i];
      if (signedTxn) {
        return encodeTransaction(signedTxn);
      }
      return encodeTransaction(txnBytes);
    });
    const payload = {
      paymentGroup,
      paymentIndex
    };
    return {
      x402Version,
      payload
    };
  }
  /**
   * Creates or retrieves an AlgorandClient for the given network.
   *
   * @param network - Network identifier (CAIP-2 format)
   * @returns AlgorandClient instance
   */
  getAlgorandClient(network) {
    if (this.config?.algorandClient) {
      return this.config.algorandClient;
    }
    if (this.config?.algodUrl) {
      return import_algorand_client.AlgorandClient.fromConfig({
        algodConfig: {
          server: this.config.algodUrl,
          token: this.config.algodToken ?? ""
        }
      });
    }
    return isTestnetNetwork(network) ? import_algorand_client.AlgorandClient.testNet() : import_algorand_client.AlgorandClient.mainNet();
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - Asset identifier from requirements
   * @param network - Network identifier
   * @returns Asset ID as string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

// src/signer.ts
var import_algorand_client2 = require("@algorandfoundation/algokit-utils/algorand-client");
var import_crypto = require("@algorandfoundation/algokit-utils/crypto");
var import_transact3 = require("@algorandfoundation/algokit-utils/transact");
var import_transaction = require("@algorandfoundation/algokit-utils/transaction");
var ALGOKIT_SIGNER = Symbol("algokit-signer");
function isAvmSignerWallet(wallet) {
  return typeof wallet === "object" && wallet !== null && "address" in wallet && typeof wallet.address === "string" && "signTransactions" in wallet && typeof wallet.signTransactions === "function";
}
function decodePrivateKey(privateKeyBase64) {
  const secretKey = Buffer.from(privateKeyBase64, "base64");
  if (secretKey.length !== 64) {
    throw new Error(
      "AVM private key must be a Base64-encoded 64-byte key (32-byte seed + 32-byte public key)"
    );
  }
  const seed = secretKey.subarray(0, 32);
  return (0, import_crypto.ed25519Generator)(seed);
}
function toClientAvmSigner(privateKeyBase64) {
  const { ed25519Pubkey, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  const algokitSigners = (0, import_transact3.generateAddressWithSigners)({ ed25519Pubkey, rawEd25519Signer });
  const address = algokitSigners.addr.toString();
  const signer = {
    address,
    signTransactions: async (txns, indexesToSign) => {
      return Promise.all(
        txns.map(async (txn, i) => {
          if (indexesToSign && !indexesToSign.includes(i)) return null;
          const decoded = (0, import_transact3.decodeTransaction)(txn);
          const signedBytes = await algokitSigners.signer([decoded], [0]);
          return signedBytes[0];
        })
      );
    }
  };
  Object.defineProperty(signer, ALGOKIT_SIGNER, {
    value: algokitSigners,
    enumerable: false,
    writable: false
  });
  return signer;
}
function getAlgokitSigner(signer) {
  const internal = signer[ALGOKIT_SIGNER];
  if (internal && "addr" in internal && "signer" in internal) {
    return { addr: internal.addr, signer: internal.signer };
  }
  return null;
}
function isTestnet(network) {
  return network === ALGORAND_TESTNET_CAIP2;
}
function toFacilitatorAvmSigner(privateKeyBase64, config) {
  const { ed25519Pubkey, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  const algokitSigners = (0, import_transact3.generateAddressWithSigners)({ ed25519Pubkey, rawEd25519Signer });
  const address = algokitSigners.addr.toString();
  const getAlgorandClientForNetwork = (network) => {
    if (isTestnet(network)) {
      if (config?.testnetUrl) {
        return import_algorand_client2.AlgorandClient.fromConfig({
          algodConfig: { server: config.testnetUrl, token: config.algodToken ?? "" }
        });
      }
      return import_algorand_client2.AlgorandClient.testNet();
    }
    if (config?.mainnetUrl) {
      return import_algorand_client2.AlgorandClient.fromConfig({
        algodConfig: { server: config.mainnetUrl, token: config.algodToken ?? "" }
      });
    }
    return import_algorand_client2.AlgorandClient.mainNet();
  };
  const clientCache = /* @__PURE__ */ new Map();
  const getClient = (network) => {
    const key = isTestnet(network) ? "testnet" : "mainnet";
    let client = clientCache.get(key);
    if (!client) {
      client = getAlgorandClientForNetwork(network);
      clientCache.set(key, client);
    }
    return client;
  };
  return {
    getAddresses: () => [address],
    signTransaction: async (txn, _) => {
      const decoded = (0, import_transact3.decodeTransaction)(txn);
      const signedBytes = await algokitSigners.signer([decoded], [0]);
      return signedBytes[0];
    },
    getAlgodClient: (network) => getClient(network).client.algod,
    simulateTransactions: async (txns, network) => {
      const algod = getClient(network).client.algod;
      return await algod.simulateRawTransactions(txns);
    },
    sendTransactions: async (signedTxns, network) => {
      const algod = getClient(network).client.algod;
      const response = await algod.sendRawTransaction(signedTxns);
      return response.txId;
    },
    waitForConfirmation: async (txId, network, waitRounds = 5) => {
      const algod = getClient(network).client.algod;
      return await (0, import_transaction.waitForConfirmation)(txId, waitRounds, algod);
    }
  };
}

// src/types.ts
function isExactAvmPayload(payload) {
  return typeof payload === "object" && payload !== null && "paymentGroup" in payload && "paymentIndex" in payload && Array.isArray(payload.paymentGroup) && typeof payload.paymentIndex === "number" && payload.paymentGroup.every((item) => typeof item === "string");
}

// src/index.ts
var import_common3 = require("@algorandfoundation/algokit-utils/common");
var import_amount2 = require("@algorandfoundation/algokit-utils/amount");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ALGOKIT_SIGNER,
  ALGORAND_ADDRESS_LENGTH,
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_MIN_TX_FEE,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS,
  ExactAvmScheme,
  MAX_REASONABLE_FEE_PER_TXN,
  USDC_CONFIG,
  USDC_DECIMALS,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  convertFromTokenAmount,
  convertToTokenAmount,
  decodeSignedTransaction,
  decodeTransaction,
  decodeUnsignedTransaction,
  encodeTransaction,
  getAlgokitSigner,
  getGenesisHashFromTransaction,
  getNetworkFromCaip2,
  getSenderFromTransaction,
  getTransactionId,
  hasSignature,
  isAlgorandNetwork,
  isAvmSignerWallet,
  isExactAvmPayload,
  isTestnetNetwork,
  isValidAlgorandAddress,
  maxReasonableGroupFee,
  toClientAvmSigner,
  toFacilitatorAvmSigner,
  validateGroupId
});
