// @algorandfoundation/algokit-utils/testing/account.mjs
import { Address } from "../packages/common/src/address.mjs";
import { Config } from "../config.mjs";
import { AlgorandClient } from "../algorand-client.mjs";
//#region src/testing/account.ts
async function getTestAccount({ suppressLog, initialFunds, accountGetter }, algodOrAlgorandClient, kmd) {
	const algorand = algodOrAlgorandClient instanceof AlgorandClient ? algodOrAlgorandClient : AlgorandClient.fromClients({
		algod: algodOrAlgorandClient,
		kmd
	});
	const account = accountGetter ? await accountGetter(algorand) : algorand.account.random();
	Config.getLogger(suppressLog).info(`New test account created with address '${account.addr}'.`);
	const dispenser = await algorand.account.dispenserFromEnvironment();
	await algorand.send.payment({
		sender: dispenser,
		receiver: account.addr,
		amount: initialFunds,
		note: "Funding test account",
		suppressLog
	});
	const accountInfo = await algorand.account.getInformation(account.addr);
	Config.getLogger(suppressLog).info("Test account funded; account balance: %d µALGO", accountInfo.balance.microAlgo);
	algorand.setSignerFromAccount(account);
	const address = Address.fromString(account.addr.toString());
	for (const key of Object.keys(account)) if (!(key in address)) address[key] = account[key];
	return address;
}
//#endregion
export { getTestAccount };
