// @algorandfoundation/algokit-utils/packages/kmd_client/src/apis/api-service.js
const require_transaction = require("../../../transact/src/transactions/transaction.js");
const require_model_runtime = require("../core/model-runtime.js");
const require_list_wallets_response = require("../models/list-wallets-response.js");
const require_export_key_response = require("../models/export-key-response.js");
const require_import_key_response = require("../models/import-key-response.js");
const require_list_keys_response = require("../models/list-keys-response.js");
const require_generate_key_response = require("../models/generate-key-response.js");
const require_export_master_key_response = require("../models/export-master-key-response.js");
const require_export_multisig_response = require("../models/export-multisig-response.js");
const require_import_multisig_response = require("../models/import-multisig-response.js");
const require_list_multisig_response = require("../models/list-multisig-response.js");
const require_sign_program_multisig_response = require("../models/sign-program-multisig-response.js");
const require_sign_multisig_response = require("../models/sign-multisig-response.js");
const require_sign_program_response = require("../models/sign-program-response.js");
const require_sign_transaction_response = require("../models/sign-transaction-response.js");
const require_wallet_info_response = require("../models/wallet-info-response.js");
const require_init_wallet_handle_token_response = require("../models/init-wallet-handle-token-response.js");
const require_rename_wallet_response = require("../models/rename-wallet-response.js");
const require_renew_wallet_handle_token_response = require("../models/renew-wallet-handle-token-response.js");
const require_create_wallet_response = require("../models/create-wallet-response.js");
const require_create_wallet_request = require("../models/create-wallet-request.js");
const require_delete_key_request = require("../models/delete-key-request.js");
const require_delete_multisig_request = require("../models/delete-multisig-request.js");
const require_export_key_request = require("../models/export-key-request.js");
const require_export_master_key_request = require("../models/export-master-key-request.js");
const require_export_multisig_request = require("../models/export-multisig-request.js");
const require_generate_key_request = require("../models/generate-key-request.js");
const require_import_key_request = require("../models/import-key-request.js");
const require_import_multisig_request = require("../models/import-multisig-request.js");
const require_init_wallet_handle_token_request = require("../models/init-wallet-handle-token-request.js");
const require_list_keys_request = require("../models/list-keys-request.js");
const require_list_multisig_request = require("../models/list-multisig-request.js");
const require_release_wallet_handle_token_request = require("../models/release-wallet-handle-token-request.js");
const require_rename_wallet_request = require("../models/rename-wallet-request.js");
const require_renew_wallet_handle_token_request = require("../models/renew-wallet-handle-token-request.js");
const require_sign_multisig_txn_request = require("../models/sign-multisig-txn-request.js");
const require_sign_program_multisig_request = require("../models/sign-program-multisig-request.js");
const require_sign_program_request = require("../models/sign-program-request.js");
const require_sign_txn_request = require("../models/sign-txn-request.js");
const require_versions_response = require("../models/versions-response.js");
const require_wallet_info_request = require("../models/wallet-info-request.js");
//#region packages/kmd_client/src/apis/api-service.ts
var KmdApi = class {
	constructor(httpRequest) {
		this.httpRequest = httpRequest;
	}
	mimeTypeFor(format) {
		return format === "json" ? "application/json" : format === "msgpack" ? "application/msgpack" : "text/plain";
	}
	/**
	* Create a new wallet (collection of keys) with the given parameters.
	*/
	async _createWallet(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_create_wallet_request.CreateWalletRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_create_wallet_response.CreateWalletResponseMeta);
	}
	/**
	* Deletes the key with the passed public key from the wallet.
	*/
	async deleteKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_delete_key_request.DeleteKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		await this.httpRequest.request({
			method: "DELETE",
			url: "/v1/key",
			path: {},
			query: {},
			headers,
			body: serializedBody
		});
	}
	/**
	* Deletes multisig preimage information for the passed address from the wallet.
	*/
	async deleteMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_delete_multisig_request.DeleteMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		await this.httpRequest.request({
			method: "DELETE",
			url: "/v1/multisig",
			path: {},
			query: {},
			headers,
			body: serializedBody
		});
	}
	/**
	* Export the secret key associated with the passed public key.
	*/
	async exportKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_export_key_request.ExportKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key/export",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_export_key_response.ExportKeyResponseMeta);
	}
	/**
	* Export the master derivation key from the wallet. This key is a master "backup" key for the underlying wallet. With it, you can regenerate all of the wallets that have been generated with this wallet's `POST /v1/key` endpoint. This key will not allow you to recover keys imported from other wallets, however.
	*/
	async exportMasterKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_export_master_key_request.ExportMasterKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/master-key/export",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_export_master_key_response.ExportMasterKeyResponseMeta);
	}
	/**
	* Given a multisig address whose preimage this wallet stores, returns the information used to generate the address, including public keys, threshold, and multisig version.
	*/
	async exportMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_export_multisig_request.ExportMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/export",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_export_multisig_response.ExportMultisigResponseMeta);
	}
	/**
	* Generates the next key in the deterministic key sequence (as determined by the master derivation key) and adds it to the wallet, returning the public key.
	*/
	async generateKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_generate_key_request.GenerateKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_generate_key_response.GenerateKeyResponseMeta);
	}
	/**
	* Import an externally generated key into the wallet. Note that if you wish to back up the imported key, you must do so by backing up the entire wallet database, because imported keys were not derived from the wallet's master derivation key.
	*/
	async importKey(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_import_key_request.ImportKeyRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key/import",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_import_key_response.ImportKeyResponseMeta);
	}
	/**
	* Generates a multisig account from the passed public keys array and multisig metadata, and stores all of this in the wallet.
	*/
	async importMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_import_multisig_request.ImportMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/import",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_import_multisig_response.ImportMultisigResponseMeta);
	}
	/**
	* Unlock the wallet and return a wallet handle token that can be used for subsequent operations. These tokens expire periodically and must be renewed. You can `POST` the token to `/v1/wallet/info` to see how much time remains until expiration, and renew it with `/v1/wallet/renew`. When you're done, you can invalidate the token with `/v1/wallet/release`.
	*/
	async initWalletHandle(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_init_wallet_handle_token_request.InitWalletHandleTokenRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/init",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_init_wallet_handle_token_response.InitWalletHandleTokenResponseMeta);
	}
	/**
	* Lists all of the public keys in this wallet. All of them have a stored private key.
	*/
	async listKeysInWallet(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_list_keys_request.ListKeysRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/key/list",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_list_keys_response.ListKeysResponseMeta);
	}
	/**
	* Lists all of the multisig accounts whose preimages this wallet stores
	*/
	async listMultisig(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_list_multisig_request.ListMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/list",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_list_multisig_response.ListMultisigResponseMeta);
	}
	/**
	* Lists all of the wallets that kmd is aware of.
	*/
	async listWallets() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/v1/wallets",
			path: {},
			query: {},
			headers,
			body: void 0
		}), require_list_wallets_response.ListWalletsResponseMeta);
	}
	/**
	* Invalidate the passed wallet handle token, making it invalid for use in subsequent requests.
	*/
	async releaseWalletHandleToken(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_release_wallet_handle_token_request.ReleaseWalletHandleTokenRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/release",
			path: {},
			query: {},
			headers,
			body: serializedBody
		});
	}
	/**
	* Rename the underlying wallet to something else
	*/
	async renameWallet(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_rename_wallet_request.RenameWalletRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/rename",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_rename_wallet_response.RenameWalletResponseMeta);
	}
	/**
	* Renew a wallet handle token, increasing its expiration duration to its initial value
	*/
	async renewWalletHandleToken(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_renew_wallet_handle_token_request.RenewWalletHandleTokenRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/renew",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_renew_wallet_handle_token_response.RenewWalletHandleTokenResponseMeta);
	}
	/**
	* Start a multisig signature, or add a signature to a partially completed multisig signature object.
	*/
	async signMultisigProgram(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_sign_program_multisig_request.SignProgramMultisigRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/signprogram",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_sign_program_multisig_response.SignProgramMultisigResponseMeta);
	}
	/**
	* Start a multisig signature, or add a signature to a partially completed multisig signature object.
	*/
	async _signMultisigTransaction(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_sign_multisig_txn_request.SignMultisigTxnRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/multisig/sign",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_sign_multisig_response.SignMultisigResponseMeta);
	}
	/**
	* Signs the passed program with a key from the wallet, determined by the account named in the request.
	*/
	async signProgram(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_sign_program_request.SignProgramRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/program/sign",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_sign_program_response.SignProgramResponseMeta);
	}
	/**
	* Signs the passed transaction with a key from the wallet, determined by the sender encoded in the transaction.
	*/
	async _signTransaction(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_sign_txn_request.SignTxnRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/transaction/sign",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_sign_transaction_response.SignTransactionResponseMeta);
	}
	async version() {
		const headers = {};
		headers["Accept"] = this.mimeTypeFor("json");
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "GET",
			url: "/versions",
			path: {},
			query: {},
			headers,
			body: void 0
		}), require_versions_response.VersionsResponseMeta);
	}
	/**
	* Returns information about the wallet associated with the passed wallet handle token. Additionally returns expiration information about the token itself.
	*/
	async walletInfo(body) {
		const headers = {};
		const responseFormat = "json";
		headers["Accept"] = this.mimeTypeFor(responseFormat);
		const bodyMeta = require_wallet_info_request.WalletInfoRequestMeta;
		const mediaType = this.mimeTypeFor(!bodyMeta ? "text" : responseFormat);
		if (mediaType) headers["Content-Type"] = mediaType;
		const serializedBody = body ? require_model_runtime.encodeJson(body, bodyMeta) : void 0;
		return require_model_runtime.decodeJson(await this.httpRequest.request({
			method: "POST",
			url: "/v1/wallet/info",
			path: {},
			query: {},
			headers,
			body: serializedBody
		}), require_wallet_info_response.WalletInfoResponseMeta);
	}
	/**
	* Create a new wallet (collection of keys) with the given parameters.
	*/
	async createWallet(body) {
		return await this._createWallet({
			...body,
			walletDriverName: body.walletDriverName ?? "sqlite"
		});
	}
	/**
	* Enables the signing of a transaction using the provided wallet and multisig info.
	* The public key is used to identify which multisig account key to use for signing.
	* When a signer is provided it is used to resolve the private key and sign the transaction, enabling rekeyed account signing.
	* @returns A multisig signature or partial signature, which can be used to form a signed transaction.
	*/
	async signMultisigTransaction(body) {
		const requestBody = {
			...body,
			transaction: require_transaction.encodeTransactionRaw(body.transaction)
		};
		return this._signMultisigTransaction(requestBody);
	}
	/**
	* Enables the signing of a transaction using the provided wallet info.
	* When a public key is provided it is used to resolve the private key and sign the transaction, enabling rekeyed account signing.
	* @returns An encoded, signed transaction.
	*/
	async signTransaction(body) {
		const requestBody = {
			...body,
			transaction: require_transaction.encodeTransactionRaw(body.transaction)
		};
		return this._signTransaction(requestBody);
	}
};
//#endregion
exports.KmdApi = KmdApi;
