// @algorandfoundation/algokit-utils/packages/transact/src/transactions/app-call.mjs
import { MAX_ARGS_SIZE, PROGRAM_PAGE_SIZE } from "../../../common/src/constants.mjs";
import { TransactionValidationErrorType } from "./common.mjs";
//#region packages/transact/src/transactions/app-call.ts
/**
* On-completion actions for application transactions.
*
* These values define what additional actions occur with the transaction.
*/
let OnApplicationComplete = /* @__PURE__ */ function(OnApplicationComplete) {
	/**
	* NoOp indicates that an app transaction will simply call its
	* approval program without any additional action.
	*/
	OnApplicationComplete[OnApplicationComplete["NoOp"] = 0] = "NoOp";
	/**
	* OptIn indicates that an app transaction will allocate some
	* local state for the app in the sender's account.
	*/
	OnApplicationComplete[OnApplicationComplete["OptIn"] = 1] = "OptIn";
	/**
	* CloseOut indicates that an app transaction will deallocate
	* some local state for the app from the user's account.
	*/
	OnApplicationComplete[OnApplicationComplete["CloseOut"] = 2] = "CloseOut";
	/**
	* ClearState is similar to CloseOut, but may never fail. This
	* allows users to reclaim their minimum balance from an app
	* they no longer wish to opt in to.
	*/
	OnApplicationComplete[OnApplicationComplete["ClearState"] = 3] = "ClearState";
	/**
	* UpdateApplication indicates that an app transaction will
	* update the approval program and clear state program for the app.
	*/
	OnApplicationComplete[OnApplicationComplete["UpdateApplication"] = 4] = "UpdateApplication";
	/**
	* DeleteApplication indicates that an app transaction will
	* delete the app parameters for the app from the creator's
	* balance record.
	*/
	OnApplicationComplete[OnApplicationComplete["DeleteApplication"] = 5] = "DeleteApplication";
	return OnApplicationComplete;
}({});
const FIELD_ARGS = "Args";
const FIELD_APPROVAL_PROGRAM = "Approval program";
const FIELD_CLEAR_STATE_PROGRAM = "Clear state program";
const FIELD_GLOBAL_STATE_SCHEMA = "Global state schema";
const FIELD_LOCAL_STATE_SCHEMA = "Local state schema";
const FIELD_EXTRA_PROGRAM_PAGES = "Extra program pages";
/**
* Validate app call transaction fields
*/
function validateAppCallTransaction(appCall) {
	const errors = new Array();
	if (appCall.appId === 0n) errors.push(...validateAppCreation(appCall));
	else errors.push(...validateAppOperation(appCall));
	errors.push(...validateAppCommonFields(appCall));
	return errors;
}
/**
* Validate app creation fields
*/
function validateAppCreation(appCall) {
	const errors = new Array();
	if (!appCall.approvalProgram || appCall.approvalProgram.length === 0) errors.push({
		type: TransactionValidationErrorType.RequiredField,
		data: FIELD_APPROVAL_PROGRAM
	});
	if (!appCall.clearStateProgram || appCall.clearStateProgram.length === 0) errors.push({
		type: TransactionValidationErrorType.RequiredField,
		data: FIELD_CLEAR_STATE_PROGRAM
	});
	const extraPages = appCall.extraProgramPages ?? 0;
	if (extraPages > 3) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: FIELD_EXTRA_PROGRAM_PAGES,
			actual: extraPages,
			max: 3,
			unit: "pages"
		}
	});
	const maxProgramSize = PROGRAM_PAGE_SIZE + extraPages * PROGRAM_PAGE_SIZE;
	if (appCall.approvalProgram && appCall.approvalProgram.length > maxProgramSize) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: FIELD_APPROVAL_PROGRAM,
			actual: appCall.approvalProgram.length,
			max: maxProgramSize,
			unit: "bytes"
		}
	});
	if (appCall.clearStateProgram && appCall.clearStateProgram.length > maxProgramSize) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: FIELD_CLEAR_STATE_PROGRAM,
			actual: appCall.clearStateProgram.length,
			max: maxProgramSize,
			unit: "bytes"
		}
	});
	const totalProgramSize = (appCall.approvalProgram?.length ?? 0) + (appCall.clearStateProgram?.length ?? 0);
	if (totalProgramSize > maxProgramSize) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Combined approval and clear state programs",
			actual: totalProgramSize,
			max: maxProgramSize,
			unit: "bytes"
		}
	});
	if (appCall.globalStateSchema) {
		if (appCall.globalStateSchema.numUints + appCall.globalStateSchema.numByteSlices > 64) errors.push({
			type: TransactionValidationErrorType.FieldTooLong,
			data: {
				field: FIELD_GLOBAL_STATE_SCHEMA,
				actual: appCall.globalStateSchema.numUints + appCall.globalStateSchema.numByteSlices,
				max: 64,
				unit: "keys"
			}
		});
	}
	if (appCall.localStateSchema) {
		if (appCall.localStateSchema.numUints + appCall.localStateSchema.numByteSlices > 16) errors.push({
			type: TransactionValidationErrorType.FieldTooLong,
			data: {
				field: FIELD_LOCAL_STATE_SCHEMA,
				actual: appCall.localStateSchema.numUints + appCall.localStateSchema.numByteSlices,
				max: 16,
				unit: "keys"
			}
		});
	}
	return errors;
}
/**
* Validate app operation (update, delete, call) fields
*/
function validateAppOperation(appCall) {
	const errors = new Array();
	if (appCall.onComplete === OnApplicationComplete.UpdateApplication) {
		if (!appCall.approvalProgram || appCall.approvalProgram.length === 0) errors.push({
			type: TransactionValidationErrorType.RequiredField,
			data: FIELD_APPROVAL_PROGRAM
		});
		if (!appCall.clearStateProgram || appCall.clearStateProgram.length === 0) errors.push({
			type: TransactionValidationErrorType.RequiredField,
			data: FIELD_CLEAR_STATE_PROGRAM
		});
	}
	if (appCall.globalStateSchema !== void 0) errors.push({
		type: TransactionValidationErrorType.ImmutableField,
		data: FIELD_GLOBAL_STATE_SCHEMA
	});
	if (appCall.localStateSchema !== void 0) errors.push({
		type: TransactionValidationErrorType.ImmutableField,
		data: FIELD_LOCAL_STATE_SCHEMA
	});
	if (appCall.extraProgramPages !== void 0) errors.push({
		type: TransactionValidationErrorType.ImmutableField,
		data: FIELD_EXTRA_PROGRAM_PAGES
	});
	return errors;
}
/**
* Validate common app call fields
*/
function validateAppCommonFields(appCall) {
	const errors = new Array();
	if (appCall.args) {
		if (appCall.args.length > 16) errors.push({
			type: TransactionValidationErrorType.FieldTooLong,
			data: {
				field: FIELD_ARGS,
				actual: appCall.args.length,
				max: 16,
				unit: "arguments"
			}
		});
		const totalArgsSize = appCall.args.reduce((sum, arg) => sum + arg.length, 0);
		if (totalArgsSize > 2048) errors.push({
			type: TransactionValidationErrorType.FieldTooLong,
			data: {
				field: "Args total size",
				actual: totalArgsSize,
				max: MAX_ARGS_SIZE,
				unit: "bytes"
			}
		});
	}
	if (appCall.accountReferences && appCall.accountReferences.length > 8) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Account references",
			actual: appCall.accountReferences.length,
			max: 8,
			unit: "refs"
		}
	});
	if (appCall.appReferences && appCall.appReferences.length > 8) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "App references",
			actual: appCall.appReferences.length,
			max: 8,
			unit: "refs"
		}
	});
	if (appCall.assetReferences && appCall.assetReferences.length > 8) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Asset references",
			actual: appCall.assetReferences.length,
			max: 8,
			unit: "refs"
		}
	});
	if (appCall.boxReferences) {
		if (appCall.boxReferences.length > 8) errors.push({
			type: TransactionValidationErrorType.FieldTooLong,
			data: {
				field: "Box references",
				actual: appCall.boxReferences.length,
				max: 8,
				unit: "refs"
			}
		});
		const appRefs = appCall.appReferences || [];
		for (const boxRef of appCall.boxReferences) if (boxRef.appId !== 0n && boxRef.appId !== appCall.appId && !appRefs.includes(boxRef.appId)) errors.push({
			type: TransactionValidationErrorType.ArbitraryConstraint,
			data: `Box reference for app ID ${boxRef.appId} must be in app references`
		});
	}
	const totalReferences = (appCall.accountReferences?.length || 0) + (appCall.appReferences?.length || 0) + (appCall.assetReferences?.length || 0) + (appCall.boxReferences?.length || 0);
	if (totalReferences > 8) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Total references",
			actual: totalReferences,
			max: 8,
			unit: "refs"
		}
	});
	return errors;
}
//#endregion
export { OnApplicationComplete, validateAppCallTransaction };
