// @algorandfoundation/algokit-utils/packages/transact/src/transactions/asset-config.mjs
import "../../../common/src/constants.mjs";
import { TransactionValidationErrorType } from "./common.mjs";
//#region packages/transact/src/transactions/asset-config.ts
/**
* Validate asset configuration transaction fields
*/
function validateAssetConfigTransaction(assetConfig) {
	const errors = new Array();
	if (assetConfig.assetId === 0n) errors.push(...validateAssetCreation(assetConfig));
	else errors.push(...validateAssetConfiguration(assetConfig));
	return errors;
}
/**
* Validate asset creation fields
*/
function validateAssetCreation(assetConfig) {
	const errors = new Array();
	if (assetConfig.total === void 0) errors.push({
		type: TransactionValidationErrorType.RequiredField,
		data: "Total"
	});
	if (assetConfig.decimals !== void 0 && assetConfig.decimals > 19) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Decimals",
			actual: assetConfig.decimals,
			max: 19,
			unit: "decimal places"
		}
	});
	if (assetConfig.unitName && assetConfig.unitName.length > 8) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Unit name",
			actual: assetConfig.unitName.length,
			max: 8,
			unit: "bytes"
		}
	});
	if (assetConfig.assetName && assetConfig.assetName.length > 32) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Asset name",
			actual: assetConfig.assetName.length,
			max: 32,
			unit: "bytes"
		}
	});
	if (assetConfig.url && assetConfig.url.length > 96) errors.push({
		type: TransactionValidationErrorType.FieldTooLong,
		data: {
			field: "Url",
			actual: assetConfig.url.length,
			max: 96,
			unit: "bytes"
		}
	});
	return errors;
}
/**
* Validate asset configuration fields
*/
function validateAssetConfiguration(assetConfig) {
	const errors = new Array();
	if (assetConfig.total !== void 0 || assetConfig.decimals !== void 0 || assetConfig.defaultFrozen !== void 0 || assetConfig.assetName !== void 0 || assetConfig.unitName !== void 0 || assetConfig.url !== void 0 || assetConfig.metadataHash !== void 0 || assetConfig.manager !== void 0 || assetConfig.reserve !== void 0 || assetConfig.freeze !== void 0 || assetConfig.clawback !== void 0) {
		if (assetConfig.total !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Total"
		});
		if (assetConfig.decimals !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Decimals"
		});
		if (assetConfig.defaultFrozen !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Default frozen"
		});
		if (assetConfig.assetName !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Asset name"
		});
		if (assetConfig.unitName !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Unit name"
		});
		if (assetConfig.url !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Url"
		});
		if (assetConfig.metadataHash !== void 0) errors.push({
			type: TransactionValidationErrorType.ImmutableField,
			data: "Metadata hash"
		});
	}
	return errors;
}
//#endregion
export { validateAssetConfigTransaction };
