// @algorandfoundation/algokit-utils/packages/transact/src/transactions/reference-types-meta.mjs
import { addressCodec } from "../../../common/src/codecs/primitives/address.mjs";
import { bigIntCodec } from "../../../common/src/codecs/primitives/bigint.mjs";
import { bytesCodec } from "../../../common/src/codecs/primitives/bytes.mjs";
//#region packages/transact/src/transactions/reference-types-meta.ts
/**
* Metadata for BoxReference
*
* Maps wire format (app, name) to BoxReference (appId, name)
*/
const BoxReferenceMeta = {
	name: "BoxReference",
	kind: "object",
	fields: [{
		name: "appId",
		wireKey: "app",
		optional: false,
		codec: bigIntCodec
	}, {
		name: "name",
		wireKey: "name",
		optional: false,
		codec: bytesCodec
	}]
};
/**
* Metadata for HoldingReference
*
* Maps wire format (account, asset) to HoldingReference (address, assetId)
*/
const HoldingReferenceMeta = {
	name: "HoldingReference",
	kind: "object",
	fields: [{
		name: "address",
		wireKey: "account",
		optional: false,
		codec: addressCodec
	}, {
		name: "assetId",
		wireKey: "asset",
		optional: false,
		codec: bigIntCodec
	}]
};
/**
* Metadata for LocalsReference
*
* Maps wire format (account, app) to LocalsReference (address, appId)
*/
const LocalsReferenceMeta = {
	name: "LocalsReference",
	kind: "object",
	fields: [{
		name: "address",
		wireKey: "account",
		optional: false,
		codec: addressCodec
	}, {
		name: "appId",
		wireKey: "app",
		optional: false,
		codec: bigIntCodec
	}]
};
//#endregion
export { BoxReferenceMeta, HoldingReferenceMeta, LocalsReferenceMeta };
