// @algorandfoundation/algokit-utils/packages/transact/src/transactions/signed-transaction.js
const require_json = require("../../../common/src/json.js");
const require_msgpack = require("../../../common/src/msgpack.js");
const require_transaction = require("./transaction.js");
const require_signed_transaction_meta = require("./signed-transaction-meta.js");
//#region packages/transact/src/transactions/signed-transaction.ts
/**
* Encode signed transactions to MsgPack for sending on the network.
*
* This method performs canonical encoding. No domain separation prefix is applicable.
*
* @param signedTransaction - The signed transaction to encode
* @returns The MsgPack encoded bytes or an error if encoding fails.
*/
function encodeSignedTransaction(signedTransaction) {
	return require_msgpack.encodeMsgpack(require_signed_transaction_meta.signedTransactionCodec.encode(signedTransaction, "msgpack"));
}
/**
* Encode signed transactions to MsgPack for sending on the network.
*
*  This method performs canonical encoding. No domain separation prefix is applicable.
*
* @param signedTransactions - A collection of signed transactions to encode
* @returns A collection of MsgPack encoded bytes or an error if encoding fails.
*/
function encodeSignedTransactions(signedTransactions) {
	return signedTransactions.map((st) => encodeSignedTransaction(st));
}
/**
* Decodes MsgPack bytes into a signed transaction.
*
* @param encodedSignedTransaction - The MsgPack encoded signed transaction bytes
* @returns The decoded SignedTransaction or an error if decoding fails.
*/
function decodeSignedTransaction(encodedSignedTransaction) {
	const decodedData = require_msgpack.decodeMsgpack(encodedSignedTransaction);
	return require_signed_transaction_meta.signedTransactionCodec.decode(decodedData, "msgpack");
}
/**
* Decodes a collection of MsgPack bytes into a signed transaction collection.
*
* @param encodedSignedTransactions - A collection of MsgPack encoded bytes, each representing a signed transaction.
* @returns A collection of decoded signed transactions or an error if decoding fails.
*/
function decodeSignedTransactions(encodedSignedTransactions) {
	return encodedSignedTransactions.map((est) => decodeSignedTransaction(est));
}
/**
* Validate a signed transaction structure
*/
function validateSignedTransaction(signedTransaction) {
	require_transaction.validateTransaction(signedTransaction.txn);
	const setSigCount = [
		signedTransaction.sig,
		signedTransaction.msig,
		signedTransaction.lsig
	].filter((sig) => sig !== void 0).length;
	if (setSigCount > 1) throw new Error(`Only one signature type can be set, found ${setSigCount}`);
	if (signedTransaction.sig && signedTransaction.sig.length !== 64) throw new Error("Signature must be 64 bytes");
}
/**
* Encode a signed transaction to wire-format JSON string.
*
* Wire format uses short keys (e.g., `amt`, `fv`, `snd`) matching the Algorand protocol.
* Useful for debugging, logging, or any scenario requiring human-readable JSON
* that preserves the canonical field names.
*
* @param signedTransaction - The signed transaction to encode
* @param space - Optional indentation for pretty-printing
* @returns JSON string in wire format
*/
function encodeSignedTransactionToJson(signedTransaction, space) {
	return require_json.stringifyJson(require_signed_transaction_meta.signedTransactionCodec.encode(signedTransaction, "json"), void 0, space);
}
//#endregion
exports.decodeSignedTransaction = decodeSignedTransaction;
exports.decodeSignedTransactions = decodeSignedTransactions;
exports.encodeSignedTransaction = encodeSignedTransaction;
exports.encodeSignedTransactionToJson = encodeSignedTransactionToJson;
exports.encodeSignedTransactions = encodeSignedTransactions;
exports.validateSignedTransaction = validateSignedTransaction;
