// @algorandfoundation/algokit-utils/packages/transact/src/transactions/transaction-type.js
//#region packages/transact/src/transactions/transaction-type.ts
/**
* Supported transaction types
*/
let TransactionType = /* @__PURE__ */ function(TransactionType) {
	/**
	* Payment transaction
	*/
	TransactionType["Payment"] = "pay";
	/**
	* Key registration transaction
	*/
	TransactionType["KeyRegistration"] = "keyreg";
	/**
	* Asset configuration transaction
	*/
	TransactionType["AssetConfig"] = "acfg";
	/**
	* Asset transfer transaction
	*/
	TransactionType["AssetTransfer"] = "axfer";
	/**
	* Asset freeze transaction
	*/
	TransactionType["AssetFreeze"] = "afrz";
	/**
	* Application transaction
	*/
	TransactionType["AppCall"] = "appl";
	/**
	* State proof transaction
	*/
	TransactionType["StateProof"] = "stpf";
	/**
	* Heartbeat transaction
	*/
	TransactionType["Heartbeat"] = "hb";
	/**
	* Unknown transaction type
	* Used when decoding transactions with unrecognized type values.
	* This should not be used when creating new transactions.
	*/
	TransactionType["Unknown"] = "unknown";
	return TransactionType;
}({});
//#endregion
exports.TransactionType = TransactionType;
