// @algorandfoundation/algokit-utils/indexer-client/indexer-lookup.mjs
//#region src/indexer-client/indexer-lookup.ts
const DEFAULT_INDEXER_MAX_API_RESOURCES_PER_ACCOUNT = 1e3;
/**
* Looks up applications that were created by the given address; will automatically paginate through all data.
* @param indexer An indexer instance
* @param address The address of the creator to look up
* @param getAll Whether or not to include deleted applications. Default true.
* @param paginationLimit The number of records to return per paginated request, default 1000
* @returns The list of application results
*/
async function lookupAccountCreatedApplicationByAddress(indexer, address, getAll = true, paginationLimit) {
	return await executePaginatedRequest((response) => {
		if ("message" in response) throw {
			status: 404,
			...response
		};
		return response.applications;
	}, (nextToken) => {
		return indexer.lookupAccountCreatedApplications(address, {
			includeAll: getAll,
			limit: paginationLimit ?? DEFAULT_INDEXER_MAX_API_RESOURCES_PER_ACCOUNT,
			...nextToken && { next: nextToken }
		});
	});
}
/**
* Looks up asset holdings for the given asset; will automatically paginate through all data.
* @param indexer An indexer instance
* @param assetId The ID of the asset to look up holdings for
* @param options Optional options to control the lookup
* @param paginationLimit The number of records to return per paginated request, default 1000
* @returns The list of application results
*/
async function lookupAssetHoldings(indexer, assetId, options, paginationLimit) {
	return await executePaginatedRequest((response) => {
		if ("message" in response) throw {
			status: 404,
			...response
		};
		return response.balances;
	}, (nextToken) => {
		return indexer.lookupAssetBalances(assetId, {
			limit: paginationLimit ?? DEFAULT_INDEXER_MAX_API_RESOURCES_PER_ACCOUNT,
			...options?.currencyGreaterThan !== void 0 && { currencyGreaterThan: options.currencyGreaterThan },
			...options?.currencyLessThan !== void 0 && { currencyLessThan: options.currencyLessThan },
			...options?.includeAll !== void 0 && { includeAll: options.includeAll },
			...nextToken && { next: nextToken }
		});
	});
}
/**
* Allows transactions to be searched for the given criteria.
* @param indexer An indexer client
* @param searchCriteria The criteria to search for
* @param paginationLimit The number of records to return per paginated request, default 1000
* @returns The search results
*/
async function searchTransactions(indexer, searchCriteria, paginationLimit) {
	let currentRound = 0n;
	const transactions = await executePaginatedRequest((response) => {
		if ("message" in response) throw {
			status: 404,
			...response
		};
		if (response.currentRound > currentRound) currentRound = response.currentRound;
		return response.transactions;
	}, (nextToken) => {
		return indexer.searchForTransactions({
			...searchCriteria,
			limit: paginationLimit ?? DEFAULT_INDEXER_MAX_API_RESOURCES_PER_ACCOUNT,
			next: nextToken
		});
	});
	return {
		currentRound,
		nextToken: void 0,
		transactions
	};
}
async function executePaginatedRequest(extractItems, buildRequest) {
	const results = [];
	let nextToken = void 0;
	while (true) {
		const response = await buildRequest(nextToken);
		const items = extractItems(response);
		if (items == null || items.length === 0) break;
		results.push(...items);
		nextToken = response["nextToken"];
		if (!nextToken) break;
	}
	return results;
}

//#endregion
export { executePaginatedRequest, lookupAccountCreatedApplicationByAddress, lookupAssetHoldings, searchTransactions };
