// @algorandfoundation/algokit-utils/packages/common/src/address.mjs
import { arrayEqual, concatArrays } from "./array.mjs";
import "./constants.mjs";
import base32 from "hi-base32";
import { sha512_256 } from "@noble/hashes/sha2.js";
//#region packages/common/src/address.ts
const ALGORAND_ADDRESS_BYTE_LENGTH = 36;
const ALGORAND_CHECKSUM_BYTE_LENGTH = 4;
const ALGORAND_ADDRESS_LENGTH = 58;
const ALGORAND_ZERO_ADDRESS_STRING = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ";
const MALFORMED_ADDRESS_ERROR_MSG = "address seems to be malformed";
const CHECKSUM_ADDRESS_ERROR_MSG = "wrong checksum for address";
function checksumFromPublicKey(publicKey) {
	return Uint8Array.from(sha512_256(publicKey).slice(28, 32));
}
function bytesToHex(bytes) {
	return Buffer.from(bytes).toString("hex");
}
function encodeUint64(num) {
	if (!(typeof num === "bigint" || Number.isInteger(num)) || num < 0 || num > BigInt("0xffffffffffffffff")) throw new Error("Input is not a 64-bit unsigned integer");
	const encoding = new Uint8Array(8);
	new DataView(encoding.buffer).setBigUint64(0, BigInt(num));
	return encoding;
}
/** Symbol used for instanceof checks across packages (CJS/ESM) */
const ADDR_SYMBOL = Symbol.for("algokit_common:Address");
/**
* Represents an Algorand address
*/
var Address = class Address {
	/**
	* The binary form of the address. For standard accounts, this is the public key.
	*/
	publicKey;
	/** @internal */
	[ADDR_SYMBOL];
	/**
	* Create a new Address object from its binary form.
	* @param publicKey - The binary form of the address. Must be 32 bytes.
	*/
	constructor(publicKey) {
		this[ADDR_SYMBOL] = true;
		if (publicKey.length !== 32) throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: 0x${bytesToHex(publicKey)}, length ${publicKey.length}`);
		this.publicKey = publicKey;
	}
	/**
	* Check if the address is equal to another address.
	*/
	equals(other) {
		return other instanceof Address && arrayEqual(this.publicKey, other.publicKey);
	}
	/**
	* Compute the 4 byte checksum of the address.
	*/
	checksum() {
		return checksumFromPublicKey(this.publicKey);
	}
	/**
	* Encode the address into a string form.
	*/
	toString() {
		return base32.encode(concatArrays(this.publicKey, this.checksum())).slice(0, 58);
	}
	/**
	* Decode an address from a string.
	* @param address - The address to decode. Must be 58 characters long.
	* @returns An Address object corresponding to the input string.
	*/
	static fromString(address) {
		if (typeof address !== "string") throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: expected string, got ${typeof address}, ${address}`);
		if (address.length !== 58) throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: expected length 58, got ${address.length}: ${address}`);
		const decoded = base32.decode.asBytes(address);
		if (decoded.length !== 36) throw new Error(`${MALFORMED_ADDRESS_ERROR_MSG}: expected byte length 36, got ${decoded.length}`);
		const pk = new Uint8Array(decoded.slice(0, 32));
		const cs = new Uint8Array(decoded.slice(32, 36));
		const checksum = checksumFromPublicKey(pk);
		if (!arrayEqual(checksum, cs)) throw new Error(`${CHECKSUM_ADDRESS_ERROR_MSG}: ${address} (${cs}, ${checksum})`);
		return new Address(pk);
	}
	/**
	* Get the zero address.
	*/
	static zeroAddress() {
		return new Address(new Uint8Array(32));
	}
};
Object.defineProperty(Address, Symbol.hasInstance, { value: function(obj) {
	if (obj instanceof Object && Object.getPrototypeOf(obj) === Address.prototype) return true;
	return Boolean(obj && typeof obj === "object" && ADDR_SYMBOL in obj && obj[ADDR_SYMBOL]);
} });
function getAddress(addr) {
	if (typeof addr == "string") return Address.fromString(addr);
	else if ("addr" in addr) return addr.addr;
	else return addr;
}
function getOptionalAddress(addr) {
	if (addr === void 0) return;
	return getAddress(addr);
}
/**
* isValidAddress checks if a string is a valid Algorand address.
* @param address - an Algorand address with checksum.
* @returns true if valid, false otherwise
*/
function isValidAddress(address) {
	try {
		Address.fromString(address);
	} catch {
		return false;
	}
	return true;
}
const APP_ID_PREFIX = new TextEncoder().encode("appID");
/**
* Get the escrow address of an application.
* @param appID - The ID of the application.
* @returns The address corresponding to that application's escrow account.
*/
function getApplicationAddress(appID) {
	return new Address(sha512_256(concatArrays(APP_ID_PREFIX, encodeUint64(appID))));
}
/**
* decodeAddress takes an Algorand address in string form and decodes it into a Uint8Array.
* @param address - an Algorand address with checksum.
* @returns the decoded form of the address's public key and checksum
*/
function decodeAddress(address) {
	return Address.fromString(address);
}
/**
* encodeAddress takes an Algorand address as a Uint8Array and encodes it into a string with checksum.
* @param address - a raw Algorand address
* @returns the address and checksum encoded as a string.
*/
function encodeAddress(address) {
	return new Address(address).toString();
}
//#endregion
export { ALGORAND_ADDRESS_BYTE_LENGTH, ALGORAND_ADDRESS_LENGTH, ALGORAND_CHECKSUM_BYTE_LENGTH, ALGORAND_ZERO_ADDRESS_STRING, Address, CHECKSUM_ADDRESS_ERROR_MSG, MALFORMED_ADDRESS_ERROR_MSG, checksumFromPublicKey, decodeAddress, encodeAddress, encodeUint64, getAddress, getApplicationAddress, getOptionalAddress, isValidAddress };
