// @algorandfoundation/algokit-utils/packages/common/src/constants.mjs
//#region packages/common/src/constants.ts
const TRANSACTION_DOMAIN_SEPARATOR = "TX";
const TRANSACTION_GROUP_DOMAIN_SEPARATOR = "TG";
const MULTISIG_DOMAIN_SEPARATOR = "MultisigAddr";
const SIGNATURE_ENCODING_INCR = 75;
const HASH_BYTES_LENGTH = 32;
const PUBLIC_KEY_BYTE_LENGTH = 32;
const MAX_TRANSACTION_GROUP_SIZE = 16;
const CHECKSUM_BYTE_LENGTH = 4;
const ADDRESS_LENGTH = 58;
const TRANSACTION_ID_LENGTH = 52;
const LENGTH_ENCODE_BYTE_SIZE = 2;
const BOOL_TRUE_BYTE = 128;
const BOOL_FALSE_BYTE = 0;
const SIGNATURE_BYTE_LENGTH = 64;
const EMPTY_SIGNATURE = new Uint8Array(64);
const MAX_EXTRA_PROGRAM_PAGES = 3;
const PROGRAM_PAGE_SIZE = 2048;
const MAX_APP_ARGS = 16;
const MAX_ARGS_SIZE = 2048;
const MAX_OVERALL_REFERENCES = 8;
const MAX_ACCOUNT_REFERENCES = 8;
const MAX_APP_REFERENCES = 8;
const MAX_ASSET_REFERENCES = 8;
const MAX_BOX_REFERENCES = 8;
const MAX_GLOBAL_STATE_KEYS = 64;
const MAX_LOCAL_STATE_KEYS = 16;
const MAX_ASSET_NAME_LENGTH = 32;
const MAX_ASSET_UNIT_NAME_LENGTH = 8;
const MAX_ASSET_URL_LENGTH = 96;
const MAX_ASSET_DECIMALS = 19;
//#endregion
export { ADDRESS_LENGTH, BOOL_FALSE_BYTE, BOOL_TRUE_BYTE, CHECKSUM_BYTE_LENGTH, EMPTY_SIGNATURE, HASH_BYTES_LENGTH, LENGTH_ENCODE_BYTE_SIZE, MAX_ACCOUNT_REFERENCES, MAX_APP_ARGS, MAX_APP_REFERENCES, MAX_ARGS_SIZE, MAX_ASSET_DECIMALS, MAX_ASSET_NAME_LENGTH, MAX_ASSET_REFERENCES, MAX_ASSET_UNIT_NAME_LENGTH, MAX_ASSET_URL_LENGTH, MAX_BOX_REFERENCES, MAX_EXTRA_PROGRAM_PAGES, MAX_GLOBAL_STATE_KEYS, MAX_LOCAL_STATE_KEYS, MAX_OVERALL_REFERENCES, MAX_TRANSACTION_GROUP_SIZE, MULTISIG_DOMAIN_SEPARATOR, PROGRAM_PAGE_SIZE, PUBLIC_KEY_BYTE_LENGTH, SIGNATURE_BYTE_LENGTH, SIGNATURE_ENCODING_INCR, TRANSACTION_DOMAIN_SEPARATOR, TRANSACTION_GROUP_DOMAIN_SEPARATOR, TRANSACTION_ID_LENGTH };
